

<!DOCTYPE html>
<html class="writer-html5" lang="en" data-content_root="./">
<head>
  <meta charset="utf-8" /><meta name="viewport" content="width=device-width, initial-scale=1" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>Reference &mdash; idtrack 0.0.4 documentation</title>
      <link rel="stylesheet" type="text/css" href="_static/pygments.css?v=03e43079" />
      <link rel="stylesheet" type="text/css" href="_static/css/theme.css?v=e59714d7" />
      <link rel="stylesheet" type="text/css" href="_static/custom_settings.css?v=15bcc2cf" />

  
      <script src="_static/jquery.js?v=5d32c60e"></script>
      <script src="_static/_sphinx_javascript_frameworks_compat.js?v=2cd50e6c"></script>
      <script src="_static/documentation_options.js?v=8c5712d9"></script>
      <script src="_static/doctools.js?v=9a2dae69"></script>
      <script src="_static/sphinx_highlight.js?v=dc90522c"></script>
      <script crossorigin="anonymous" integrity="sha256-Ae2Vz/4ePdIu6ZyI/5ZGsYnb+m0JlOmKPjt6XZ9JJkA=" src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
    <script src="_static/js/theme.js"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="Contributor Guide" href="contributing.html" />
    <link rel="prev" title="Comparison with Available Tools" href="_notebooks/comparison.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >

          
          
          <a href="index.html" class="icon icon-home">
            idtrack
              <img src="_static/logo.png" class="logo" alt="Logo"/>
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <p class="caption" role="heading"><span class="caption-text">Contents:</span></p>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="installation.html">Installation</a></li>
<li class="toctree-l1"><a class="reference internal" href="usage.html">Usage</a></li>
<li class="toctree-l1 current"><a class="current reference internal" href="#">Reference</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#idtrack._api.API"><code class="docutils literal notranslate"><span class="pre">API</span></code></a><ul>
<li class="toctree-l3"><a class="reference internal" href="#idtrack._api.API.available_available_external_databases"><code class="docutils literal notranslate"><span class="pre">API.available_available_external_databases()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#idtrack._api.API.calculate_graph_caches"><code class="docutils literal notranslate"><span class="pre">API.calculate_graph_caches()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#idtrack._api.API.classify_multiple_conversion"><code class="docutils literal notranslate"><span class="pre">API.classify_multiple_conversion()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#idtrack._api.API.configure_logger"><code class="docutils literal notranslate"><span class="pre">API.configure_logger()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#idtrack._api.API.convert_identifier"><code class="docutils literal notranslate"><span class="pre">API.convert_identifier()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#idtrack._api.API.convert_identifier_multiple"><code class="docutils literal notranslate"><span class="pre">API.convert_identifier_multiple()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#idtrack._api.API.get_database_manager"><code class="docutils literal notranslate"><span class="pre">API.get_database_manager()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#idtrack._api.API.get_ensembl_organism"><code class="docutils literal notranslate"><span class="pre">API.get_ensembl_organism()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#idtrack._api.API.infer_identifier_source"><code class="docutils literal notranslate"><span class="pre">API.infer_identifier_source()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#idtrack._api.API.initialize_graph"><code class="docutils literal notranslate"><span class="pre">API.initialize_graph()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#idtrack._api.API.print_binned_conversion"><code class="docutils literal notranslate"><span class="pre">API.print_binned_conversion()</span></code></a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#idtrack._database_manager.DatabaseManager"><code class="docutils literal notranslate"><span class="pre">DatabaseManager</span></code></a><ul>
<li class="toctree-l3"><a class="reference internal" href="#idtrack._database_manager.DatabaseManager._create_relation_helper"><code class="docutils literal notranslate"><span class="pre">DatabaseManager._create_relation_helper()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#idtrack._database_manager.DatabaseManager._determine_usecols_ids"><code class="docutils literal notranslate"><span class="pre">DatabaseManager._determine_usecols_ids()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#idtrack._database_manager.DatabaseManager.available_releases"><code class="docutils literal notranslate"><span class="pre">DatabaseManager.available_releases</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#idtrack._database_manager.DatabaseManager.available_releases_no_save"><code class="docutils literal notranslate"><span class="pre">DatabaseManager.available_releases_no_save</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#idtrack._database_manager.DatabaseManager.available_releases_versions"><code class="docutils literal notranslate"><span class="pre">DatabaseManager.available_releases_versions()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#idtrack._database_manager.DatabaseManager.available_tables_mysql"><code class="docutils literal notranslate"><span class="pre">DatabaseManager.available_tables_mysql()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#idtrack._database_manager.DatabaseManager.change_assembly"><code class="docutils literal notranslate"><span class="pre">DatabaseManager.change_assembly()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#idtrack._database_manager.DatabaseManager.change_form"><code class="docutils literal notranslate"><span class="pre">DatabaseManager.change_form()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#idtrack._database_manager.DatabaseManager.change_release"><code class="docutils literal notranslate"><span class="pre">DatabaseManager.change_release()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#idtrack._database_manager.DatabaseManager.check_if_change_assembly_works"><code class="docutils literal notranslate"><span class="pre">DatabaseManager.check_if_change_assembly_works()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#idtrack._database_manager.DatabaseManager.check_version_info"><code class="docutils literal notranslate"><span class="pre">DatabaseManager.check_version_info()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#idtrack._database_manager.DatabaseManager.create_available_databases"><code class="docutils literal notranslate"><span class="pre">DatabaseManager.create_available_databases()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#idtrack._database_manager.DatabaseManager.create_database_content"><code class="docutils literal notranslate"><span class="pre">DatabaseManager.create_database_content()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#idtrack._database_manager.DatabaseManager.create_external_all"><code class="docutils literal notranslate"><span class="pre">DatabaseManager.create_external_all()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#idtrack._database_manager.DatabaseManager.create_external_db"><code class="docutils literal notranslate"><span class="pre">DatabaseManager.create_external_db()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#idtrack._database_manager.DatabaseManager.create_id_history"><code class="docutils literal notranslate"><span class="pre">DatabaseManager.create_id_history()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#idtrack._database_manager.DatabaseManager.create_id_history_fixed"><code class="docutils literal notranslate"><span class="pre">DatabaseManager.create_id_history_fixed()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#idtrack._database_manager.DatabaseManager.create_ids"><code class="docutils literal notranslate"><span class="pre">DatabaseManager.create_ids()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#idtrack._database_manager.DatabaseManager.create_relation_archive"><code class="docutils literal notranslate"><span class="pre">DatabaseManager.create_relation_archive()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#idtrack._database_manager.DatabaseManager.create_relation_current"><code class="docutils literal notranslate"><span class="pre">DatabaseManager.create_relation_current()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#idtrack._database_manager.DatabaseManager.create_release_id"><code class="docutils literal notranslate"><span class="pre">DatabaseManager.create_release_id()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#idtrack._database_manager.DatabaseManager.create_version_info"><code class="docutils literal notranslate"><span class="pre">DatabaseManager.create_version_info()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#idtrack._database_manager.DatabaseManager.download_table"><code class="docutils literal notranslate"><span class="pre">DatabaseManager.download_table()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#idtrack._database_manager.DatabaseManager.external_inst"><code class="docutils literal notranslate"><span class="pre">DatabaseManager.external_inst</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#idtrack._database_manager.DatabaseManager.file_name"><code class="docutils literal notranslate"><span class="pre">DatabaseManager.file_name()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#idtrack._database_manager.DatabaseManager.get_db"><code class="docutils literal notranslate"><span class="pre">DatabaseManager.get_db()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#idtrack._database_manager.DatabaseManager.get_release_date"><code class="docutils literal notranslate"><span class="pre">DatabaseManager.get_release_date()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#idtrack._database_manager.DatabaseManager.get_table"><code class="docutils literal notranslate"><span class="pre">DatabaseManager.get_table()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#idtrack._database_manager.DatabaseManager.id_ver_from_df"><code class="docutils literal notranslate"><span class="pre">DatabaseManager.id_ver_from_df()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#idtrack._database_manager.DatabaseManager.mysql_database"><code class="docutils literal notranslate"><span class="pre">DatabaseManager.mysql_database</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#idtrack._database_manager.DatabaseManager.node_dict_maker"><code class="docutils literal notranslate"><span class="pre">DatabaseManager.node_dict_maker()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#idtrack._database_manager.DatabaseManager.node_name_maker"><code class="docutils literal notranslate"><span class="pre">DatabaseManager.node_name_maker()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#idtrack._database_manager.DatabaseManager.tables_in_disk"><code class="docutils literal notranslate"><span class="pre">DatabaseManager.tables_in_disk()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#idtrack._database_manager.DatabaseManager.version_fix"><code class="docutils literal notranslate"><span class="pre">DatabaseManager.version_fix()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#idtrack._database_manager.DatabaseManager.version_fix_incomplete"><code class="docutils literal notranslate"><span class="pre">DatabaseManager.version_fix_incomplete()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#idtrack._database_manager.DatabaseManager.version_uniformize"><code class="docutils literal notranslate"><span class="pre">DatabaseManager.version_uniformize()</span></code></a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#idtrack._external_databases.ExternalDatabases"><code class="docutils literal notranslate"><span class="pre">ExternalDatabases</span></code></a><ul>
<li class="toctree-l3"><a class="reference internal" href="#idtrack._external_databases.ExternalDatabases.create_template_yaml"><code class="docutils literal notranslate"><span class="pre">ExternalDatabases.create_template_yaml()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#idtrack._external_databases.ExternalDatabases.file_name_modified_yaml"><code class="docutils literal notranslate"><span class="pre">ExternalDatabases.file_name_modified_yaml()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#idtrack._external_databases.ExternalDatabases.file_name_template_yaml"><code class="docutils literal notranslate"><span class="pre">ExternalDatabases.file_name_template_yaml()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#idtrack._external_databases.ExternalDatabases.give_list_for_case"><code class="docutils literal notranslate"><span class="pre">ExternalDatabases.give_list_for_case()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#idtrack._external_databases.ExternalDatabases.load_modified_yaml"><code class="docutils literal notranslate"><span class="pre">ExternalDatabases.load_modified_yaml()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#idtrack._external_databases.ExternalDatabases.validate_yaml_file_up_to_date"><code class="docutils literal notranslate"><span class="pre">ExternalDatabases.validate_yaml_file_up_to_date()</span></code></a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#idtrack._the_graph.TheGraph"><code class="docutils literal notranslate"><span class="pre">TheGraph</span></code></a><ul>
<li class="toctree-l3"><a class="reference internal" href="#idtrack._the_graph.TheGraph._combined_edges"><code class="docutils literal notranslate"><span class="pre">TheGraph._combined_edges()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#idtrack._the_graph.TheGraph._combined_edges_genes_helper"><code class="docutils literal notranslate"><span class="pre">TheGraph._combined_edges_genes_helper()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#idtrack._the_graph.TheGraph._get_active_ranges_of_id"><code class="docutils literal notranslate"><span class="pre">TheGraph._get_active_ranges_of_id()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#idtrack._the_graph.TheGraph._node_trios"><code class="docutils literal notranslate"><span class="pre">TheGraph._node_trios()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#idtrack._the_graph.TheGraph.attach_included_forms"><code class="docutils literal notranslate"><span class="pre">TheGraph.attach_included_forms()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#idtrack._the_graph.TheGraph.available_external_databases"><code class="docutils literal notranslate"><span class="pre">TheGraph.available_external_databases</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#idtrack._the_graph.TheGraph.available_external_databases_assembly"><code class="docutils literal notranslate"><span class="pre">TheGraph.available_external_databases_assembly</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#idtrack._the_graph.TheGraph.available_genome_assemblies"><code class="docutils literal notranslate"><span class="pre">TheGraph.available_genome_assemblies</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#idtrack._the_graph.TheGraph.available_releases_given_database_assembly"><code class="docutils literal notranslate"><span class="pre">TheGraph.available_releases_given_database_assembly()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#idtrack._the_graph.TheGraph.calculate_caches"><code class="docutils literal notranslate"><span class="pre">TheGraph.calculate_caches()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#idtrack._the_graph.TheGraph.combined_edges"><code class="docutils literal notranslate"><span class="pre">TheGraph.combined_edges</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#idtrack._the_graph.TheGraph.combined_edges_assembly_specific_genes"><code class="docutils literal notranslate"><span class="pre">TheGraph.combined_edges_assembly_specific_genes</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#idtrack._the_graph.TheGraph.combined_edges_genes"><code class="docutils literal notranslate"><span class="pre">TheGraph.combined_edges_genes</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#idtrack._the_graph.TheGraph.compact_ranges"><code class="docutils literal notranslate"><span class="pre">TheGraph.compact_ranges()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#idtrack._the_graph.TheGraph.external_database_connection_form"><code class="docutils literal notranslate"><span class="pre">TheGraph.external_database_connection_form</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#idtrack._the_graph.TheGraph.get_active_ranges_of_base_id_alternative"><code class="docutils literal notranslate"><span class="pre">TheGraph.get_active_ranges_of_base_id_alternative()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#idtrack._the_graph.TheGraph.get_active_ranges_of_id"><code class="docutils literal notranslate"><span class="pre">TheGraph.get_active_ranges_of_id</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#idtrack._the_graph.TheGraph.get_active_ranges_of_id_ensembl_all_inclusive"><code class="docutils literal notranslate"><span class="pre">TheGraph.get_active_ranges_of_id_ensembl_all_inclusive()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#idtrack._the_graph.TheGraph.get_external_database_nodes"><code class="docutils literal notranslate"><span class="pre">TheGraph.get_external_database_nodes()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#idtrack._the_graph.TheGraph.get_id_list"><code class="docutils literal notranslate"><span class="pre">TheGraph.get_id_list()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#idtrack._the_graph.TheGraph.get_intersecting_ranges"><code class="docutils literal notranslate"><span class="pre">TheGraph.get_intersecting_ranges()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#idtrack._the_graph.TheGraph.get_next_edge_releases"><code class="docutils literal notranslate"><span class="pre">TheGraph.get_next_edge_releases()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#idtrack._the_graph.TheGraph.get_two_nodes_coinciding_releases"><code class="docutils literal notranslate"><span class="pre">TheGraph.get_two_nodes_coinciding_releases()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#idtrack._the_graph.TheGraph.hyperconnective_nodes"><code class="docutils literal notranslate"><span class="pre">TheGraph.hyperconnective_nodes</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#idtrack._the_graph.TheGraph.is_point_in_range"><code class="docutils literal notranslate"><span class="pre">TheGraph.is_point_in_range()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#idtrack._the_graph.TheGraph.list_to_ranges"><code class="docutils literal notranslate"><span class="pre">TheGraph.list_to_ranges()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#idtrack._the_graph.TheGraph.lower_chars_graph"><code class="docutils literal notranslate"><span class="pre">TheGraph.lower_chars_graph</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#idtrack._the_graph.TheGraph.node_name_alternatives"><code class="docutils literal notranslate"><span class="pre">TheGraph.node_name_alternatives()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#idtrack._the_graph.TheGraph.node_trios"><code class="docutils literal notranslate"><span class="pre">TheGraph.node_trios</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#idtrack._the_graph.TheGraph.ranges_to_list"><code class="docutils literal notranslate"><span class="pre">TheGraph.ranges_to_list()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#idtrack._the_graph.TheGraph.rev"><code class="docutils literal notranslate"><span class="pre">TheGraph.rev</span></code></a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#idtrack._graph_maker.GraphMaker"><code class="docutils literal notranslate"><span class="pre">GraphMaker</span></code></a><ul>
<li class="toctree-l3"><a class="reference internal" href="#idtrack._graph_maker.GraphMaker.construct_graph"><code class="docutils literal notranslate"><span class="pre">GraphMaker.construct_graph()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#idtrack._graph_maker.GraphMaker.construct_graph_form"><code class="docutils literal notranslate"><span class="pre">GraphMaker.construct_graph_form()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#idtrack._graph_maker.GraphMaker.create_file_name"><code class="docutils literal notranslate"><span class="pre">GraphMaker.create_file_name()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#idtrack._graph_maker.GraphMaker.export_disk"><code class="docutils literal notranslate"><span class="pre">GraphMaker.export_disk()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#idtrack._graph_maker.GraphMaker.get_graph"><code class="docutils literal notranslate"><span class="pre">GraphMaker.get_graph()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#idtrack._graph_maker.GraphMaker.initialize_downloads"><code class="docutils literal notranslate"><span class="pre">GraphMaker.initialize_downloads()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#idtrack._graph_maker.GraphMaker.read_exported"><code class="docutils literal notranslate"><span class="pre">GraphMaker.read_exported()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#idtrack._graph_maker.GraphMaker.remove_non_gene_trees"><code class="docutils literal notranslate"><span class="pre">GraphMaker.remove_non_gene_trees()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#idtrack._graph_maker.GraphMaker.split_id"><code class="docutils literal notranslate"><span class="pre">GraphMaker.split_id()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#idtrack._graph_maker.GraphMaker.update_graph_with_the_new_release"><code class="docutils literal notranslate"><span class="pre">GraphMaker.update_graph_with_the_new_release()</span></code></a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#idtrack._verify_organism.VerifyOrganism"><code class="docutils literal notranslate"><span class="pre">VerifyOrganism</span></code></a><ul>
<li class="toctree-l3"><a class="reference internal" href="#idtrack._verify_organism.VerifyOrganism.fetch_organism_and_latest_release"><code class="docutils literal notranslate"><span class="pre">VerifyOrganism.fetch_organism_and_latest_release()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#idtrack._verify_organism.VerifyOrganism.get_formal_name"><code class="docutils literal notranslate"><span class="pre">VerifyOrganism.get_formal_name()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#idtrack._verify_organism.VerifyOrganism.get_latest_release"><code class="docutils literal notranslate"><span class="pre">VerifyOrganism.get_latest_release()</span></code></a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#idtrack._track.Track"><code class="docutils literal notranslate"><span class="pre">Track</span></code></a><ul>
<li class="toctree-l3"><a class="reference internal" href="#idtrack._track.Track.graph"><code class="docutils literal notranslate"><span class="pre">Track.graph</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#idtrack._track.Track.version_info"><code class="docutils literal notranslate"><span class="pre">Track.version_info</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#idtrack._track.Track._external_entrance_placeholder"><code class="docutils literal notranslate"><span class="pre">Track._external_entrance_placeholder</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#idtrack._track.Track._external_entrance_placeholders"><code class="docutils literal notranslate"><span class="pre">Track._external_entrance_placeholders</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#idtrack._track.Track._calculate_node_scores_helper"><code class="docutils literal notranslate"><span class="pre">Track._calculate_node_scores_helper</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#idtrack._track.Track._choose_relevant_synonym_helper"><code class="docutils literal notranslate"><span class="pre">Track._choose_relevant_synonym_helper()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#idtrack._track.Track._create_priority_list_ensembl"><code class="docutils literal notranslate"><span class="pre">Track._create_priority_list_ensembl()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#idtrack._track.Track._final_conversion"><code class="docutils literal notranslate"><span class="pre">Track._final_conversion()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#idtrack._track.Track._final_conversion_dict_prepare"><code class="docutils literal notranslate"><span class="pre">Track._final_conversion_dict_prepare()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#idtrack._track.Track._minimum_assembly_jumps_helper"><code class="docutils literal notranslate"><span class="pre">Track._minimum_assembly_jumps_helper()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#idtrack._track.Track._path_score_sorter_all_targets"><code class="docutils literal notranslate"><span class="pre">Track._path_score_sorter_all_targets()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#idtrack._track.Track._path_score_sorter_single_target"><code class="docutils literal notranslate"><span class="pre">Track._path_score_sorter_single_target()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#idtrack._track.Track._recursive_synonymous"><code class="docutils literal notranslate"><span class="pre">Track._recursive_synonymous()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#idtrack._track.Track.calculate_node_scores"><code class="docutils literal notranslate"><span class="pre">Track.calculate_node_scores()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#idtrack._track.Track.calculate_score_and_select"><code class="docutils literal notranslate"><span class="pre">Track.calculate_score_and_select()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#idtrack._track.Track.choose_relevant_synonym"><code class="docutils literal notranslate"><span class="pre">Track.choose_relevant_synonym()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#idtrack._track.Track.convert"><code class="docutils literal notranslate"><span class="pre">Track.convert()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#idtrack._track.Track.convert_optimized_multiple"><code class="docutils literal notranslate"><span class="pre">Track.convert_optimized_multiple()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#idtrack._track.Track.edge_key_orientor"><code class="docutils literal notranslate"><span class="pre">Track.edge_key_orientor()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#idtrack._track.Track.get_from_release_and_reverse_vars"><code class="docutils literal notranslate"><span class="pre">Track.get_from_release_and_reverse_vars()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#idtrack._track.Track.get_next_edges"><code class="docutils literal notranslate"><span class="pre">Track.get_next_edges()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#idtrack._track.Track.get_possible_paths"><code class="docutils literal notranslate"><span class="pre">Track.get_possible_paths()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#idtrack._track.Track.identify_source"><code class="docutils literal notranslate"><span class="pre">Track.identify_source()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#idtrack._track.Track.minimum_assembly_jumps"><code class="docutils literal notranslate"><span class="pre">Track.minimum_assembly_jumps()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#idtrack._track.Track.path_search"><code class="docutils literal notranslate"><span class="pre">Track.path_search()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#idtrack._track.Track.path_step_possible_assembly_jumps"><code class="docutils literal notranslate"><span class="pre">Track.path_step_possible_assembly_jumps()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#idtrack._track.Track.should_graph_reversed"><code class="docutils literal notranslate"><span class="pre">Track.should_graph_reversed()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#idtrack._track.Track.synonymous_nodes"><code class="docutils literal notranslate"><span class="pre">Track.synonymous_nodes()</span></code></a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#idtrack._track_tests.TrackTests"><code class="docutils literal notranslate"><span class="pre">TrackTests</span></code></a><ul>
<li class="toctree-l3"><a class="reference internal" href="#idtrack._track_tests.TrackTests.history_travel_testing"><code class="docutils literal notranslate"><span class="pre">TrackTests.history_travel_testing()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#idtrack._track_tests.TrackTests.history_travel_testing_random"><code class="docutils literal notranslate"><span class="pre">TrackTests.history_travel_testing_random()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#idtrack._track_tests.TrackTests.history_travel_testing_random_arguments_generator"><code class="docutils literal notranslate"><span class="pre">TrackTests.history_travel_testing_random_arguments_generator()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#idtrack._track_tests.TrackTests.how_many_corresponding_path_ensembl"><code class="docutils literal notranslate"><span class="pre">TrackTests.how_many_corresponding_path_ensembl()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#idtrack._track_tests.TrackTests.is_base_is_range_correct"><code class="docutils literal notranslate"><span class="pre">TrackTests.is_base_is_range_correct()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#idtrack._track_tests.TrackTests.is_combined_edges_dicts_overlapping_and_complete"><code class="docutils literal notranslate"><span class="pre">TrackTests.is_combined_edges_dicts_overlapping_and_complete()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#idtrack._track_tests.TrackTests.is_edge_with_same_nts_only_at_backbone_nodes"><code class="docutils literal notranslate"><span class="pre">TrackTests.is_edge_with_same_nts_only_at_backbone_nodes()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#idtrack._track_tests.TrackTests.is_external_conversion_robust"><code class="docutils literal notranslate"><span class="pre">TrackTests.is_external_conversion_robust()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#idtrack._track_tests.TrackTests.is_id_functions_consistent_ensembl"><code class="docutils literal notranslate"><span class="pre">TrackTests.is_id_functions_consistent_ensembl()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#idtrack._track_tests.TrackTests.is_id_functions_consistent_ensembl_2"><code class="docutils literal notranslate"><span class="pre">TrackTests.is_id_functions_consistent_ensembl_2()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#idtrack._track_tests.TrackTests.is_id_functions_consistent_external"><code class="docutils literal notranslate"><span class="pre">TrackTests.is_id_functions_consistent_external()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#idtrack._track_tests.TrackTests.is_node_consistency_robust"><code class="docutils literal notranslate"><span class="pre">TrackTests.is_node_consistency_robust()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#idtrack._track_tests.TrackTests.is_range_functions_robust"><code class="docutils literal notranslate"><span class="pre">TrackTests.is_range_functions_robust()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#idtrack._track_tests.TrackTests.random_dataset_source_generator"><code class="docutils literal notranslate"><span class="pre">TrackTests.random_dataset_source_generator()</span></code></a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#idtrack._db.DB"><code class="docutils literal notranslate"><span class="pre">DB</span></code></a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="contributing.html">Contributor Guide</a></li>
<li class="toctree-l1"><a class="reference internal" href="authors.html">Credits</a></li>
<li class="toctree-l1"><a class="reference internal" href="code_of_conduct.html">Contributor Covenant Code of Conduct</a></li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="index.html">idtrack</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="index.html" class="icon icon-home" aria-label="Home"></a></li>
      <li class="breadcrumb-item active">Reference</li>
      <li class="wy-breadcrumbs-aside">
            <a href="_sources/reference.rst.txt" rel="nofollow"> View page source</a>
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <section id="reference">
<h1>Reference<a class="headerlink" href="#reference" title="Link to this heading"></a></h1>
<dl class="py class" id="module-idtrack._api">
<dt class="sig sig-object py" id="idtrack._api.API">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">API</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">local_repository</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/idtrack/_api.html#API"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#idtrack._api.API" title="Link to this definition"></a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></p>
<p>Application programming interface for simple operations using <code class="docutils literal notranslate"><span class="pre">idtrack</span></code> package.</p>
<p>Class initialization.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>local_repository</strong> (<em>str</em>) – An absolute path in local machine to store downloaded and preprocessed content.</p>
</dd>
</dl>
<dl class="py method">
<dt class="sig sig-object py" id="idtrack._api.API.available_available_external_databases">
<span class="sig-name descname"><span class="pre">available_available_external_databases</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/idtrack/_api.html#API.available_available_external_databases"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#idtrack._api.API.available_available_external_databases" title="Link to this definition"></a></dt>
<dd><p>Show which organisms are available to be used in the scope of this package.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>Set of strings saying which external databases are present in the graph.</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p>set[str]</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="idtrack._api.API.calculate_graph_caches">
<span class="sig-name descname"><span class="pre">calculate_graph_caches</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/idtrack/_api.html#API.calculate_graph_caches"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#idtrack._api.API.calculate_graph_caches" title="Link to this definition"></a></dt>
<dd><p>Calculate cached variables of the graph object using only one method.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="idtrack._api.API.classify_multiple_conversion">
<span class="sig-name descname"><span class="pre">classify_multiple_conversion</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">matchings</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/idtrack/_api.html#API.classify_multiple_conversion"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#idtrack._api.API.classify_multiple_conversion" title="Link to this definition"></a></dt>
<dd><p>Create a dictionary by classifying the results into different bins.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>matchings</strong> (<em>list</em><em>[</em><em>dict</em><em>[</em><em>str</em><em>, </em><em>Any</em><em>]</em><em>]</em>) – The output of <code class="docutils literal notranslate"><span class="pre">convert_identifier_multiple</span></code> method.</p>
</dd>
<dt class="field-even">Raises<span class="colon">:</span></dt>
<dd class="field-even"><p><strong>ValueError</strong> – Unexpected program error</p>
</dd>
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>List of <code class="docutils literal notranslate"><span class="pre">convert_identifier</span></code> method outputs.</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p>dict[str, list[dict]]</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="idtrack._api.API.configure_logger">
<span class="sig-name descname"><span class="pre">configure_logger</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">level</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/idtrack/_api.html#API.configure_logger"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#idtrack._api.API.configure_logger" title="Link to this definition"></a></dt>
<dd><p>Configure logger in a way that shows the logs in a specified format.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="idtrack._api.API.convert_identifier">
<span class="sig-name descname"><span class="pre">convert_identifier</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">identifier</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">from_release</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">to_release</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">final_database</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">prioritize_to_one_filter</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">return_path</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/idtrack/_api.html#API.convert_identifier"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#idtrack._api.API.convert_identifier" title="Link to this definition"></a></dt>
<dd><p>Finds corresponding identifier in specified target using the constructed graph and pathfinder algorithm.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>identifier</strong> (<em>str</em>) – Query ID.</p></li>
<li><p><strong>from_release</strong> (<em>int</em><em> | </em><em>None</em>) – Query ID is from which Ensembl release, if provided.</p></li>
<li><p><strong>to_release</strong> (<em>int</em><em> | </em><em>None</em>) – Ensembl release for target gene set.
Which Ensembl release the user wants to convert the ID into. The default is the latest Ensembl release.</p></li>
<li><p><strong>final_database</strong> (<em>str</em><em> | </em><em>None</em>) – Database for the target gene set.
Which database the user wants to convert the ID into. The default is ‘ensembl_gene’.</p></li>
<li><p><strong>prioritize_to_one_filter</strong> (<em>bool</em>) – Decide to use a series of filters to score the possible paths, and ideally choose
a single target at the end.</p></li>
<li><p><strong>return_path</strong> (<em>bool</em>) – If <code class="docutils literal notranslate"><span class="pre">True</span></code>, returns the path from source to query ID.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><p>A dictionary with following keys</p>
<ul class="simple">
<li><p><code class="docutils literal notranslate"><span class="pre">&quot;target_id&quot;</span></code>: Final IDs after conversion.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">&quot;last_node&quot;</span></code>: The last node in history travel, so it is an Ensembl gene ID.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">&quot;final_database&quot;</span></code>: Final database of the final IDs.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">&quot;graph_id&quot;</span></code>: The corresponding ID in the graph for the query ID.
For example, if the query ID is ‘actb’, this will be ‘ACTB’.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">&quot;query_id&quot;</span></code>: The input query ID.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">&quot;no_corresponding&quot;</span></code>: If <code class="docutils literal notranslate"><span class="pre">True</span></code>, there is no such ID in the graph.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">&quot;no_conversion&quot;</span></code>: If <code class="docutils literal notranslate"><span class="pre">True</span></code>, it is not possible to convert into the target. It is 1-to-0 matching.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">&quot;no_target&quot;</span></code>: If <code class="docutils literal notranslate"><span class="pre">True</span></code>, there is no matching in the described final database, but the history
travel was successful until ‘last_node’. In other words, ‘final_conversion’ failed, no mathing identifier
is found in the target database but there is mathing identifier in Ensembl.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">&quot;the_path&quot;</span></code>: The path from source to query ID. (If <cite>return_path</cite> is set to <code class="docutils literal notranslate"><span class="pre">True</span></code>.)</p></li>
</ul>
</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>dict</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="idtrack._api.API.convert_identifier_multiple">
<span class="sig-name descname"><span class="pre">convert_identifier_multiple</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">identifier_list</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">verbose</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">pbar_prefix</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">''</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/idtrack/_api.html#API.convert_identifier_multiple"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#idtrack._api.API.convert_identifier_multiple" title="Link to this definition"></a></dt>
<dd><p>Basically <code class="docutils literal notranslate"><span class="pre">convert_identifier</span></code> method for multiple conversion procedure with progress bar.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>identifier_list</strong> – List of query IDs to feed the <code class="docutils literal notranslate"><span class="pre">convert_identifier</span></code> method.</p></li>
<li><p><strong>kwargs</strong> – Keyword arguments to pass into <code class="docutils literal notranslate"><span class="pre">convert_identifier</span></code> method.</p></li>
<li><p><strong>verbose</strong> (<em>bool</em>) – If <code class="docutils literal notranslate"><span class="pre">True</span></code>, shows the progress.</p></li>
<li><p><strong>pbar_prefix</strong> (<em>str</em>) – The string to be put before the progress bar.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>List of <code class="docutils literal notranslate"><span class="pre">convert_identifier</span></code> method outputs.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>list[dict]</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="idtrack._api.API.get_database_manager">
<span class="sig-name descname"><span class="pre">get_database_manager</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">organism_name</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">last_ensembl_release</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/idtrack/_api.html#API.get_database_manager"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#idtrack._api.API.get_database_manager" title="Link to this definition"></a></dt>
<dd><p>Instantiate and return a DatabaseManager object for a specified organism and Ensembl release.</p>
<p>This method sets up a DatabaseManager configured to ignore data after a given Ensembl release.
It uses a deep copy of the default database backbone form and stores data in the local repository
specified during API initialization.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>organism_name</strong> (<em>str</em>) – The formal name of the organism (e.g., ‘homo_sapiens’).</p></li>
<li><p><strong>last_ensembl_release</strong> (<em>int</em>) – The most recent Ensembl release to include. Data after this release will be ignored.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>An instance of the DatabaseManager class configured with the provided organism and release settings.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="idtrack._api.API.get_ensembl_organism">
<span class="sig-name descname"><span class="pre">get_ensembl_organism</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">tentative_organism_name</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/idtrack/_api.html#API.get_ensembl_organism"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#idtrack._api.API.get_ensembl_organism" title="Link to this definition"></a></dt>
<dd><p>Make sure the user enters correct organism name and retrieves latest Ensembl release with ease.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>tentative_organism_name</strong> (<em>str</em>) – Organism of interest.
This does not have to be formal name. For example, ‘human’, ‘hsapiens’ as well as formal name
‘homo_sapiens’ is accepted by the program.</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>Formal organism name and associated latest Ensembl release.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>tuple[str, int]</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="idtrack._api.API.infer_identifier_source">
<span class="sig-name descname"><span class="pre">infer_identifier_source</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">id_list</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">mode</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">'assembly_ensembl_release'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">report_only_winner</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/idtrack/_api.html#API.infer_identifier_source"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#idtrack._api.API.infer_identifier_source" title="Link to this definition"></a></dt>
<dd><p>Infer the source of given set of IDs, by comparing which source covers most of the query IDs.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>id_list</strong> (<em>list</em>) – List of query IDs.</p></li>
<li><p><strong>mode</strong> (<em>str</em>) – <ul>
<li><p><code class="docutils literal notranslate"><span class="pre">&quot;complete&quot;</span></code>: Looks for the best match in terms of database, assembly and Ensembl release.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">&quot;ensembl_release&quot;</span></code>: Looks for the best match in terms of Ensembl release only.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">&quot;assembly&quot;</span></code>: Looks for the best match in terms of genome assembly only.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">&quot;assembly_ensembl_release&quot;</span></code>: Looks for the best match in terms of Ensembl release and assembly.</p></li>
</ul>
</p></li>
<li><p><strong>report_only_winner</strong> (<em>bool</em>) – If <code class="docutils literal notranslate"><span class="pre">True</span></code>, return only the winner. If <code class="docutils literal notranslate"><span class="pre">False</span></code>, return each
possible source with corresponding scores.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>Result based on <cite>mode</cite> parameter.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="idtrack._api.API.initialize_graph">
<span class="sig-name descname"><span class="pre">initialize_graph</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">organism_name</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">last_ensembl_release</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">return_test</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/idtrack/_api.html#API.initialize_graph"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#idtrack._api.API.initialize_graph" title="Link to this definition"></a></dt>
<dd><p>Creates a graph and initializes pathfinder class.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>organism_name</strong> (<em>str</em>) – Formal organism name as an output of <code class="docutils literal notranslate"><span class="pre">get_ensembl_organism</span></code> method.</p></li>
<li><p><strong>last_ensembl_release</strong> (<em>int</em>) – Ensembl release of interest. The object will work on only given Ensembl release,
but some methods does not care which form the DatabaseManager is defined to.
The latest possible Ensembl release is the best choice for graph building with no drawbacks.</p></li>
<li><p><strong>return_test</strong> (<em>bool</em>) – If <code class="docutils literal notranslate"><span class="pre">True</span></code>, return the <code class="docutils literal notranslate"><span class="pre">TrackTest</span></code> object instead, which has some functions to test the
pathfinder performance and graph integrity.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="idtrack._api.API.print_binned_conversion">
<span class="sig-name descname"><span class="pre">print_binned_conversion</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">binned_conversion</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/idtrack/_api.html#API.print_binned_conversion"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#idtrack._api.API.print_binned_conversion" title="Link to this definition"></a></dt>
<dd><p>Print the output of <code class="docutils literal notranslate"><span class="pre">classify_multiple_conversion</span></code> method.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>binned_conversion</strong> – The output of <code class="docutils literal notranslate"><span class="pre">classify_multiple_conversion</span></code> method.</p>
</dd>
</dl>
</dd></dl>

</dd></dl>

<dl class="py class" id="module-idtrack._database_manager">
<dt class="sig sig-object py" id="idtrack._database_manager.DatabaseManager">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">DatabaseManager</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">organism</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">form</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">local_repository</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ensembl_release</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ignore_before</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ignore_after</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">store_raw_always</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">genome_assembly</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/idtrack/_database_manager.html#DatabaseManager"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#idtrack._database_manager.DatabaseManager" title="Link to this definition"></a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></p>
<p>Downloads, preprocesses, and stores the necessary source files for the program.</p>
<p>Class initialization.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>organism</strong> (<em>str</em>) – Formal organism name. The output provided by <code class="xref py py-class docutils literal notranslate"><span class="pre">VerifyOrganism</span></code> would be
a perfect choice.</p></li>
<li><p><strong>ensembl_release</strong> (<em>int</em><em> | </em><em>None</em>) – Ensembl release of interest. The object will work on only given Ensembl release, but some
methods does not care which form the DatabaseManager is defined to.
The latest possible Ensembl release is the best choice for graph building with no drawbacks.</p></li>
<li><p><strong>form</strong> (<em>str</em>) – Either ‘gene’, ‘transcript’ or ‘translation’. The object will work on only given form, but some
methods does not care which form the DatabaseManager is defined to.</p></li>
<li><p><strong>local_repository</strong> (<em>str</em>) – An absolute path in local machine to store downloaded and preprocessed content.</p></li>
<li><p><strong>ignore_before</strong> (<em>int</em><em> | </em><em>None</em>) – Ensembl release as the lower limit to include in the downloaded contents. The object will
ignore all Ensembl release lower than this integer.</p></li>
<li><p><strong>ignore_after</strong> (<em>int</em><em> | </em><em>float</em><em> | </em><em>None</em>) – Similar to ‘ignore_before’ but as the upper limit as the name suggest.</p></li>
<li><p><strong>store_raw_always</strong> (<em>bool</em>) – If <code class="docutils literal notranslate"><span class="pre">True</span></code>, the raw MySQL tables will be also saved in the disk.</p></li>
<li><p><strong>genome_assembly</strong> (<em>int</em><em> | </em><em>None</em>) – Genome assembly of interest. The selection should be one of the keys in the
<code class="xref py py-attr docutils literal notranslate"><span class="pre">DB.assembly_mysqlport_priority</span></code> dictionary. The object will work on only given assembly.
The default is the latest genome assembly (also called highest priority assembly).</p></li>
</ul>
</dd>
<dt class="field-even">Raises<span class="colon">:</span></dt>
<dd class="field-even"><ul class="simple">
<li><p><strong>ValueError</strong> – When the input parameters are not in the specified format.</p></li>
<li><p><strong>NotImplementedError</strong> – Currently only supports ‘homo_sapiens’ as the organism name.</p></li>
</ul>
</dd>
</dl>
<dl class="py method">
<dt class="sig sig-object py" id="idtrack._database_manager.DatabaseManager._create_relation_helper">
<span class="sig-name descname"><span class="pre">_create_relation_helper</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">df</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/idtrack/_database_manager.html#DatabaseManager._create_relation_helper"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#idtrack._database_manager.DatabaseManager._create_relation_helper" title="Link to this definition"></a></dt>
<dd><p>Helper method for creating ‘relationship’ dataframes for two methods.</p>
<p>The two methods is <a class="reference internal" href="#idtrack._database_manager.DatabaseManager.create_relation_current" title="idtrack._database_manager.DatabaseManager.create_relation_current"><code class="xref py py-meth docutils literal notranslate"><span class="pre">DatabaseManager.create_relation_current()</span></code></a> and
<a class="reference internal" href="#idtrack._database_manager.DatabaseManager.create_relation_archive" title="idtrack._database_manager.DatabaseManager.create_relation_archive"><code class="xref py py-meth docutils literal notranslate"><span class="pre">DatabaseManager.create_relation_archive()</span></code></a>. The method is not expected to be used by
the user.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>df</strong> (<em>DataFrame</em>) – Output of these two methods mentioned.</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>Dataframe of three columns <cite>gene</cite>, <cite>transcript</cite>, and <cite>translation</cite>. Note that there are some empty
cells in <cite>translation</cite> column as not all transcripts has translations.</p>
</dd>
<dt class="field-odd">Raises<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>ValueError</strong> – If the input dataframe is not with the correct number of columns or column names.</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><em>DataFrame</em></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="idtrack._database_manager.DatabaseManager._determine_usecols_ids">
<em class="property"><span class="pre">static</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">_determine_usecols_ids</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">form</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/idtrack/_database_manager.html#DatabaseManager._determine_usecols_ids"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#idtrack._database_manager.DatabaseManager._determine_usecols_ids" title="Link to this definition"></a></dt>
<dd><p>Helper method to guide which columns are interesting for each form.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>form</strong> (<em>str</em>) – Form of interest</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>Tuple of lists to be used further in the main methods.</p>
</dd>
<dt class="field-odd">Raises<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>ValueError</strong> – If form is not either ‘gene’, ‘transcript’, or ‘translation’.</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p>tuple[list[str], list[str], list[str]]</p>
</dd>
</dl>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="idtrack._database_manager.DatabaseManager.available_releases">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">available_releases</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="pre">list</span><span class="p"><span class="pre">[</span></span><span class="pre">int</span><span class="p"><span class="pre">]</span></span></em><a class="headerlink" href="#idtrack._database_manager.DatabaseManager.available_releases" title="Link to this definition"></a></dt>
<dd><p>List Ensembl releases that can be queried <strong>and</strong> meet the ignore-window constraints.</p>
<p>The list is computed once (via :py:meth`available_releases_versions`) and then cached for the rest of
the session.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>Sorted list of integer release numbers.</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p>list[int]</p>
</dd>
</dl>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="idtrack._database_manager.DatabaseManager.available_releases_no_save">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">available_releases_no_save</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="pre">list</span><span class="p"><span class="pre">[</span></span><span class="pre">int</span><span class="p"><span class="pre">]</span></span></em><a class="headerlink" href="#idtrack._database_manager.DatabaseManager.available_releases_no_save" title="Link to this definition"></a></dt>
<dd><p>List Ensembl releases reachable on the remote MySQL server.</p>
<p>Unlike <a class="reference internal" href="#idtrack._database_manager.DatabaseManager.available_releases" title="idtrack._database_manager.DatabaseManager.available_releases"><code class="xref py py-attr docutils literal notranslate"><span class="pre">available_releases</span></code></a>, this helper never writes the
discovered list to disk; it simply caches the value in memory for the
lifetime of the current <a class="reference internal" href="#idtrack._database_manager.DatabaseManager" title="idtrack._database_manager.DatabaseManager"><code class="xref py py-class docutils literal notranslate"><span class="pre">DatabaseManager</span></code></a> instance.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p><dl class="simple">
<dt>Sorted list of release numbers that</dt><dd><ul class="simple">
<li><p>match the organism and genome assembly tied to the instance,</p></li>
<li><p>fall within the <code class="docutils literal notranslate"><span class="pre">ignore_before</span></code> / <code class="docutils literal notranslate"><span class="pre">ignore_after</span></code> window.</p></li>
</ul>
</dd>
</dl>
</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p>list[int]</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="idtrack._database_manager.DatabaseManager.available_releases_versions">
<span class="sig-name descname"><span class="pre">available_releases_versions</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/idtrack/_database_manager.html#DatabaseManager.available_releases_versions"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#idtrack._database_manager.DatabaseManager.available_releases_versions" title="Link to this definition"></a></dt>
<dd><p>Define available Ensembl releases for the DatabaseManager instance to work on.</p>
<p>Discover which releases exist on the configured MySQL mirror.
It looks on the MySQL server results to determine which Ensembl releases are available. The method does not
return directly the Ensembl releases defined by ‘ignore_after’ and ‘ignore_before’ parameters, and looks on the
server as a verification.</p>
<p>A <code class="docutils literal notranslate"><span class="pre">SHOW</span> <span class="pre">DATABASES</span></code> query is issued, the result is filtered with a
regex of the form <code class="docutils literal notranslate"><span class="pre">^{organism}_core_&lt;release&gt;_.*$</span></code>, and the surviving
release numbers are cleaned and range-checked.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>kwargs</strong> – Passed straight through to <a class="reference internal" href="#idtrack._database_manager.DatabaseManager.get_db" title="idtrack._database_manager.DatabaseManager.get_db"><code class="xref py py-meth docutils literal notranslate"><span class="pre">DatabaseManager.get_db()</span></code></a>.</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>List of integers indicating which Ensembl releases are available.</p>
</dd>
<dt class="field-odd">Raises<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>ValueError</strong> – Unexpected error in regex functions.</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p>list[int]</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="idtrack._database_manager.DatabaseManager.available_tables_mysql">
<span class="sig-name descname"><span class="pre">available_tables_mysql</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/idtrack/_database_manager.html#DatabaseManager.available_tables_mysql"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#idtrack._database_manager.DatabaseManager.available_tables_mysql" title="Link to this definition"></a></dt>
<dd><p>Fetches all the tables in the MySQL database of the instance.</p>
<dl class="field-list simple">
<dt class="field-odd">Raises<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>NotImplementedError</strong> – Not implemented.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="idtrack._database_manager.DatabaseManager.change_assembly">
<span class="sig-name descname"><span class="pre">change_assembly</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">genome_assembly</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">last_possible_ensembl_release</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/idtrack/_database_manager.html#DatabaseManager.change_assembly"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#idtrack._database_manager.DatabaseManager.change_assembly" title="Link to this definition"></a></dt>
<dd><p>Clone the manager while targeting a new genome assembly.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>genome_assembly</strong> (<em>int</em>) – Key from <code class="xref py py-data docutils literal notranslate"><span class="pre">DB.assembly_mysqlport_priority</span></code> (e.g. <code class="docutils literal notranslate"><span class="pre">38</span></code> for <em>GRCh38</em>).</p></li>
<li><p><strong>last_possible_ensembl_release</strong> (<em>bool</em>) – If <code class="docutils literal notranslate"><span class="pre">True</span></code>, the new instance automatically picks the
newest Ensembl release available for <em>genome_assembly</em> instead of
re-using <code class="docutils literal notranslate"><span class="pre">self.ensembl_release</span></code>. Defaults to <code class="docutils literal notranslate"><span class="pre">False</span></code>.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>Independent manager configured for the requested assembly.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference internal" href="#idtrack._database_manager.DatabaseManager" title="idtrack._database_manager.DatabaseManager">DatabaseManager</a></p>
</dd>
</dl>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>The original object remains unchanged; remember to switch your variable reference to the returned instance.</p>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="idtrack._database_manager.DatabaseManager.change_form">
<span class="sig-name descname"><span class="pre">change_form</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">form</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/idtrack/_database_manager.html#DatabaseManager.change_form"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#idtrack._database_manager.DatabaseManager.change_form" title="Link to this definition"></a></dt>
<dd><p>Changes the form of DatabaseManager instance with passing all other variables unchanged.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>form</strong> (<em>str</em>) – New form of interest. Refer to <code class="xref py py-attr docutils literal notranslate"><span class="pre">DatabaseManager.__init__.form</span></code></p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>New instance of DatabaseManager with only ‘form’ is changed.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference internal" href="#idtrack._database_manager.DatabaseManager" title="idtrack._database_manager.DatabaseManager"><em>DatabaseManager</em></a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="idtrack._database_manager.DatabaseManager.change_release">
<span class="sig-name descname"><span class="pre">change_release</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">ensembl_release</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/idtrack/_database_manager.html#DatabaseManager.change_release"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#idtrack._database_manager.DatabaseManager.change_release" title="Link to this definition"></a></dt>
<dd><p>Changes the Ensembl release of DatabaseManager instance with passing all other variables unchanged.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>ensembl_release</strong> (<em>int</em>) – New Ensembl release of interest.
Refer to <code class="xref py py-attr docutils literal notranslate"><span class="pre">DatabaseManager.__init__.ensembl_release</span></code></p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>New instance of DatabaseManager with only ‘ensembl_release’ is changed.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference internal" href="#idtrack._database_manager.DatabaseManager" title="idtrack._database_manager.DatabaseManager"><em>DatabaseManager</em></a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="idtrack._database_manager.DatabaseManager.check_if_change_assembly_works">
<span class="sig-name descname"><span class="pre">check_if_change_assembly_works</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">db_manager</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">target_assembly</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/idtrack/_database_manager.html#DatabaseManager.check_if_change_assembly_works"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#idtrack._database_manager.DatabaseManager.check_if_change_assembly_works" title="Link to this definition"></a></dt>
<dd><p>Test whether <code class="docutils literal notranslate"><span class="pre">change_assembly</span></code> succeeds for <em>target_assembly</em>.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>db_manager</strong> (<a class="reference internal" href="#idtrack._database_manager.DatabaseManager" title="idtrack._database_manager.DatabaseManager"><em>DatabaseManager</em></a>) – A working manager to clone.</p></li>
<li><p><strong>target_assembly</strong> (<em>int</em>) – Assembly code to probe.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><code class="docutils literal notranslate"><span class="pre">True</span></code> if the new manager could be instantiated without
raising <code class="docutils literal notranslate"><span class="pre">ValueError</span></code>; <code class="docutils literal notranslate"><span class="pre">False</span></code> otherwise.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>bool</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="idtrack._database_manager.DatabaseManager.check_version_info">
<span class="sig-name descname"><span class="pre">check_version_info</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/idtrack/_database_manager.html#DatabaseManager.check_version_info"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#idtrack._database_manager.DatabaseManager.check_version_info" title="Link to this definition"></a></dt>
<dd><p>Look at across all Ensembl releases and decide the ‘version_info’ variable to be used in the program.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>Either <cite>without_version</cite> when all Ensembl releases has identifiers without ‘Versions’,
<cite>with_version</cite> when all Ensembl releases has identifiers with ‘Versions’,
or <cite>add_version</cite> when some Ensembl releases has identifiers with ‘Versions’.</p>
</dd>
<dt class="field-even">Raises<span class="colon">:</span></dt>
<dd class="field-even"><p><strong>ValueError</strong> – If the dataframe obtained by <code class="docutils literal notranslate"><span class="pre">self.get_db(&quot;versioninfo&quot;)</span></code> has a problematic column.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>str</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="idtrack._database_manager.DatabaseManager.create_available_databases">
<span class="sig-name descname"><span class="pre">create_available_databases</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/idtrack/_database_manager.html#DatabaseManager.create_available_databases"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#idtrack._database_manager.DatabaseManager.create_available_databases" title="Link to this definition"></a></dt>
<dd><p>Fetches all the databases in the MySQL server of the instance.</p>
<p>Filters out the query <code class="docutils literal notranslate"><span class="pre">SHOW</span> <span class="pre">databases</span></code> based on matching to a specific regex string
<code class="docutils literal notranslate"><span class="pre">f&quot;^{self.organism}_core_[0-9]+_.+$&quot;</span></code></p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>All available ‘MySQL databases’ in the server for given assembly (which is a parameter in the
<cite>DatabaseManager</cite> instance).</p>
</dd>
<dt class="field-even">Raises<span class="colon">:</span></dt>
<dd class="field-even"><p><strong>ValueError</strong> – If the response has unexpected format or length.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><em>Series</em></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="idtrack._database_manager.DatabaseManager.create_database_content">
<span class="sig-name descname"><span class="pre">create_database_content</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">just_download</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/idtrack/_database_manager.html#DatabaseManager.create_database_content"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#idtrack._database_manager.DatabaseManager.create_database_content" title="Link to this definition"></a></dt>
<dd><p>Retrives all External database information from the server to feed the <code class="docutils literal notranslate"><span class="pre">ExternalDatabase</span></code> class.</p>
<p>It is quite costly operation, potentially takes time to be completed. It helps <code class="docutils literal notranslate"><span class="pre">ExternalDatabase</span></code> class to
create the <code class="docutils literal notranslate"><span class="pre">yaml</span></code> file mentioned. It downloads for all assemblies, all Ensembl releases and all forms
available.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>just_download</strong> (<em>bool</em>) – <ul class="simple">
<li><p><code class="docutils literal notranslate"><span class="pre">False</span></code> (default) - concatenate all intermediate results
into one DataFrame and return it.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">True</span></code> - perform the downloads so they are cached on disk
but return an <strong>empty</strong> DataFrame.</p></li>
</ul>
</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>The output of <code class="docutils literal notranslate"><span class="pre">create_external_db</span></code> when <cite>filter_mode</cite> <cite>relevant-database</cite>. Adds <cite>assembly</cite>, <cite>release</cite>,
and <cite>form</cite> columns to the resulting dataframe.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><em>DataFrame</em></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="idtrack._database_manager.DatabaseManager.create_external_all">
<span class="sig-name descname"><span class="pre">create_external_all</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">return_mode</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/idtrack/_database_manager.html#DatabaseManager.create_external_all"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#idtrack._database_manager.DatabaseManager.create_external_all" title="Link to this definition"></a></dt>
<dd><p>Download external databases for all assemblies.</p>
<p>The method considers the ‘priority’ of the assemblies indicated in
<code class="xref py py-attr docutils literal notranslate"><span class="pre">DB.assembly_mysqlport_priority</span></code>. The method is not found in ‘get_db’ so the result is not saved.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>return_mode</strong> (<em>str</em>) – Either one of three choices <cite>all</cite>, <cite>unique</cite>, <cite>duplicated</cite>. There is currently no use case for
<cite>unique</cite>, <cite>duplicated</cite> by the program.</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>Returns the relationship information indicated by <code class="docutils literal notranslate"><span class="pre">ExternalDatabases</span></code> class.
The result is a dataframe with ‘release’, ‘graph_id’, ‘id_db’, ‘name_db’, ‘ensembl_identity’, and
‘xref_identity’, and also ‘assembly’.</p>
</dd>
<dt class="field-odd">Raises<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>ValueError</strong> – If <cite>return_mode</cite> is not among the possible ones.</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><em>DataFrame</em></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="idtrack._database_manager.DatabaseManager.create_external_db">
<span class="sig-name descname"><span class="pre">create_external_db</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">filter_mode</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/idtrack/_database_manager.html#DatabaseManager.create_external_db"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#idtrack._database_manager.DatabaseManager.create_external_db" title="Link to this definition"></a></dt>
<dd><p>Retrieves External identifier relationship. Returns the connections or databases only.</p>
<p>The method actually based on a rationele to the following MySQL query, not exactly but quite close to the
following query.</p>
<div class="highlight-sql notranslate"><div class="highlight"><pre><span></span><span class="k">SELECT</span><span class="w"> </span><span class="k">g</span><span class="p">.</span><span class="n">stable_id</span><span class="p">,</span><span class="w"> </span><span class="n">t</span><span class="p">.</span><span class="n">stable_id</span><span class="p">,</span><span class="w"> </span><span class="n">tr</span><span class="p">.</span><span class="n">stable_id</span><span class="p">,</span><span class="w"> </span><span class="n">x</span><span class="p">.</span><span class="n">dbprimary_acc</span><span class="p">,</span><span class="w"> </span><span class="n">edb</span><span class="p">.</span><span class="n">db_name</span><span class="p">,</span><span class="w"> </span><span class="n">es</span><span class="p">.</span><span class="n">synonym</span><span class="p">,</span><span class="w"> </span><span class="n">ix</span><span class="p">.</span><span class="o">*</span>
<span class="k">FROM</span><span class="w"> </span><span class="n">gene</span><span class="w"> </span><span class="k">g</span>
<span class="k">JOIN</span><span class="w"> </span><span class="n">transcript</span><span class="w"> </span><span class="n">t</span><span class="w"> </span><span class="k">USING</span><span class="w"> </span><span class="p">(</span><span class="n">gene_id</span><span class="p">)</span>
<span class="k">JOIN</span><span class="w"> </span><span class="k">translation</span><span class="w"> </span><span class="n">tr</span><span class="w"> </span><span class="k">USING</span><span class="w"> </span><span class="p">(</span><span class="n">transcript_id</span><span class="p">)</span>
<span class="k">JOIN</span><span class="w"> </span><span class="n">object_xref</span><span class="w"> </span><span class="n">ox</span><span class="w"> </span><span class="k">ON</span><span class="w"> </span><span class="p">(</span><span class="k">g</span><span class="p">.</span><span class="n">gene_id</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">ox</span><span class="p">.</span><span class="n">ensembl_id</span><span class="w"> </span><span class="k">AND</span><span class="w"> </span><span class="n">ox</span><span class="p">.</span><span class="n">ensembl_object_type</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="ss">&quot;Gene&quot;</span><span class="p">)</span>
<span class="k">JOIN</span><span class="w"> </span><span class="n">xref</span><span class="w"> </span><span class="n">x</span><span class="w"> </span><span class="k">ON</span><span class="w"> </span><span class="p">(</span><span class="n">ox</span><span class="p">.</span><span class="n">xref_id</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">x</span><span class="p">.</span><span class="n">xref_id</span><span class="p">)</span>
<span class="k">LEFT</span><span class="w"> </span><span class="k">JOIN</span><span class="w"> </span><span class="n">external_db</span><span class="w"> </span><span class="n">edb</span><span class="w"> </span><span class="k">ON</span><span class="w"> </span><span class="p">(</span><span class="n">x</span><span class="p">.</span><span class="n">external_db_id</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">edb</span><span class="p">.</span><span class="n">external_db_id</span><span class="p">)</span>
<span class="k">LEFT</span><span class="w"> </span><span class="k">JOIN</span><span class="w"> </span><span class="n">identity_xref</span><span class="w"> </span><span class="n">ix</span><span class="w"> </span><span class="k">ON</span><span class="w"> </span><span class="p">(</span><span class="n">ox</span><span class="p">.</span><span class="n">object_xref_id</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">ix</span><span class="p">.</span><span class="n">object_xref_id</span><span class="p">)</span>
<span class="k">LEFT</span><span class="w"> </span><span class="k">JOIN</span><span class="w"> </span><span class="n">external_synonym</span><span class="w"> </span><span class="n">es</span><span class="w"> </span><span class="k">ON</span><span class="w"> </span><span class="p">(</span><span class="n">x</span><span class="p">.</span><span class="n">xref_id</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">es</span><span class="p">.</span><span class="n">xref_id</span><span class="p">)</span>
<span class="k">LIMIT</span><span class="w"> </span><span class="mi">10</span><span class="p">;</span>
</pre></div>
</div>
<p>Instead of <code class="docutils literal notranslate"><span class="pre">FROM</span> <span class="pre">gene</span> <span class="pre">g</span></code>, it is possible to be a bit more specific by replacing the following:</p>
<div class="highlight-sql notranslate"><div class="highlight"><pre><span></span><span class="k">FROM</span><span class="w"> </span><span class="n">coord_system</span><span class="w"> </span><span class="n">cs</span>
<span class="k">JOIN</span><span class="w"> </span><span class="n">seq_region</span><span class="w"> </span><span class="n">sr</span><span class="w"> </span><span class="k">USING</span><span class="w"> </span><span class="p">(</span><span class="n">coord_system_id</span><span class="p">)</span>
<span class="k">JOIN</span><span class="w"> </span><span class="n">gene</span><span class="w"> </span><span class="k">g</span><span class="w"> </span><span class="k">USING</span><span class="w"> </span><span class="p">(</span><span class="n">seq_region_id</span><span class="p">)</span>
</pre></div>
</div>
<p>The MySQL server can be following for the experimentation purpose.</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>mysql<span class="w"> </span>--user<span class="o">=</span>anonymous<span class="w"> </span>--host<span class="o">=</span>ensembldb.ensembl.org<span class="w"> </span>-D<span class="w"> </span>homo_sapiens_core_105_38<span class="w"> </span>-A
<span class="c1"># As written in the following link.</span>
<span class="c1"># https://m.ensembl.org/info/docs/api/core/core_schema.html</span>
</pre></div>
</div>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>filter_mode</strong> (<em>str</em>) – <p>Determine what to return after retrieving the data.
One of the followings: <cite>relevant</cite>, <cite>all</cite>, <cite>database</cite>, <cite>relevant-database</cite>.</p>
<ul class="simple">
<li><p><cite>relevant</cite> and <cite>all</cite>:
Returns the relationship information. ‘all’ returns all relationships while
‘relevant’ returns only those indicated by <code class="docutils literal notranslate"><span class="pre">ExternalDatabases</span></code> class.
The result is a dataframe with ‘release’, ‘graph_id’, ‘id_db’, ‘name_db’, ‘ensembl_identity’, and
‘xref_identity’.</p></li>
<li><p><cite>database</cite> and <cite>relevant-database</cite>:
Returns the a dataframe indicating databases. The resulting dataframe
contain two columns: ‘name_db’ and ‘count’.</p></li>
</ul>
</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>Dataframe of specifified type via <code class="docutils literal notranslate"><span class="pre">filter_mode</span></code>.</p>
</dd>
<dt class="field-odd">Raises<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>ValueError</strong> – If incorrect entry for <code class="docutils literal notranslate"><span class="pre">filter_mode</span></code> parameter, or if there is an inconsistency between
    external <cite>yaml</cite> file and current state of <code class="docutils literal notranslate"><span class="pre">DatabaseManager</span></code>.</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><em>DataFrame</em></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="idtrack._database_manager.DatabaseManager.create_id_history">
<span class="sig-name descname"><span class="pre">create_id_history</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">narrow</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/idtrack/_database_manager.html#DatabaseManager.create_id_history"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#idtrack._database_manager.DatabaseManager.create_id_history" title="Link to this definition"></a></dt>
<dd><p>Retrieves historical releationship between Ensembl IDs of a given form.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>narrow</strong> (<em>bool</em>) – Determine whether a some more information should be added between Ensembl gene IDs. For example,
which genome assembly is used, or when was the connection is established. For usual uses, no need to
set it <code class="docutils literal notranslate"><span class="pre">True</span></code>.</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>Dataframe of following columns; <cite>old_stable_id</cite>, <cite>old_version</cite>, <cite>new_stable_id</cite>, <cite>new_version</cite>, <cite>score</cite>,
<cite>old_release</cite>, <cite>new_release</cite>. Note that there are some empty cells in new and old Ensembl IDs, since
there are ‘retirement’ events or ‘birth’ events of IDs.</p>
</dd>
<dt class="field-odd">Raises<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>ValueError</strong> – If the delimiter <code class="xref py py-attr docutils literal notranslate"><span class="pre">DB.id_ver_delimiter</span></code> is in Ensembl IDs.</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><em>DataFrame</em></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="idtrack._database_manager.DatabaseManager.create_id_history_fixed">
<span class="sig-name descname"><span class="pre">create_id_history_fixed</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">narrow</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">inspect</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/idtrack/_database_manager.html#DatabaseManager.create_id_history_fixed"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#idtrack._database_manager.DatabaseManager.create_id_history_fixed" title="Link to this definition"></a></dt>
<dd><p>Depracated method alternative for ‘create_id_history’, which also corrects the resulting dataframe.</p>
<p>This method created a fixed version of idhistory. It fixes the problems seen in some exceptional cases like
Homo sapiens <code class="docutils literal notranslate"><span class="pre">ENSG00000232423</span></code> at release 105. Raw version gives the order of versions as follows: 1-2, 2-3,
1-2, 2-3 and so on. However, after second connection, version 1 should be already lost, the last active version
should be 3. For this reason, 1-2 connection should be corrected as 3-2. This method does this.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>narrow</strong> (<em>bool</em>) – The same parameter in <code class="xref py py-attr docutils literal notranslate"><span class="pre">DatabaseManager.create_id_history.narrow</span></code>.</p></li>
<li><p><strong>inspect</strong> (<em>bool</em>) – If inspect is <code class="docutils literal notranslate"><span class="pre">True</span></code>, then add the newly created columns as new ones into the existing one.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><cite>old_stable_id</cite>, <cite>old_version</cite>, <cite>new_stable_id</cite>, <cite>new_version</cite>, <cite>score</cite>,
<cite>old_release</cite>, <cite>new_release</cite>.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>Dataframe of following columns</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="idtrack._database_manager.DatabaseManager.create_ids">
<span class="sig-name descname"><span class="pre">create_ids</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">form</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/idtrack/_database_manager.html#DatabaseManager.create_ids"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#idtrack._database_manager.DatabaseManager.create_ids" title="Link to this definition"></a></dt>
<dd><p>Retrieves the Ensembl IDs.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>form</strong> (<em>str</em>) – Form of interest, either ‘gene’, ‘transcript’, or ‘translation’.</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><cite>gene_id</cite>, <cite>gene_stable_id</cite>, and <cite>gene_version</cite>. The ‘ID’ and ‘Version’ need to
be concatanated afterwards.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>Dataframe of three columns</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="idtrack._database_manager.DatabaseManager.create_relation_archive">
<span class="sig-name descname"><span class="pre">create_relation_archive</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/idtrack/_database_manager.html#DatabaseManager.create_relation_archive"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#idtrack._database_manager.DatabaseManager.create_relation_archive" title="Link to this definition"></a></dt>
<dd><p>Retrieves the relationship between different forms of Ensembl IDs for all Ensembl releases.</p>
<p>It is not recommended as there are some missing rows. Instead use
<a class="reference internal" href="#idtrack._database_manager.DatabaseManager.create_relation_current" title="idtrack._database_manager.DatabaseManager.create_relation_current"><code class="xref py py-meth docutils literal notranslate"><span class="pre">DatabaseManager.create_relation_current()</span></code></a> for all Ensembl releases separately,
and concatanate the resulting data frames.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p><cite>gene</cite>, <cite>transcript</cite>, and <cite>translation</cite>. Note that there are some empty cells
in <cite>translation</cite> column as not all transcripts has translations.</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p>Dataframe of three columns</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="idtrack._database_manager.DatabaseManager.create_relation_current">
<span class="sig-name descname"><span class="pre">create_relation_current</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/idtrack/_database_manager.html#DatabaseManager.create_relation_current"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#idtrack._database_manager.DatabaseManager.create_relation_current" title="Link to this definition"></a></dt>
<dd><p>Retrieves the relationship between different forms of Ensembl IDs.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p><cite>gene</cite>, <cite>transcript</cite>, and <cite>translation</cite>. Note that there are some empty cells
in <cite>translation</cite> column as not all transcripts has translations.</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p>Dataframe of three columns</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="idtrack._database_manager.DatabaseManager.create_release_id">
<span class="sig-name descname"><span class="pre">create_release_id</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/idtrack/_database_manager.html#DatabaseManager.create_release_id"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#idtrack._database_manager.DatabaseManager.create_release_id" title="Link to this definition"></a></dt>
<dd><p>Retrieves the Ensembl IDs and applies <cite>version_fix</cite> method to refine it.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>A dataframe with <code class="docutils literal notranslate"><span class="pre">f'{form}_stable_id'</span></code> and <code class="docutils literal notranslate"><span class="pre">f'{form}_version'</span></code>.</p>
</dd>
<dt class="field-even">Raises<span class="colon">:</span></dt>
<dd class="field-even"><p><strong>ValueError</strong> – If the delimiter <code class="xref py py-attr docutils literal notranslate"><span class="pre">DB.id_ver_delimiter</span></code> is in Ensembl IDs, or if the stable IDs are
    not unique.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><em>DataFrame</em></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="idtrack._database_manager.DatabaseManager.create_version_info">
<span class="sig-name descname"><span class="pre">create_version_info</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/idtrack/_database_manager.html#DatabaseManager.create_version_info"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#idtrack._database_manager.DatabaseManager.create_version_info" title="Link to this definition"></a></dt>
<dd><p>Check whether all Ensembl release has Ensembl IDs with versions or without versions.</p>
<p>Some organisms such as S. cerevisiae has Ensembl identifiers that has no ‘Version’, but only ‘ID’. The method
looks whether this is the case for all the Ensembl releases for a given organism.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>A dataframe with two columns as <cite>ensembl_release</cite>, <cite>version_info</cite>.</p>
</dd>
<dt class="field-even">Raises<span class="colon">:</span></dt>
<dd class="field-even"><p><strong>NotImplementedError</strong> – If some Ensembl identifiers without versions, some are not.”</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><em>DataFrame</em></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="idtrack._database_manager.DatabaseManager.download_table">
<span class="sig-name descname"><span class="pre">download_table</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">table_key</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">usecols</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/idtrack/_database_manager.html#DatabaseManager.download_table"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#idtrack._database_manager.DatabaseManager.download_table" title="Link to this definition"></a></dt>
<dd><p>Downloads the raw table from MySQL server and extracts requested columns.</p>
<p>The method is not generally expected to be used by the user. User is expected to use ‘get_table’ instead.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>table_key</strong> (<em>str</em>) – The raw MySQL database (table) of interest. For example <cite>mapping_session</cite>, <cite>xref</cite>, <cite>gene</cite>.</p></li>
<li><p><strong>usecols</strong> (<em>list</em><em> | </em><em>None</em>) – The column of interest in the specified table. Short for ‘use columns’.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>Raw table as a pandas DataFrame.</p>
</dd>
<dt class="field-odd">Raises<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>ValueError</strong> – If there is no column in the table.</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><em>DataFrame</em></p>
</dd>
</dl>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="idtrack._database_manager.DatabaseManager.external_inst">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">external_inst</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><a class="reference internal" href="#idtrack._external_databases.ExternalDatabases" title="idtrack._external_databases.ExternalDatabases"><span class="pre">ExternalDatabases</span></a></em><a class="headerlink" href="#idtrack._database_manager.DatabaseManager.external_inst" title="Link to this definition"></a></dt>
<dd><p>Create an instance of <code class="xref py py-class docutils literal notranslate"><span class="pre">ExternalDatabases</span></code> and set as a property.</p>
<p>The parameters of the DatabaseManager instance will be directly passed in the creation of the
<cite>ExternalDatabases</cite> instance so they will be consistent with each other.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>An ExternalDatabases instance.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="idtrack._database_manager.DatabaseManager.file_name">
<span class="sig-name descname"><span class="pre">file_name</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">df_type</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">*</span></span><span class="n"><span class="pre">args</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ensembl_release</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/idtrack/_database_manager.html#DatabaseManager.file_name"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#idtrack._database_manager.DatabaseManager.file_name" title="Link to this definition"></a></dt>
<dd><p>Determine file name for reading/writing into h5 file based on dataframe type.</p>
<p>The method is not expected to be used by the user.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>df_type</strong> (<em>str</em>) – Either <cite>processed</cite>, <cite>mysql</cite>, <cite>common</cite>.</p></li>
<li><p><strong>args</strong> – Arguments to pass into the functions, which are specific to each <code class="docutils literal notranslate"><span class="pre">df_type</span></code>.</p></li>
<li><p><strong>ensembl_release</strong> (<em>int</em><em> | </em><em>None</em>) – Associated Ensembl release of the file of interest.</p></li>
<li><p><strong>kwargs</strong> – Keyword arguments to pass into the functions, which are specific to each <code class="docutils literal notranslate"><span class="pre">df_type</span></code>.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>The file name for h5 file and the key to be used inside.</p>
</dd>
<dt class="field-odd">Raises<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>ValueError</strong> – If the parameter <cite>df_type</cite> is not in specified format.</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p>tuple[str, str]</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="idtrack._database_manager.DatabaseManager.get_db">
<span class="sig-name descname"><span class="pre">get_db</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">df_indicator</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">create_even_if_exist</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">save_after_calculation</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">overwrite_even_if_exist</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/idtrack/_database_manager.html#DatabaseManager.get_db"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#idtrack._database_manager.DatabaseManager.get_db" title="Link to this definition"></a></dt>
<dd><p>For saving, exporting, and naming convention. Main method to retrieve the data sources.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>df_indicator</strong> (<em>str</em>) – A string indicating which dataframe the user is requested to access.
It should contain only one or zero ‘_’ character.</p></li>
<li><p><strong>create_even_if_exist</strong> (<em>bool</em>) – If <code class="docutils literal notranslate"><span class="pre">True</span></code>, independent of the status in the disk, the table will be downloaded.</p></li>
<li><p><strong>save_after_calculation</strong> (<em>bool</em>) – If <code class="docutils literal notranslate"><span class="pre">True</span></code>, the downloaded table will be stored in the disk, under a <cite>h5</cite> file at
the local directory specified in init method.</p></li>
<li><p><strong>overwrite_even_if_exist</strong> (<em>bool</em>) – If <code class="docutils literal notranslate"><span class="pre">True</span></code>, regardless of whether it is already saved in the disk, the program
re-saves removing the previous table with the same name.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>The data of interest depending on <cite>df_indicator</cite> parameter.</p>
</dd>
<dt class="field-odd">Raises<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>ValueError</strong> – If <cite>df_indicator</cite> parameter is not in specified format.</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><em>DataFrame</em> | <em>Series</em></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="idtrack._database_manager.DatabaseManager.get_release_date">
<span class="sig-name descname"><span class="pre">get_release_date</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/idtrack/_database_manager.html#DatabaseManager.get_release_date"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#idtrack._database_manager.DatabaseManager.get_release_date" title="Link to this definition"></a></dt>
<dd><p>Get the associated date of release for each Ensembl release.</p>
<dl class="field-list simple">
<dt class="field-odd">Raises<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>NotImplementedError</strong> – Note implemented.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="idtrack._database_manager.DatabaseManager.get_table">
<span class="sig-name descname"><span class="pre">get_table</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">table_key</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">usecols</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">create_even_if_exist</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">save_after_calculation</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">overwrite_even_if_exist</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/idtrack/_database_manager.html#DatabaseManager.get_table"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#idtrack._database_manager.DatabaseManager.get_table" title="Link to this definition"></a></dt>
<dd><p>Master method for MySQL processes. Downloads, stores, or retrieves the database of raw MySQL results.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>table_key</strong> (<em>str</em>) – The raw MySQL database (table) of interest. For example <cite>mapping_session</cite>, <cite>xref</cite>, <cite>gene</cite>.</p></li>
<li><p><strong>usecols</strong> (<em>list</em><em> | </em><em>None</em>) – The column of interest in the specified table. Short for ‘use columns’.</p></li>
<li><p><strong>create_even_if_exist</strong> (<em>bool</em>) – If <code class="docutils literal notranslate"><span class="pre">True</span></code>, independent of the status in the disk, the table will be downloaded.</p></li>
<li><p><strong>save_after_calculation</strong> (<em>bool</em>) – If <code class="docutils literal notranslate"><span class="pre">True</span></code>, the downloaded table will be stored in the disk, under a <cite>h5</cite> file at
the local directory specified in init method.</p></li>
<li><p><strong>overwrite_even_if_exist</strong> (<em>bool</em>) – If <code class="docutils literal notranslate"><span class="pre">True</span></code>, regardless of whether it is already saved in the disk, the program
re-saves removing the previous table with the same name.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>Raw table as a pandas DataFrame.</p>
</dd>
<dt class="field-odd">Raises<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>ValueError</strong> – If the ‘usecols’ is not specified correctly.</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><em>DataFrame</em></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="idtrack._database_manager.DatabaseManager.id_ver_from_df">
<span class="sig-name descname"><span class="pre">id_ver_from_df</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">dbm_the_ids</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/idtrack/_database_manager.html#DatabaseManager.id_ver_from_df"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#idtrack._database_manager.DatabaseManager.id_ver_from_df" title="Link to this definition"></a></dt>
<dd><p>Creates node names given the dataframe of IDs, which has different columns for ‘ID’ and ‘Version’.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>dbm_the_ids</strong> (<em>DataFrame</em>) – Dataframe of IDs, typically the output of <code class="docutils literal notranslate"><span class="pre">db_manager.get_db(&quot;ids&quot;)</span></code>.</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>Node name as “ID.Version” if there is ‘Version’, else only “ID”.</p>
</dd>
<dt class="field-odd">Raises<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>ValueError</strong> – If column names of the <cite>dbm_the_ids</cite> is not as expected.</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p>list[str]</p>
</dd>
</dl>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="idtrack._database_manager.DatabaseManager.mysql_database">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">mysql_database</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="pre">str</span></em><a class="headerlink" href="#idtrack._database_manager.DatabaseManager.mysql_database" title="Link to this definition"></a></dt>
<dd><p>The program has to choose a ‘MySQL database’ based on ensembl release and organism name in the MYSQL server.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>The string specifying the ‘MySQL database’ in the format determined by the regex pattern.</p>
</dd>
<dt class="field-even">Raises<span class="colon">:</span></dt>
<dd class="field-even"><p><strong>ValueError</strong> – If there is more than one such match.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="idtrack._database_manager.DatabaseManager.node_dict_maker">
<em class="property"><span class="pre">static</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">node_dict_maker</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">id_entry</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">version_entry</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/idtrack/_database_manager.html#DatabaseManager.node_dict_maker"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#idtrack._database_manager.DatabaseManager.node_dict_maker" title="Link to this definition"></a></dt>
<dd><p>Create a dict for ID and Version.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>id_entry</strong> (<em>str</em>) – For example, the first part of (<cite>ENSG00000000001</cite>) in <cite>ENSG00000000001.1</cite> before delimiter.</p></li>
<li><p><strong>version_entry</strong> (<em>Any</em>) – For example, the second part of (<cite>1</cite>) in <cite>ENSG00000000001.1</cite> before delimiter.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>Dictionary in the format of <code class="docutils literal notranslate"><span class="pre">{&quot;ID&quot;:</span> <span class="pre">id_entry,</span> <span class="pre">&quot;Version&quot;:</span> <span class="pre">version_entry}</span></code></p>
</dd>
<dt class="field-odd">Raises<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>ValueError</strong> – If ‘Version’ is not floating number (cannot be converted into integer).</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p>dict[str, <em>Any</em>]</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="idtrack._database_manager.DatabaseManager.node_name_maker">
<em class="property"><span class="pre">static</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">node_name_maker</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">node_dict</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/idtrack/_database_manager.html#DatabaseManager.node_name_maker"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#idtrack._database_manager.DatabaseManager.node_name_maker" title="Link to this definition"></a></dt>
<dd><p>This function creates ID-Version.</p>
<p>If the Version information is not there, it only uses ID, which is necessary for some organisms which
does not have versioned IDs.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>node_dict</strong> (<em>dict</em><em>[</em><em>str</em><em>, </em><em>Any</em><em>]</em>) – The output of <a class="reference internal" href="#idtrack._database_manager.DatabaseManager.node_dict_maker" title="idtrack._database_manager.DatabaseManager.node_dict_maker"><code class="xref py py-meth docutils literal notranslate"><span class="pre">DatabaseManager.node_dict_maker()</span></code></a>.</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>Node name as “ID.Version” if there is ‘Version’, else only “ID”.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>str</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="idtrack._database_manager.DatabaseManager.tables_in_disk">
<span class="sig-name descname"><span class="pre">tables_in_disk</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/idtrack/_database_manager.html#DatabaseManager.tables_in_disk"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#idtrack._database_manager.DatabaseManager.tables_in_disk" title="Link to this definition"></a></dt>
<dd><p>Retrieves the keys in the h5 file, which is associated with the DatabaseManager instance.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>List of keys (also called hierarchy) in the h5 file.</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p>list[str]</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="idtrack._database_manager.DatabaseManager.version_fix">
<span class="sig-name descname"><span class="pre">version_fix</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">df</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">version_str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">version_info</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/idtrack/_database_manager.html#DatabaseManager.version_fix"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#idtrack._database_manager.DatabaseManager.version_fix" title="Link to this definition"></a></dt>
<dd><p>Depending the version information of the organism, uniformize the identifiers.</p>
<p>The method is not expected to be used by the user.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>df</strong> (<em>DataFrame</em>) – Input dataframe to fix the version information.</p></li>
<li><p><strong>version_str</strong> (<em>str</em>) – Which column contain the ‘Version’ information.</p></li>
<li><p><strong>version_info</strong> (<em>str</em><em> | </em><em>None</em>) – The program parameter obtained somehow by <code class="docutils literal notranslate"><span class="pre">check_version_info</span></code> method.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>Modified dataframe in terms of version information.</p>
</dd>
<dt class="field-odd">Raises<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>ValueError</strong> – If <cite>version_info</cite> is not one of the output of <cite>check_version_info</cite> method.</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><em>DataFrame</em></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="idtrack._database_manager.DatabaseManager.version_fix_incomplete">
<span class="sig-name descname"><span class="pre">version_fix_incomplete</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">df_fx</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">id_col_fx</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ver_col_fx</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/idtrack/_database_manager.html#DatabaseManager.version_fix_incomplete"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#idtrack._database_manager.DatabaseManager.version_fix_incomplete" title="Link to this definition"></a></dt>
<dd><p>The same logic with <code class="docutils literal notranslate"><span class="pre">version_fix</span></code> method, but do different process if associated ‘ID’ is <code class="docutils literal notranslate"><span class="pre">np.nan</span></code>.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>df_fx</strong> (<em>DataFrame</em>) – Input dataframe to fix the version information.</p></li>
<li><p><strong>id_col_fx</strong> (<em>str</em>) – Which column contain the ‘ID’ information.</p></li>
<li><p><strong>ver_col_fx</strong> (<em>str</em>) – Which column contain the ‘Version’ information.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>Modified dataframe in terms of version information.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><em>DataFrame</em></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="idtrack._database_manager.DatabaseManager.version_uniformize">
<span class="sig-name descname"><span class="pre">version_uniformize</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">df</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">version_str</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/idtrack/_database_manager.html#DatabaseManager.version_uniformize"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#idtrack._database_manager.DatabaseManager.version_uniformize" title="Link to this definition"></a></dt>
<dd><p>Final operation for <a class="reference internal" href="#idtrack._database_manager.DatabaseManager.create_ids" title="idtrack._database_manager.DatabaseManager.create_ids"><code class="xref py py-meth docutils literal notranslate"><span class="pre">DatabaseManager.create_ids()</span></code></a>.</p>
<p>The method is not expected to be used by the user.</p>
<p>Make the ‘Version’ column, integer if there is ‘Version’ information, else, make put <code class="docutils literal notranslate"><span class="pre">np.nan</span></code> instead. The
operation is in line with <a class="reference internal" href="#idtrack._database_manager.DatabaseManager.create_version_info" title="idtrack._database_manager.DatabaseManager.create_version_info"><code class="xref py py-meth docutils literal notranslate"><span class="pre">DatabaseManager.create_version_info()</span></code></a> method.
Note that <code class="docutils literal notranslate"><span class="pre">np.nan</span></code> values will be used by <cite>node_name_maker</cite> and <cite>node_name_maker</cite> methods.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>df</strong> (<em>DataFrame</em>) – Input dataframe, see the <cite>create_ids</cite> method for more detail.</p></li>
<li><p><strong>version_str</strong> (<em>str</em>) – Which column contain the ‘Version’ information</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>Corrected version of the dataframe in terms of ‘Version’ information.</p>
</dd>
<dt class="field-odd">Raises<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>NotImplementedError</strong> – If some Ensembl identifiers without versions, some are not.”</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><em>DataFrame</em></p>
</dd>
</dl>
</dd></dl>

</dd></dl>

<dl class="py class" id="module-idtrack._external_databases">
<dt class="sig sig-object py" id="idtrack._external_databases.ExternalDatabases">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">ExternalDatabases</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">organism</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ensembl_release</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">form</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">local_repository</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">genome_assembly</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/idtrack/_external_databases.html#ExternalDatabases"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#idtrack._external_databases.ExternalDatabases" title="Link to this definition"></a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></p>
<p>Creates and manages external database <cite>yaml</cite> file.</p>
<p>Class initialization.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>organism</strong> (<em>str</em>) – Refer to <code class="xref py py-attr docutils literal notranslate"><span class="pre">DatabaseManager.__init__.organism</span></code>.</p></li>
<li><p><strong>ensembl_release</strong> (<em>int</em>) – Refer to <code class="xref py py-attr docutils literal notranslate"><span class="pre">DatabaseManager.__init__.ensembl_release</span></code>.</p></li>
<li><p><strong>form</strong> (<em>str</em>) – Refer to <code class="xref py py-attr docutils literal notranslate"><span class="pre">DatabaseManager.__init__.form</span></code>.</p></li>
<li><p><strong>local_repository</strong> (<em>str</em>) – Refer to <code class="xref py py-attr docutils literal notranslate"><span class="pre">DatabaseManager.__init__.local_repository</span></code>.</p></li>
<li><p><strong>genome_assembly</strong> (<em>int</em>) – Refer to <code class="xref py py-attr docutils literal notranslate"><span class="pre">DatabaseManager.__init__.genome_assembly</span></code>.</p></li>
</ul>
</dd>
</dl>
<dl class="py method">
<dt class="sig sig-object py" id="idtrack._external_databases.ExternalDatabases.create_template_yaml">
<span class="sig-name descname"><span class="pre">create_template_yaml</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">df</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/idtrack/_external_databases.html#ExternalDatabases.create_template_yaml"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#idtrack._external_databases.ExternalDatabases.create_template_yaml" title="Link to this definition"></a></dt>
<dd><p>Creates the templete <cite>yaml</cite> file, indicating each external database possible.</p>
<p>Example output file will start as following, of course, with much more database under <cite>gene</cite>.</p>
<div class="highlight-yaml notranslate"><div class="highlight"><pre><span></span><span class="nt">homo_sapiens</span><span class="p">:</span>
<span class="w">    </span><span class="nt">gene</span><span class="p">:</span>
<span class="w">        </span><span class="nt">ArrayExpress</span><span class="p">:</span>
<span class="w">            </span><span class="nt">Assembly</span><span class="p">:</span>
<span class="w">                </span><span class="s">&quot;37&quot;</span><span class="p p-Indicator">:</span>
<span class="w">                    </span><span class="nt">Ensembl release</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">79,80,81,82,83,84,85,86,87,88,89</span>
<span class="w">                    </span><span class="nt">Include</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">false</span>
<span class="w">                </span><span class="s">&quot;38&quot;</span><span class="p p-Indicator">:</span>
<span class="w">                    </span><span class="nt">Ensembl release</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">79,80,81,82,83,84,85,86,87,88,89</span>
<span class="w">                    </span><span class="nt">Include</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">false</span>
<span class="w">            </span><span class="nt">Database Index</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">0</span>
<span class="w">            </span><span class="nt">Potential Synonymous</span><span class="p">:</span><span class="w"> </span><span class="s">&quot;&quot;</span>
<span class="w">        </span><span class="nt">Clone-based (Ensembl)</span><span class="p">:</span>
<span class="w">            </span><span class="nt">Assembly</span><span class="p">:</span>
<span class="w">                </span><span class="s">&quot;37&quot;</span><span class="p p-Indicator">:</span>
<span class="w">                    </span><span class="nt">Ensembl release</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">79,80,81,82,83,84,85</span>
<span class="w">                    </span><span class="nt">Include</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">false</span>
<span class="w">                </span><span class="s">&quot;38&quot;</span><span class="p p-Indicator">:</span>
<span class="w">                    </span><span class="nt">Ensembl release</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">79,80,81,82,83,84,85</span>
<span class="w">                    </span><span class="nt">Include</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">false</span>
<span class="w">            </span><span class="nt">Database Index</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">5</span>
<span class="w">            </span><span class="nt">Potential Synonymous</span><span class="p">:</span><span class="w"> </span><span class="s">&quot;&quot;</span>
</pre></div>
</div>
<p>Note that the <cite>Potential Synonymous</cite> is now all empty. In the following versions, it is aimed to integrate
a feature that prevent to heve synonymous databases in the list. Likewise, <cite>Database Index</cite> has now no use case,
in the program.</p>
<p>It is important to follow the final warning raised by the method. “Please edit the file based on
requested external databases and add ‘_modified’ to the file name.”. The editing should be done by converting
<cite>Include</cite> sections from <cite>false</cite> to <cite>true</cite>. It is recommended to make the change for each assembly for a given
database.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>df</strong> (<em>DataFrame</em>) – The output of following operation <code class="docutils literal notranslate"><span class="pre">db_manager.create_database_content()</span></code>.</p>
</dd>
<dt class="field-even">Raises<span class="colon">:</span></dt>
<dd class="field-even"><p><strong>ValueError</strong> – Unexpected error.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="idtrack._external_databases.ExternalDatabases.file_name_modified_yaml">
<span class="sig-name descname"><span class="pre">file_name_modified_yaml</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">mode</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/idtrack/_external_databases.html#ExternalDatabases.file_name_modified_yaml"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#idtrack._external_databases.ExternalDatabases.file_name_modified_yaml" title="Link to this definition"></a></dt>
<dd><p>File name creator, intended to be used only <a class="reference internal" href="#idtrack._external_databases.ExternalDatabases.load_modified_yaml" title="idtrack._external_databases.ExternalDatabases.load_modified_yaml"><code class="xref py py-meth docutils literal notranslate"><span class="pre">ExternalDatabases.load_modified_yaml()</span></code></a>.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>mode</strong> (<em>str</em>) – Decide to retrieve whether the <cite>default</cite> yaml file for the given organism
or <cite>configured</cite> yaml file, which is created and modified by the user.</p>
</dd>
<dt class="field-even">Raises<span class="colon">:</span></dt>
<dd class="field-even"><p><strong>ValueError</strong> – If ‘mode’ is not either ‘configured’ or ‘default’.</p>
</dd>
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>File name as an absolute path.</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p>str</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="idtrack._external_databases.ExternalDatabases.file_name_template_yaml">
<span class="sig-name descname"><span class="pre">file_name_template_yaml</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/idtrack/_external_databases.html#ExternalDatabases.file_name_template_yaml"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#idtrack._external_databases.ExternalDatabases.file_name_template_yaml" title="Link to this definition"></a></dt>
<dd><p>File name creator, intended to be used only <a class="reference internal" href="#idtrack._external_databases.ExternalDatabases.create_template_yaml" title="idtrack._external_databases.ExternalDatabases.create_template_yaml"><code class="xref py py-meth docutils literal notranslate"><span class="pre">ExternalDatabases.create_template_yaml()</span></code></a>.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>Absolute path for the template yaml file.</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p>str</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="idtrack._external_databases.ExternalDatabases.give_list_for_case">
<span class="sig-name descname"><span class="pre">give_list_for_case</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">give_type</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/idtrack/_external_databases.html#ExternalDatabases.give_list_for_case"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#idtrack._external_databases.ExternalDatabases.give_list_for_case" title="Link to this definition"></a></dt>
<dd><p>Retrieve some simple information from <cite>yaml</cite> file.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>give_type</strong> (<em>str</em>) – <p>Either <code class="docutils literal notranslate"><span class="pre">'db'</span></code> or <code class="docutils literal notranslate"><span class="pre">'assembly'</span></code>.</p>
<ul class="simple">
<li><p><cite>db</cite>: the method gives associated external database names in <cite>yaml</cite> file for given
organism, assembly, Ensembl release and form.</p></li>
<li><p><cite>assembly</cite>: the method gives possible Ensembl assemblies that at least one external database is
defined in <code class="docutils literal notranslate"><span class="pre">yaml</span></code> config file.</p></li>
</ul>
</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>List of strings of databases if ‘db’ is <cite>give_type</cite>, or list of integers if ‘assembly’ is <cite>give_type</cite>.</p>
</dd>
<dt class="field-odd">Raises<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>ValueError</strong> – If <cite>give_type</cite> is not in specified format.</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p>list</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="idtrack._external_databases.ExternalDatabases.load_modified_yaml">
<span class="sig-name descname"><span class="pre">load_modified_yaml</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/idtrack/_external_databases.html#ExternalDatabases.load_modified_yaml"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#idtrack._external_databases.ExternalDatabases.load_modified_yaml" title="Link to this definition"></a></dt>
<dd><p>Reads the modified <cite>yaml</cite> file if created by the user, else read the default for the given organism.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>Loaded yaml file as the dictionary</p>
</dd>
<dt class="field-even">Raises<span class="colon">:</span></dt>
<dd class="field-even"><p><strong>FileNotFoundError</strong> – When neither default nor configured <cite>yaml</cite> file exist.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>dict</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="idtrack._external_databases.ExternalDatabases.validate_yaml_file_up_to_date">
<span class="sig-name descname"><span class="pre">validate_yaml_file_up_to_date</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">read_yaml_file</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/idtrack/_external_databases.html#ExternalDatabases.validate_yaml_file_up_to_date"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#idtrack._external_databases.ExternalDatabases.validate_yaml_file_up_to_date" title="Link to this definition"></a></dt>
<dd><p>Validates that the current Ensembl release is present in the YAML configuration file.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>read_yaml_file</strong> (<em>dict</em>) – read_yaml_file (dict): YAML file loaded as a dictionary,
typically from <a class="reference internal" href="#idtrack._external_databases.ExternalDatabases.load_modified_yaml" title="idtrack._external_databases.ExternalDatabases.load_modified_yaml"><code class="xref py py-meth docutils literal notranslate"><span class="pre">ExternalDatabases.load_modified_yaml()</span></code></a>.</p>
</dd>
<dt class="field-even">Raises<span class="colon">:</span></dt>
<dd class="field-even"><p><strong>ValueError</strong> – If the current Ensembl release is not found in the YAML configuration.</p>
</dd>
</dl>
</dd></dl>

</dd></dl>

<dl class="py class" id="module-idtrack._the_graph">
<dt class="sig sig-object py" id="idtrack._the_graph.TheGraph">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">TheGraph</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="o"><span class="pre">*</span></span><span class="n"><span class="pre">args</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/idtrack/_the_graph.html#TheGraph"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#idtrack._the_graph.TheGraph" title="Link to this definition"></a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">MultiDiGraph</span></code></p>
<p>Graph object containing all the information about the relationships between bio-IDs.</p>
<p>It is a subclass of <code class="xref py py-class docutils literal notranslate"><span class="pre">networkx.MultiDiGraph</span></code>, with some additional methods to aid the pathfinder
algorithm in <code class="xref py py-class docutils literal notranslate"><span class="pre">Track</span></code>. The object is constructed by <code class="xref py py-class docutils literal notranslate"><span class="pre">GraphMaker</span></code>.</p>
<p>Class initialization.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>args</strong> – Arguments to be passed to <code class="xref py py-class docutils literal notranslate"><span class="pre">networkx.MultiDiGraph</span></code>.</p></li>
<li><p><strong>kwargs</strong> – Keyword arguments to be passed to <code class="xref py py-class docutils literal notranslate"><span class="pre">networkx.MultiDiGraph</span></code>.</p></li>
</ul>
</dd>
</dl>
<dl class="py method">
<dt class="sig sig-object py" id="idtrack._the_graph.TheGraph._combined_edges">
<em class="property"><span class="pre">static</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">_combined_edges</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">node_list</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">the_graph</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/idtrack/_the_graph.html#TheGraph._combined_edges"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#idtrack._the_graph.TheGraph._combined_edges" title="Link to this definition"></a></dt>
<dd><p>Combines the database, assembly, ensembl release information from all edges from a given node.</p>
<p>The method fills the dictionary with nodes provided as a parameter. The edge data coming from the same node
types (the edge between the same node type can exist only between backbone nodes as tested in
<code class="xref py py-meth docutils literal notranslate"><span class="pre">TrackTest.is_edge_with_same_nts_only_at_backbone_nodes()</span></code>) are excluded.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>node_list</strong> (<em>NodeView</em><em> | </em><em>list</em>) – Node names to be calculated.</p></li>
<li><p><strong>the_graph</strong> (<em>MultiDiGraph</em>) – The graph to be used for the calculation. It is generally <cite>self</cite> or <cite>self.rev</cite> depending on
how it has been wanted to calculate.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>A dictionary with following format
<code class="docutils literal notranslate"><span class="pre">{node_name:</span> <span class="pre">{database_name:</span> <span class="pre">{assembly:</span> <span class="pre">{ensembl</span> <span class="pre">release</span> <span class="pre">set}}}}</span></code>.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>dict</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="idtrack._the_graph.TheGraph._combined_edges_genes_helper">
<em class="property"><span class="pre">static</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">_combined_edges_genes_helper</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">the_result</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/idtrack/_the_graph.html#TheGraph._combined_edges_genes_helper"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#idtrack._the_graph.TheGraph._combined_edges_genes_helper" title="Link to this definition"></a></dt>
<dd><p>Helper method is called for two methods only.</p>
<p>The method is called by <a class="reference internal" href="#idtrack._the_graph.TheGraph.combined_edges_assembly_specific_genes" title="idtrack._the_graph.TheGraph.combined_edges_assembly_specific_genes"><code class="xref py py-meth docutils literal notranslate"><span class="pre">TheGraph.combined_edges_assembly_specific_genes()</span></code></a> and
<a class="reference internal" href="#idtrack._the_graph.TheGraph.combined_edges_genes" title="idtrack._the_graph.TheGraph.combined_edges_genes"><code class="xref py py-meth docutils literal notranslate"><span class="pre">TheGraph.combined_edges_genes()</span></code></a> only. The aim of the method is to merge all the data
coming from multiple external databases.</p>
<p>Note that the main reason of having a separate methods for these two is basically due to the fact that <cite>gene</cite>
nodes has no outgoing edges except to another <cite>gene</cite> node. For this reason,
<a class="reference internal" href="#idtrack._the_graph.TheGraph._combined_edges" title="idtrack._the_graph.TheGraph._combined_edges"><code class="xref py py-meth docutils literal notranslate"><span class="pre">TheGraph._combined_edges()</span></code></a> method has been used with reversed graph. In order to have the
information for Ensembl genes or Ensembl assembly genes, the information from all edges should be merged and
the name for the database has to be edited accordingly.</p>
<p>Here, <cite>ensembl_gene</cite> is renamed with <cite>assembly_X_ensembl_gene</cite> for more intuitive understanding of the source of
the gene, and also to make the naming consistent.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>the_result</strong> – A dictionary as the output of <a class="reference internal" href="#idtrack._the_graph.TheGraph._combined_edges" title="idtrack._the_graph.TheGraph._combined_edges"><code class="xref py py-meth docutils literal notranslate"><span class="pre">TheGraph._combined_edges()</span></code></a>.</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>A dictionary with following format
<code class="docutils literal notranslate"><span class="pre">{database_name:</span> <span class="pre">{assembly:</span> <span class="pre">{ensembl</span> <span class="pre">release</span> <span class="pre">set}}}</span></code>.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>dict</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="idtrack._the_graph.TheGraph._get_active_ranges_of_id">
<span class="sig-name descname"><span class="pre">_get_active_ranges_of_id</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">input_id</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/idtrack/_the_graph.html#TheGraph._get_active_ranges_of_id"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#idtrack._the_graph.TheGraph._get_active_ranges_of_id" title="Link to this definition"></a></dt>
<dd><p>Calculating the range of active ensembl releases of nodes separately for backbone nodes and others.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>input_id</strong> (<em>str</em>) – Query ID.</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>Ranges of the ID as list of lists. Outputs should always be increasing, inclusive ranges.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>list[list]</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="idtrack._the_graph.TheGraph._node_trios">
<span class="sig-name descname"><span class="pre">_node_trios</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">the_id</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/idtrack/_the_graph.html#TheGraph._node_trios"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#idtrack._the_graph.TheGraph._node_trios" title="Link to this definition"></a></dt>
<dd><p>Calculates the unique tuple called <cite>trios</cite> (database, assembly, Ensembl release) for a given ID.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>the_id</strong> (<em>str</em>) – Query ID.</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>Set of trios.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>set[tuple]</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="idtrack._the_graph.TheGraph.attach_included_forms">
<span class="sig-name descname"><span class="pre">attach_included_forms</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">available_forms</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/idtrack/_the_graph.html#TheGraph.attach_included_forms"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#idtrack._the_graph.TheGraph.attach_included_forms" title="Link to this definition"></a></dt>
<dd><p>Set the <code class="docutils literal notranslate"><span class="pre">available_forms</span></code> variable for the instance.</p>
<p>This is a separate function and mainly aimed to be used by <code class="xref py py-class docutils literal notranslate"><span class="pre">GraphMaker</span></code> during the
process of constructing the graph. Just after the graphs for the forms of interests (typically <cite>gene</cite>,
<cite>transcript</cite>, and <cite>protein</cite>) are merged, this method is called to store which forms are included in the graph.
Setting the variable before merging the graphs causes errors and inconsistencies.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>available_forms</strong> (<em>list</em>) – Determine which forms (transcript, translation, gene) are included.</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p>None</p>
</dd>
</dl>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="idtrack._the_graph.TheGraph.available_external_databases">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">available_external_databases</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="pre">set</span></em><a class="headerlink" href="#idtrack._the_graph.TheGraph.available_external_databases" title="Link to this definition"></a></dt>
<dd><p>Find the available external databases found in the graph.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>Set of all external databases in the graph.</p>
</dd>
</dl>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="idtrack._the_graph.TheGraph.available_external_databases_assembly">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">available_external_databases_assembly</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="pre">dict</span><span class="p"><span class="pre">[</span></span><span class="pre">int</span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="pre">set</span><span class="p"><span class="pre">]</span></span></em><a class="headerlink" href="#idtrack._the_graph.TheGraph.available_external_databases_assembly" title="Link to this definition"></a></dt>
<dd><p>Find the available external databases found in the graph for each assembly separately.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>Dict of all external databases in the graph, assemblies as keys.</p>
</dd>
</dl>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="idtrack._the_graph.TheGraph.available_genome_assemblies">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">available_genome_assemblies</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="pre">set</span><span class="p"><span class="pre">[</span></span><span class="pre">int</span><span class="p"><span class="pre">]</span></span></em><a class="headerlink" href="#idtrack._the_graph.TheGraph.available_genome_assemblies" title="Link to this definition"></a></dt>
<dd><p>Find the genome assemblies found in the graph by iterating through all nodes.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>Genome assemblies, which is also found in <code class="xref py py-attr docutils literal notranslate"><span class="pre">DB.assembly_mysqlport_priority</span></code>.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="idtrack._the_graph.TheGraph.available_releases_given_database_assembly">
<span class="sig-name descname"><span class="pre">available_releases_given_database_assembly</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">database_name</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">assembly</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/idtrack/_the_graph.html#TheGraph.available_releases_given_database_assembly"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#idtrack._the_graph.TheGraph.available_releases_given_database_assembly" title="Link to this definition"></a></dt>
<dd><p>Possible Ensembl releases defined for a given database and assembly.</p>
<p>The method uses <cite>node_trios</cite> unnecessarily method, which consumes a lot of memory and hinders high
computational efficiency. However, this method is used only in testing purposes, when the speed and memory is
not of a concern.</p>
<p>It is important to note that not all databases are defined in all Ensembl release. To see for more information,
have a look at the <code class="xref py py-class docutils literal notranslate"><span class="pre">ExternalDatabases</span></code>.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>database_name</strong> (<em>str</em>) – External database or node type (except <cite>external</cite>) it should be one of the item from
<a class="reference internal" href="#idtrack._the_graph.TheGraph.available_external_databases" title="idtrack._the_graph.TheGraph.available_external_databases"><code class="xref py py-meth docutils literal notranslate"><span class="pre">TheGraph.available_external_databases()</span></code></a>. The method also works with <cite>node types</cite>
(except <cite>external</cite>), since they are also defined in <cite>node_trios</cite>. Important to note that the <cite>node type</cite>
for Ensembl should follow <code class="xref py py-attr docutils literal notranslate"><span class="pre">DB.nts_assembly</span></code> or <code class="xref py py-attr docutils literal notranslate"><span class="pre">DB.nts_base_ensembl</span></code>.</p></li>
<li><p><strong>assembly</strong> (<em>int</em>) – Assembly, it should be one of the item from
<a class="reference internal" href="#idtrack._the_graph.TheGraph.available_genome_assemblies" title="idtrack._the_graph.TheGraph.available_genome_assemblies"><code class="xref py py-meth docutils literal notranslate"><span class="pre">TheGraph.available_genome_assemblies()</span></code></a>.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>Available Ensembl releases as set of integers.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>set[int]</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="idtrack._the_graph.TheGraph.calculate_caches">
<span class="sig-name descname"><span class="pre">calculate_caches</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/idtrack/_the_graph.html#TheGraph.calculate_caches"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#idtrack._the_graph.TheGraph.calculate_caches" title="Link to this definition"></a></dt>
<dd><p>Calculate cached variables using only one method.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="idtrack._the_graph.TheGraph.combined_edges">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">combined_edges</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="pre">dict</span></em><a class="headerlink" href="#idtrack._the_graph.TheGraph.combined_edges" title="Link to this definition"></a></dt>
<dd><p>Combines all the edge data for a given node except <cite>Ensembl gene</cite> and <cite>assembly specific Ensembl gene</cite> nodes.</p>
<p>It iterates all the edges of a given node to construct a dictionary with a specified format.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>A dictionary with following format
<code class="docutils literal notranslate"><span class="pre">{node_name:</span> <span class="pre">{database_name:</span> <span class="pre">{assembly:</span> <span class="pre">{ensembl</span> <span class="pre">release</span> <span class="pre">set}}}}</span></code>.</p>
</dd>
</dl>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="idtrack._the_graph.TheGraph.combined_edges_assembly_specific_genes">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">combined_edges_assembly_specific_genes</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="pre">dict</span></em><a class="headerlink" href="#idtrack._the_graph.TheGraph.combined_edges_assembly_specific_genes" title="Link to this definition"></a></dt>
<dd><p>Combines all the edge data for a given node for only <cite>assembly specific Ensembl gene</cite> nodes.</p>
<p>It iterates all the edges of a given node to construct a dictionary with a specified format.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>A dictionary with following format
<code class="docutils literal notranslate"><span class="pre">{node_name:</span> <span class="pre">{database_name:</span> <span class="pre">{assembly:</span> <span class="pre">{ensembl</span> <span class="pre">release</span> <span class="pre">set}}}}</span></code>.
But ut has only one assembly, whose hint is given in the database_name.</p>
</dd>
</dl>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="idtrack._the_graph.TheGraph.combined_edges_genes">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">combined_edges_genes</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="pre">dict</span></em><a class="headerlink" href="#idtrack._the_graph.TheGraph.combined_edges_genes" title="Link to this definition"></a></dt>
<dd><p>Combines all the edge data for a given node for only <cite>Ensembl gene</cite> nodes.</p>
<p>It iterates all the edges of a given node to construct a dictionary with a specified format.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>A dictionary with following format
<code class="docutils literal notranslate"><span class="pre">{node_name:</span> <span class="pre">{database_name:</span> <span class="pre">{assembly:</span> <span class="pre">{ensembl</span> <span class="pre">release</span> <span class="pre">set}}}}</span></code>.
Can have multiple assemblies as some main assembly genes are shared across different assemblies.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="idtrack._the_graph.TheGraph.compact_ranges">
<em class="property"><span class="pre">static</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">compact_ranges</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">list_of_ranges</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/idtrack/_the_graph.html#TheGraph.compact_ranges"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#idtrack._the_graph.TheGraph.compact_ranges" title="Link to this definition"></a></dt>
<dd><p>Reduce the list of ranges into least possible number of ranges.</p>
<p>O(n) time and space complexity: a forward in place compaction and copying back the elements,
as then each inner step is O(1) (get/set instead of del).</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>list_of_ranges</strong> (<em>list</em><em>[</em><em>list</em><em>]</em>) – List of increasing, inclusive ranges. The elements are positive integers. Note that there
should be no overlapping elements.
The output of <a class="reference internal" href="#idtrack._the_graph.TheGraph.list_to_ranges" title="idtrack._the_graph.TheGraph.list_to_ranges"><code class="xref py py-meth docutils literal notranslate"><span class="pre">TheGraph.list_to_ranges()</span></code></a> or
<a class="reference internal" href="#idtrack._the_graph.TheGraph._get_active_ranges_of_id" title="idtrack._the_graph.TheGraph._get_active_ranges_of_id"><code class="xref py py-meth docutils literal notranslate"><span class="pre">TheGraph._get_active_ranges_of_id()</span></code></a> are a perfect input here.</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>List of ranges as defined in <code class="xref py py-attr docutils literal notranslate"><span class="pre">TheGraph.compact_ranges.list_of_ranges</span></code>.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>list[list]</p>
</dd>
</dl>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="idtrack._the_graph.TheGraph.external_database_connection_form">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">external_database_connection_form</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="pre">dict</span><span class="p"><span class="pre">[</span></span><span class="pre">str</span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="pre">str</span><span class="p"><span class="pre">]</span></span></em><a class="headerlink" href="#idtrack._the_graph.TheGraph.external_database_connection_form" title="Link to this definition"></a></dt>
<dd><p>Finds which form of Ensembl ID the external database identifiers are connected to.</p>
<p>Each external database connects to a specific form (gene, transcript, translation) Ensembl ID. The relevant
form is chosen by <code class="xref py py-class docutils literal notranslate"><span class="pre">ExternalDatabases</span></code> class.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>Dictionary mapping external database name into the associated form (gene, transcript, translation).</p>
</dd>
<dt class="field-even">Raises<span class="colon">:</span></dt>
<dd class="field-even"><p><strong>ValueError</strong> – If non-Ensembl node is connected.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="idtrack._the_graph.TheGraph.get_active_ranges_of_base_id_alternative">
<span class="sig-name descname"><span class="pre">get_active_ranges_of_base_id_alternative</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">base_id</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/idtrack/_the_graph.html#TheGraph.get_active_ranges_of_base_id_alternative"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#idtrack._the_graph.TheGraph.get_active_ranges_of_base_id_alternative" title="Link to this definition"></a></dt>
<dd><p>Get the range of an base ID based on the child IDs it is connected to.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>base_id</strong> (<em>str</em>) – Query ID. Should be with node type of Ensembl base ID.</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>Ranges of the ID as list of lists. Outputs should always be increasing, inclusive ranges.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>list[list]</p>
</dd>
</dl>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="idtrack._the_graph.TheGraph.get_active_ranges_of_id">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">get_active_ranges_of_id</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="pre">dict</span><span class="p"><span class="pre">[</span></span><span class="pre">str</span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="pre">list</span><span class="p"><span class="pre">[</span></span><span class="pre">list</span><span class="p"><span class="pre">]</span></span><span class="p"><span class="pre">]</span></span></em><a class="headerlink" href="#idtrack._the_graph.TheGraph.get_active_ranges_of_id" title="Link to this definition"></a></dt>
<dd><p>Returns the range of active ensembl releases of nodes, ignoring which assembly the release is coming from.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>A dictionary with following format
<code class="docutils literal notranslate"><span class="pre">{id:</span> <span class="pre">list_of_ranges</span></code>.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="idtrack._the_graph.TheGraph.get_active_ranges_of_id_ensembl_all_inclusive">
<span class="sig-name descname"><span class="pre">get_active_ranges_of_id_ensembl_all_inclusive</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">the_id</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/idtrack/_the_graph.html#TheGraph.get_active_ranges_of_id_ensembl_all_inclusive"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#idtrack._the_graph.TheGraph.get_active_ranges_of_id_ensembl_all_inclusive" title="Link to this definition"></a></dt>
<dd><p>Generate active ranges of Ensembl gene nodes with all assemblies.</p>
<p>Note that <a class="reference internal" href="#idtrack._the_graph.TheGraph.get_active_ranges_of_id" title="idtrack._the_graph.TheGraph.get_active_ranges_of_id"><code class="xref py py-meth docutils literal notranslate"><span class="pre">TheGraph.get_active_ranges_of_id()</span></code></a> method provided the range for main assembly
that the graph is build on. This method combines the other assemblies together. Also, it verifies whether
the <a class="reference internal" href="#idtrack._the_graph.TheGraph.combined_edges" title="idtrack._the_graph.TheGraph.combined_edges"><code class="xref py py-meth docutils literal notranslate"><span class="pre">TheGraph.combined_edges()</span></code></a> and <a class="reference internal" href="#idtrack._the_graph.TheGraph.get_active_ranges_of_id" title="idtrack._the_graph.TheGraph.get_active_ranges_of_id"><code class="xref py py-meth docutils literal notranslate"><span class="pre">TheGraph.get_active_ranges_of_id()</span></code></a>
methods provides consistent results.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>the_id</strong> (<em>str</em>) – Query ID. Should be either Ensembl gene or assembly specific Ensembl gene IDs.</p>
</dd>
<dt class="field-even">Raises<span class="colon">:</span></dt>
<dd class="field-even"><p><strong>ValueError</strong> – If there is inconsistency between the outputs of these two functions.
    If the query is not one of the specified node type.</p>
</dd>
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>Ranges of the ID as list of lists. Outputs should always be increasing, inclusive ranges.</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p>list[list]</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="idtrack._the_graph.TheGraph.get_external_database_nodes">
<span class="sig-name descname"><span class="pre">get_external_database_nodes</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">database_name</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/idtrack/_the_graph.html#TheGraph.get_external_database_nodes"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#idtrack._the_graph.TheGraph.get_external_database_nodes" title="Link to this definition"></a></dt>
<dd><p>For a given external database, returns set of all node names defined at least once in that database.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>database_name</strong> (<em>str</em>) – External database, it should be one of the item from
<a class="reference internal" href="#idtrack._the_graph.TheGraph.available_external_databases" title="idtrack._the_graph.TheGraph.available_external_databases"><code class="xref py py-meth docutils literal notranslate"><span class="pre">TheGraph.available_external_databases()</span></code></a>.</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>Node names (identifiers) set.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>set[str]</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="idtrack._the_graph.TheGraph.get_id_list">
<span class="sig-name descname"><span class="pre">get_id_list</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">database</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">assembly</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">release</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/idtrack/_the_graph.html#TheGraph.get_id_list"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#idtrack._the_graph.TheGraph.get_id_list" title="Link to this definition"></a></dt>
<dd><p>Given a trio (database, assembly, release), generates a list of node names (identifiers).</p>
<p>Similar to <a class="reference internal" href="#idtrack._the_graph.TheGraph.available_releases_given_database_assembly" title="idtrack._the_graph.TheGraph.available_releases_given_database_assembly"><code class="xref py py-meth docutils literal notranslate"><span class="pre">TheGraph.available_releases_given_database_assembly()</span></code></a>, the method uses
<cite>node_trios</cite> unnecessarily method, which consumes a lot of memory and hinders high
computational efficiency. However, this method is used only in testing purposes, when the speed and memory is
not of a concern.</p>
<p>It is imporant to note that the Ensembl IDs with versions <code class="xref py py-attr docutils literal notranslate"><span class="pre">DB.alternative_versions</span></code> will be also
returned if the database is the Ensembl gene with the main assembly, and the assembly is the main assembly.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>database</strong> (<em>str</em>) – External database name if the <cite>node type</cite> is <cite>external</cite>, else <cite>node type</cite>. The <cite>node type</cite>
for Ensembl should follow <code class="xref py py-attr docutils literal notranslate"><span class="pre">DB.nts_assembly</span></code> or <code class="xref py py-attr docutils literal notranslate"><span class="pre">DB.nts_base_ensembl</span></code>.</p></li>
<li><p><strong>assembly</strong> (<em>int</em>) – Assembly, it should be one of the item from
<a class="reference internal" href="#idtrack._the_graph.TheGraph.available_genome_assemblies" title="idtrack._the_graph.TheGraph.available_genome_assemblies"><code class="xref py py-meth docutils literal notranslate"><span class="pre">TheGraph.available_genome_assemblies()</span></code></a>.</p></li>
<li><p><strong>release</strong> (<em>int</em>) – Requested Ensembl releases as an integer.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>Node names (identifiers) list.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>list[str]</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="idtrack._the_graph.TheGraph.get_intersecting_ranges">
<em class="property"><span class="pre">static</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">get_intersecting_ranges</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">lor1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">lor2</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">compact</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/idtrack/_the_graph.html#TheGraph.get_intersecting_ranges"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#idtrack._the_graph.TheGraph.get_intersecting_ranges" title="Link to this definition"></a></dt>
<dd><p>As the name suggest, calculates the intersecting ranges of two list of ranges.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>lor1</strong> (<em>list</em><em>[</em><em>list</em><em>]</em>) – List of ranges as defined in <code class="xref py py-attr docutils literal notranslate"><span class="pre">TheGraph.compact_ranges.list_of_ranges</span></code>.</p></li>
<li><p><strong>lor2</strong> (<em>list</em><em>[</em><em>list</em><em>]</em>) – List of ranges as defined in <code class="xref py py-attr docutils literal notranslate"><span class="pre">TheGraph.compact_ranges.list_of_ranges</span></code>.</p></li>
<li><p><strong>compact</strong> (<em>bool</em>) – If set <cite>True</cite>, returns the reduced list of ranges via
<a class="reference internal" href="#idtrack._the_graph.TheGraph.compact_ranges" title="idtrack._the_graph.TheGraph.compact_ranges"><code class="xref py py-meth docutils literal notranslate"><span class="pre">TheGraph.compact_ranges()</span></code></a> method.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>List of ranges as defined in <code class="xref py py-attr docutils literal notranslate"><span class="pre">TheGraph.compact_ranges.list_of_ranges</span></code>.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>list[list]</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="idtrack._the_graph.TheGraph.get_next_edge_releases">
<span class="sig-name descname"><span class="pre">get_next_edge_releases</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">from_id</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">reverse</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/idtrack/_the_graph.html#TheGraph.get_next_edge_releases"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#idtrack._the_graph.TheGraph.get_next_edge_releases" title="Link to this definition"></a></dt>
<dd><p>Retrieves the next edge releases from a node, depending on the directionality of the graph.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>from_id</strong> (<em>str</em>) – Query ID. Should be with node type of Ensembl gene.</p></li>
<li><p><strong>reverse</strong> (<em>bool</em>) – The direction of desired next edges.
If <code class="docutils literal notranslate"><span class="pre">True</span></code>, previous edges are returned. If <code class="docutils literal notranslate"><span class="pre">False</span></code>, next edges are returned.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>The Ensembl releases of next (or previous if reverse is <code class="docutils literal notranslate"><span class="pre">True</span></code>) edges.</p>
</dd>
<dt class="field-odd">Raises<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>ValueError</strong> – If the query ID is not with the node type of graph backbone.</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p>list</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="idtrack._the_graph.TheGraph.get_two_nodes_coinciding_releases">
<span class="sig-name descname"><span class="pre">get_two_nodes_coinciding_releases</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">id1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">id2</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">compact</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/idtrack/_the_graph.html#TheGraph.get_two_nodes_coinciding_releases"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#idtrack._the_graph.TheGraph.get_two_nodes_coinciding_releases" title="Link to this definition"></a></dt>
<dd><p>Find the intersecting range of two nodes in the graph.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>id1</strong> (<em>str</em>) – First Query ID.</p></li>
<li><p><strong>id2</strong> (<em>str</em>) – Second Query ID.</p></li>
<li><p><strong>compact</strong> (<em>bool</em>) – Parameter to pass into <a class="reference internal" href="#idtrack._the_graph.TheGraph.get_intersecting_ranges" title="idtrack._the_graph.TheGraph.get_intersecting_ranges"><code class="xref py py-meth docutils literal notranslate"><span class="pre">TheGraph.get_intersecting_ranges()</span></code></a> method.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>List of ranges as defined in <code class="xref py py-attr docutils literal notranslate"><span class="pre">TheGraph.compact_ranges.list_of_ranges</span></code>.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>list[list]</p>
</dd>
</dl>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="idtrack._the_graph.TheGraph.hyperconnective_nodes">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">hyperconnective_nodes</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="pre">dict</span></em><a class="headerlink" href="#idtrack._the_graph.TheGraph.hyperconnective_nodes" title="Link to this definition"></a></dt>
<dd><p>Cached property for hyper-connective nodes as keys and number of connections as values.</p>
<p>Hyper-connective nodes are defined as external nodes with more than
<code class="xref py py-attr docutils literal notranslate"><span class="pre">DB.hyperconnecting_threshold</span></code>. This nodes creates a huge bottleneck in finding the synonyms
of a node and hence the pathfinder algorithm. The pathfinder algoritm is hence speed-up by ignoring these nodes
in the search process. This is in theory sacrificing the precision of the algorithm a little bit, but
in practice these nodes generally does not have not very precise ID matching. Most of the case, the relavant
IDs are matched via some other external ID, so the performance of the pathfinder algorithm actually improves
if hyper-connective nodes are ignored.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>Dictionary of hyper-connective nodes as keys and number of connections as values.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="idtrack._the_graph.TheGraph.is_point_in_range">
<em class="property"><span class="pre">static</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">is_point_in_range</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">lor</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">p</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/idtrack/_the_graph.html#TheGraph.is_point_in_range"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#idtrack._the_graph.TheGraph.is_point_in_range" title="Link to this definition"></a></dt>
<dd><p>Simple method to determine whether a given integer is covered by list of ranges.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>lor</strong> (<em>list</em><em>[</em><em>list</em><em>]</em>) – List of ranges as defined in <code class="xref py py-attr docutils literal notranslate"><span class="pre">TheGraph.compact_ranges.list_of_ranges</span></code>.</p></li>
<li><p><strong>p</strong> (<em>int</em>) – A positive integer.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><code class="docutils literal notranslate"><span class="pre">True</span></code> if it is in the range, <code class="docutils literal notranslate"><span class="pre">False</span></code> otherwise.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>bool</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="idtrack._the_graph.TheGraph.list_to_ranges">
<em class="property"><span class="pre">static</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">list_to_ranges</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">lst</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/idtrack/_the_graph.html#TheGraph.list_to_ranges"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#idtrack._the_graph.TheGraph.list_to_ranges" title="Link to this definition"></a></dt>
<dd><p>Convert sorted non-repeating list of integers into list of inclusive non-overlapping ranges.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>lst</strong> (<em>list</em><em>[</em><em>int</em><em>]</em>) – List of integers. It should be sorted in increasing order. Repeating element is not allowed.
The output of <a class="reference internal" href="#idtrack._the_graph.TheGraph.ranges_to_list" title="idtrack._the_graph.TheGraph.ranges_to_list"><code class="xref py py-meth docutils literal notranslate"><span class="pre">TheGraph.ranges_to_list()</span></code></a> is a perfect input here.</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>Ranges as list of lists. Outputs should always be increasing, inclusive ranges. With positive integers.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>list[list]</p>
</dd>
</dl>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="idtrack._the_graph.TheGraph.lower_chars_graph">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">lower_chars_graph</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="pre">dict</span></em><a class="headerlink" href="#idtrack._the_graph.TheGraph.lower_chars_graph" title="Link to this definition"></a></dt>
<dd><p>A simple dictionary mapping the node name with its lower characters into the original node name.</p>
<dl class="field-list simple">
<dt class="field-odd">Raises<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>ValueError</strong> – If there are multiple nodes which becomes the same after lower character conversion.</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>A dictionary with following format
<code class="docutils literal notranslate"><span class="pre">{lower_char_id:</span> <span class="pre">id}</span></code></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="idtrack._the_graph.TheGraph.node_name_alternatives">
<span class="sig-name descname"><span class="pre">node_name_alternatives</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">identifier</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/idtrack/_the_graph.html#TheGraph.node_name_alternatives"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#idtrack._the_graph.TheGraph.node_name_alternatives" title="Link to this definition"></a></dt>
<dd><p>Matching a query ID into the ID found in the graph based on some criteria and priorities.</p>
<p>A query ID is sometimes not found exactly in the graph due to the format it has. However, very slight
modifications of the string of query ID could help the pathfinder locate the ID of interest. For example,
<cite>actb</cite> can be queried for the pathfinder, but nothing is found as upper character version of the query,
<cite>ACTB</cite>, is found in the graph instead.</p>
<p>Priority is as follows: (1) try to find directly in the graph (2) look for lower-char version (3) look for
initial substring before separators (4) check all possible variations of dash and underscore for each of which
follow the priority list above.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>identifier</strong> (<em>str</em>) – A bio-ID of interest. This could be an ID in the graph or some other ID to be matched
with the corresponding ID in the graph.</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><dl class="simple">
<dt>A tuple with the first element is the bio-ID in the graph if the query is found somehow, else <code class="docutils literal notranslate"><span class="pre">None</span></code>.</dt><dd><p>The second element is to show whether the query ID is found without any modifications or not.</p>
</dd>
</dl>
</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>tuple[str | None, bool]</p>
</dd>
</dl>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="idtrack._the_graph.TheGraph.node_trios">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">node_trios</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="pre">dict</span><span class="p"><span class="pre">[</span></span><span class="pre">str</span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="pre">set</span><span class="p"><span class="pre">[</span></span><span class="pre">tuple</span><span class="p"><span class="pre">]</span></span><span class="p"><span class="pre">]</span></span></em><a class="headerlink" href="#idtrack._the_graph.TheGraph.node_trios" title="Link to this definition"></a></dt>
<dd><p>Creates a dict for all nodes with <cite>node_trios</cite> calculated by <a class="reference internal" href="#idtrack._the_graph.TheGraph._node_trios" title="idtrack._the_graph.TheGraph._node_trios"><code class="xref py py-meth docutils literal notranslate"><span class="pre">TheGraph._node_trios()</span></code></a>.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>A memory intensive dictionary with node name as the key and calculated <cite>node_trios</cite> as the value.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="idtrack._the_graph.TheGraph.ranges_to_list">
<span class="sig-name descname"><span class="pre">ranges_to_list</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">lor</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/idtrack/_the_graph.html#TheGraph.ranges_to_list"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#idtrack._the_graph.TheGraph.ranges_to_list" title="Link to this definition"></a></dt>
<dd><p>Convert list of inclusive non-overlapping ranges into sorted sorted non-repeating list of integers.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>lor</strong> (<em>list</em><em>[</em><em>list</em><em>]</em>) – Ranges as list of lists. Should always be increasing, inclusive ranges. With positive integers.
The output of <a class="reference internal" href="#idtrack._the_graph.TheGraph.list_to_ranges" title="idtrack._the_graph.TheGraph.list_to_ranges"><code class="xref py py-meth docutils literal notranslate"><span class="pre">TheGraph.list_to_ranges()</span></code></a> is a perfect input here.</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>Sorted non-repeating list of integers.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>list[int]</p>
</dd>
</dl>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="idtrack._the_graph.TheGraph.rev">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">rev</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><a class="reference internal" href="#idtrack._the_graph.TheGraph" title="idtrack._the_graph.TheGraph"><span class="pre">TheGraph</span></a></em><a class="headerlink" href="#idtrack._the_graph.TheGraph.rev" title="Link to this definition"></a></dt>
<dd><p>The same graph but the edges are in reverse orientation.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p><code class="docutils literal notranslate"><span class="pre">TheGraph</span></code> object without copying another one in the memory.</p>
</dd>
</dl>
</dd></dl>

</dd></dl>

<dl class="py class" id="module-idtrack._graph_maker">
<dt class="sig sig-object py" id="idtrack._graph_maker.GraphMaker">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">GraphMaker</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">db_manager</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/idtrack/_graph_maker.html#GraphMaker"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#idtrack._graph_maker.GraphMaker" title="Link to this definition"></a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></p>
<p>Creates ID history graph.</p>
<p>It includes Ensembl gene ID history. Ensembl ID history is obtained from Ensembl
resources, which shows the connection between different Ensembl base IDs or different versions of the same Ensembl
base ID. Ensembl transcripts (with base IDs and versions) are connected to gene, and Ensembl proteins are
connected to transcripts. Additionally, a selected set of external databases are connected to the related Ensembl
IDs: for example UniProt IDs are associated with proteins, while RefSeq transcript IDs are associated with
transcripts. The <code class="docutils literal notranslate"><span class="pre">GraphMaker</span></code> class also saves the resulting graph into the defined temporary directory for later
calculations.</p>
<p>Class initialization.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>db_manager</strong> (<a class="reference internal" href="#idtrack._database_manager.DatabaseManager" title="idtrack._database_manager.DatabaseManager"><em>DatabaseManager</em></a>) – Needed to download all necessary tables and data frames.
It contains the temporary directory to save the resultant graph.</p>
</dd>
<dt class="field-even">Raises<span class="colon">:</span></dt>
<dd class="field-even"><p><strong>ValueError</strong> – <code class="docutils literal notranslate"><span class="pre">GraphMaker</span></code> has to be created with the latest release possible in <code class="docutils literal notranslate"><span class="pre">db_manager</span></code>. If not,
    the exception is raised.</p>
</dd>
</dl>
<dl class="py method">
<dt class="sig sig-object py" id="idtrack._graph_maker.GraphMaker.construct_graph">
<span class="sig-name descname"><span class="pre">construct_graph</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">narrow</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">form_list</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">narrow_external</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/idtrack/_graph_maker.html#GraphMaker.construct_graph"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#idtrack._graph_maker.GraphMaker.construct_graph" title="Link to this definition"></a></dt>
<dd><p>Main method to construct the graph.</p>
<p>It creates the graph with Ensembl gene, transcript and protein information. It also adds
<code class="docutils literal notranslate"><span class="pre">DB.nts_base_ensembl[f]</span></code> nodes into the graph, which has only base Ensembl gene ID (no version).
External database entries described in <code class="docutils literal notranslate"><span class="pre">ExternalDatabases</span></code> will be part of the graph. Normally, user
is not expected to use this method, as the method is utilized in <code class="docutils literal notranslate"><span class="pre">get_graph</span></code> method.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>narrow</strong> (<em>bool</em>) – Determine whether a some more information should be added between Ensembl gene IDs. For example,
which genome assembly is used, or when was the connection is established. For usual uses, no need to
set it <code class="docutils literal notranslate"><span class="pre">True</span></code>.</p></li>
<li><p><strong>form_list</strong> (<em>list</em><em> | </em><em>None</em>) – Determine which forms (transcript, translation, gene) should be included. If <code class="docutils literal notranslate"><span class="pre">None</span></code>, then
include all possible forms defined in <code class="docutils literal notranslate"><span class="pre">DatabaseManager</span></code>.
It has to be list composed of following strings: ‘gene’, ‘transcript’, ‘translation’.</p></li>
<li><p><strong>narrow_external</strong> (<em>bool</em>) – If set <code class="docutils literal notranslate"><span class="pre">False</span></code>, all possible external databases defined in Ensembl MySQL server will be
included into the graph. The graph will be immensely larger, and the ID history travel calculation will
be very slow. Additionally, the success of ID conversion under such a setting it has not been
tested yet.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>Resultant multiedge directed graph.</p>
</dd>
<dt class="field-odd">Raises<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>ValueError</strong> – Unexpected error.</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><a class="reference internal" href="#idtrack._the_graph.TheGraph" title="idtrack._the_graph.TheGraph"><em>TheGraph</em></a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="idtrack._graph_maker.GraphMaker.construct_graph_form">
<span class="sig-name descname"><span class="pre">construct_graph_form</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">narrow</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">db_manager</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/idtrack/_graph_maker.html#GraphMaker.construct_graph_form"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#idtrack._graph_maker.GraphMaker.construct_graph_form" title="Link to this definition"></a></dt>
<dd><p>Creates a graph with connected nodes based on historical relationships between each Ensembl IDs.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>narrow</strong> (<em>bool</em>) – See parameter in <code class="xref py py-attr docutils literal notranslate"><span class="pre">Graph.construct_graph.narrow</span></code></p></li>
<li><p><strong>db_manager</strong> (<a class="reference internal" href="#idtrack._database_manager.DatabaseManager" title="idtrack._database_manager.DatabaseManager"><em>DatabaseManager</em></a>) – The method reads ID history dataframe, and Ensembl IDs lists at each Ensembl release,
provided by <code class="docutils literal notranslate"><span class="pre">DatabaseManager</span></code>.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>Resultant multi edge directed graph.</p>
</dd>
<dt class="field-odd">Raises<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>ValueError</strong> – Unexpected error.</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><a class="reference internal" href="#idtrack._the_graph.TheGraph" title="idtrack._the_graph.TheGraph"><em>TheGraph</em></a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="idtrack._graph_maker.GraphMaker.create_file_name">
<span class="sig-name descname"><span class="pre">create_file_name</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">narrow</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/idtrack/_graph_maker.html#GraphMaker.create_file_name"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#idtrack._graph_maker.GraphMaker.create_file_name" title="Link to this definition"></a></dt>
<dd><p>File name creator which includes some information regarding the construction process.</p>
<p>Facilitates to recognize the graph based on file name.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>narrow</strong> (<em>bool</em>) – See parameter in <code class="xref py py-attr docutils literal notranslate"><span class="pre">Graph.construct_graph.narrow</span></code></p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>Absolute file path in the temporary directory provided by <code class="docutils literal notranslate"><span class="pre">DatabaseManager</span></code>.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>str</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="idtrack._graph_maker.GraphMaker.export_disk">
<span class="sig-name descname"><span class="pre">export_disk</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">g</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">file_path</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">overwrite</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/idtrack/_graph_maker.html#GraphMaker.export_disk"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#idtrack._graph_maker.GraphMaker.export_disk" title="Link to this definition"></a></dt>
<dd><p>Write the <cite>pickle</cite> file in the provided file path, which contains the graph.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>g</strong> (<a class="reference internal" href="#idtrack._the_graph.TheGraph" title="idtrack._the_graph.TheGraph"><em>TheGraph</em></a>) – Multi edge directed graph object to stor in the disk.</p></li>
<li><p><strong>file_path</strong> (<em>str</em>) – Absolute target path, provided by <code class="xref py py-meth docutils literal notranslate"><span class="pre">Graph.create_file_name()</span></code></p></li>
<li><p><strong>overwrite</strong> (<em>bool</em>) – See parameter in <code class="xref py py-attr docutils literal notranslate"><span class="pre">Graph.get_graph.overwrite_even_if_exist</span></code></p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="idtrack._graph_maker.GraphMaker.get_graph">
<span class="sig-name descname"><span class="pre">get_graph</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">narrow</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">create_even_if_exist</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">save_after_calculation</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">overwrite_even_if_exist</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/idtrack/_graph_maker.html#GraphMaker.get_graph"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#idtrack._graph_maker.GraphMaker.get_graph" title="Link to this definition"></a></dt>
<dd><p>Simplifies the graph construction process.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>narrow</strong> (<em>bool</em>) – See parameter in <code class="xref py py-attr docutils literal notranslate"><span class="pre">Graph.construct_graph.narrow</span></code></p></li>
<li><p><strong>create_even_if_exist</strong> (<em>bool</em>) – Determine whether create the graph even if it exists. If there is no graph in the
provided temporary directory, the graph will be created regardless.</p></li>
<li><p><strong>save_after_calculation</strong> (<em>bool</em>) – Determine whether resultant graph will be saved or not.</p></li>
<li><p><strong>overwrite_even_if_exist</strong> (<em>bool</em>) – If the graph will be saved, determine whether the program should overwrite.
If <code class="docutils literal notranslate"><span class="pre">False</span></code>, it does not re-saves the calculated (or loaded) graph.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>Resultant multi edge directed graph, which can be used in all future calculations.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference internal" href="#idtrack._the_graph.TheGraph" title="idtrack._the_graph.TheGraph"><em>TheGraph</em></a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="idtrack._graph_maker.GraphMaker.initialize_downloads">
<span class="sig-name descname"><span class="pre">initialize_downloads</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/idtrack/_graph_maker.html#GraphMaker.initialize_downloads"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#idtrack._graph_maker.GraphMaker.initialize_downloads" title="Link to this definition"></a></dt>
<dd><p>Initialize the external database downloads.</p>
<dl class="field-list simple">
<dt class="field-odd">Raises<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>NotImplementedError</strong> – Not implemented yet. Currently, the necessary data sources are downloaded when needed
    during the graph construction process.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="idtrack._graph_maker.GraphMaker.read_exported">
<em class="property"><span class="pre">static</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">read_exported</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">file_path</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/idtrack/_graph_maker.html#GraphMaker.read_exported"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#idtrack._graph_maker.GraphMaker.read_exported" title="Link to this definition"></a></dt>
<dd><p>Read the <cite>pickle</cite> file in the provided file path, which contains the graph.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>file_path</strong> (<em>str</em>) – Absolute path of the file of interest.</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>Resultant multi edge directed graph.</p>
</dd>
<dt class="field-odd">Raises<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>FileNotFoundError</strong> – When there is no file in the provided directory.</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><a class="reference internal" href="#idtrack._the_graph.TheGraph" title="idtrack._the_graph.TheGraph"><em>TheGraph</em></a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="idtrack._graph_maker.GraphMaker.remove_non_gene_trees">
<em class="property"><span class="pre">static</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">remove_non_gene_trees</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">graph</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">forms_remove</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/idtrack/_graph_maker.html#GraphMaker.remove_non_gene_trees"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#idtrack._graph_maker.GraphMaker.remove_non_gene_trees" title="Link to this definition"></a></dt>
<dd><p>Removes the edges between the nodes with the same <cite>node type</cite> and removes abstract nodes (Void and Retired).</p>
<p>The nodes between two the same <code class="xref py py-attr docutils literal notranslate"><span class="pre">DB.node_type_str</span></code> will be removed. Also, the nodes with versions
<code class="xref py py-attr docutils literal notranslate"><span class="pre">DB.no_new_node_id</span></code> and <code class="xref py py-attr docutils literal notranslate"><span class="pre">DB.no_old_node_id</span></code> will be also removed.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>graph</strong> (<a class="reference internal" href="#idtrack._the_graph.TheGraph" title="idtrack._the_graph.TheGraph"><em>TheGraph</em></a>) – The output of <code class="xref py py-attr docutils literal notranslate"><span class="pre">Graph.construct_graph</span></code> or
<code class="xref py py-attr docutils literal notranslate"><span class="pre">Graph.construct_graph_form</span></code>.</p></li>
<li><p><strong>forms_remove</strong> (<em>list</em><em> | </em><em>None</em>) – Determine which <cite>node type</cite> are of interest.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>Resultant multi edge directed graph.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference internal" href="#idtrack._the_graph.TheGraph" title="idtrack._the_graph.TheGraph"><em>TheGraph</em></a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="idtrack._graph_maker.GraphMaker.split_id">
<em class="property"><span class="pre">static</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">split_id</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">id_to_split</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">which_part</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/idtrack/_graph_maker.html#GraphMaker.split_id"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#idtrack._graph_maker.GraphMaker.split_id" title="Link to this definition"></a></dt>
<dd><p>Simpler method to retrieve ID or Version part of a node name.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>id_to_split</strong> (<em>str</em>) – Query node name.</p></li>
<li><p><strong>which_part</strong> (<em>str</em>) – Either ‘Version’ or ‘ID’.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>The requested substring of the node name.</p>
</dd>
<dt class="field-odd">Raises<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>ValueError</strong> – If ‘which_part’ is assigned to some other value than ‘Version’, or ‘ID’.</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p>str | float</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="idtrack._graph_maker.GraphMaker.update_graph_with_the_new_release">
<span class="sig-name descname"><span class="pre">update_graph_with_the_new_release</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/idtrack/_graph_maker.html#GraphMaker.update_graph_with_the_new_release"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#idtrack._graph_maker.GraphMaker.update_graph_with_the_new_release" title="Link to this definition"></a></dt>
<dd><p>When new release arrive, just add new nodes.</p>
<dl class="field-list simple">
<dt class="field-odd">Raises<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>NotImplementedError</strong> – Not implemented yet. Currently, the user is expected to recreate whole graph using
    <code class="docutils literal notranslate"><span class="pre">get_graph</span></code> method. Note that not all databases need to be re-downloaded, the program will only
    download the new release, and re-construct the graph.</p>
</dd>
</dl>
</dd></dl>

</dd></dl>

<dl class="py class" id="module-idtrack._verify_organism">
<dt class="sig sig-object py" id="idtrack._verify_organism.VerifyOrganism">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">VerifyOrganism</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">organism_query</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/idtrack/_verify_organism.html#VerifyOrganism"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#idtrack._verify_organism.VerifyOrganism" title="Link to this definition"></a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></p>
<p>Finds the formal organism name and associated latest Ensembl release given a tentative organism name.</p>
<p>This class is designed to aid the user regarding the organism name accepted by the program and latest Ensembl
release possible without manually looking online resources. Every time of use, the class uses Ensembl REST API to
get the information regarding the organisms and associated Ensembl releases, so the class requires an active
Internet connection.</p>
<p>Class initialization.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>organism_query</strong> (<em>str</em>) – Organism of interest.
This does not have to be formal name. For example, ‘human’, ‘hsapiens’ as well as formal name
‘homo_sapiens’ is accepted by the program.</p>
</dd>
</dl>
<dl class="py method">
<dt class="sig sig-object py" id="idtrack._verify_organism.VerifyOrganism.fetch_organism_and_latest_release">
<span class="sig-name descname"><span class="pre">fetch_organism_and_latest_release</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">connect_timeout</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">read_timeout</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/idtrack/_verify_organism.html#VerifyOrganism.fetch_organism_and_latest_release"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#idtrack._verify_organism.VerifyOrganism.fetch_organism_and_latest_release" title="Link to this definition"></a></dt>
<dd><p>Creates a query and process the data to construct two dataframes for Ensembl release and organism names.</p>
<p>The query to get the name of the organisms defined in the database, but also to get the latest
ensembl release to go through. This method is actually main functionality of the class; in general uses, the
user is not expected to use this method directly.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>connect_timeout</strong> (<em>int</em>) – The number of seconds the requests will wait to establish a connection to a remote machine.</p></li>
<li><p><strong>read_timeout</strong> (<em>int</em>) – The number of seconds the client will wait for the server to send a response</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>Two dataframes to be used by the instance methods later on.</p>
</dd>
<dt class="field-odd">Raises<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>ValueError</strong> – Unexpected error.</p></li>
<li><p><strong>TimeoutError</strong> – If allowed time is passed.</p></li>
</ul>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p>tuple[<em>DataFrame</em>, <em>DataFrame</em>]</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="idtrack._verify_organism.VerifyOrganism.get_formal_name">
<span class="sig-name descname"><span class="pre">get_formal_name</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/idtrack/_verify_organism.html#VerifyOrganism.get_formal_name"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#idtrack._verify_organism.VerifyOrganism.get_formal_name" title="Link to this definition"></a></dt>
<dd><p>Using the API response and queried organism name, find the formal organism name to be used for the program.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>Formal organism name.</p>
</dd>
<dt class="field-even">Raises<span class="colon">:</span></dt>
<dd class="field-even"><ul class="simple">
<li><p><strong>KeyError</strong> – If the query is not defined in the pre-created dataframes.</p></li>
<li><p><strong>ValueError</strong> – If there are multiple entries for query and hence the query is ambiguous.”</p></li>
</ul>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>str</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="idtrack._verify_organism.VerifyOrganism.get_latest_release">
<span class="sig-name descname"><span class="pre">get_latest_release</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/idtrack/_verify_organism.html#VerifyOrganism.get_latest_release"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#idtrack._verify_organism.VerifyOrganism.get_latest_release" title="Link to this definition"></a></dt>
<dd><p>Using the API response and queried organism name, find the latest Ensembl release to be used for the program.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>Latest Ensembl release.</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p>int</p>
</dd>
</dl>
</dd></dl>

</dd></dl>

<dl class="py class" id="module-idtrack._track">
<dt class="sig sig-object py" id="idtrack._track.Track">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">Track</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">db_manager</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/idtrack/_track.html#Track"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#idtrack._track.Track" title="Link to this definition"></a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></p>
<p>Bidirectional path-finding resolver for biological identifiers.</p>
<p><cite>Track</cite> builds and queries a <em>bio-ID</em> multigraph that stitches together
Ensembl history edges (genes, transcripts, proteins) and cross-reference
edges to external databases (UniProt, RefSeq, …).  Given a <strong>source
identifier</strong>, a <strong>target Ensembl release</strong>, and/or a <strong>target database</strong>,
the class:</p>
<ol class="arabic simple">
<li><p><strong>Normalises</strong> the source to an Ensembl <em>gene</em> node when necessary.</p></li>
<li><p><strong>Time-travels</strong> through historical edges—forward or backward—until it
reaches the requested release, optionally “beaming-up” through external
IDs when the backbone is disconnected.</p></li>
<li><p><strong>Converts</strong> the resolved Ensembl gene into the requested external
database (or returns the gene itself) while annotating the result with
confidence scores and the full traversal path.</p></li>
</ol>
<p>Two mutually-recursive engines power the search:</p>
<ul class="simple">
<li><p><cite>_recursive_function</cite>   — depth-first search along temporal edges.</p></li>
<li><p><cite>_recursive_synonymous</cite> — search for synonymous nodes at a single
release to enable the external “beam-up”.</p></li>
</ul>
<dl class="py attribute">
<dt class="sig sig-object py" id="idtrack._track.Track.graph">
<span class="sig-name descname"><span class="pre">graph</span></span><a class="headerlink" href="#idtrack._track.Track.graph" title="Link to this definition"></a></dt>
<dd><p>The pre-computed bio-ID graph produced by
<code class="xref py py-class docutils literal notranslate"><span class="pre">_graph_maker.GraphMaker</span></code>.</p>
<dl class="field-list simple">
<dt class="field-odd">Type<span class="colon">:</span></dt>
<dd class="field-odd"><p>networkx.MultiDiGraph</p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="idtrack._track.Track.version_info">
<span class="sig-name descname"><span class="pre">version_info</span></span><a class="headerlink" href="#idtrack._track.Track.version_info" title="Link to this definition"></a></dt>
<dd><p>Metadata about the graph build (Ensembl releases included, build date,
Git commit, etc.).</p>
<dl class="field-list simple">
<dt class="field-odd">Type<span class="colon">:</span></dt>
<dd class="field-odd"><p>dict</p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="idtrack._track.Track._external_entrance_placeholder">
<span class="sig-name descname"><span class="pre">_external_entrance_placeholder</span></span><a class="headerlink" href="#idtrack._track.Track._external_entrance_placeholder" title="Link to this definition"></a></dt>
<dd><p>Sentinel node IDs that mark artificial edges used when an external ID
is pulled onto the Ensembl backbone (<cite>False → -1</cite>, <cite>True → 10001</cite>).</p>
<dl class="field-list simple">
<dt class="field-odd">Type<span class="colon">:</span></dt>
<dd class="field-odd"><p>dict[bool, int]</p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="idtrack._track.Track._external_entrance_placeholders">
<span class="sig-name descname"><span class="pre">_external_entrance_placeholders</span></span><a class="headerlink" href="#idtrack._track.Track._external_entrance_placeholders" title="Link to this definition"></a></dt>
<dd><p>Sorted list of the sentinel values above.</p>
<dl class="field-list simple">
<dt class="field-odd">Type<span class="colon">:</span></dt>
<dd class="field-odd"><p>list[int]</p>
</dd>
</dl>
</dd></dl>

<p>Create a <cite>Track</cite> resolver and load (or build) its graph.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>db_manager</strong> (<a class="reference internal" href="#idtrack._database_manager.DatabaseManager" title="idtrack._database_manager.DatabaseManager"><em>DatabaseManager</em></a>) – Connection manager that knows how to fetch Ensembl and
cross-reference tables from a local cache or a live MySQL mirror.
The same instance is forwarded to <code class="xref py py-class docutils literal notranslate"><span class="pre">_graph_maker.GraphMaker</span></code>.</p></li>
<li><p><strong>kwargs</strong> – Additional keyword arguments forwarded verbatim to
<code class="xref py py-meth docutils literal notranslate"><span class="pre">_graph_maker.GraphMaker.get_graph()</span></code>.  Common flags include
<cite>force_rebuild</cite> (recompute the graph from scratch), <cite>species</cite>
(restrict to one taxon), and <cite>cache_dir</cite> (override on-disk cache location).</p></li>
</ul>
</dd>
</dl>
<dl class="py property">
<dt class="sig sig-object py" id="idtrack._track.Track._calculate_node_scores_helper">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">_calculate_node_scores_helper</span></span><a class="headerlink" href="#idtrack._track.Track._calculate_node_scores_helper" title="Link to this definition"></a></dt>
<dd><p>Build and cache helper look-ups for node-scoring.</p>
<p>The property constructs two complementary data structures:</p>
<ul class="simple">
<li><dl class="simple">
<dt><strong>filter_set</strong> - the union of</dt><dd><p>(1) every external-database node-type present in the graph, and (2) every Ensembl-specific
node-type (gene, transcript, translation, …) across <em>all</em> assemblies.
This set can therefore be passed unmodified to
<a class="reference internal" href="#idtrack._track.Track.synonymous_nodes" title="idtrack._track.Track.synonymous_nodes"><code class="xref py py-meth docutils literal notranslate"><span class="pre">synonymous_nodes()</span></code></a> to ask for “anything that is not an assembly-less backbone gene”.</p>
</dd>
</dl>
</li>
<li><dl class="simple">
<dt><strong>ensembl_include</strong> - a mapping</dt><dd><p><cite>{form → set(node_type_str)}</cite> where each value lists the node-types
that should be considered equivalent to that <em>form</em> (e.g. <em>gene</em>, <em>transcript</em>, <em>translation</em>) when
computing richness metrics.</p>
</dd>
</dl>
</li>
</ul>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p><cite>(filter_set, ensembl_include)</cite> exactly as described above.</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p>tuple[set[str], dict[str, set[str]]]</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="idtrack._track.Track._choose_relevant_synonym_helper">
<span class="sig-name descname"><span class="pre">_choose_relevant_synonym_helper</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">from_id</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">synonym_ids</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">to_release</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">from_release</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">mode</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/idtrack/_track.html#Track._choose_relevant_synonym_helper"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#idtrack._track.Track._choose_relevant_synonym_helper" title="Link to this definition"></a></dt>
<dd><p>Select the <strong>most temporally relevant</strong> synonym(s) for an Ensembl gene-ID family.</p>
<p>The method evaluates each candidate in <cite>synonym_ids</cite> against the
<em>target</em> release <cite>to_release</cite> and, when applicable, the <em>source</em>
release <cite>from_release</cite>.  Its job is to decide <strong>where</strong> the path
should <em>enter</em> the Ensembl backbone and <strong>whether</strong> the remainder of
the traversal must run in <em>reverse</em> (new → old) order.</p>
<section id="selection-strategy">
<h2>Selection strategy<a class="headerlink" href="#selection-strategy" title="Link to this heading"></a></h2>
<ol class="arabic simple">
<li><p><strong>Fixed `from_release`</strong> - If the caller already knows the release
of the starting node, every candidate is paired with that same
release and the correct <em>reverse</em> flag is derived trivially.</p></li>
<li><p><strong>Non-backbone start</strong> - When the starting node is not an
Ensembl-gene backbone ID, the synonym whose <strong>active range edge</strong> is
<em>closest</em> (or <em>farthest</em>, per <cite>mode</cite>) to <cite>to_release</cite> is chosen.</p></li>
<li><p><strong>Backbone start</strong> - If the query is itself an Ensembl-gene, the
algorithm first looks for <em>overlapping</em> ranges between the query
and each synonym; if none overlap, it falls back to the distance
rule described in step 2.</p></li>
</ol>
<dl class="field-list simple">
<dt class="field-odd">param from_id<span class="colon">:</span></dt>
<dd class="field-odd"><p>Identifier from which the path search will start.</p>
</dd>
<dt class="field-even">type from_id<span class="colon">:</span></dt>
<dd class="field-even"><p>str</p>
</dd>
<dt class="field-odd">param synonym_ids<span class="colon">:</span></dt>
<dd class="field-odd"><p>Ensembl IDs considered synonyms of
<cite>from_id</cite> (typically the <em>same</em> gene with different version
numbers).</p>
</dd>
<dt class="field-even">type synonym_ids<span class="colon">:</span></dt>
<dd class="field-even"><p>Sequence[str]</p>
</dd>
<dt class="field-odd">param to_release<span class="colon">:</span></dt>
<dd class="field-odd"><p>Target Ensembl release that the overall conversion
aims for.</p>
</dd>
<dt class="field-even">type to_release<span class="colon">:</span></dt>
<dd class="field-even"><p>int</p>
</dd>
<dt class="field-odd">param from_release<span class="colon">:</span></dt>
<dd class="field-odd"><p>Release in which <cite>from_id</cite> is known to
be active.  If <em>None</em>, the method infers a suitable release for
each candidate.</p>
</dd>
<dt class="field-even">type from_release<span class="colon">:</span></dt>
<dd class="field-even"><p>int | None</p>
</dd>
<dt class="field-odd">param mode<span class="colon">:</span></dt>
<dd class="field-odd"><p>Either <cite>‘closest’</cite> or <cite>‘distant’</cite>—controls whether
the synonym chosen should minimise or maximise its distance to
<cite>to_release</cite>.</p>
</dd>
<dt class="field-even">type mode<span class="colon">:</span></dt>
<dd class="field-even"><p>str</p>
</dd>
<dt class="field-odd">returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>One or more triplets of the form
<cite>[synonym_id, entry_release, reverse]</cite> where:</p>
<ul class="simple">
<li><p><cite>synonym_id</cite> - the chosen synonym,</p></li>
<li><p><cite>entry_release</cite> - release at which to join the backbone, and</p></li>
<li><p><cite>reverse</cite> - <em>True</em> if the subsequent history walk must run
backwards in time.</p></li>
</ul>
</dd>
<dt class="field-even">rtype<span class="colon">:</span></dt>
<dd class="field-even"><p>list[list[Union[str, int, bool]]]</p>
</dd>
<dt class="field-odd">raises ValueError<span class="colon">:</span></dt>
<dd class="field-odd"><p>If no synonym satisfies the distance/overlap criteria
    or if <cite>mode</cite> is invalid.</p>
</dd>
</dl>
</section>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>to_release</strong> (<em>int</em>)</p></li>
<li><p><strong>from_release</strong> (<em>int</em><em> | </em><em>None</em>)</p></li>
<li><p><strong>mode</strong> (<em>str</em>)</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="idtrack._track.Track._create_priority_list_ensembl">
<span class="sig-name descname"><span class="pre">_create_priority_list_ensembl</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">from_id</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">to_release</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/idtrack/_track.html#Track._create_priority_list_ensembl"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#idtrack._track.Track._create_priority_list_ensembl" title="Link to this definition"></a></dt>
<dd><p>Build a priority list of assemblies in which <cite>from_id</cite> is active.</p>
<p>The priorities are the numeric <strong>assembly rankings</strong> defined in
<code class="xref py py-attr docutils literal notranslate"><span class="pre">DB.assembly_mysqlport_priority</span></code> (smaller numbers mean
higher priority).</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>from_id</strong> (<em>str</em>) – Ensembl gene identifier.</p></li>
<li><p><strong>to_release</strong> (<em>int</em>) – Target Ensembl release; only assemblies that
contain this release are considered.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>Sorted list of priority values (ascending).</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>list[int]</p>
</dd>
<dt class="field-even">Raises<span class="colon">:</span></dt>
<dd class="field-even"><p><strong>ValueError</strong> – If the gene is not active in <em>any</em> assembly at
    <cite>to_release</cite>.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="idtrack._track.Track._final_conversion">
<span class="sig-name descname"><span class="pre">_final_conversion</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">converted</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cnvt</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">final_database</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ens_release</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">return_path</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">return_ensembl_alternative</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/idtrack/_track.html#Track._final_conversion"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#idtrack._track.Track._final_conversion" title="Link to this definition"></a></dt>
<dd><p>Convert an Ensembl gene node to the requested external database.</p>
<p>Convert an Ensembl gene node to the requested external database and
merge the result back into <cite>converted</cite>.</p>
<p>The routine:</p>
<ol class="arabic simple">
<li><p>Builds every legal synonym path from <cite>cnvt</cite> to
<cite>final_database</cite> that is active in <cite>ens_release</cite> (or in any
release as a fallback).</p></li>
<li><p>Computes assembly-jump penalties for each path.</p></li>
<li><p>Calls <a class="reference internal" href="#idtrack._track.Track._final_conversion_dict_prepare" title="idtrack._track.Track._final_conversion_dict_prepare"><code class="xref py py-meth docutils literal notranslate"><span class="pre">_final_conversion_dict_prepare()</span></code></a> to create the
conversion sub-dict.</p></li>
<li><p>Optionally falls back to returning the Ensembl gene itself when no
synonym exists and <cite>return_ensembl_alternative</cite> is True.</p></li>
</ol>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>converted</strong> (<em>dict</em>) – The current accumulator being built by
<a class="reference internal" href="#idtrack._track.Track.convert" title="idtrack._track.Track.convert"><code class="xref py py-meth docutils literal notranslate"><span class="pre">convert()</span></code></a>.</p></li>
<li><p><strong>cnvt</strong> (<em>str</em>) – Ensembl gene identifier that is undergoing final
conversion.</p></li>
<li><p><strong>final_database</strong> (<em>str</em>) – Target external database.</p></li>
<li><p><strong>ens_release</strong> (<em>int</em>) – Target Ensembl release.</p></li>
<li><p><strong>return_path</strong> (<em>bool</em>) – If True, embed the path(s) that lead to each
synonym.</p></li>
<li><p><strong>return_ensembl_alternative</strong> (<em>bool</em>) – When no synonym can be found,
add a fallback entry that keeps the Ensembl gene.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>The same <cite>converted</cite> dict, updated in place (and also
returned for convenience).</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>dict</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="idtrack._track.Track._final_conversion_dict_prepare">
<em class="property"><span class="pre">static</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">_final_conversion_dict_prepare</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">confidence</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">sysns</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">paths</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">min_priority_list</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">len_priority_list</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">add_ass_jump_list</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">final_database</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/idtrack/_track.html#Track._final_conversion_dict_prepare"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#idtrack._track.Track._final_conversion_dict_prepare" title="Link to this definition"></a></dt>
<dd><p>Assemble the final-conversion section that will be attached to a candidate path.</p>
<p>The section contains a global conversion-confidence flag plus one
entry per synonym that survived the path-finding stage.  When
<cite>paths</cite> is None the structure is identical but omits the
<cite>‘the_path’</cite> member to save memory.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>confidence</strong> (<em>int</em><em> | </em><em>float</em>) – Heuristic confidence for the <em>whole</em>
conversion step - <cite>0</cite> for “perfect”, larger values for
fallback scenarios, <cite>np.inf</cite> when no conversion was
possible.</p></li>
<li><p><strong>sysns</strong> (<em>list</em>) – List of synonym identifiers <em>in the same order</em>
as the metric lists below.</p></li>
<li><p><strong>paths</strong> (<em>list</em><em>[</em><em>list</em><em>] </em><em>| </em><em>None</em>) – One walk (edge list) per synonym, or
<em>None</em> if the caller does not want to expose paths.</p></li>
<li><p><strong>min_priority_list</strong> (<em>list</em>) – Minimum assembly priority reached
by each walk.</p></li>
<li><p><strong>len_priority_list</strong> (<em>list</em>) – Number of distinct assembly
priorities encountered by each walk.</p></li>
<li><p><strong>add_ass_jump_list</strong> (<em>list</em>) – Additional assembly-jump penalty
incurred during the synonym hop itself.</p></li>
<li><p><strong>final_database</strong> (<em>str</em>) – Name of the database these synonyms belong
to (e.g. <cite>‘uniprot’</cite> or <cite>DB.nts_ensembl[“gene”]</cite>).</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>Nested dictionary ready to be stored under the key
<cite>‘final_conversion’</cite>.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>dict</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="idtrack._track.Track._minimum_assembly_jumps_helper">
<em class="property"><span class="pre">static</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">_minimum_assembly_jumps_helper</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">step_pri</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">current_priority</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">priorities</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/idtrack/_track.html#Track._minimum_assembly_jumps_helper"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#idtrack._track.Track._minimum_assembly_jumps_helper" title="Link to this definition"></a></dt>
<dd><p>Internal worker for <a class="reference internal" href="#idtrack._track.Track.minimum_assembly_jumps" title="idtrack._track.Track.minimum_assembly_jumps"><code class="xref py py-meth docutils literal notranslate"><span class="pre">minimum_assembly_jumps()</span></code></a>.</p>
<p>Given the priority sets for the remaining edges, iterate until all have
been consumed while updating the <em>current</em> assembly priority and counting
how often it must drop.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>step_pri</strong> (<em>list</em><em>[</em><em>int</em><em>]</em>) – Priority values of the edge currently under
consideration.</p></li>
<li><p><strong>current_priority</strong> (<em>int</em>) – Priority value inherited from previous steps.</p></li>
<li><p><strong>priorities</strong> (<em>list</em><em>[</em><em>list</em><em>[</em><em>int</em><em>]</em><em>]</em>) – Priority lists for the <em>rest</em> of the path,
<strong>already sorted</strong> for correct bisecting.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>Same three-tuple as documented in
<a class="reference internal" href="#idtrack._track.Track.minimum_assembly_jumps" title="idtrack._track.Track.minimum_assembly_jumps"><code class="xref py py-meth docutils literal notranslate"><span class="pre">minimum_assembly_jumps()</span></code></a>.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>tuple[int, list[int], int]</p>
</dd>
<dt class="field-even">Raises<span class="colon">:</span></dt>
<dd class="field-even"><p><strong>ValueError</strong> – If the priority lattice enters an impossible state.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="idtrack._track.Track._path_score_sorter_all_targets">
<span class="sig-name descname"><span class="pre">_path_score_sorter_all_targets</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">dict_of_dict</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">from_id</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">to_release</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/idtrack/_track.html#Track._path_score_sorter_all_targets"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#idtrack._track.Track._path_score_sorter_all_targets" title="Link to this definition"></a></dt>
<dd><p>Select the overall best target(s).</p>
<p>Select the <strong>overall best target(s)</strong> once every candidate Ensembl node
has itself been reduced to its single best path.</p>
<p>The method linearises several per-path metrics into an <em>importance
order</em> (see the tuple at the top of the function), then:</p>
<ol class="arabic simple">
<li><p>Computes that ordered score for <strong>each</strong> pair <cite>(ensembl_gene, final_target)</cite>.</p></li>
<li><dl class="simple">
<dt>Finds the global minimum; if multiple pairs tie:</dt><dd><ul class="simple">
<li><p>Prefer the target whose identifier is identical to <cite>from_id</cite>.</p></li>
<li><dl class="simple">
<dt>If more than one Ensembl gene still tie, fall back on</dt><dd><p><a class="reference internal" href="#idtrack._track.Track.calculate_node_scores" title="idtrack._track.Track.calculate_node_scores"><code class="xref py py-meth docutils literal notranslate"><span class="pre">calculate_node_scores()</span></code></a> to favour the “richer” node.</p>
</dd>
</dl>
</li>
</ul>
</dd>
</dl>
</li>
<li><p>Returns a <em>pruned</em> copy of <cite>dict_of_dict</cite> that contains only the
surviving Ensembl genes, each with only the winning <cite>final_elements</cite> entry.
Additional provenance is written to <cite>filter_scores</cite>.</p></li>
</ol>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>dict_of_dict</strong> (<em>dict</em>) – Nested result of
<a class="reference internal" href="#idtrack._track.Track.calculate_score_and_select" title="idtrack._track.Track.calculate_score_and_select"><code class="xref py py-meth docutils literal notranslate"><span class="pre">calculate_score_and_select()</span></code></a>.  Keys are candidate
Ensembl genes; values are dictionaries that already contain
<em>one</em> best path per final target.</p></li>
<li><p><strong>from_id</strong> (<em>str</em>) – Original query identifier; used to break ties in
favour of “same as input”.</p></li>
<li><p><strong>to_release</strong> (<em>int</em>) – Target Ensembl release; forwarded to
<a class="reference internal" href="#idtrack._track.Track.calculate_node_scores" title="idtrack._track.Track.calculate_node_scores"><code class="xref py py-meth docutils literal notranslate"><span class="pre">calculate_node_scores()</span></code></a> during tie-breaking.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><dl class="simple">
<dt>A reduced version of <cite>dict_of_dict</cite> holding only the winner(s) and enriched with a</dt><dd><p><cite>final_elements[*][‘filter_scores’]</cite> sub-dict that records the filters applied.</p>
</dd>
</dl>
</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>dict</p>
</dd>
<dt class="field-even">Raises<span class="colon">:</span></dt>
<dd class="field-even"><p><strong>ValueError</strong> – If <cite>dict_of_dict</cite> is empty.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="idtrack._track.Track._path_score_sorter_single_target">
<em class="property"><span class="pre">static</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">_path_score_sorter_single_target</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">lst_of_dict</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/idtrack/_track.html#Track._path_score_sorter_single_target"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#idtrack._track.Track._path_score_sorter_single_target" title="Link to this definition"></a></dt>
<dd><p>Select the best score dictionary for <em>one</em> conversion target.</p>
<p>The input is a list of dictionaries produced by
<a class="reference internal" href="#idtrack._track.Track.calculate_score_and_select" title="idtrack._track.Track.calculate_score_and_select"><code class="xref py py-meth docutils literal notranslate"><span class="pre">calculate_score_and_select()</span></code></a>.  Each dictionary is converted into a
tuple according to the <em>lexicographic importance order</em></p>
<p><cite>(“assembly_jump”, “external_jump”, “external_step”,
“edge_scores_reduced”, “ensembl_step”)</cite></p>
<p>and the dictionary with the <strong>smallest</strong> tuple is returned.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>lst_of_dict</strong> (<em>list</em><em>[</em><em>dict</em><em>]</em>) – Candidate score dictionaries for this target.</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>The chosen “winner” score dictionary.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>dict</p>
</dd>
<dt class="field-even">Raises<span class="colon">:</span></dt>
<dd class="field-even"><p><strong>ValueError</strong> – If the input list is empty.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="idtrack._track.Track._recursive_synonymous">
<span class="sig-name descname"><span class="pre">_recursive_synonymous</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">_the_id</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">synonymous_ones</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">synonymous_ones_db</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">filter_node_type</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">the_path</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">the_path_db</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">depth_max</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">from_release</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ensembl_backbone_shallow_search</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/idtrack/_track.html#Track._recursive_synonymous"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#idtrack._track.Track._recursive_synonymous" title="Link to this definition"></a></dt>
<dd><p>Helper method to be used in <code class="xref py py-meth docutils literal notranslate"><span class="pre">_graph.Track.synonymous_nodes()</span></code>.</p>
<p>Recursively explore the bio-ID graph to collect <strong>synonymous paths</strong>
starting at <cite>_the_id</cite> and ending on a node whose <em>type</em> is a member of <cite>filter_node_type</cite>.</p>
<p>A <em>path</em> is a list of node identifiers (<cite>_the_path</cite>) together with a
parallel list of their node-type strings (<cite>_the_path_db</cite>).
The search is <strong>breadth-limited</strong>: the <em>depth</em> of a path is defined as
the <strong>maximum count of any single node-type</strong> it contains
(e.g. a path with three <cite>external</cite> nodes has depth 3).  Recursion
stops when that depth would exceed <cite>depth_max</cite>.</p>
<p>Additional pruning rules:</p>
<ul class="simple">
<li><p>The walk never visits the same node twice (no cycles).</p></li>
<li><dl class="simple">
<dt>It never traverses two consecutive edges whose source and target</dt><dd><p>share the <strong>same node-type</strong>—this prevents “time-travel” within the
Ensembl history backbone.</p>
</dd>
</dl>
</li>
<li><dl class="simple">
<dt>When <cite>ensembl_backbone_shallow_search</cite> is <em>True</em>, the search is</dt><dd><p>restricted to the <strong>reverse</strong> direction except for node-types listed
in <code class="xref py py-attr docutils literal notranslate"><span class="pre">DB.nts_bidirectional_synonymous_search</span></code>.</p>
</dd>
</dl>
</li>
</ul>
<p>On reaching a terminating node the method <em>appends</em> the discovered
paths to <cite>synonymous_ones</cite> and <cite>synonymous_ones_db</cite>. It does not
return anything. Results are accumulated in <cite>synonymous_ones</cite> and <cite>synonymous_ones_db</cite>.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>_the_id</strong> (<em>str</em>) – Identifier of the starting node (Ensembl or
external).</p></li>
<li><p><strong>synonymous_ones</strong> (<em>list</em>) – Mutable list that will receive
each successful identifier path.</p></li>
<li><p><strong>synonymous_ones_db</strong> (<em>list</em>) – Mutable list that will receive
the corresponding node-type paths.</p></li>
<li><p><strong>filter_node_type</strong> (<em>set</em><em>[</em><em>str</em><em>]</em>) – Allowed node-types for the <strong>final</strong>
node of a path (e.g. <cite>{‘ensembl_gene’}</cite>).</p></li>
<li><p><strong>the_path</strong> (<em>list</em><em> | </em><em>None</em>) – Current path leading to
<cite>_the_id</cite>; <em>None</em> for the root invocation.</p></li>
<li><p><strong>the_path_db</strong> (<em>list</em><em> | </em><em>None</em>) – Node-type counterpart of
<cite>the_path</cite>; <em>None</em> for the root invocation.</p></li>
<li><p><strong>depth_max</strong> (<em>int</em>) – Maximum allowed depth as defined above.</p></li>
<li><p><strong>from_release</strong> (<em>int</em><em> | </em><em>None</em>) – If given, only keep terminal
nodes that are <em>active</em> in this Ensembl release.</p></li>
<li><p><strong>ensembl_backbone_shallow_search</strong> (<em>bool</em>) – Activate the
shallow, mostly-reverse search mode described above.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="idtrack._track.Track.calculate_node_scores">
<span class="sig-name descname"><span class="pre">calculate_node_scores</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">the_id</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ens_release</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/idtrack/_track.html#Track.calculate_node_scores"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#idtrack._track.Track.calculate_node_scores" title="Link to this definition"></a></dt>
<dd><p>Rank competing Ensembl targets by the “richness” of their synonyms.</p>
<p>The method counts, within a radius of <em>two</em> synonym hops, how many unique
identifiers of various categories point to each candidate and returns the
counts as <strong>negative integers</strong> so that <em>smaller</em> is <em>better</em> for the
up-stream sorter.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>the_id</strong> (<em>str</em>) – Identifier that is being converted.</p></li>
<li><p><strong>ens_release</strong> (<em>int</em>) – Target Ensembl release; only synonyms active in this
release are considered.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><dl class="simple">
<dt><cite>[-ext, -form₁, -form₂]</cite> where</dt><dd><ul class="simple">
<li><p><cite>ext</cite>  - number of distinct <strong>external-database</strong> synonyms.</p></li>
<li><p><cite>form₁</cite> - number of distinct synonyms of the most important Ensembl form (typically <em>gene</em>).</p></li>
<li><p><cite>form₂</cite> - number of distinct synonyms of the second form (typically <em>transcript</em> or <em>translation</em>).</p></li>
</ul>
</dd>
</dl>
</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>list</p>
</dd>
<dt class="field-even">Raises<span class="colon">:</span></dt>
<dd class="field-even"><p><strong>ValueError</strong> – If the graph does not expose exactly the two expected
    non-backbone forms, or if a synonym node’s type cannot be mapped to
    <em>external</em>, <em>form₁</em>, or <em>form₂</em>.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="idtrack._track.Track.calculate_score_and_select">
<span class="sig-name descname"><span class="pre">calculate_score_and_select</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">all_possible_paths</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">reduction</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">remove_na</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">from_releases</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">to_release</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">score_of_the_queried_item</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">return_path</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">from_id</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/idtrack/_track.html#Track.calculate_score_and_select"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#idtrack._track.Track.calculate_score_and_select" title="Link to this definition"></a></dt>
<dd><p>Collapse a set of candidate paths into the single best path per target.</p>
<p>For each path produced by the search engine the function:</p>
<p>1. Computes an <strong>edge-score aggregate</strong> using <cite>reduction</cite> while handling
missing values as directed by <cite>remove_na</cite>.
2. Tallies <em>external</em> statistics (steps, jumps, initial conversion
confidence) and <em>assembly</em> statistics (number of priority drops, final
priority).
3. Packs all metrics into a dictionary and stores it under the key of the
path’s final destination node.
4. Keeps only the lexicographically “smallest” dictionary per destination
via <a class="reference internal" href="#idtrack._track.Track._path_score_sorter_single_target" title="idtrack._track.Track._path_score_sorter_single_target"><code class="xref py py-meth docutils literal notranslate"><span class="pre">_path_score_sorter_single_target()</span></code></a>.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>all_possible_paths</strong> (<em>tuple</em>) – Sequence of edge-lists representing
every admissible walk returned by the path-finder.</p></li>
<li><p><strong>reduction</strong> (<em>Callable</em>) – Function such as
<cite>np.mean</cite> or <cite>sum</cite> used to collapse edge weights into one number.</p></li>
<li><p><strong>remove_na</strong> (<em>str</em>) – How to treat <em>NaN</em> edge weights - one of
<cite>‘omit’</cite>, <cite>‘to_1’</cite>, <cite>‘to_0’</cite>.</p></li>
<li><p><strong>from_releases</strong> (<em>Iterable</em><em>[</em><em>int</em><em>]</em>) – Release that each path starts from;
must align with <cite>all_possible_paths</cite>.</p></li>
<li><p><strong>to_release</strong> (<em>int</em>) – Target release - needed to know whether an edge is
traversed forward or reverse.</p></li>
<li><p><strong>score_of_the_queried_item</strong> (<em>float</em>) – Fallback weight for the implicit edge
that represents the <em>query ID</em> itself.</p></li>
<li><p><strong>return_path</strong> (<em>bool</em>) – If <em>True</em>, embed the full edge-list inside each
score dict under the key <cite>‘the_path’</cite>.</p></li>
<li><p><strong>from_id</strong> (<em>str</em>) – Original identifier being converted - echoed back in the
score dict for traceability.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>Mapping
<cite>{destination_id → best_score_dict}</cite>.
Each <em>score dict</em> contains (inter alia)
<cite>assembly_jump</cite>, <cite>external_jump</cite>, <cite>external_step</cite>,
<cite>edge_scores_reduced</cite>, and <cite>ensembl_step</cite>.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>dict</p>
</dd>
<dt class="field-even">Raises<span class="colon">:</span></dt>
<dd class="field-even"><p><strong>ValueError</strong> – If an unexpected edge encoding is encountered, if an edge
    score is invalid/∞, or if <cite>remove_na</cite> is set to an unknown mode.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="idtrack._track.Track.choose_relevant_synonym">
<span class="sig-name descname"><span class="pre">choose_relevant_synonym</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">the_id</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">depth_max</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">to_release</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">filter_node_type</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">from_release</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/idtrack/_track.html#Track.choose_relevant_synonym"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#idtrack._track.Track.choose_relevant_synonym" title="Link to this definition"></a></dt>
<dd><p>Wrapper that <strong>discovers, clusters, and ranks</strong> synonymous Ensembl candidates for a given identifier.</p>
<p>The function performs three steps:</p>
<ol class="arabic simple">
<li><p><strong>Discover</strong> paths to <em>all</em> Ensembl-gene nodes that share the same
biological identity (<cite>synonymous_nodes</cite>).</p></li>
<li><p><strong>Cluster</strong> those paths by <strong>gene ID</strong> (ignoring version).</p></li>
<li><p><strong>Rank</strong> each cluster with
<a class="reference internal" href="#idtrack._track.Track._choose_relevant_synonym_helper" title="idtrack._track.Track._choose_relevant_synonym_helper"><code class="xref py py-meth docutils literal notranslate"><span class="pre">_choose_relevant_synonym_helper()</span></code></a>, selecting the entry release
(and direction) that best suits <cite>to_release</cite>.</p></li>
</ol>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>the_id</strong> (<em>str</em>) – Source identifier (Ensembl or external).</p></li>
<li><p><strong>depth_max</strong> (<em>int</em>) – Maximum depth passed to
<a class="reference internal" href="#idtrack._track.Track.synonymous_nodes" title="idtrack._track.Track.synonymous_nodes"><code class="xref py py-meth docutils literal notranslate"><span class="pre">synonymous_nodes()</span></code></a>; governs how far the synonym search is
allowed to roam through external nodes.</p></li>
<li><p><strong>to_release</strong> (<em>int</em>) – Target Ensembl release required by the overall
conversion.</p></li>
<li><p><strong>filter_node_type</strong> (<em>set</em><em>[</em><em>str</em><em>]</em>) – Node-types that the synonym search
must terminate on (usually <cite>{‘ensembl_gene’}</cite>).</p></li>
<li><p><strong>from_release</strong> (<em>int</em><em> | </em><em>None</em>) – Known active release of <cite>the_id</cite>.  If
<em>None</em>, the helper will infer one.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><p>A list whose elements are</p>
<p><cite>[synonym_id, entry_release, reverse, identifier_path, node_type_path]</cite></p>
<p>where the last two items reproduce the path returned by
<a class="reference internal" href="#idtrack._track.Track.synonymous_nodes" title="idtrack._track.Track.synonymous_nodes"><code class="xref py py-meth docutils literal notranslate"><span class="pre">synonymous_nodes()</span></code></a>.</p>
</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>list[list[Any]]</p>
</dd>
</dl>
<p class="rubric">Notes</p>
<p><em>The method purposefully keeps **all*</em> equally-ranked candidates;
further tie-breaking is deferred to the main path-scoring routine.*</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="idtrack._track.Track.convert">
<span class="sig-name descname"><span class="pre">convert</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">from_id</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">from_release=None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">to_release=None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">final_database=None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">reduction=&lt;function</span> <span class="pre">mean&gt;</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">remove_na='omit'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">score_of_the_queried_item=nan</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">go_external=True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">prioritize_to_one_filter=False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">return_path=False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">deprioritize_lrg_genes=True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">return_ensembl_alternative=True</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/idtrack/_track.html#Track.convert"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#idtrack._track.Track.convert" title="Link to this definition"></a></dt>
<dd><p>End-to-end ID conversion workflow.</p>
<p>Starting from <cite>from_id</cite> the routine</p>
<ol class="arabic simple">
<li><p>Determines the correct <strong>time-travel direction</strong> if
<cite>from_release</cite> is unspecified.</p></li>
<li><p>Enumerates <em>all</em> admissible paths with
<a class="reference internal" href="#idtrack._track.Track.get_possible_paths" title="idtrack._track.Track.get_possible_paths"><code class="xref py py-meth docutils literal notranslate"><span class="pre">get_possible_paths()</span></code></a> (forward and/or reverse).</p></li>
<li><p>Collapses those paths with <a class="reference internal" href="#idtrack._track.Track.calculate_score_and_select" title="idtrack._track.Track.calculate_score_and_select"><code class="xref py py-meth docutils literal notranslate"><span class="pre">calculate_score_and_select()</span></code></a>.</p></li>
<li><p>Optionally converts the surviving Ensembl gene(s) into
<cite>final_database</cite> via <a class="reference internal" href="#idtrack._track.Track._final_conversion" title="idtrack._track.Track._final_conversion"><code class="xref py py-meth docutils literal notranslate"><span class="pre">_final_conversion()</span></code></a>.</p></li>
<li><p>Optionally applies a final global selection with
<a class="reference internal" href="#idtrack._track.Track._path_score_sorter_all_targets" title="idtrack._track.Track._path_score_sorter_all_targets"><code class="xref py py-meth docutils literal notranslate"><span class="pre">_path_score_sorter_all_targets()</span></code></a>.</p></li>
</ol>
<p>The output structure mirrors this decision tree and, when
<cite>return_path</cite> is <em>True</em>, embeds the full edge list so that callers
can audit every hop.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>from_id</strong> (<em>str</em>) – Source identifier (Ensembl, UniProt, RefSeq, …).</p></li>
<li><p><strong>from_release</strong> (<em>int</em><em> | </em><em>None</em>) – Starting Ensembl release.
<em>None</em> → infer from the graph.</p></li>
<li><p><strong>to_release</strong> (<em>int</em><em> | </em><em>None</em>) – Target Ensembl release.
Defaults to the newest release contained in the graph.</p></li>
<li><p><strong>final_database</strong> (<em>str</em><em> | </em><em>None</em>) – External database to
convert into.  <em>None</em> → stay on the Ensembl gene.</p></li>
<li><p><strong>reduction</strong> (<em>Callable</em>) – Function (e.g. <cite>numpy.mean</cite>)
used to collapse per-edge weights.  Must accept an iterable of
floats and return a float.</p></li>
<li><p><strong>remove_na</strong> (<em>str</em>) – Strategy for <em>NaN</em> edge weights -
<cite>‘omit’</cite>, <cite>‘to_1’</cite>, or <cite>‘to_0’</cite>.</p></li>
<li><p><strong>score_of_the_queried_item</strong> (<em>float</em>) – Weight assigned to
the implicit edge that represents <cite>from_id</cite> itself.</p></li>
<li><p><strong>go_external</strong> (<em>bool</em>) – Allow jumps through external
databases when the backbone is disconnected.</p></li>
<li><p><strong>prioritize_to_one_filter</strong> (<em>bool</em>) – After all scoring,
keep only the single globally best target.</p></li>
<li><p><strong>return_path</strong> (<em>bool</em>) – Embed the full edge list(s) in the
returned dictionary.</p></li>
<li><p><strong>deprioritize_lrg_genes</strong> (<em>bool</em>) – If <em>True</em> and other
results exist, drop LRG_* genomic regions from the final set.</p></li>
<li><p><strong>return_ensembl_alternative</strong> (<em>bool</em>) – When converting to an
external database, also return the Ensembl gene as a fallback.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><ul class="simple">
<li><p><strong>dict</strong> - Structured result as described above.</p></li>
<li><p><strong>None</strong> - No admissible path was found.</p></li>
</ul>
</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>dict | None</p>
</dd>
<dt class="field-even">Raises<span class="colon">:</span></dt>
<dd class="field-even"><p><strong>ValueError</strong> – For non-callable <cite>reduction</cite>, unsupported
    <cite>remove_na</cite> modes, unknown <cite>final_database</cite> values, or
    logical inconsistencies detected during processing.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="idtrack._track.Track.convert_optimized_multiple">
<span class="sig-name descname"><span class="pre">convert_optimized_multiple</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/idtrack/_track.html#Track.convert_optimized_multiple"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#idtrack._track.Track.convert_optimized_multiple" title="Link to this definition"></a></dt>
<dd><p>Placeholder for a <strong>batch-optimised</strong> converter.</p>
<p>The intended behaviour is to accept <em>multiple</em> query IDs and choose a
conversion target for each such that cross-sample clashes (e.g.
duplicate loci) are minimised.</p>
<dl class="field-list simple">
<dt class="field-odd">Raises<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>NotImplementedError</strong> – Always - the optimisation strategy is not yet implemented.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="idtrack._track.Track.edge_key_orientor">
<span class="sig-name descname"><span class="pre">edge_key_orientor</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">n1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">n2</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">n3</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/idtrack/_track.html#Track.edge_key_orientor"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#idtrack._track.Track.edge_key_orientor" title="Link to this definition"></a></dt>
<dd><p>Return the stored orientation of a multigraph edge.</p>
<p>For multigraphs every logical edge is stored <em>once</em>, but the caller may
hold <cite>(u, v, k)</cite> or <cite>(v, u, k)</cite>.
This helper resolves the ambiguity so that subsequent attribute look-ups
succeed.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>n1</strong> (<em>str</em>) – One endpoint of the edge.</p></li>
<li><p><strong>n2</strong> (<em>str</em>) – The other endpoint.</p></li>
<li><p><strong>n3</strong> (<em>int</em>) – Edge key (index) within the <cite>networkx</cite> multi-edge.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>A triple that is guaranteed to exist as written
in <cite>self.graph</cite>.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>tuple[str, str, int]</p>
</dd>
<dt class="field-even">Raises<span class="colon">:</span></dt>
<dd class="field-even"><p><strong>AssertionError</strong> – If neither orientation is present in the graph.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="idtrack._track.Track.get_from_release_and_reverse_vars">
<em class="property"><span class="pre">static</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">get_from_release_and_reverse_vars</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">lor</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">p</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">mode</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/idtrack/_track.html#Track.get_from_release_and_reverse_vars"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#idtrack._track.Track.get_from_release_and_reverse_vars" title="Link to this definition"></a></dt>
<dd><p>Derive a list of <cite>(release, reverse)</cite> tuples.</p>
<p>Derive a list of <cite>(release, reverse)</cite> tuples that indicate <em>which</em>
Ensembl release to start the graph walk from and <em>whether</em> that walk
should move <strong>backwards in time</strong>.</p>
<p>Given a collection of active-range intervals <cite>lor</cite> and a pivot
release <cite>p</cite>, the algorithm selects one or two release points per
interval depending on <cite>mode</cite>:</p>
<ul class="simple">
<li><dl class="simple">
<dt>‘closest’ - choose the release <strong>nearest</strong> to <cite>p</cite> within or at</dt><dd><p>the ends of the interval.</p>
</dd>
</dl>
</li>
<li><dl class="simple">
<dt>‘distant’ - choose the release <strong>farthest</strong> from <cite>p</cite> within the</dt><dd><p>interval.</p>
</dd>
</dl>
</li>
</ul>
<p>The boolean in each tuple is <em>True</em> when the walk should start <strong>after</strong>
the selected release and move backwards (i.e. “reverse mode”), and
<em>False</em> when it should move forwards.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>lor</strong> (<em>list</em>) – List of inclusive
<cite>(first_release, last_release)</cite> intervals in ascending order.</p></li>
<li><p><strong>p</strong> (<em>int</em>) – Pivot release around which “closest” or “distant” is
evaluated.</p></li>
<li><p><strong>mode</strong> (<em>str</em>) – Either <cite>‘closest’</cite> or <cite>‘distant’</cite>.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>Release / reverse-flag pairs, ordered in
the sequence they should be tried by the path-finder.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>list[tuple[int, bool]]</p>
</dd>
<dt class="field-even">Raises<span class="colon">:</span></dt>
<dd class="field-even"><p><strong>ValueError</strong> – If an interval in <cite>lor</cite> is malformed, <cite>mode</cite> is
    not recognised, or internal consistency checks fail.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="idtrack._track.Track.get_next_edges">
<span class="sig-name descname"><span class="pre">get_next_edges</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">from_id</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">from_release</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">reverse</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">debugging</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/idtrack/_track.html#Track.get_next_edges"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#idtrack._track.Track.get_next_edges" title="Link to this definition"></a></dt>
<dd><p>Enumerate <strong>chronologically admissible</strong> history edges from a node.</p>
<p>Starting at <cite>from_id</cite> and release <cite>from_release</cite>, the method scans
outgoing (or incoming, when <cite>reverse</cite> is <em>True</em>) edges whose
timestamps allow the path to <strong>advance</strong> in the desired temporal
direction.  It collapses duplicate “same-ID” transitions and flags
self-loops so that later heuristics can treat branch points and tips
differently.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>from_id</strong> (<em>str</em>) – Current node from which the search will step.</p></li>
<li><p><strong>from_release</strong> (<em>int</em>) – Release at which the current node is known to
exist.</p></li>
<li><p><strong>reverse</strong> (<em>bool</em>) – <em>False</em> to walk <strong>forward</strong> in history
(old → new), <em>True</em> to walk <strong>backward</strong> (new → old).</p></li>
<li><p><strong>debugging</strong> (<em>bool</em>) – If set, disables the duplicate-edge
collapse so that unit tests can inspect the raw edge set.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><p>Sorted list of edge
descriptors, each of which is</p>
<p><cite>[edge_release, is_self_loop, src_node, dst_node, multiedge_key]</cite>.</p>
</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>list[list[Union[int, bool, str, int]]]</p>
</dd>
<dt class="field-even">Raises<span class="colon">:</span></dt>
<dd class="field-even"><p><strong>ValueError</strong> – If inconsistent multi-edges (same nodes, same release)
    are detected—this signals a corrupted graph build.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="idtrack._track.Track.get_possible_paths">
<span class="sig-name descname"><span class="pre">get_possible_paths</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">from_id</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">from_release</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">to_release</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">reverse</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">go_external</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">increase_depth_until</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">2</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">increase_jump_until</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">from_release_inferred</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/idtrack/_track.html#Track.get_possible_paths"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#idtrack._track.Track.get_possible_paths" title="Link to this definition"></a></dt>
<dd><p>Run <a class="reference internal" href="#idtrack._track.Track.path_search" title="idtrack._track.Track.path_search"><code class="xref py py-meth docutils literal notranslate"><span class="pre">path_search()</span></code></a> under <strong>progressively relaxed settings</strong>.</p>
<p>Run <a class="reference internal" href="#idtrack._track.Track.path_search" title="idtrack._track.Track.path_search"><code class="xref py py-meth docutils literal notranslate"><span class="pre">path_search()</span></code></a> under <strong>progressively relaxed settings</strong> until at
least one viable path is found—or every relaxation level is exhausted.</p>
<p>Four search stages are attempted in order:</p>
<ol class="arabic simple">
<li><p><strong>Backbone-only</strong> - external jumps disabled.</p></li>
<li><dl class="simple">
<dt><strong>External enabled</strong> - allow external jumps; increment synonym depth</dt><dd><p>and jump limit after each failure up to
<cite>increase_depth_until</cite>/<cite>increase_jump_until</cite>.</p>
</dd>
</dl>
</li>
<li><dl class="simple">
<dt><strong>Backbone with multiple-Ensembl transition</strong> - external disabled but</dt><dd><p>permit starting release to shift on external nodes.</p>
</dd>
</dl>
</li>
<li><dl class="simple">
<dt><strong>External + multiple-transition</strong> - most permissive search, with</dt><dd><p>iterative depth/jump relaxation as in stage 2.</p>
</dd>
</dl>
</li>
</ol>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>from_id</strong> (<em>str</em>) – Identifier to convert.</p></li>
<li><p><strong>from_release</strong> (<em>int</em>) – Release at which the search begins.</p></li>
<li><p><strong>to_release</strong> (<em>int</em>) – Desired target release.</p></li>
<li><p><strong>reverse</strong> (<em>bool</em>) – Traverse the Ensembl history backwards if <em>True</em>,
forwards otherwise.</p></li>
<li><p><strong>go_external</strong> (<em>bool</em>) – If <em>False</em>, skip any stage that
requires external jumps.</p></li>
<li><p><strong>increase_depth_until</strong> (<em>int</em>) – Additional synonym-search
depth to allow beyond the default.</p></li>
<li><p><strong>increase_jump_until</strong> (<em>int</em>) – Additional external-jump count
to allow beyond the default.</p></li>
<li><p><strong>from_release_inferred</strong> (<em>bool</em>) – <em>Reserved for future use.</em>
Indicates that <em>from_release</em> was chosen automatically rather
than provided by the user.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>All paths discovered by the most restrictive stage that yielded
<strong>at least one</strong> result, returned as an immutable tuple.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>tuple[tuple[tuple[str, str, int]]]</p>
</dd>
</dl>
<p class="rubric">Notes</p>
<p>The function copies and mutates
<code class="xref py py-attr docutils literal notranslate"><span class="pre">DB.external_search_settings</span></code> internally; the caller’s
copy is not modified.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="idtrack._track.Track.identify_source">
<span class="sig-name descname"><span class="pre">identify_source</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">dataset_ids</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">mode</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/idtrack/_track.html#Track.identify_source"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#idtrack._track.Track.identify_source" title="Link to this definition"></a></dt>
<dd><p>Infer the most likely origin (assembly and/or Ensembl release) of a heterogeneous identifier list.</p>
<p>The function tallies how often each origin triple appears among
<cite>dataset_ids</cite> and returns the counts sorted in descending order.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>dataset_ids</strong> (<em>list</em><em>[</em><em>str</em><em>]</em>) – Collection of identifiers to analyse.</p></li>
<li><p><strong>mode</strong> (<em>str</em>) – Granularity of the origin to extract - one of
- ‘complete’                  → (assembly, db, release)
- ‘ensembl_release’           → release only
- ‘assembly’                  → assembly only
- ‘assembly_ensembl_release’  → (assembly, release)</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>Pairs <cite>(origin, count)</cite> sorted by
frequency.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>list[tuple[Any, int]]</p>
</dd>
<dt class="field-even">Raises<span class="colon">:</span></dt>
<dd class="field-even"><p><strong>ValueError</strong> – If <cite>mode</cite> is not one of the recognised values.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="idtrack._track.Track.minimum_assembly_jumps">
<span class="sig-name descname"><span class="pre">minimum_assembly_jumps</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">the_path</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">step_pri</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">current_priority</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/idtrack/_track.html#Track.minimum_assembly_jumps"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#idtrack._track.Track.minimum_assembly_jumps" title="Link to this definition"></a></dt>
<dd><p>Compute the penalty incurred by assembly downgrades along a path.</p>
<p>Each path step may be annotated with one or more candidate assemblies.
These are translated into priority values via
<cite>DB.assembly_mysqlport_priority</cite>.  The algorithm walks the path,
tracking the current priority and counting how many times it must drop
to a lower priority value—each drop constitutes an “assembly jump”
penalty.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>the_path</strong> (<em>Iterable</em><em>[</em><em>tuple</em><em>]</em>) – Sequence of edge descriptors; each element is
either <cite>(n1, n2, k)</cite> or <cite>(n1, n2, k, release)</cite>.</p></li>
<li><p><strong>step_pri</strong> (<em>list</em><em>[</em><em>int</em><em>] </em><em>| </em><em>None</em><em>, </em><em>optional</em>) – Priority list for the first edge.
If None, it is derived from <cite>the_path</cite>.</p></li>
<li><p><strong>current_priority</strong> (<em>int</em><em> | </em><em>None</em><em>, </em><em>optional</em>) – Starting priority.  If None,
initialised to <cite>max(step_pri)</cite>.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><ul class="simple">
<li><p><cite>assembly_jump</cite> - total number of priority drops.</p></li>
<li><p><cite>step_pri</cite> - priority list of the last processed edge.</p></li>
<li><p><cite>current_priority</cite> - priority value after the final edge.</p></li>
</ul>
</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>tuple[int, list[int], int]</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="idtrack._track.Track.path_search">
<span class="sig-name descname"><span class="pre">path_search</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">from_id</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">from_release</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">to_release</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">reverse</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">external_settings</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">external_jump</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">multiple_ensembl_transition</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/idtrack/_track.html#Track.path_search"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#idtrack._track.Track.path_search" title="Link to this definition"></a></dt>
<dd><p>Enumerate every admissible <strong>history path</strong> from <em>from_id</em> at <em>from_release</em> to <em>to_release</em>.</p>
<p>The algorithm performs a depth-first traversal of the Ensembl history
edges.  Whenever it becomes stranded on a non-backbone node it may
“beam-up” via a <em>synonym path</em> through an external database, subject to
the constraints in <cite>external_settings</cite>:</p>
<ul class="simple">
<li><p><cite>synonymous_max_depth</cite> - maximum depth of a synonym search.</p></li>
<li><p><cite>jump_limit</cite> - maximum number of external “beam-up” jumps allowed.</p></li>
<li><p><cite>nts_backbone</cite> - canonical node-type of the Ensembl backbone.</p></li>
</ul>
<p>Additional flags control the initial conditions:</p>
<ul class="simple">
<li><dl class="simple">
<dt>Setting <cite>external_jump</cite> to <em>np.inf</em> <strong>disables</strong> external jumps.</dt><dd><p>Setting it to <em>None</em> <strong>enables</strong> them with the counter reset to <cite>0</cite>.</p>
</dd>
</dl>
</li>
<li><dl class="simple">
<dt><cite>multiple_ensembl_transition</cite> allows the algorithm to time-travel to</dt><dd><p>a <em>different</em> release while still on an external node; this is useful
when <em>from_release</em> was inferred and might not actually connect.</p>
</dd>
</dl>
</li>
</ul>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>from_id</strong> (<em>str</em>) – Identifier to start the search from.</p></li>
<li><p><strong>from_release</strong> (<em>int</em>) – Release number where <em>from_id</em> is considered
active.</p></li>
<li><p><strong>to_release</strong> (<em>int</em>) – Target Ensembl release.</p></li>
<li><p><strong>reverse</strong> (<em>bool</em>) – If <em>True</em>, traverse the graph <strong>backwards</strong> in time;
otherwise forwards.</p></li>
<li><p><strong>external_settings</strong> (<em>dict</em>) – Copy of
<code class="xref py py-attr docutils literal notranslate"><span class="pre">DB.external_search_settings</span></code> that governs depth,
jump limits, and backbone node-type.</p></li>
<li><p><strong>external_jump</strong> (<em>float</em><em> | </em><em>None</em>) – Current external-jump count
(<em>None</em> starts from zero, <em>np.inf</em> forbids any jump).</p></li>
<li><p><strong>multiple_ensembl_transition</strong> (<em>bool</em>) – Permit the synonym
engine to select a <em>different</em> release for an external node when
no path exists at <em>from_release</em>.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>A set of <strong>edge-lists</strong>.  Each edge is stored as
<cite>(src, dst, key)</cite>; an empty walk that terminates immediately
is represented by <cite>((None, from_id, None),)</cite>.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>set[tuple[tuple[str, str, int]]]</p>
</dd>
</dl>
<p class="rubric">Notes</p>
<p><em>The method is intentionally side-effect free; it constructs all
intermediate data on the stack and returns a fresh set.</em></p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="idtrack._track.Track.path_step_possible_assembly_jumps">
<span class="sig-name descname"><span class="pre">path_step_possible_assembly_jumps</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">n1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">n2</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">n3</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">n4</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/idtrack/_track.html#Track.path_step_possible_assembly_jumps"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#idtrack._track.Track.path_step_possible_assembly_jumps" title="Link to this definition"></a></dt>
<dd><p>Return the genome assemblies that can legally be used for a single edge.</p>
<p>The helper inspects the edge that connects <cite>n1</cite> → <cite>n2</cite> and filters the
assemblies recorded on that edge against the <em>release</em> constraint <cite>n4</cite>:</p>
<ul class="simple">
<li><dl class="simple">
<dt><strong>None</strong> - the edge is treated as backbone history; the result is the</dt><dd><p>graph-wide default assembly (usually the build on which the backbone was
constructed).</p>
</dd>
</dl>
</li>
<li><p><strong>int</strong> - keep only assemblies whose <em>release set</em> contains that single release.</p></li>
<li><p><strong>set[int]</strong> - keep assemblies whose release set intersects the provided set.</p></li>
</ul>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>n1</strong> (<em>str</em>) – Source node identifier.</p></li>
<li><p><strong>n2</strong> (<em>str</em>) – Destination node identifier.</p></li>
<li><p><strong>n3</strong> (<em>int</em>) – Edge key within the NetworkX multigraph.</p></li>
<li><p><strong>n4</strong> (<em>int</em><em> | </em><em>set</em><em>[</em><em>int</em><em>] </em><em>| </em><em>None</em><em>, </em><em>optional</em>) – Release filter as described above.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>Sorted list of assembly names (e.g. <cite>[“GRCh37”, “GRCh38”]</cite>).</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>list[str]</p>
</dd>
<dt class="field-even">Raises<span class="colon">:</span></dt>
<dd class="field-even"><p><strong>ValueError</strong> – If <cite>n4</cite> is of an unsupported type.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="idtrack._track.Track.should_graph_reversed">
<span class="sig-name descname"><span class="pre">should_graph_reversed</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">from_id</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">to_release</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/idtrack/_track.html#Track.should_graph_reversed"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#idtrack._track.Track.should_graph_reversed" title="Link to this definition"></a></dt>
<dd><p>Determine the <strong>temporal orientation</strong> of the graph walk.</p>
<p>Given an identifier that is active in one or more release intervals,
the routine decides whether the subsequent path-finder must move
<strong>forward in time</strong>, <strong>backward in time</strong>, or explore <em>both</em> directions
in parallel in order to reach the target release.</p>
<p>The decision is based on the <em>closest</em> boundary of every active
interval returned by
<a class="reference internal" href="#idtrack._track.Track.get_from_release_and_reverse_vars" title="idtrack._track.Track.get_from_release_and_reverse_vars"><code class="xref py py-meth docutils literal notranslate"><span class="pre">Track.get_from_release_and_reverse_vars()</span></code></a> (<cite>mode=’closest’</cite>).</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>from_id</strong> (<em>str</em>) – The starting identifier (Ensembl gene, transcript,
protein, or external ID).</p></li>
<li><p><strong>to_release</strong> (<em>int</em>) – The Ensembl release the user wishes to convert
<em>to</em>.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><ul class="simple">
<li><dl class="simple">
<dt><cite>‘forward’</cite> - walk old → new, starting at the <strong>earliest</strong></dt><dd><p>release in which <em>from_id</em> is active
&amp;nbsp;&amp;nbsp;&amp;nbsp;→ return <cite>(‘forward’, start_release)</cite></p>
</dd>
</dl>
</li>
<li><dl class="simple">
<dt><cite>’reverse’</cite> - walk new → old, starting at the <strong>latest</strong></dt><dd><p>active release
&amp;nbsp;&amp;nbsp;&amp;nbsp;→ return <cite>(‘reverse’, start_release)</cite></p>
</dd>
</dl>
</li>
<li><dl class="simple">
<dt><cite>’both’</cite> - split search: one forward walk and one reverse</dt><dd><p>walk
&amp;nbsp;&amp;nbsp;&amp;nbsp;→ return <cite>(‘both’, (forward_start,
reverse_start))</cite></p>
</dd>
</dl>
</li>
</ul>
</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>tuple[str, Union[int, tuple[int, int]]]</p>
</dd>
<dt class="field-even">Raises<span class="colon">:</span></dt>
<dd class="field-even"><p><strong>ValueError</strong> – If <em>from_id</em> is never active in or around
    <cite>to_release</cite> (i.e. no viable starting release can be found).</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="idtrack._track.Track.synonymous_nodes">
<span class="sig-name descname"><span class="pre">synonymous_nodes</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">the_id</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">depth_max</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">filter_node_type</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">from_release</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ensembl_backbone_shallow_search</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/idtrack/_track.html#Track.synonymous_nodes"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#idtrack._track.Track.synonymous_nodes" title="Link to this definition"></a></dt>
<dd><p>Public wrapper around <a class="reference internal" href="#idtrack._track.Track._recursive_synonymous" title="idtrack._track.Track._recursive_synonymous"><code class="xref py py-meth docutils literal notranslate"><span class="pre">_recursive_synonymous()</span></code></a>.</p>
<p>The method returns <strong>all minimal-length synonym paths</strong> emanating from <cite>the_id</cite>.</p>
<p>The function first runs a <em>default</em> depth search determined by
<cite>DB.external_search_settings[‘synonymous_max_depth’]</cite>.
If no synonym is found and <cite>depth_max</cite> is greater than that default,
a second, deeper search is attempted.</p>
<p>For every distinct <em>target</em> node the shortest path is kept; longer
paths to the same target are discarded.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>the_id</strong> (<em>str</em>) – Source identifier.</p></li>
<li><p><strong>depth_max</strong> (<em>int</em>) – Maximum search depth to try if the default search fails.</p></li>
<li><p><strong>filter_node_type</strong> (<em>set</em><em>[</em><em>str</em><em>]</em>) – Node-types that are acceptable for the
<em>target</em> node(s).  Must <em>not</em> include the generic
<cite>‘external’</cite> type—specify the concrete external DB instead.</p></li>
<li><p><strong>from_release</strong> (<em>int</em><em> | </em><em>None</em>) – Constrain targets to those active in this Ensembl release.</p></li>
<li><p><strong>ensembl_backbone_shallow_search</strong> (<em>bool</em>) – If <em>True</em>,
restricts the graph traversal as explained in <a class="reference internal" href="#idtrack._track.Track._recursive_synonymous" title="idtrack._track.Track._recursive_synonymous"><code class="xref py py-meth docutils literal notranslate"><span class="pre">_recursive_synonymous()</span></code></a>.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><dl class="simple">
<dt>A list whose elements are <cite>[identifier_path, node_type_path]</cite> pairs,</dt><dd><p>each representing the minimal route to one synonymous node.</p>
</dd>
</dl>
</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>list[list[list[str]]]</p>
</dd>
<dt class="field-even">Raises<span class="colon">:</span></dt>
<dd class="field-even"><p><strong>ValueError</strong> – If <cite>filter_node_type</cite> improperly contains the generic
    external type, or if <cite>depth_max</cite> is incompatible with <cite>ensembl_backbone_shallow_search</cite>.</p>
</dd>
</dl>
</dd></dl>

</dd></dl>

<dl class="py class" id="module-idtrack._track_tests">
<dt class="sig sig-object py" id="idtrack._track_tests.TrackTests">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">TrackTests</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="o"><span class="pre">*</span></span><span class="n"><span class="pre">args</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/idtrack/_track_tests.html#TrackTests"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#idtrack._track_tests.TrackTests" title="Link to this definition"></a></dt>
<dd><p>Bases: <a class="reference internal" href="#idtrack._track.Track" title="idtrack._track.Track"><code class="xref py py-class docutils literal notranslate"><span class="pre">Track</span></code></a>, <code class="xref py py-class docutils literal notranslate"><span class="pre">ABC</span></code></p>
<p>Developer-facing integrity-test harness for <code class="xref py py-class docutils literal notranslate"><span class="pre">Track</span></code>.</p>
<p>This module defines <a class="reference internal" href="#idtrack._track_tests.TrackTests" title="idtrack._track_tests.TrackTests"><code class="xref py py-class docutils literal notranslate"><span class="pre">TrackTests</span></code></a>, a mix-in that adds an extensive
white-box test suite to a populated <code class="xref py py-class docutils literal notranslate"><span class="pre">idtrack.Track</span></code> instance.  The
class is <strong>for developers only</strong>; it should never be used in production
pipelines.  Every public method beginning with <code class="docutils literal notranslate"><span class="pre">is_</span></code> returns a boolean
that tells whether a specific invariant holds.  Methods beginning with
<code class="docutils literal notranslate"><span class="pre">history_</span></code> execute heavier, end-to-end conversions and collect rich
statistics. The class is intended to be <em>mixin-ed</em> into a concrete Track
subclass—or instantiated standalone—<strong>after</strong> the underlying graph and
lookup tables have been fully built.  It performs purely read-only
operations and therefore imposes no risk of mutating state.</p>
<p>All test methods share the following contract:</p>
<ul class="simple">
<li><p>They never raise on failure—<em>return-value only</em>—so they can be run in
bulk without interrupting your session.</p></li>
<li><p>A return value of <code class="docutils literal notranslate"><span class="pre">True</span></code> means the invariant holds; <code class="docutils literal notranslate"><span class="pre">False</span></code> means
a violation was detected.</p></li>
<li><p>Where useful, a <em>verbose</em> flag gives a tqdm progress bar so long-
running checks remain user-friendly.</p></li>
</ul>
<p>Typical use:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">tests</span> <span class="o">=</span> <span class="n">TrackTests</span><span class="p">(</span><span class="o">...</span><span class="p">)</span>
<span class="n">tests</span><span class="o">.</span><span class="n">is_id_functions_consistent_ensembl</span><span class="p">()</span>  <span class="c1"># Raises if inconsistent.</span>
</pre></div>
</div>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>The class is <em>not</em> designed for production; instantiate it only in test
suites or interactive debugging sessions.</p>
</div>
<p>Initialize the test harness.</p>
<p>All positional and keyword arguments are forwarded verbatim to
<code class="xref py py-class docutils literal notranslate"><span class="pre">__init__</span></code>.  Besides constructing the underlying
graph, the initializer sets up a dedicated <code class="xref py py-data docutils literal notranslate"><span class="pre">logging.Logger</span></code>
named <code class="docutils literal notranslate"><span class="pre">&quot;track_tests&quot;</span></code> so individual test routines can emit structured
diagnostics without polluting the main application log.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>args</strong> – Positional arguments accepted by <code class="xref py py-class docutils literal notranslate"><span class="pre">__init__</span></code>.</p></li>
<li><p><strong>kwargs</strong> – Keyword arguments accepted by <code class="xref py py-class docutils literal notranslate"><span class="pre">__init__</span></code>.</p></li>
</ul>
</dd>
</dl>
<dl class="py method">
<dt class="sig sig-object py" id="idtrack._track_tests.TrackTests.history_travel_testing">
<span class="sig-name descname"><span class="pre">history_travel_testing</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">from_release</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">from_assembly</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">from_database</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">to_release</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">to_database</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">go_external</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">prioritize_to_one_filter</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">convert_using_release</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">from_fraction</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1.0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">verbose</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">verbose_detailed</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">return_ensembl_alternative</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/idtrack/_track_tests.html#TrackTests.history_travel_testing"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#idtrack._track_tests.TrackTests.history_travel_testing" title="Link to this definition"></a></dt>
<dd><p>End-to-end conversion regression test.</p>
<p>This is the <em>work-horse</em> diagnostic routine: it converts a (possibly
random) subset of source IDs to <em>to_database</em>/<em>to_release</em> using the
full path-finder and records detailed metrics along the way.</p>
<p>The function is intentionally permissive - it never aborts on a single
failed conversion but rather builds up a comprehensive <em>metrics</em>
dictionary for post-mortem analysis.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>from_release</strong> (<em>int</em>) – Source Ensembl release.</p></li>
<li><p><strong>from_assembly</strong> (<em>int</em>) – Source genome assembly.</p></li>
<li><p><strong>from_database</strong> (<em>str</em>) – Name of the starting database (Ensembl backbone or
external).</p></li>
<li><p><strong>to_release</strong> (<em>int</em>) – Destination Ensembl release.</p></li>
<li><p><strong>to_database</strong> (<em>str</em>) – Name of the destination database.</p></li>
<li><p><strong>go_external</strong> (<em>bool</em>) – Allow the path-finder to leave the Ensembl lineage at
intermediate steps (recommended <em>True</em> unless you explicitly
want a pure Ensembl mapping).</p></li>
<li><p><strong>prioritize_to_one_filter</strong> (<em>bool</em>) – When <em>True</em> the converter ranks 1→1 paths
above many→many alternatives.  This mimics the behaviour of the
public conversion API.</p></li>
<li><p><strong>convert_using_release</strong> (<em>bool</em>) – If <em>True</em> the source release is passed to
<code class="xref py py-meth docutils literal notranslate"><span class="pre">idtrack.Track.convert()</span></code> so the path-finder need not
<cite>guess</cite> the starting point.</p></li>
<li><p><strong>from_fraction</strong> (<em>float</em>) – Fraction (0.0 - 1.0] of <em>ids_from</em> to sample.  Use
&lt; 1.0 for a quicker, stochastic smoke test.</p></li>
<li><p><strong>verbose</strong> (<em>bool</em>) – Show coarse progress information via tqdm.</p></li>
<li><p><strong>verbose_detailed</strong> (<em>bool</em>) – Include additional per-ID counters in the tqdm
suffix to watch the metrics grow in real time.</p></li>
<li><p><strong>return_ensembl_alternative</strong> (<em>bool</em>) – Forwarded to
<code class="xref py py-meth docutils literal notranslate"><span class="pre">idtrack.Track.convert()</span></code>; when <em>True</em> the converter also
reports Ensembl gene alternatives that would lead to the same
final IDs.</p></li>
</ul>
</dd>
<dt class="field-even">Raises<span class="colon">:</span></dt>
<dd class="field-even"><p><strong>ValueError</strong> – <em>from_database</em> or <em>to_database</em> is an Ensembl
    node-type that should be addressed through the backbone helper
    or <em>from_fraction</em> is outside (0.0, 1.0].</p>
</dd>
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>A nested dictionary with keys <code class="docutils literal notranslate"><span class="pre">parameters</span></code>, <code class="docutils literal notranslate"><span class="pre">time</span></code> and
multiple sub-sections such as <code class="docutils literal notranslate"><span class="pre">lost_item</span></code>, <code class="docutils literal notranslate"><span class="pre">one_to_one_ids</span></code>,
<code class="docutils literal notranslate"><span class="pre">one_to_multiple_ids</span></code>, <code class="docutils literal notranslate"><span class="pre">conversion</span></code> as described in the source
code.</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p>dict</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="idtrack._track_tests.TrackTests.history_travel_testing_random">
<span class="sig-name descname"><span class="pre">history_travel_testing_random</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">from_fraction</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">strict_forward</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">convert_using_release</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">prioritize_to_one_filter</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">verbose</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">verbose_detailed</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/idtrack/_track_tests.html#TrackTests.history_travel_testing_random"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#idtrack._track_tests.TrackTests.history_travel_testing_random" title="Link to this definition"></a></dt>
<dd><p>Convenience wrapper around <a class="reference internal" href="#idtrack._track_tests.TrackTests.history_travel_testing" title="idtrack._track_tests.TrackTests.history_travel_testing"><code class="xref py py-meth docutils literal notranslate"><span class="pre">history_travel_testing()</span></code></a>.</p>
<p>The routine generates a <em>random</em> but internally consistent test case
via <a class="reference internal" href="#idtrack._track_tests.TrackTests.history_travel_testing_random_arguments_generator" title="idtrack._track_tests.TrackTests.history_travel_testing_random_arguments_generator"><code class="xref py py-meth docutils literal notranslate"><span class="pre">history_travel_testing_random_arguments_generator()</span></code></a>, logs
the chosen parameters (unless <em>verbose</em> is <em>False</em>) and delegates the
heavy lifting to <a class="reference internal" href="#idtrack._track_tests.TrackTests.history_travel_testing" title="idtrack._track_tests.TrackTests.history_travel_testing"><code class="xref py py-meth docutils literal notranslate"><span class="pre">history_travel_testing()</span></code></a>.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>from_fraction</strong> (<em>float</em>) – Fraction of IDs to sample from the source set.</p></li>
<li><p><strong>strict_forward</strong> (<em>bool</em>) – Forwarded to the argument generator.</p></li>
<li><p><strong>convert_using_release</strong> (<em>bool</em>) – Forwarded to
<a class="reference internal" href="#idtrack._track_tests.TrackTests.history_travel_testing" title="idtrack._track_tests.TrackTests.history_travel_testing"><code class="xref py py-meth docutils literal notranslate"><span class="pre">history_travel_testing()</span></code></a>.</p></li>
<li><p><strong>prioritize_to_one_filter</strong> (<em>bool</em>) – Forwarded to
<a class="reference internal" href="#idtrack._track_tests.TrackTests.history_travel_testing" title="idtrack._track_tests.TrackTests.history_travel_testing"><code class="xref py py-meth docutils literal notranslate"><span class="pre">history_travel_testing()</span></code></a>.</p></li>
<li><p><strong>verbose</strong> (<em>bool</em>) – Show coarse progress information.</p></li>
<li><p><strong>verbose_detailed</strong> (<em>bool</em>) – Include extended per-ID counters in the progress
bar.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>The <em>metrics</em> dictionary returned by
<a class="reference internal" href="#idtrack._track_tests.TrackTests.history_travel_testing" title="idtrack._track_tests.TrackTests.history_travel_testing"><code class="xref py py-meth docutils literal notranslate"><span class="pre">history_travel_testing()</span></code></a>.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>dict</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="idtrack._track_tests.TrackTests.history_travel_testing_random_arguments_generator">
<span class="sig-name descname"><span class="pre">history_travel_testing_random_arguments_generator</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">strict_forward</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/idtrack/_track_tests.html#TrackTests.history_travel_testing_random_arguments_generator"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#idtrack._track_tests.TrackTests.history_travel_testing_random_arguments_generator" title="Link to this definition"></a></dt>
<dd><p>Generate a plausible random parameter set for <a class="reference internal" href="#idtrack._track_tests.TrackTests.history_travel_testing" title="idtrack._track_tests.TrackTests.history_travel_testing"><code class="xref py py-meth docutils literal notranslate"><span class="pre">history_travel_testing()</span></code></a>.</p>
<p>The helper picks <strong>compatible</strong> source/target assemblies, releases and
databases so the subsequent conversion test has a realistic chance to
succeed.  When <em>strict_forward</em> is <em>True</em> the target release is
guaranteed to be <strong>≥</strong> the source release (no time-travel back).</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>strict_forward</strong> (<em>bool</em>) – Enforce a non-decreasing release direction.</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>Keys <code class="docutils literal notranslate"><span class="pre">from_assembly</span></code>, <code class="docutils literal notranslate"><span class="pre">from_release</span></code>, <code class="docutils literal notranslate"><span class="pre">to_release</span></code>,
<code class="docutils literal notranslate"><span class="pre">from_database</span></code>, <code class="docutils literal notranslate"><span class="pre">to_database</span></code> ready to be splatted into
<a class="reference internal" href="#idtrack._track_tests.TrackTests.history_travel_testing" title="idtrack._track_tests.TrackTests.history_travel_testing"><code class="xref py py-meth docutils literal notranslate"><span class="pre">history_travel_testing()</span></code></a>.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>dict</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="idtrack._track_tests.TrackTests.how_many_corresponding_path_ensembl">
<span class="sig-name descname"><span class="pre">how_many_corresponding_path_ensembl</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">from_release</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">from_assembly</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">to_release</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">go_external</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">verbose</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/idtrack/_track_tests.html#TrackTests.how_many_corresponding_path_ensembl"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#idtrack._track_tests.TrackTests.how_many_corresponding_path_ensembl" title="Link to this definition"></a></dt>
<dd><p>Count history paths between two Ensembl releases.</p>
<p>The method iterates over <strong>all</strong> Ensembl-gene stable IDs that exist in
<em>from_release</em>/<em>from_assembly</em>.  For every ID that is present in the
graph it calls <code class="xref py py-meth docutils literal notranslate"><span class="pre">idtrack.Track.get_possible_paths()</span></code> and records
how many distinct paths the searcher finds to <em>to_release</em>.</p>
<p>The routine is <strong>non-destructive</strong>; it merely provides a quick way to
gauge the <cite>density</cite> of the history graph or to spot releases where
path-finding was unexpectedly difficult.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>from_release</strong> (<em>int</em>) – Source Ensembl release number.</p></li>
<li><p><strong>from_assembly</strong> (<em>int</em>) – Source genome assembly code.</p></li>
<li><p><strong>to_release</strong> (<em>int</em>) – Target Ensembl release number.</p></li>
<li><p><strong>go_external</strong> (<em>bool</em>) – If <em>True</em> history paths are allowed to temporarily
leave the Ensembl lineage via external databases.</p></li>
<li><p><strong>verbose</strong> (<em>bool</em>) – Show a tqdm progress bar (default <em>True</em>).</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><p>A list of two-element sub-lists
<code class="docutils literal notranslate"><span class="pre">[stable_id,</span> <span class="pre">n_paths]</span></code> where <code class="docutils literal notranslate"><span class="pre">n_paths</span></code> is</p>
<blockquote>
<div><ul class="simple">
<li><p>an <em>int</em> ≥ 0 when the ID was in the graph, or</p></li>
<li><p><em>None</em> when the source ID was absent.</p></li>
</ul>
</div></blockquote>
</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>list[list[Union[str, int, None]]]</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="idtrack._track_tests.TrackTests.is_base_is_range_correct">
<span class="sig-name descname"><span class="pre">is_base_is_range_correct</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">verbose</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/idtrack/_track_tests.html#TrackTests.is_base_is_range_correct"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#idtrack._track_tests.TrackTests.is_base_is_range_correct" title="Link to this definition"></a></dt>
<dd><p>Verify consistency of <em>base-gene</em> active-range calculations.</p>
<p>Each “base Ensembl gene” node (<code class="docutils literal notranslate"><span class="pre">node_type</span> <span class="pre">==</span> <span class="pre">'base_ensembl_gene'</span></code>)
has an <em>active release range</em>—the list of Ensembl releases during
which descendants of the gene were present.  There are two
independent ways to obtain this information:</p>
<ol class="arabic simple">
<li><p><strong>High-level helper</strong>
<code class="docutils literal notranslate"><span class="pre">graph.get_active_ranges_of_base_id_alternative</span></code> - a cached
convenience wrapper.</p></li>
<li><p><strong>Low-level reconstruction</strong> by aggregating the <em>combined_edges</em>
table and converting the set of releases into compact
<code class="docutils literal notranslate"><span class="pre">[start,</span> <span class="pre">end]</span></code> slices via <code class="docutils literal notranslate"><span class="pre">graph.list_to_ranges</span></code>.</p></li>
</ol>
<p>This test iterates through <strong>all</strong> base-gene nodes and asserts that
the two methods deliver byte-identical results.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>verbose</strong> (<em>bool</em>) – If <em>True</em> (default) show a tqdm progress
bar that updates with the current node under inspection.</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><code class="docutils literal notranslate"><span class="pre">True</span></code> if <em>every</em> base-gene has matching ranges; <code class="docutils literal notranslate"><span class="pre">False</span></code>
as soon as a single mismatch is encountered.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>bool</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="idtrack._track_tests.TrackTests.is_combined_edges_dicts_overlapping_and_complete">
<span class="sig-name descname"><span class="pre">is_combined_edges_dicts_overlapping_and_complete</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/idtrack/_track_tests.html#TrackTests.is_combined_edges_dicts_overlapping_and_complete"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#idtrack._track_tests.TrackTests.is_combined_edges_dicts_overlapping_and_complete" title="Link to this definition"></a></dt>
<dd><p>Check edge-dictionary partitioning invariants.</p>
<p>The Track graph materialises three <em>edge caches</em>—<code class="docutils literal notranslate"><span class="pre">combined_edges</span></code>
and its two specialised siblings—each storing adjacency and release
metadata for a different subset of nodes:</p>
<ul class="simple">
<li><p><code class="docutils literal notranslate"><span class="pre">combined_edges</span></code> - all nodes, including backbone genes.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">combined_edges_genes</span></code> - stable Ensembl genes (non-assembly-specific).</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">combined_edges_assembly_specific_genes</span></code> - genes that exist only
on a single assembly.</p></li>
</ul>
<p>The design contract says:</p>
<ol class="arabic simple">
<li><p><strong>Disjointness</strong> - No node key may appear in more than one
dictionary.</p></li>
<li><p><strong>Completeness</strong> - The <em>union</em> of the dictionaries must cover all
graph nodes <strong>except</strong> those that represent <em>alternative database
versions</em> (e.g. “EnsemblMetazoa”) which are intentionally kept
separate.</p></li>
</ol>
<p>This routine enforces both rules.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p><code class="docutils literal notranslate"><span class="pre">True</span></code> if the dictionaries are pair-wise disjoint <strong>and</strong>
collectively cover every eligible node; <code class="docutils literal notranslate"><span class="pre">False</span></code> otherwise.</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p>bool</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="idtrack._track_tests.TrackTests.is_edge_with_same_nts_only_at_backbone_nodes">
<span class="sig-name descname"><span class="pre">is_edge_with_same_nts_only_at_backbone_nodes</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/idtrack/_track_tests.html#TrackTests.is_edge_with_same_nts_only_at_backbone_nodes"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#idtrack._track_tests.TrackTests.is_edge_with_same_nts_only_at_backbone_nodes" title="Link to this definition"></a></dt>
<dd><p>Assert <em>same-node-type</em> edges exist <strong>only</strong> between backbone genes.</p>
<p>The graph is a <em>multilayer</em> network where nodes of different</p>
<p>This method traverses every base-gene node and checks that condition.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p><code class="docutils literal notranslate"><span class="pre">True</span></code> when no overlaps are found; <code class="docutils literal notranslate"><span class="pre">False</span></code> otherwise.
Each offending base ID triggers a warning with the conflicting
ranges.</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p>bool</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="idtrack._track_tests.TrackTests.is_external_conversion_robust">
<span class="sig-name descname"><span class="pre">is_external_conversion_robust</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">convert_using_release</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">verbose</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/idtrack/_track_tests.html#TrackTests.is_external_conversion_robust"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#idtrack._track_tests.TrackTests.is_external_conversion_robust" title="Link to this definition"></a></dt>
<dd><p>Validate Ensembl→external conversion against MySQL ground truth.</p>
<p>A random external database is chosen for <strong>every</strong> genome assembly. For
the selected combination the method grabs the authoritative mapping
table (graph-ID → external ID set) from MySQL and converts the same
graph-IDs with <code class="xref py py-meth docutils literal notranslate"><span class="pre">idtrack.Track.convert()</span></code>.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>convert_using_release</strong> (<em>bool</em>) – Whether to pin the <em>from_release</em> when
calling the converter.  Keeping this <em>True</em> usually speeds up
the search and mimics user-facing behaviour.</p></li>
<li><p><strong>verbose</strong> (<em>bool</em>) – Print the current assembly/database/release being tested.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><em>True</em> if every converted set equals the MySQL reference,
<em>False</em> upon the first deviation.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>bool</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="idtrack._track_tests.TrackTests.is_id_functions_consistent_ensembl">
<span class="sig-name descname"><span class="pre">is_id_functions_consistent_ensembl</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">verbose</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/idtrack/_track_tests.html#TrackTests.is_id_functions_consistent_ensembl"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#idtrack._track_tests.TrackTests.is_id_functions_consistent_ensembl" title="Link to this definition"></a></dt>
<dd><p>Ensure Ensembl ID-list helpers agree with the SQL back-end.</p>
<p>For every release listed in <code class="docutils literal notranslate"><span class="pre">graph.graph[&quot;confident_for_release&quot;]</span></code>
the test compares two independent sources of Ensembl-gene IDs for the
<em>current genome assembly</em>:</p>
<ol class="arabic simple">
<li><p>IDs retrieved directly from MySQL via
<code class="xref py py-class docutils literal notranslate"><span class="pre">DatabaseManager</span></code>.</p></li>
<li><p>IDs returned by <code class="xref py py-meth docutils literal notranslate"><span class="pre">idtrack.Track.get_id_list()</span></code> from the graph.</p></li>
</ol>
<p>A mismatch means either the graph was built incompletely or the
helper functions drift out of sync with the database schema.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>verbose</strong> (<em>bool</em>) – If <code class="docutils literal notranslate"><span class="pre">True</span></code> (default) show a tqdm progress-bar while
iterating through the releases.</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><code class="docutils literal notranslate"><span class="pre">True</span></code> when <strong>all</strong> releases produce identical sets; else
<code class="docutils literal notranslate"><span class="pre">False</span></code> (a descriptive warning is logged).</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>bool</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="idtrack._track_tests.TrackTests.is_id_functions_consistent_ensembl_2">
<span class="sig-name descname"><span class="pre">is_id_functions_consistent_ensembl_2</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">verbose</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/idtrack/_track_tests.html#TrackTests.is_id_functions_consistent_ensembl_2"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#idtrack._track_tests.TrackTests.is_id_functions_consistent_ensembl_2" title="Link to this definition"></a></dt>
<dd><p>Cross-check Ensembl ID <em>range</em> helpers against raw edge data.</p>
<p>For every <strong>backbone</strong> Ensembl-gene node the routine computes the list
of active release ranges in two distinct ways:</p>
<ol class="arabic simple">
<li><p><strong>Raw computation</strong> - by flattening
<code class="docutils literal notranslate"><span class="pre">graph.combined_edges_genes</span></code> and compacting the releases with
<code class="xref py py-meth docutils literal notranslate"><span class="pre">idtrack.Track.list_to_ranges()</span></code>.</p></li>
<li><p><strong>Cached lookup</strong> - via the lazily built dictionary
<code class="docutils literal notranslate"><span class="pre">graph.get_active_ranges_of_id</span></code>.</p></li>
</ol>
<p>The two lists must match exactly.  A divergence would indicate that
the cached helper is out of sync with the authoritative edge
structure.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>verbose</strong> (<em>bool</em>) – If <code class="docutils literal notranslate"><span class="pre">True</span></code> (default) wrap the iteration in a tqdm bar.</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><code class="docutils literal notranslate"><span class="pre">True</span></code> if <em>all</em> gene nodes pass; <code class="docutils literal notranslate"><span class="pre">False</span></code> after the first
failure (a warning is emitted).</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>bool</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="idtrack._track_tests.TrackTests.is_id_functions_consistent_external">
<span class="sig-name descname"><span class="pre">is_id_functions_consistent_external</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">verbose</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/idtrack/_track_tests.html#TrackTests.is_id_functions_consistent_external"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#idtrack._track_tests.TrackTests.is_id_functions_consistent_external" title="Link to this definition"></a></dt>
<dd><p>Check external-ID list helpers against the raw MySQL tables.</p>
<p>For <strong>every</strong> combination of <em>assembly</em>, <em>Ensembl release</em> (limited to
<code class="xref py py-attr docutils literal notranslate"><span class="pre">graph.graph[&quot;confident_for_release&quot;]</span></code>) and <em>external
database</em> this test performs the following steps:</p>
<ol class="arabic simple">
<li><p>Query the authoritative list of external IDs directly from the
MySQL snapshot via <code class="xref py py-class docutils literal notranslate"><span class="pre">DatabaseManager</span></code>.</p></li>
<li><p>Ask the in-memory graph for the same list via
<code class="xref py py-meth docutils literal notranslate"><span class="pre">idtrack.Track.get_id_list()</span></code>.</p></li>
<li><p>Normalise node names through
<code class="xref py py-meth docutils literal notranslate"><span class="pre">idtrack.Track.node_name_alternatives()</span></code> to cope with the
occasional “_1” suffix.</p></li>
<li><p>Compare the two sets.  A mismatch is logged and the method returns
<code class="docutils literal notranslate"><span class="pre">False</span></code> immediately.</p></li>
</ol>
<p>The exhaustive traversal is expensive (minutes for large genomes) but
ensures the graph`s indexing helpers never drift from the actual
database content.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>verbose</strong> (<em>bool</em>) – If <em>True</em> (default) display a tqdm progress bar and emit
log messages at INFO level.  When <em>False</em> the method runs
silently.</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><em>True</em> when every single comparison matched, <em>False</em> as soon
as an inconsistency is encountered.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>bool</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="idtrack._track_tests.TrackTests.is_node_consistency_robust">
<span class="sig-name descname"><span class="pre">is_node_consistency_robust</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">verbose</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/idtrack/_track_tests.html#TrackTests.is_node_consistency_robust"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#idtrack._track_tests.TrackTests.is_node_consistency_robust" title="Link to this definition"></a></dt>
<dd><p>Check for illegal neighbour relationships and multi-edges.</p>
<p>The graph may contain <strong>exactly one</strong> edge between nodes of <em>different</em>
node-types.  Nodes of the <em>same</em> node-type are only allowed when that
type is the Ensembl backbone (<code class="docutils literal notranslate"><span class="pre">ensembl_gene</span></code>).  Any deviation - a
lateral same-type connection or &gt;1 multi-edge - is logged and aborts
the test.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>verbose</strong> (<em>bool</em>) – Print offending nodes when a violation is detected.</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><em>True</em> when the graph satisfies the topology rules, <em>False</em>
otherwise.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>bool</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="idtrack._track_tests.TrackTests.is_range_functions_robust">
<span class="sig-name descname"><span class="pre">is_range_functions_robust</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">verbose</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/idtrack/_track_tests.html#TrackTests.is_range_functions_robust"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#idtrack._track_tests.TrackTests.is_range_functions_robust" title="Link to this definition"></a></dt>
<dd><p>Detect overlapping release ranges among sibling Ensembl IDs.</p>
<p>A <em>base</em> Ensembl-gene ID is the stable identifier that groups multiple
versioned Ensembl-gene records (siblings).  The gene-history model
requires that the release ranges of sibling IDs <strong>never overlap</strong> -
each release must be covered by <em>exactly one</em> child stable ID.</p>
<p>This method traverses every base-gene node and checks that condition.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>verbose</strong> (<em>bool</em>) – If <code class="docutils literal notranslate"><span class="pre">True</span></code> (default) display a tqdm progress-bar.</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><code class="docutils literal notranslate"><span class="pre">True</span></code> when no overlaps are found; <code class="docutils literal notranslate"><span class="pre">False</span></code> otherwise.
Each offending base ID triggers a warning with the conflicting
ranges.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>bool</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="idtrack._track_tests.TrackTests.random_dataset_source_generator">
<span class="sig-name descname"><span class="pre">random_dataset_source_generator</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">assembly</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">include_ensembl</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">release_lower_limit</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">form</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/idtrack/_track_tests.html#TrackTests.random_dataset_source_generator"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#idtrack._track_tests.TrackTests.random_dataset_source_generator" title="Link to this definition"></a></dt>
<dd><p>Pick a random (&lt;database&gt;, &lt;assembly&gt;, &lt;release&gt;) tuple.</p>
<p>The function guarantees that the triple actually exists in the graph
and - if <em>release_lower_limit</em> is provided - honours the minimum
release constraint.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>assembly</strong> (<em>int</em>) – NCBI assembly code (integer).</p></li>
<li><p><strong>include_ensembl</strong> (<em>bool</em>) – Whether Ensembl backbone databases may be returned
as <em>database</em>.</p></li>
<li><p><strong>release_lower_limit</strong> (<em>int</em><em> | </em><em>None</em>) – Smallest permissible Ensembl release number
for the returned triple.  <em>None</em> disables the filter.</p></li>
<li><p><strong>form</strong> (<em>str</em><em> | </em><em>None</em>) – Restrict the draw to a particular connection <em>form</em>
(protein/coding/gene).  <em>None</em> means no restriction.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><code class="docutils literal notranslate"><span class="pre">(&lt;database&gt;,</span> <span class="pre">&lt;assembly&gt;,</span> <span class="pre">&lt;release&gt;)</span></code> or <em>None</em> when
no matching release exists.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>tuple | None</p>
</dd>
</dl>
</dd></dl>

</dd></dl>

<dl class="py class" id="module-idtrack._db">
<dt class="sig sig-object py" id="idtrack._db.DB">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">DB</span></span><a class="reference internal" href="_modules/idtrack/_db.html#DB"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#idtrack._db.DB" title="Link to this definition"></a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></p>
<p>Keeps some constants shared across different classes.</p>
</dd></dl>

</section>


           </div>
          </div>
          <footer><div class="rst-footer-buttons" role="navigation" aria-label="Footer">
        <a href="_notebooks/comparison.html" class="btn btn-neutral float-left" title="Comparison with Available Tools" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
        <a href="contributing.html" class="btn btn-neutral float-right" title="Contributor Guide" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2025, Kemal Inecik.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>