

<!DOCTYPE html>
<html class="writer-html5" lang="en" data-content_root="./">
<head>
  <meta charset="utf-8" /><meta name="viewport" content="width=device-width, initial-scale=1" />

  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>Reference &mdash; idtrack 0.0.4 documentation</title>
      <link rel="stylesheet" type="text/css" href="_static/pygments.css?v=03e43079" />
      <link rel="stylesheet" type="text/css" href="_static/css/theme.css?v=e59714d7" />
      <link rel="stylesheet" type="text/css" href="_static/custom_settings.css?v=15bcc2cf" />

  
      <script src="_static/jquery.js?v=5d32c60e"></script>
      <script src="_static/_sphinx_javascript_frameworks_compat.js?v=2cd50e6c"></script>
      <script src="_static/documentation_options.js?v=8c5712d9"></script>
      <script src="_static/doctools.js?v=9a2dae69"></script>
      <script src="_static/sphinx_highlight.js?v=dc90522c"></script>
      <script crossorigin="anonymous" integrity="sha256-Ae2Vz/4ePdIu6ZyI/5ZGsYnb+m0JlOmKPjt6XZ9JJkA=" src="https://cdnjs.cloudflare.com/ajax/libs/require.js/2.3.4/require.min.js"></script>
    <script src="_static/js/theme.js"></script>
    <link rel="index" title="Index" href="genindex.html" />
    <link rel="search" title="Search" href="search.html" />
    <link rel="next" title="Contributor Guide" href="contributing.html" />
    <link rel="prev" title="Comparison with Available Tools" href="_notebooks/comparison.html" /> 
</head>

<body class="wy-body-for-nav"> 
  <div class="wy-grid-for-nav">
    <nav data-toggle="wy-nav-shift" class="wy-nav-side">
      <div class="wy-side-scroll">
        <div class="wy-side-nav-search" >

          
          
          <a href="index.html" class="icon icon-home">
            idtrack
              <img src="_static/logo.png" class="logo" alt="Logo"/>
          </a>
<div role="search">
  <form id="rtd-search-form" class="wy-form" action="search.html" method="get">
    <input type="text" name="q" placeholder="Search docs" aria-label="Search docs" />
    <input type="hidden" name="check_keywords" value="yes" />
    <input type="hidden" name="area" value="default" />
  </form>
</div>
        </div><div class="wy-menu wy-menu-vertical" data-spy="affix" role="navigation" aria-label="Navigation menu">
              <p class="caption" role="heading"><span class="caption-text">Contents:</span></p>
<ul class="current">
<li class="toctree-l1"><a class="reference internal" href="installation.html">Installation</a></li>
<li class="toctree-l1"><a class="reference internal" href="usage.html">Usage</a></li>
<li class="toctree-l1 current"><a class="current reference internal" href="#">Reference</a><ul>
<li class="toctree-l2"><a class="reference internal" href="#idtrack._database_manager.DatabaseManager"><code class="docutils literal notranslate"><span class="pre">DatabaseManager</span></code></a><ul>
<li class="toctree-l3"><a class="reference internal" href="#idtrack._database_manager.DatabaseManager._create_relation_helper"><code class="docutils literal notranslate"><span class="pre">DatabaseManager._create_relation_helper()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#idtrack._database_manager.DatabaseManager._determine_usecols_ids"><code class="docutils literal notranslate"><span class="pre">DatabaseManager._determine_usecols_ids()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#idtrack._database_manager.DatabaseManager.available_releases"><code class="docutils literal notranslate"><span class="pre">DatabaseManager.available_releases</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#idtrack._database_manager.DatabaseManager.available_releases_no_save"><code class="docutils literal notranslate"><span class="pre">DatabaseManager.available_releases_no_save</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#idtrack._database_manager.DatabaseManager.available_releases_versions"><code class="docutils literal notranslate"><span class="pre">DatabaseManager.available_releases_versions()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#idtrack._database_manager.DatabaseManager.available_tables_mysql"><code class="docutils literal notranslate"><span class="pre">DatabaseManager.available_tables_mysql()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#idtrack._database_manager.DatabaseManager.change_assembly"><code class="docutils literal notranslate"><span class="pre">DatabaseManager.change_assembly()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#idtrack._database_manager.DatabaseManager.change_form"><code class="docutils literal notranslate"><span class="pre">DatabaseManager.change_form()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#idtrack._database_manager.DatabaseManager.change_release"><code class="docutils literal notranslate"><span class="pre">DatabaseManager.change_release()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#idtrack._database_manager.DatabaseManager.check_if_change_assembly_works"><code class="docutils literal notranslate"><span class="pre">DatabaseManager.check_if_change_assembly_works()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#idtrack._database_manager.DatabaseManager.check_version_info"><code class="docutils literal notranslate"><span class="pre">DatabaseManager.check_version_info()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#idtrack._database_manager.DatabaseManager.create_available_databases"><code class="docutils literal notranslate"><span class="pre">DatabaseManager.create_available_databases()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#idtrack._database_manager.DatabaseManager.create_database_content"><code class="docutils literal notranslate"><span class="pre">DatabaseManager.create_database_content()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#idtrack._database_manager.DatabaseManager.create_external_all"><code class="docutils literal notranslate"><span class="pre">DatabaseManager.create_external_all()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#idtrack._database_manager.DatabaseManager.create_external_db"><code class="docutils literal notranslate"><span class="pre">DatabaseManager.create_external_db()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#idtrack._database_manager.DatabaseManager.create_id_history"><code class="docutils literal notranslate"><span class="pre">DatabaseManager.create_id_history()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#idtrack._database_manager.DatabaseManager.create_id_history_fixed"><code class="docutils literal notranslate"><span class="pre">DatabaseManager.create_id_history_fixed()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#idtrack._database_manager.DatabaseManager.create_ids"><code class="docutils literal notranslate"><span class="pre">DatabaseManager.create_ids()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#idtrack._database_manager.DatabaseManager.create_relation_archive"><code class="docutils literal notranslate"><span class="pre">DatabaseManager.create_relation_archive()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#idtrack._database_manager.DatabaseManager.create_relation_current"><code class="docutils literal notranslate"><span class="pre">DatabaseManager.create_relation_current()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#idtrack._database_manager.DatabaseManager.create_release_id"><code class="docutils literal notranslate"><span class="pre">DatabaseManager.create_release_id()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#idtrack._database_manager.DatabaseManager.create_version_info"><code class="docutils literal notranslate"><span class="pre">DatabaseManager.create_version_info()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#idtrack._database_manager.DatabaseManager.download_table"><code class="docutils literal notranslate"><span class="pre">DatabaseManager.download_table()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#idtrack._database_manager.DatabaseManager.external_inst"><code class="docutils literal notranslate"><span class="pre">DatabaseManager.external_inst</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#idtrack._database_manager.DatabaseManager.file_name"><code class="docutils literal notranslate"><span class="pre">DatabaseManager.file_name()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#idtrack._database_manager.DatabaseManager.get_db"><code class="docutils literal notranslate"><span class="pre">DatabaseManager.get_db()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#idtrack._database_manager.DatabaseManager.get_release_date"><code class="docutils literal notranslate"><span class="pre">DatabaseManager.get_release_date()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#idtrack._database_manager.DatabaseManager.get_table"><code class="docutils literal notranslate"><span class="pre">DatabaseManager.get_table()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#idtrack._database_manager.DatabaseManager.id_ver_from_df"><code class="docutils literal notranslate"><span class="pre">DatabaseManager.id_ver_from_df()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#idtrack._database_manager.DatabaseManager.mysql_database"><code class="docutils literal notranslate"><span class="pre">DatabaseManager.mysql_database</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#idtrack._database_manager.DatabaseManager.node_dict_maker"><code class="docutils literal notranslate"><span class="pre">DatabaseManager.node_dict_maker()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#idtrack._database_manager.DatabaseManager.node_name_maker"><code class="docutils literal notranslate"><span class="pre">DatabaseManager.node_name_maker()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#idtrack._database_manager.DatabaseManager.tables_in_disk"><code class="docutils literal notranslate"><span class="pre">DatabaseManager.tables_in_disk()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#idtrack._database_manager.DatabaseManager.version_fix"><code class="docutils literal notranslate"><span class="pre">DatabaseManager.version_fix()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#idtrack._database_manager.DatabaseManager.version_fix_incomplete"><code class="docutils literal notranslate"><span class="pre">DatabaseManager.version_fix_incomplete()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#idtrack._database_manager.DatabaseManager.version_uniformize"><code class="docutils literal notranslate"><span class="pre">DatabaseManager.version_uniformize()</span></code></a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#idtrack._external_databases.ExternalDatabases"><code class="docutils literal notranslate"><span class="pre">ExternalDatabases</span></code></a><ul>
<li class="toctree-l3"><a class="reference internal" href="#idtrack._external_databases.ExternalDatabases.create_template_yaml"><code class="docutils literal notranslate"><span class="pre">ExternalDatabases.create_template_yaml()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#idtrack._external_databases.ExternalDatabases.file_name_modified_yaml"><code class="docutils literal notranslate"><span class="pre">ExternalDatabases.file_name_modified_yaml()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#idtrack._external_databases.ExternalDatabases.file_name_template_yaml"><code class="docutils literal notranslate"><span class="pre">ExternalDatabases.file_name_template_yaml()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#idtrack._external_databases.ExternalDatabases.give_list_for_case"><code class="docutils literal notranslate"><span class="pre">ExternalDatabases.give_list_for_case()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#idtrack._external_databases.ExternalDatabases.load_modified_yaml"><code class="docutils literal notranslate"><span class="pre">ExternalDatabases.load_modified_yaml()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#idtrack._external_databases.ExternalDatabases.validate_yaml_file_up_to_date"><code class="docutils literal notranslate"><span class="pre">ExternalDatabases.validate_yaml_file_up_to_date()</span></code></a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#idtrack._the_graph.TheGraph"><code class="docutils literal notranslate"><span class="pre">TheGraph</span></code></a><ul>
<li class="toctree-l3"><a class="reference internal" href="#idtrack._the_graph.TheGraph._attach_included_forms"><code class="docutils literal notranslate"><span class="pre">TheGraph._attach_included_forms()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#idtrack._the_graph.TheGraph._combined_edges"><code class="docutils literal notranslate"><span class="pre">TheGraph._combined_edges()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#idtrack._the_graph.TheGraph._combined_edges_genes_helper"><code class="docutils literal notranslate"><span class="pre">TheGraph._combined_edges_genes_helper()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#idtrack._the_graph.TheGraph._get_active_ranges_of_id"><code class="docutils literal notranslate"><span class="pre">TheGraph._get_active_ranges_of_id()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#idtrack._the_graph.TheGraph._node_trios"><code class="docutils literal notranslate"><span class="pre">TheGraph._node_trios()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#idtrack._the_graph.TheGraph.available_external_databases"><code class="docutils literal notranslate"><span class="pre">TheGraph.available_external_databases</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#idtrack._the_graph.TheGraph.available_external_databases_assembly"><code class="docutils literal notranslate"><span class="pre">TheGraph.available_external_databases_assembly</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#idtrack._the_graph.TheGraph.available_genome_assemblies"><code class="docutils literal notranslate"><span class="pre">TheGraph.available_genome_assemblies</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#idtrack._the_graph.TheGraph.available_releases_given_database_assembly"><code class="docutils literal notranslate"><span class="pre">TheGraph.available_releases_given_database_assembly</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#idtrack._the_graph.TheGraph.calculate_caches"><code class="docutils literal notranslate"><span class="pre">TheGraph.calculate_caches()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#idtrack._the_graph.TheGraph.combined_edges"><code class="docutils literal notranslate"><span class="pre">TheGraph.combined_edges</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#idtrack._the_graph.TheGraph.combined_edges_assembly_specific_genes"><code class="docutils literal notranslate"><span class="pre">TheGraph.combined_edges_assembly_specific_genes</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#idtrack._the_graph.TheGraph.combined_edges_genes"><code class="docutils literal notranslate"><span class="pre">TheGraph.combined_edges_genes</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#idtrack._the_graph.TheGraph.compact_ranges"><code class="docutils literal notranslate"><span class="pre">TheGraph.compact_ranges()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#idtrack._the_graph.TheGraph.external_database_connection_form"><code class="docutils literal notranslate"><span class="pre">TheGraph.external_database_connection_form</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#idtrack._the_graph.TheGraph.get_active_ranges_of_base_id_alternative"><code class="docutils literal notranslate"><span class="pre">TheGraph.get_active_ranges_of_base_id_alternative()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#idtrack._the_graph.TheGraph.get_active_ranges_of_id"><code class="docutils literal notranslate"><span class="pre">TheGraph.get_active_ranges_of_id</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#idtrack._the_graph.TheGraph.get_active_ranges_of_id_ensembl_all_inclusive"><code class="docutils literal notranslate"><span class="pre">TheGraph.get_active_ranges_of_id_ensembl_all_inclusive()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#idtrack._the_graph.TheGraph.get_external_database_nodes"><code class="docutils literal notranslate"><span class="pre">TheGraph.get_external_database_nodes()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#idtrack._the_graph.TheGraph.get_id_list"><code class="docutils literal notranslate"><span class="pre">TheGraph.get_id_list()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#idtrack._the_graph.TheGraph.get_intersecting_ranges"><code class="docutils literal notranslate"><span class="pre">TheGraph.get_intersecting_ranges()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#idtrack._the_graph.TheGraph.get_next_edge_releases"><code class="docutils literal notranslate"><span class="pre">TheGraph.get_next_edge_releases()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#idtrack._the_graph.TheGraph.get_two_nodes_coinciding_releases"><code class="docutils literal notranslate"><span class="pre">TheGraph.get_two_nodes_coinciding_releases()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#idtrack._the_graph.TheGraph.hyperconnective_nodes"><code class="docutils literal notranslate"><span class="pre">TheGraph.hyperconnective_nodes</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#idtrack._the_graph.TheGraph.is_point_in_range"><code class="docutils literal notranslate"><span class="pre">TheGraph.is_point_in_range()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#idtrack._the_graph.TheGraph.list_to_ranges"><code class="docutils literal notranslate"><span class="pre">TheGraph.list_to_ranges()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#idtrack._the_graph.TheGraph.lower_chars_graph"><code class="docutils literal notranslate"><span class="pre">TheGraph.lower_chars_graph</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#idtrack._the_graph.TheGraph.node_name_alternatives"><code class="docutils literal notranslate"><span class="pre">TheGraph.node_name_alternatives()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#idtrack._the_graph.TheGraph.node_trios"><code class="docutils literal notranslate"><span class="pre">TheGraph.node_trios</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#idtrack._the_graph.TheGraph.ranges_to_list"><code class="docutils literal notranslate"><span class="pre">TheGraph.ranges_to_list()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#idtrack._the_graph.TheGraph.rev"><code class="docutils literal notranslate"><span class="pre">TheGraph.rev</span></code></a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#idtrack._graph_maker.GraphMaker"><code class="docutils literal notranslate"><span class="pre">GraphMaker</span></code></a><ul>
<li class="toctree-l3"><a class="reference internal" href="#idtrack._graph_maker.GraphMaker.construct_graph"><code class="docutils literal notranslate"><span class="pre">GraphMaker.construct_graph()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#idtrack._graph_maker.GraphMaker.construct_graph_form"><code class="docutils literal notranslate"><span class="pre">GraphMaker.construct_graph_form()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#idtrack._graph_maker.GraphMaker.create_file_name"><code class="docutils literal notranslate"><span class="pre">GraphMaker.create_file_name()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#idtrack._graph_maker.GraphMaker.export_disk"><code class="docutils literal notranslate"><span class="pre">GraphMaker.export_disk()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#idtrack._graph_maker.GraphMaker.get_graph"><code class="docutils literal notranslate"><span class="pre">GraphMaker.get_graph()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#idtrack._graph_maker.GraphMaker.initialize_downloads"><code class="docutils literal notranslate"><span class="pre">GraphMaker.initialize_downloads()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#idtrack._graph_maker.GraphMaker.read_exported"><code class="docutils literal notranslate"><span class="pre">GraphMaker.read_exported()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#idtrack._graph_maker.GraphMaker.remove_non_gene_trees"><code class="docutils literal notranslate"><span class="pre">GraphMaker.remove_non_gene_trees()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#idtrack._graph_maker.GraphMaker.split_id"><code class="docutils literal notranslate"><span class="pre">GraphMaker.split_id()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#idtrack._graph_maker.GraphMaker.update_graph_with_the_new_release"><code class="docutils literal notranslate"><span class="pre">GraphMaker.update_graph_with_the_new_release()</span></code></a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#idtrack._verify_organism.VerifyOrganism"><code class="docutils literal notranslate"><span class="pre">VerifyOrganism</span></code></a><ul>
<li class="toctree-l3"><a class="reference internal" href="#idtrack._verify_organism.VerifyOrganism.fetch_organism_and_latest_release"><code class="docutils literal notranslate"><span class="pre">VerifyOrganism.fetch_organism_and_latest_release()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#idtrack._verify_organism.VerifyOrganism.get_formal_name"><code class="docutils literal notranslate"><span class="pre">VerifyOrganism.get_formal_name()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#idtrack._verify_organism.VerifyOrganism.get_latest_release"><code class="docutils literal notranslate"><span class="pre">VerifyOrganism.get_latest_release()</span></code></a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#idtrack._track.Track"><code class="docutils literal notranslate"><span class="pre">Track</span></code></a><ul>
<li class="toctree-l3"><a class="reference internal" href="#idtrack._track.Track.graph"><code class="docutils literal notranslate"><span class="pre">Track.graph</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#idtrack._track.Track.version_info"><code class="docutils literal notranslate"><span class="pre">Track.version_info</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#idtrack._track.Track._external_entrance_placeholder"><code class="docutils literal notranslate"><span class="pre">Track._external_entrance_placeholder</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#idtrack._track.Track._external_entrance_placeholders"><code class="docutils literal notranslate"><span class="pre">Track._external_entrance_placeholders</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#idtrack._track.Track._calculate_node_scores_helper"><code class="docutils literal notranslate"><span class="pre">Track._calculate_node_scores_helper</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#idtrack._track.Track._choose_relevant_synonym_helper"><code class="docutils literal notranslate"><span class="pre">Track._choose_relevant_synonym_helper()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#idtrack._track.Track._create_priority_list_ensembl"><code class="docutils literal notranslate"><span class="pre">Track._create_priority_list_ensembl()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#idtrack._track.Track._final_conversion"><code class="docutils literal notranslate"><span class="pre">Track._final_conversion()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#idtrack._track.Track._final_conversion_dict_prepare"><code class="docutils literal notranslate"><span class="pre">Track._final_conversion_dict_prepare()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#idtrack._track.Track._minimum_assembly_jumps_helper"><code class="docutils literal notranslate"><span class="pre">Track._minimum_assembly_jumps_helper()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#idtrack._track.Track._path_score_sorter_all_targets"><code class="docutils literal notranslate"><span class="pre">Track._path_score_sorter_all_targets()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#idtrack._track.Track._path_score_sorter_single_target"><code class="docutils literal notranslate"><span class="pre">Track._path_score_sorter_single_target()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#idtrack._track.Track._recursive_synonymous"><code class="docutils literal notranslate"><span class="pre">Track._recursive_synonymous()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#idtrack._track.Track.calculate_node_scores"><code class="docutils literal notranslate"><span class="pre">Track.calculate_node_scores()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#idtrack._track.Track.calculate_score_and_select"><code class="docutils literal notranslate"><span class="pre">Track.calculate_score_and_select()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#idtrack._track.Track.choose_relevant_synonym"><code class="docutils literal notranslate"><span class="pre">Track.choose_relevant_synonym()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#idtrack._track.Track.convert"><code class="docutils literal notranslate"><span class="pre">Track.convert()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#idtrack._track.Track.convert_optimized_multiple"><code class="docutils literal notranslate"><span class="pre">Track.convert_optimized_multiple()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#idtrack._track.Track.edge_key_orientor"><code class="docutils literal notranslate"><span class="pre">Track.edge_key_orientor()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#idtrack._track.Track.get_from_release_and_reverse_vars"><code class="docutils literal notranslate"><span class="pre">Track.get_from_release_and_reverse_vars()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#idtrack._track.Track.get_next_edges"><code class="docutils literal notranslate"><span class="pre">Track.get_next_edges()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#idtrack._track.Track.get_possible_paths"><code class="docutils literal notranslate"><span class="pre">Track.get_possible_paths()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#idtrack._track.Track.identify_source"><code class="docutils literal notranslate"><span class="pre">Track.identify_source()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#idtrack._track.Track.minimum_assembly_jumps"><code class="docutils literal notranslate"><span class="pre">Track.minimum_assembly_jumps()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#idtrack._track.Track.path_search"><code class="docutils literal notranslate"><span class="pre">Track.path_search()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#idtrack._track.Track.path_step_possible_assembly_jumps"><code class="docutils literal notranslate"><span class="pre">Track.path_step_possible_assembly_jumps()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#idtrack._track.Track.should_graph_reversed"><code class="docutils literal notranslate"><span class="pre">Track.should_graph_reversed()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#idtrack._track.Track.synonymous_nodes"><code class="docutils literal notranslate"><span class="pre">Track.synonymous_nodes()</span></code></a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#idtrack._track_tests.TrackTests"><code class="docutils literal notranslate"><span class="pre">TrackTests</span></code></a><ul>
<li class="toctree-l3"><a class="reference internal" href="#idtrack._track_tests.TrackTests.format_history_travel_testing_report"><code class="docutils literal notranslate"><span class="pre">TrackTests.format_history_travel_testing_report()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#idtrack._track_tests.TrackTests.format_history_travel_testing_report_header"><code class="docutils literal notranslate"><span class="pre">TrackTests.format_history_travel_testing_report_header()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#idtrack._track_tests.TrackTests.history_travel_testing"><code class="docutils literal notranslate"><span class="pre">TrackTests.history_travel_testing()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#idtrack._track_tests.TrackTests.history_travel_testing_random"><code class="docutils literal notranslate"><span class="pre">TrackTests.history_travel_testing_random()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#idtrack._track_tests.TrackTests.history_travel_testing_random_arguments_generator"><code class="docutils literal notranslate"><span class="pre">TrackTests.history_travel_testing_random_arguments_generator()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#idtrack._track_tests.TrackTests.how_many_corresponding_path_ensembl"><code class="docutils literal notranslate"><span class="pre">TrackTests.how_many_corresponding_path_ensembl()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#idtrack._track_tests.TrackTests.is_base_is_range_correct"><code class="docutils literal notranslate"><span class="pre">TrackTests.is_base_is_range_correct()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#idtrack._track_tests.TrackTests.is_combined_edges_dicts_overlapping_and_complete"><code class="docutils literal notranslate"><span class="pre">TrackTests.is_combined_edges_dicts_overlapping_and_complete()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#idtrack._track_tests.TrackTests.is_edge_with_same_nts_only_at_backbone_nodes"><code class="docutils literal notranslate"><span class="pre">TrackTests.is_edge_with_same_nts_only_at_backbone_nodes()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#idtrack._track_tests.TrackTests.is_final_external_conversion_robust"><code class="docutils literal notranslate"><span class="pre">TrackTests.is_final_external_conversion_robust()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#idtrack._track_tests.TrackTests.is_id_functions_consistent_ensembl"><code class="docutils literal notranslate"><span class="pre">TrackTests.is_id_functions_consistent_ensembl()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#idtrack._track_tests.TrackTests.is_id_functions_consistent_ensembl_2"><code class="docutils literal notranslate"><span class="pre">TrackTests.is_id_functions_consistent_ensembl_2()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#idtrack._track_tests.TrackTests.is_id_functions_consistent_external"><code class="docutils literal notranslate"><span class="pre">TrackTests.is_id_functions_consistent_external()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#idtrack._track_tests.TrackTests.is_node_consistency_robust"><code class="docutils literal notranslate"><span class="pre">TrackTests.is_node_consistency_robust()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#idtrack._track_tests.TrackTests.is_range_functions_robust"><code class="docutils literal notranslate"><span class="pre">TrackTests.is_range_functions_robust()</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#idtrack._track_tests.TrackTests.random_dataset_source_generator"><code class="docutils literal notranslate"><span class="pre">TrackTests.random_dataset_source_generator()</span></code></a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#idtrack._db.DB"><code class="docutils literal notranslate"><span class="pre">DB</span></code></a><ul>
<li class="toctree-l3"><a class="reference internal" href="#idtrack._db.DB.id_ver_delimiter"><code class="docutils literal notranslate"><span class="pre">DB.id_ver_delimiter</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#idtrack._db.DB.first_version"><code class="docutils literal notranslate"><span class="pre">DB.first_version</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#idtrack._db.DB.connection_timeout"><code class="docutils literal notranslate"><span class="pre">DB.connection_timeout</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#idtrack._db.DB.reading_timeout"><code class="docutils literal notranslate"><span class="pre">DB.reading_timeout</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#idtrack._db.DB.ensembl_ftp_base"><code class="docutils literal notranslate"><span class="pre">DB.ensembl_ftp_base</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#idtrack._db.DB.rest_server_api"><code class="docutils literal notranslate"><span class="pre">DB.rest_server_api</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#idtrack._db.DB.rest_server_ext"><code class="docutils literal notranslate"><span class="pre">DB.rest_server_ext</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#idtrack._db.DB.mysql_host"><code class="docutils literal notranslate"><span class="pre">DB.mysql_host</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#idtrack._db.DB.myqsl_user"><code class="docutils literal notranslate"><span class="pre">DB.myqsl_user</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#idtrack._db.DB.mysql_togo"><code class="docutils literal notranslate"><span class="pre">DB.mysql_togo</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#idtrack._db.DB.assembly_mysqlport_priority"><code class="docutils literal notranslate"><span class="pre">DB.assembly_mysqlport_priority</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#idtrack._db.DB.assembly_priority"><code class="docutils literal notranslate"><span class="pre">DB.assembly_priority</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#idtrack._db.DB.sorted_assemblies"><code class="docutils literal notranslate"><span class="pre">DB.sorted_assemblies</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#idtrack._db.DB.main_assembly"><code class="docutils literal notranslate"><span class="pre">DB.main_assembly</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#idtrack._db.DB.synonym_id_nodes_prefix"><code class="docutils literal notranslate"><span class="pre">DB.synonym_id_nodes_prefix</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#idtrack._db.DB.no_old_node_id"><code class="docutils literal notranslate"><span class="pre">DB.no_old_node_id</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#idtrack._db.DB.no_new_node_id"><code class="docutils literal notranslate"><span class="pre">DB.no_new_node_id</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#idtrack._db.DB.alternative_versions"><code class="docutils literal notranslate"><span class="pre">DB.alternative_versions</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#idtrack._db.DB.hyperconnecting_threshold"><code class="docutils literal notranslate"><span class="pre">DB.hyperconnecting_threshold</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#idtrack._db.DB.node_type_str"><code class="docutils literal notranslate"><span class="pre">DB.node_type_str</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#idtrack._db.DB.nts_external"><code class="docutils literal notranslate"><span class="pre">DB.nts_external</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#idtrack._db.DB.forms_in_order"><code class="docutils literal notranslate"><span class="pre">DB.forms_in_order</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#idtrack._db.DB.backbone_form"><code class="docutils literal notranslate"><span class="pre">DB.backbone_form</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#idtrack._db.DB.nts_ensembl"><code class="docutils literal notranslate"><span class="pre">DB.nts_ensembl</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#idtrack._db.DB.nts_ensembl_reverse"><code class="docutils literal notranslate"><span class="pre">DB.nts_ensembl_reverse</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#idtrack._db.DB.nts_assembly"><code class="docutils literal notranslate"><span class="pre">DB.nts_assembly</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#idtrack._db.DB.nts_assembly_reverse"><code class="docutils literal notranslate"><span class="pre">DB.nts_assembly_reverse</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#idtrack._db.DB.nts_base_ensembl"><code class="docutils literal notranslate"><span class="pre">DB.nts_base_ensembl</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#idtrack._db.DB.nts_base_ensembl_reverse"><code class="docutils literal notranslate"><span class="pre">DB.nts_base_ensembl_reverse</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#idtrack._db.DB.nts_bidirectional_synonymous_search"><code class="docutils literal notranslate"><span class="pre">DB.nts_bidirectional_synonymous_search</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#idtrack._db.DB.nts_assembly_gene"><code class="docutils literal notranslate"><span class="pre">DB.nts_assembly_gene</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#idtrack._db.DB.connection_dict"><code class="docutils literal notranslate"><span class="pre">DB.connection_dict</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#idtrack._db.DB.conn_dict_str_ensembl_base"><code class="docutils literal notranslate"><span class="pre">DB.conn_dict_str_ensembl_base</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#idtrack._db.DB.external_search_settings"><code class="docutils literal notranslate"><span class="pre">DB.external_search_settings</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#idtrack._db.DB.placeholder_na"><code class="docutils literal notranslate"><span class="pre">DB.placeholder_na</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#idtrack._db.DB.UTF8"><code class="docutils literal notranslate"><span class="pre">DB.UTF8</span></code></a></li>
<li class="toctree-l3"><a class="reference internal" href="#idtrack._db.DB.UTF8_STR"><code class="docutils literal notranslate"><span class="pre">DB.UTF8_STR</span></code></a></li>
</ul>
</li>
<li class="toctree-l2"><a class="reference internal" href="#idtrack._db.EmptyConversionMetricsError"><code class="docutils literal notranslate"><span class="pre">EmptyConversionMetricsError</span></code></a></li>
</ul>
</li>
<li class="toctree-l1"><a class="reference internal" href="contributing.html">Contributor Guide</a></li>
<li class="toctree-l1"><a class="reference internal" href="authors.html">Credits</a></li>
<li class="toctree-l1"><a class="reference internal" href="code_of_conduct.html">Contributor Covenant Code of Conduct</a></li>
</ul>

        </div>
      </div>
    </nav>

    <section data-toggle="wy-nav-shift" class="wy-nav-content-wrap"><nav class="wy-nav-top" aria-label="Mobile navigation menu" >
          <i data-toggle="wy-nav-top" class="fa fa-bars"></i>
          <a href="index.html">idtrack</a>
      </nav>

      <div class="wy-nav-content">
        <div class="rst-content">
          <div role="navigation" aria-label="Page navigation">
  <ul class="wy-breadcrumbs">
      <li><a href="index.html" class="icon icon-home" aria-label="Home"></a></li>
      <li class="breadcrumb-item active">Reference</li>
      <li class="wy-breadcrumbs-aside">
            <a href="_sources/reference.rst.txt" rel="nofollow"> View page source</a>
      </li>
  </ul>
  <hr/>
</div>
          <div role="main" class="document" itemscope="itemscope" itemtype="http://schema.org/Article">
           <div itemprop="articleBody">
             
  <section id="reference">
<h1>Reference<a class="headerlink" href="#reference" title="Link to this heading"></a></h1>
<dl class="py class" id="module-idtrack._database_manager">
<dt class="sig sig-object py" id="idtrack._database_manager.DatabaseManager">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">DatabaseManager</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">organism</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">form</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">local_repository</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ensembl_release</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ignore_before</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ignore_after</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">store_raw_always</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">genome_assembly</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/idtrack/_database_manager.html#DatabaseManager"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#idtrack._database_manager.DatabaseManager" title="Link to this definition"></a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></p>
<p>Manage retrieval, preprocessing, and storage of Ensembl Core and related external datasets.</p>
<p>The <em>DatabaseManager</em> centralizes all low-level operations required for ID-track analyses, including
discovering which Ensembl releases are available for a given organism/assembly, downloading the
corresponding MySQL tables, normalizing column names, persisting raw and processed files under a
local cache directory, and orchestrating auxiliary look-ups to third-party resources via
<code class="xref py py-class docutils literal notranslate"><span class="pre">ExternalDatabases</span></code>.  By funnelling every data-access path through a single object the
wider package gains:</p>
<ul class="simple">
<li><p><strong>Stable, reproducible builds</strong> - every graph, lookup table, or ID-history file is anchored to
the exact <em>Ensembl release</em>, <em>genome assembly</em>, and <em>form</em> (gene, transcript, translation, …)
with which the manager was configured.</p></li>
<li><p><strong>Transparent caching</strong> - expensive downloads happen once; subsequent requests are served from
disk, making large iterative analyses feasible on modest hardware.</p></li>
<li><p><strong>Unified version logic</strong> - helper methods such as
<a class="reference internal" href="#idtrack._database_manager.DatabaseManager.version_uniformize" title="idtrack._database_manager.DatabaseManager.version_uniformize"><code class="xref py py-meth docutils literal notranslate"><span class="pre">version_uniformize()</span></code></a> and
<a class="reference internal" href="#idtrack._database_manager.DatabaseManager.check_version_info" title="idtrack._database_manager.DatabaseManager.check_version_info"><code class="xref py py-meth docutils literal notranslate"><span class="pre">check_version_info()</span></code></a> guarantee that cross-release identifier changes
are captured and resolved consistently across the codebase.</p></li>
</ul>
<section id="key-public-methods-attributes">
<h2>Key public methods/attributes<a class="headerlink" href="#key-public-methods-attributes" title="Link to this heading"></a></h2>
<ul class="simple">
<li><p><a class="reference internal" href="#idtrack._database_manager.DatabaseManager.available_releases" title="idtrack._database_manager.DatabaseManager.available_releases"><code class="xref py py-meth docutils literal notranslate"><span class="pre">available_releases()</span></code></a> — list releases that can be queried <em>and</em> saved locally.</p></li>
<li><p><a class="reference internal" href="#idtrack._database_manager.DatabaseManager.change_release" title="idtrack._database_manager.DatabaseManager.change_release"><code class="xref py py-meth docutils literal notranslate"><span class="pre">change_release()</span></code></a> — switch the manager to another Ensembl release in-place.</p></li>
<li><p><a class="reference internal" href="#idtrack._database_manager.DatabaseManager.download_table" title="idtrack._database_manager.DatabaseManager.download_table"><code class="xref py py-meth docutils literal notranslate"><span class="pre">download_table()</span></code></a> — fetch a single MySQL table and write it to <code class="docutils literal notranslate"><span class="pre">local_repository</span></code>.</p></li>
<li><p><a class="reference internal" href="#idtrack._database_manager.DatabaseManager.create_external_all" title="idtrack._database_manager.DatabaseManager.create_external_all"><code class="xref py py-meth docutils literal notranslate"><span class="pre">create_external_all()</span></code></a> — pull every supported external resource (UniProt, RefSeq, …).</p></li>
<li><p><code class="xref py py-attr docutils literal notranslate"><span class="pre">organism</span></code>, <code class="xref py py-attr docutils literal notranslate"><span class="pre">form</span></code>, <code class="xref py py-attr docutils literal notranslate"><span class="pre">ensembl_release</span></code>,
<code class="xref py py-attr docutils literal notranslate"><span class="pre">genome_assembly</span></code> — core configuration knobs, surfaced for quick inspection.</p></li>
</ul>
<p>The class is <strong>stateful</strong>: change-mutating helpers update internal cached properties so that the
instance always reflects its current configuration.  Use the built-in <code class="xref py py-meth docutils literal notranslate"><span class="pre">__str__()</span></code> for a
concise, human-readable dump of that state.</p>
<p>Initialize a <a class="reference internal" href="#idtrack._database_manager.DatabaseManager" title="idtrack._database_manager.DatabaseManager"><code class="xref py py-class docutils literal notranslate"><span class="pre">DatabaseManager</span></code></a> for a specific organism, release, and assembly.</p>
<dl class="field-list simple">
<dt class="field-odd">param organism<span class="colon">:</span></dt>
<dd class="field-odd"><p>Canonical species name in Ensembl schema (e.g. <code class="docutils literal notranslate"><span class="pre">&quot;homo_sapiens&quot;</span></code> or
<code class="docutils literal notranslate"><span class="pre">&quot;mus_musculus&quot;</span></code>). Anything else raises <code class="xref py py-class docutils literal notranslate"><span class="pre">NotImplementedError</span></code>.</p>
</dd>
<dt class="field-even">type organism<span class="colon">:</span></dt>
<dd class="field-even"><p>str</p>
</dd>
<dt class="field-odd">param form<span class="colon">:</span></dt>
<dd class="field-odd"><p>Biological entity level of interest—one of <code class="docutils literal notranslate"><span class="pre">&quot;gene&quot;</span></code>, <code class="docutils literal notranslate"><span class="pre">&quot;transcript&quot;</span></code>,
<code class="docutils literal notranslate"><span class="pre">&quot;translation&quot;</span></code>, …—governing which stable-ID columns will be expected downstream.</p>
</dd>
<dt class="field-even">type form<span class="colon">:</span></dt>
<dd class="field-even"><p>str</p>
</dd>
<dt class="field-odd">param local_repository<span class="colon">:</span></dt>
<dd class="field-odd"><p>Absolute or relative path to a writable directory that will hold
all downloaded MySQL dumps, intermediate parquet/Feather files, and ready-to-use
artefacts. The directory must already exist and be both readable and writable.</p>
</dd>
<dt class="field-even">type local_repository<span class="colon">:</span></dt>
<dd class="field-even"><p>str</p>
</dd>
<dt class="field-odd">param ensembl_release<span class="colon">:</span></dt>
<dd class="field-odd"><p>Target Ensembl release number. If <code class="docutils literal notranslate"><span class="pre">None</span></code> the most
recent release available for <em>genome_assembly</em> is selected automatically.</p>
</dd>
<dt class="field-even">type ensembl_release<span class="colon">:</span></dt>
<dd class="field-even"><p>Optional[int]</p>
</dd>
<dt class="field-odd">param ignore_before<span class="colon">:</span></dt>
<dd class="field-odd"><p>Earliest release to include when building cross-release ID
histories. Defaults to the minimum release supported by the selected assembly.</p>
</dd>
<dt class="field-even">type ignore_before<span class="colon">:</span></dt>
<dd class="field-even"><p>Optional[int]</p>
</dd>
<dt class="field-odd">param ignore_after<span class="colon">:</span></dt>
<dd class="field-odd"><p>Latest release to include when building histories.
<code class="docutils literal notranslate"><span class="pre">np.inf</span></code> (the default) disables the upper bound and includes all newer releases.</p>
</dd>
<dt class="field-even">type ignore_after<span class="colon">:</span></dt>
<dd class="field-even"><p>Optional[int | float]</p>
</dd>
<dt class="field-odd">param store_raw_always<span class="colon">:</span></dt>
<dd class="field-odd"><p>When <code class="docutils literal notranslate"><span class="pre">True</span></code> raw MySQL tables are <em>always</em> copied to
<code class="docutils literal notranslate"><span class="pre">local_repository</span></code> before conversion; when <code class="docutils literal notranslate"><span class="pre">False</span></code> they are kept only in memory.</p>
</dd>
<dt class="field-even">type store_raw_always<span class="colon">:</span></dt>
<dd class="field-even"><p>bool</p>
</dd>
<dt class="field-odd">param genome_assembly<span class="colon">:</span></dt>
<dd class="field-odd"><p>NCBI assembly version (e.g. <code class="docutils literal notranslate"><span class="pre">38</span></code> for GRCh38). If
omitted, the assembly with the highest priority for <em>organism</em> is used.</p>
</dd>
<dt class="field-even">type genome_assembly<span class="colon">:</span></dt>
<dd class="field-even"><p>Optional[int]</p>
</dd>
<dt class="field-odd">raises ValueError<span class="colon">:</span></dt>
<dd class="field-odd"><p>If <em>form</em> is not in the supported list or <em>local_repository</em> fails basic
    path/read/write checks.</p>
</dd>
<dt class="field-even">raises NotImplementedError<span class="colon">:</span></dt>
<dd class="field-even"><p>If <em>organism</em> is not yet supported by the package.</p>
</dd>
</dl>
<dl class="py method">
<dt class="sig sig-object py" id="idtrack._database_manager.DatabaseManager._create_relation_helper">
<span class="sig-name descname"><span class="pre">_create_relation_helper</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">df</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/idtrack/_database_manager.html#DatabaseManager._create_relation_helper"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#idtrack._database_manager.DatabaseManager._create_relation_helper" title="Link to this definition"></a></dt>
<dd><p>Convert an ID/version matrix into the canonical three-column relationship table.</p>
<p>The helper is shared by <a class="reference internal" href="#idtrack._database_manager.DatabaseManager.create_relation_current" title="idtrack._database_manager.DatabaseManager.create_relation_current"><code class="xref py py-meth docutils literal notranslate"><span class="pre">DatabaseManager.create_relation_current()</span></code></a>
and <a class="reference internal" href="#idtrack._database_manager.DatabaseManager.create_relation_archive" title="idtrack._database_manager.DatabaseManager.create_relation_archive"><code class="xref py py-meth docutils literal notranslate"><span class="pre">DatabaseManager.create_relation_archive()</span></code></a> and is not intended
for direct use.  It validates the incoming frame, fixes inconsistent
version numbers (via <a class="reference internal" href="#idtrack._database_manager.DatabaseManager.version_fix" title="idtrack._database_manager.DatabaseManager.version_fix"><code class="xref py py-meth docutils literal notranslate"><span class="pre">DatabaseManager.version_fix()</span></code></a> and
<a class="reference internal" href="#idtrack._database_manager.DatabaseManager.version_fix_incomplete" title="idtrack._database_manager.DatabaseManager.version_fix_incomplete"><code class="xref py py-meth docutils literal notranslate"><span class="pre">DatabaseManager.version_fix_incomplete()</span></code></a>), converts missing
translations to <code class="docutils literal notranslate"><span class="pre">NaN</span></code>-compatible floats, casts all stable-ID columns to
string, and finally compresses each <em>ID + version</em> pair into the compact
node label used throughout <em>ID-track</em> graphs.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>df</strong> (<em>pandas.DataFrame</em>) – A six-column frame with exactly the following
names (order irrelevant): <code class="docutils literal notranslate"><span class="pre">gene_stable_id</span></code>, <code class="docutils literal notranslate"><span class="pre">gene_version</span></code>,
<code class="docutils literal notranslate"><span class="pre">transcript_stable_id</span></code>, <code class="docutils literal notranslate"><span class="pre">transcript_version</span></code>,
<code class="docutils literal notranslate"><span class="pre">translation_stable_id</span></code>, <code class="docutils literal notranslate"><span class="pre">translation_version</span></code>.</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><dl class="simple">
<dt>Three columns—<code class="docutils literal notranslate"><span class="pre">gene</span></code>, <code class="docutils literal notranslate"><span class="pre">transcript</span></code>,</dt><dd><p><code class="docutils literal notranslate"><span class="pre">translation</span></code>—deduplicated and index-reset, ready for graph
construction.</p>
</dd>
</dl>
</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>pandas.DataFrame</p>
</dd>
<dt class="field-even">Raises<span class="colon">:</span></dt>
<dd class="field-even"><p><strong>ValueError</strong> – If <em>df</em> does not contain the required six columns or if
    version columns cannot be coerced to the expected numeric dtype.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="idtrack._database_manager.DatabaseManager._determine_usecols_ids">
<em class="property"><span class="pre">static</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">_determine_usecols_ids</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">form</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/idtrack/_database_manager.html#DatabaseManager._determine_usecols_ids"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#idtrack._database_manager.DatabaseManager._determine_usecols_ids" title="Link to this definition"></a></dt>
<dd><p>Determine column subsets needed to fetch identifier tables for a given Ensembl molecular form.</p>
<p>The helper translates a user-facing <em>form</em> string (<code class="docutils literal notranslate"><span class="pre">&quot;gene&quot;</span></code>, <code class="docutils literal notranslate"><span class="pre">&quot;transcript&quot;</span></code>, or
<code class="docutils literal notranslate"><span class="pre">&quot;translation&quot;</span></code>) into three ordered lists that drive low-level SQL selects throughout
<em>ID-track</em>.  Splitting the information this way lets public routines such as
<a class="reference internal" href="#idtrack._database_manager.DatabaseManager.create_ids" title="idtrack._database_manager.DatabaseManager.create_ids"><code class="xref py py-meth docutils literal notranslate"><span class="pre">DatabaseManager.create_ids()</span></code></a> assemble the minimal column set required for each
organism/release while still keeping associated keys available for later joins.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>form</strong> (<em>str</em>) – Molecular form whose identifier columns are requested. Must be one of
<a class="reference internal" href="#idtrack._db.DB.forms_in_order" title="idtrack._db.DB.forms_in_order"><code class="xref py py-data docutils literal notranslate"><span class="pre">idtrack._db.DB.forms_in_order</span></code></a> (<code class="docutils literal notranslate"><span class="pre">&quot;gene&quot;</span></code>, <code class="docutils literal notranslate"><span class="pre">&quot;transcript&quot;</span></code>, or
<code class="docutils literal notranslate"><span class="pre">&quot;translation&quot;</span></code>).</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><ul class="simple">
<li><p><strong>stable_id_version</strong> - always <code class="docutils literal notranslate"><span class="pre">[&quot;stable_id&quot;,</span> <span class="pre">&quot;version&quot;]</span></code>; the canonical ID and its
version counter.</p></li>
<li><p><strong>usecols_core</strong> - primary-key column for <em>form</em> plus <code class="docutils literal notranslate"><span class="pre">stable_id_version</span></code>.</p></li>
<li><p><strong>usecols_asso</strong> - foreign-key columns linking <em>form</em> to upstream forms, enabling
later joins (e.g., <code class="docutils literal notranslate"><span class="pre">[&quot;transcript_id&quot;,</span> <span class="pre">&quot;gene_id&quot;]</span></code> for transcripts).</p></li>
</ul>
</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>tuple[list[str], list[str], list[str]]</p>
</dd>
<dt class="field-even">Raises<span class="colon">:</span></dt>
<dd class="field-even"><p><strong>ValueError</strong> – If <em>form</em> is not in <code class="docutils literal notranslate"><span class="pre">{&quot;gene&quot;,</span> <span class="pre">&quot;transcript&quot;,</span> <span class="pre">&quot;translation&quot;}</span></code>.</p>
</dd>
</dl>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="idtrack._database_manager.DatabaseManager.available_releases">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">available_releases</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="pre">list</span><span class="p"><span class="pre">[</span></span><span class="pre">int</span><span class="p"><span class="pre">]</span></span></em><a class="headerlink" href="#idtrack._database_manager.DatabaseManager.available_releases" title="Link to this definition"></a></dt>
<dd><p>Return Ensembl releases that are both reachable and within the ignore window.</p>
<p>The set is discovered via <a class="reference internal" href="#idtrack._database_manager.DatabaseManager.available_releases_versions" title="idtrack._database_manager.DatabaseManager.available_releases_versions"><code class="xref py py-meth docutils literal notranslate"><span class="pre">available_releases_versions()</span></code></a>, filtered against
<code class="docutils literal notranslate"><span class="pre">ignore_before</span></code> / <code class="docutils literal notranslate"><span class="pre">ignore_after</span></code>, sorted in ascending order, and cached for the lifetime of
this <a class="reference internal" href="#idtrack._database_manager.DatabaseManager" title="idtrack._database_manager.DatabaseManager"><code class="xref py py-class docutils literal notranslate"><span class="pre">DatabaseManager</span></code></a> instance.  The resulting list represents releases that can
safely be queried <strong>and</strong> cached locally, guaranteeing reproducible downstream analyses.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>Sorted release numbers satisfying reachability and ignore-window constraints.</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p>list[int]</p>
</dd>
</dl>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="idtrack._database_manager.DatabaseManager.available_releases_no_save">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">available_releases_no_save</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="pre">list</span><span class="p"><span class="pre">[</span></span><span class="pre">int</span><span class="p"><span class="pre">]</span></span></em><a class="headerlink" href="#idtrack._database_manager.DatabaseManager.available_releases_no_save" title="Link to this definition"></a></dt>
<dd><p>Return reachable Ensembl releases without persisting the discovery to disk.</p>
<p>Functionally identical to <a class="reference internal" href="#idtrack._database_manager.DatabaseManager.available_releases" title="idtrack._database_manager.DatabaseManager.available_releases"><code class="xref py py-meth docutils literal notranslate"><span class="pre">available_releases()</span></code></a>, except that the discovered list is <strong>not</strong>
written to the on-disk YAML cache. This helper is useful when users want a quick, read-only view
of server availability—e.g., inside CI pipelines—without contaminating the persistent cache.
The value is still memoized in memory for the current <a class="reference internal" href="#idtrack._database_manager.DatabaseManager" title="idtrack._database_manager.DatabaseManager"><code class="xref py py-class docutils literal notranslate"><span class="pre">DatabaseManager</span></code></a> instance.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>Sorted release numbers reachable on the remote MySQL server and compliant with the
ignore window.</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p>list[int]</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="idtrack._database_manager.DatabaseManager.available_releases_versions">
<span class="sig-name descname"><span class="pre">available_releases_versions</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/idtrack/_database_manager.html#DatabaseManager.available_releases_versions"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#idtrack._database_manager.DatabaseManager.available_releases_versions" title="Link to this definition"></a></dt>
<dd><p>Discover valid Ensembl releases for the configured organism and assembly.</p>
<p>A <code class="docutils literal notranslate"><span class="pre">SHOW</span> <span class="pre">DATABASES</span></code> query is issued against the configured MySQL mirror.  Results are matched
against the pattern <code class="docutils literal notranslate"><span class="pre">^{organism}_core_&lt;release&gt;_.*$</span></code>; the captured <code class="docutils literal notranslate"><span class="pre">&lt;release&gt;</span></code> component is
converted to <code class="docutils literal notranslate"><span class="pre">int</span></code> (floating-point labels are rejected), range-checked against the manager’s
<code class="docutils literal notranslate"><span class="pre">ignore_before</span></code> / <code class="docutils literal notranslate"><span class="pre">ignore_after</span></code> bounds, and finally sorted. Optional keyword arguments are
forwarded verbatim to <a class="reference internal" href="#idtrack._database_manager.DatabaseManager.get_db" title="idtrack._database_manager.DatabaseManager.get_db"><code class="xref py py-meth docutils literal notranslate"><span class="pre">DatabaseManager.get_db()</span></code></a>, allowing callers to tweak connection or
caching behaviour.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>kwargs</strong> – Arbitrary keyword arguments passed straight through to
<a class="reference internal" href="#idtrack._database_manager.DatabaseManager.get_db" title="idtrack._database_manager.DatabaseManager.get_db"><code class="xref py py-meth docutils literal notranslate"><span class="pre">DatabaseManager.get_db()</span></code></a> (e.g., <code class="docutils literal notranslate"><span class="pre">mysql_conn</span></code>, <code class="docutils literal notranslate"><span class="pre">force_refresh</span></code>).</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>Sorted list of release numbers that exist on the mirror and comply with the ignore window.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>list[int]</p>
</dd>
<dt class="field-even">Raises<span class="colon">:</span></dt>
<dd class="field-even"><p><strong>ValueError</strong> – If a database name does not match the expected regex, if floating-point release
    labels are encountered, or if other inconsistencies arise while parsing server results.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="idtrack._database_manager.DatabaseManager.available_tables_mysql">
<span class="sig-name descname"><span class="pre">available_tables_mysql</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/idtrack/_database_manager.html#DatabaseManager.available_tables_mysql"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#idtrack._database_manager.DatabaseManager.available_tables_mysql" title="Link to this definition"></a></dt>
<dd><p>Enumerate tables present in the selected Ensembl MySQL schema.</p>
<p>Intended to complement <code class="xref py py-meth docutils literal notranslate"><span class="pre">available_databases_mysql()</span></code>: while that
method lists <em>databases</em> (one per organism/release/assembly), this one will
drill into the active database and return the table names themselves, such
as <code class="docutils literal notranslate"><span class="pre">&quot;gene&quot;</span></code>, <code class="docutils literal notranslate"><span class="pre">&quot;transcript&quot;</span></code>, <code class="docutils literal notranslate"><span class="pre">&quot;xref&quot;</span></code>, and so on.</p>
<dl class="field-list simple">
<dt class="field-odd">Raises<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>NotImplementedError</strong> – Always - the table enumeration logic has not yet been written.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="idtrack._database_manager.DatabaseManager.change_assembly">
<span class="sig-name descname"><span class="pre">change_assembly</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">genome_assembly</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">last_possible_ensembl_release</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/idtrack/_database_manager.html#DatabaseManager.change_assembly"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#idtrack._database_manager.DatabaseManager.change_assembly" title="Link to this definition"></a></dt>
<dd><p>Clone the manager while targeting a new genome assembly (e.g. GRCh38 → GRCh37).</p>
<p>Genome assemblies are encoded as integers in Ensembl’s schema naming (<code class="docutils literal notranslate"><span class="pre">38</span></code> for <em>GRCh38</em>,
<code class="docutils literal notranslate"><span class="pre">37</span></code> for <em>GRCh37</em>, <code class="docutils literal notranslate"><span class="pre">102</span></code> for <em>GRCm39</em>, …).  When <em>last_possible_ensembl_release</em> is <code class="docutils literal notranslate"><span class="pre">True</span></code> the
method automatically picks the most recent Ensembl release that <strong>still</strong> provides MySQL dumps for the
requested assembly, ensuring compatibility.  All other settings are copied verbatim.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>genome_assembly</strong> (<em>int</em>) – Key from <code class="xref py py-data docutils literal notranslate"><span class="pre">DB.assembly_mysqlport_priority</span></code> mapping—see Ensembl
documentation for valid values.</p></li>
<li><p><strong>last_possible_ensembl_release</strong> (<em>bool</em>) – When <code class="docutils literal notranslate"><span class="pre">True</span></code> override <em>ensembl_release</em> with the
newest version available for <em>genome_assembly</em>.  Defaults to <code class="docutils literal notranslate"><span class="pre">False</span></code>.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>New manager tied to the requested assembly (and possibly a recalculated release).</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference internal" href="#idtrack._database_manager.DatabaseManager" title="idtrack._database_manager.DatabaseManager">DatabaseManager</a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="idtrack._database_manager.DatabaseManager.change_form">
<span class="sig-name descname"><span class="pre">change_form</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">form</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/idtrack/_database_manager.html#DatabaseManager.change_form"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#idtrack._database_manager.DatabaseManager.change_form" title="Link to this definition"></a></dt>
<dd><p>Clone the manager while switching the biological form of interest.</p>
<p>A “form” denotes the identifier namespace to track—<code class="docutils literal notranslate"><span class="pre">gene</span></code>, <code class="docutils literal notranslate"><span class="pre">transcript</span></code>, <code class="docutils literal notranslate"><span class="pre">translation</span></code>, etc.  This
method preserves every other configuration knob (organism, release, assembly, cache directory, ignore
windows, …) and returns a brand-new instance so that the original object remains unaffected.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>form</strong> (<em>str</em>) – Target form/namespace recognised by <code class="xref py py-meth docutils literal notranslate"><span class="pre">__init__()</span></code>.  Typical values
are <code class="docutils literal notranslate"><span class="pre">&quot;gene&quot;</span></code>, <code class="docutils literal notranslate"><span class="pre">&quot;transcript&quot;</span></code>, or <code class="docutils literal notranslate"><span class="pre">&quot;translation&quot;</span></code>.</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>An independent manager identical to <em>self</em> except for <code class="xref py py-attr docutils literal notranslate"><span class="pre">form</span></code>.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference internal" href="#idtrack._database_manager.DatabaseManager" title="idtrack._database_manager.DatabaseManager">DatabaseManager</a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="idtrack._database_manager.DatabaseManager.change_release">
<span class="sig-name descname"><span class="pre">change_release</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">ensembl_release</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/idtrack/_database_manager.html#DatabaseManager.change_release"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#idtrack._database_manager.DatabaseManager.change_release" title="Link to this definition"></a></dt>
<dd><p>Produce a new manager that targets a different Ensembl release.</p>
<p>The returned instance inherits organism, form, assembly, and all caching parameters, but points every
subsequent query (MySQL, FTP, or REST) to <em>ensembl_release</em>.  This is the recommended way to traverse
releases in scripted analyses without mutating objects in-place.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>ensembl_release</strong> (<em>int</em>) – Desired Ensembl release number (e.g. <code class="docutils literal notranslate"><span class="pre">111</span></code>).  Must be available for the
current genome assembly or a <code class="xref py py-data docutils literal notranslate"><span class="pre">NotImplementedError</span></code> may be raised further down the call
stack when data retrieval is attempted.</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>Fresh manager initialised for <em>ensembl_release</em>.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference internal" href="#idtrack._database_manager.DatabaseManager" title="idtrack._database_manager.DatabaseManager">DatabaseManager</a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="idtrack._database_manager.DatabaseManager.check_if_change_assembly_works">
<span class="sig-name descname"><span class="pre">check_if_change_assembly_works</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">db_manager</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">target_assembly</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/idtrack/_database_manager.html#DatabaseManager.check_if_change_assembly_works"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#idtrack._database_manager.DatabaseManager.check_if_change_assembly_works" title="Link to this definition"></a></dt>
<dd><p>Evaluate whether <em>db_manager</em> can be cloned to operate on <em>target_assembly</em>.</p>
<p>A lightweight health-check that calls <a class="reference internal" href="#idtrack._database_manager.DatabaseManager.change_assembly" title="idtrack._database_manager.DatabaseManager.change_assembly"><code class="xref py py-meth docutils literal notranslate"><span class="pre">DatabaseManager.change_assembly()</span></code></a> inside a
<code class="docutils literal notranslate"><span class="pre">try/except</span></code> block and converts the outcome to a boolean flag rather than letting the exception
propagate.  It allows batch workflows to skip assemblies that are unavailable or invalid without
interrupting processing.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>db_manager</strong> (<a class="reference internal" href="#idtrack._database_manager.DatabaseManager" title="idtrack._database_manager.DatabaseManager"><em>DatabaseManager</em></a>) – Manager instance to probe.</p></li>
<li><p><strong>target_assembly</strong> (<em>int</em>) – Genome-assembly code to test (a key of
<a class="reference internal" href="#idtrack._db.DB.assembly_mysqlport_priority" title="idtrack._db.DB.assembly_mysqlport_priority"><code class="xref py py-data docutils literal notranslate"><span class="pre">idtrack._db.DB.assembly_mysqlport_priority</span></code></a>).</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><code class="docutils literal notranslate"><span class="pre">True</span></code> if the assembly switch succeeds without raising <code class="xref py py-class docutils literal notranslate"><span class="pre">ValueError</span></code>; <code class="docutils literal notranslate"><span class="pre">False</span></code> otherwise.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>bool</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="idtrack._database_manager.DatabaseManager.check_version_info">
<span class="sig-name descname"><span class="pre">check_version_info</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/idtrack/_database_manager.html#DatabaseManager.check_version_info"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#idtrack._database_manager.DatabaseManager.check_version_info" title="Link to this definition"></a></dt>
<dd><p>Infer whether the organism’s Ensembl IDs come <strong>with</strong>, <strong>without</strong>, or <strong>mixed</strong> versions.</p>
<p>The method scans all releases available for the current genome assembly and inspects the boolean
flag in the <code class="docutils literal notranslate"><span class="pre">version_info</span></code> column of a pre-computed table (<code class="docutils literal notranslate"><span class="pre">get_db(&quot;versioninfo&quot;)</span></code>).  Three
mutually exclusive scenarios exist:</p>
<ul class="simple">
<li><p>All releases lack version suffixes: <code class="docutils literal notranslate"><span class="pre">&quot;without_version&quot;</span></code></p></li>
<li><p>All releases include suffixes: <code class="docutils literal notranslate"><span class="pre">&quot;with_version&quot;</span></code></p></li>
<li><p>A mixture of both states: <code class="docutils literal notranslate"><span class="pre">&quot;add_version&quot;</span></code> (synthetic versions will be injected)</p></li>
</ul>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p><dl class="simple">
<dt>One of <code class="docutils literal notranslate"><span class="pre">&quot;without_version&quot;</span></code>, <code class="docutils literal notranslate"><span class="pre">&quot;with_version&quot;</span></code>, or <code class="docutils literal notranslate"><span class="pre">&quot;add_version&quot;</span></code>.  Callers use the</dt><dd><p>string to decide how to standardise identifier columns.</p>
</dd>
</dl>
</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p>str</p>
</dd>
<dt class="field-odd">Raises<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>ValueError</strong> – If the <em>version_info</em> column in the source table is not strictly boolean, signalling
    a corrupted download or schema drift.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="idtrack._database_manager.DatabaseManager.create_available_databases">
<span class="sig-name descname"><span class="pre">create_available_databases</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/idtrack/_database_manager.html#DatabaseManager.create_available_databases"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#idtrack._database_manager.DatabaseManager.create_available_databases" title="Link to this definition"></a></dt>
<dd><p>Discover MySQL databases for the configured organism/assembly.</p>
<p>The manager issues a <code class="docutils literal notranslate"><span class="pre">SHOW</span> <span class="pre">DATABASES</span></code> query against the Ensembl public MySQL mirror and filters
names that match <code class="docutils literal notranslate"><span class="pre">^{organism}_core_[0-9]+_.*$</span></code>.  The resulting list is returned as a single-column
dataframe so that callers can seamlessly chain further pandas operations or persist the result.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p><dl class="simple">
<dt>One column named <code class="docutils literal notranslate"><span class="pre">&quot;available_databases&quot;</span></code> listing all databases that match</dt><dd><p>the organism, irrespective of Ensembl release or genome assembly.</p>
</dd>
</dl>
</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p>pandas.DataFrame</p>
</dd>
<dt class="field-odd">Raises<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>ValueError</strong> – If the server response is not a sequence of single-field tuples <strong>or</strong> if any tuple
    element is not a string.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="idtrack._database_manager.DatabaseManager.create_database_content">
<span class="sig-name descname"><span class="pre">create_database_content</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">just_download</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/idtrack/_database_manager.html#DatabaseManager.create_database_content"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#idtrack._database_manager.DatabaseManager.create_database_content" title="Link to this definition"></a></dt>
<dd><p>Retrieve and optionally cache external-database metadata for every assembly, release, and form.</p>
<p>The helper iterates over <em>all</em> genome assemblies defined in
<a class="reference internal" href="#idtrack._db.DB.assembly_mysqlport_priority" title="idtrack._db.DB.assembly_mysqlport_priority"><code class="xref py py-data docutils literal notranslate"><span class="pre">idtrack._db.DB.assembly_mysqlport_priority</span></code></a>, every available Ensembl release for each assembly,
and every identifier <em>form</em> supported by the package, downloading the <code class="docutils literal notranslate"><span class="pre">external_database</span></code> table for
each combination.  The resulting frames are concatenated, enriched with <code class="docutils literal notranslate"><span class="pre">assembly</span></code>, <code class="docutils literal notranslate"><span class="pre">release</span></code>,
<code class="docutils literal notranslate"><span class="pre">form</span></code>, and <code class="docutils literal notranslate"><span class="pre">organism</span></code> columns, and returned to the caller.  When <code class="docutils literal notranslate"><span class="pre">just_download</span></code> is <code class="docutils literal notranslate"><span class="pre">True</span></code> the
downloads are still performed (ensuring they are cached on disk for future runs) but an <strong>empty</strong>
dataframe is returned to avoid unnecessary memory use.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>just_download</strong> (<em>bool</em>) – <ul class="simple">
<li><p><strong>False</strong> - concatenate intermediate results and return the union dataframe (default).</p></li>
<li><p><strong>True</strong> - download and cache each frame but return an empty dataframe.</p></li>
</ul>
</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><dl class="simple">
<dt>External-database relationships augmented with assembly, release, form, and organism</dt><dd><p>columns.  Empty when <code class="docutils literal notranslate"><span class="pre">just_download</span></code> is <code class="docutils literal notranslate"><span class="pre">True</span></code>.</p>
</dd>
</dl>
</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>pandas.DataFrame</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="idtrack._database_manager.DatabaseManager.create_external_all">
<span class="sig-name descname"><span class="pre">create_external_all</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">return_mode</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/idtrack/_database_manager.html#DatabaseManager.create_external_all"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#idtrack._database_manager.DatabaseManager.create_external_all" title="Link to this definition"></a></dt>
<dd><p>Download and collate cross-reference mappings from every supported genome assembly.</p>
<p>The manager cycles through every genome assembly recognised for the current organism (ordered by
<a class="reference internal" href="#idtrack._db.DB.assembly_mysqlport_priority" title="idtrack._db.DB.assembly_mysqlport_priority"><code class="xref py py-data docutils literal notranslate"><span class="pre">idtrack._db.DB.assembly_mysqlport_priority</span></code></a>), fetches the <em>external_relevant</em> mapping table
for each via <a class="reference internal" href="#idtrack._database_manager.DatabaseManager.get_db" title="idtrack._database_manager.DatabaseManager.get_db"><code class="xref py py-meth docutils literal notranslate"><span class="pre">get_db()</span></code></a>, labels every row with its source assembly, and finally
concatenates the tables.  Because this helper is intended for <strong>ad-hoc inspection only</strong>, it bypasses
the <a class="reference internal" href="#idtrack._database_manager.DatabaseManager.get_db" title="idtrack._database_manager.DatabaseManager.get_db"><code class="xref py py-meth docutils literal notranslate"><span class="pre">get_db()</span></code></a> caching layer and therefore <strong>never writes</strong> the result to the
local repository.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>return_mode</strong> (<em>str</em>) – <p>Strategy for handling rows that appear in more than one assembly.</p>
<ul class="simple">
<li><dl class="simple">
<dt><code class="docutils literal notranslate"><span class="pre">&quot;all&quot;</span></code></dt><dd><p>Keep one copy of every unique
<code class="docutils literal notranslate"><span class="pre">(release,</span> <span class="pre">graph_id,</span> <span class="pre">id_db,</span> <span class="pre">name_db,</span> <span class="pre">ensembl_identity,</span> <span class="pre">xref_identity,</span> <span class="pre">assembly)</span></code>
combination.  Duplicates are resolved <em>within</em> each assembly only.</p>
</dd>
</dl>
</li>
<li><dl class="simple">
<dt><code class="docutils literal notranslate"><span class="pre">&quot;unique&quot;</span></code></dt><dd><p>Keep one copy of every unique
<code class="docutils literal notranslate"><span class="pre">(release,</span> <span class="pre">graph_id,</span> <span class="pre">id_db,</span> <span class="pre">name_db,</span> <span class="pre">ensembl_identity,</span> <span class="pre">xref_identity)</span></code> combination <em>across</em>
<strong>all</strong> assemblies, preferring the assembly with the highest
priority. <em>(Currently no downstream use case.)</em></p>
</dd>
</dl>
</li>
<li><dl class="simple">
<dt><code class="docutils literal notranslate"><span class="pre">&quot;duplicated&quot;</span></code></dt><dd><p>Return <strong>only</strong> the rows that occur in more than one assembly as a
<code class="xref py py-class docutils literal notranslate"><span class="pre">pandas.core.groupby.generic.DataFrameGroupBy</span></code>, keyed by the same column set used for
<code class="docutils literal notranslate"><span class="pre">&quot;unique&quot;</span></code>. <em>(Currently no downstream use case.)</em></p>
</dd>
</dl>
</li>
</ul>
</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><ul class="simple">
<li><dl class="simple">
<dt>If <em>return_mode</em> is <code class="docutils literal notranslate"><span class="pre">&quot;all&quot;</span></code> or <code class="docutils literal notranslate"><span class="pre">&quot;unique&quot;</span></code>,</dt><dd><p>a de-duplicated cross-reference table with the
columns <code class="docutils literal notranslate"><span class="pre">release</span></code>, <code class="docutils literal notranslate"><span class="pre">graph_id</span></code>, <code class="docutils literal notranslate"><span class="pre">id_db</span></code>, <code class="docutils literal notranslate"><span class="pre">name_db</span></code>, <code class="docutils literal notranslate"><span class="pre">ensembl_identity</span></code>, <code class="docutils literal notranslate"><span class="pre">xref_identity</span></code>,
and <code class="docutils literal notranslate"><span class="pre">assembly</span></code>.</p>
</dd>
</dl>
</li>
<li><dl class="simple">
<dt>If <em>return_mode</em> is <code class="docutils literal notranslate"><span class="pre">&quot;duplicated&quot;</span></code>,</dt><dd><p>a group-by view containing only duplicated entries.</p>
</dd>
</dl>
</li>
</ul>
</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>Union[pandas.DataFrame, pandas.core.groupby.generic.DataFrameGroupBy]</p>
</dd>
<dt class="field-even">Raises<span class="colon">:</span></dt>
<dd class="field-even"><p><strong>ValueError</strong> – If <em>return_mode</em> is not <code class="docutils literal notranslate"><span class="pre">&quot;all&quot;</span></code>, <code class="docutils literal notranslate"><span class="pre">&quot;unique&quot;</span></code>, or <code class="docutils literal notranslate"><span class="pre">&quot;duplicated&quot;</span></code>.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="idtrack._database_manager.DatabaseManager.create_external_db">
<span class="sig-name descname"><span class="pre">create_external_db</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">filter_mode</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/idtrack/_database_manager.html#DatabaseManager.create_external_db"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#idtrack._database_manager.DatabaseManager.create_external_db" title="Link to this definition"></a></dt>
<dd><p>Retrieve Ensembl-external-ID relationships and/or database statistics.</p>
<p>This consolidates a complex SQL join—spanning Ensembl core tables <em>gene</em>, <em>transcript</em>, <em>translation</em> and
the cross-reference tables <em>xref</em>, <em>object_xref</em>, <em>identity_xref</em>, <em>external_db</em>, and <em>external_synonym</em>—into a
single pandas dataframe. It enables downstream analyses such as mapping Ensembl gene models to
UniProt, RefSeq, or CCDS identifiers, or summarising which external sources are represented in a given Ensembl
release.  The result type and granularity are controlled by <em>filter_mode</em>, allowing either the raw relationship
rows or a per-database count to be returned.</p>
<p>The query executed is conceptually equivalent to the (simplified) MySQL statement below,
though the actual SQL is constructed programmatically for flexibility and performance:</p>
<div class="highlight-sql notranslate"><div class="highlight"><pre><span></span><span class="k">SELECT</span><span class="w"> </span><span class="k">g</span><span class="p">.</span><span class="n">stable_id</span><span class="p">,</span><span class="w"> </span><span class="n">t</span><span class="p">.</span><span class="n">stable_id</span><span class="p">,</span><span class="w"> </span><span class="n">tr</span><span class="p">.</span><span class="n">stable_id</span><span class="p">,</span><span class="w"> </span><span class="n">x</span><span class="p">.</span><span class="n">dbprimary_acc</span><span class="p">,</span><span class="w"> </span><span class="n">edb</span><span class="p">.</span><span class="n">db_name</span><span class="p">,</span><span class="w"> </span><span class="n">es</span><span class="p">.</span><span class="n">synonym</span><span class="p">,</span><span class="w"> </span><span class="n">ix</span><span class="p">.</span><span class="o">*</span>
<span class="k">FROM</span><span class="w"> </span><span class="n">gene</span><span class="w"> </span><span class="k">g</span>
<span class="k">JOIN</span><span class="w"> </span><span class="n">transcript</span><span class="w"> </span><span class="n">t</span><span class="w"> </span><span class="k">USING</span><span class="w"> </span><span class="p">(</span><span class="n">gene_id</span><span class="p">)</span>
<span class="k">JOIN</span><span class="w"> </span><span class="k">translation</span><span class="w"> </span><span class="n">tr</span><span class="w"> </span><span class="k">USING</span><span class="w"> </span><span class="p">(</span><span class="n">transcript_id</span><span class="p">)</span>
<span class="k">JOIN</span><span class="w"> </span><span class="n">object_xref</span><span class="w"> </span><span class="n">ox</span><span class="w"> </span><span class="k">ON</span><span class="w"> </span><span class="p">(</span><span class="k">g</span><span class="p">.</span><span class="n">gene_id</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">ox</span><span class="p">.</span><span class="n">ensembl_id</span><span class="w"> </span><span class="k">AND</span><span class="w"> </span><span class="n">ox</span><span class="p">.</span><span class="n">ensembl_object_type</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="ss">&quot;Gene&quot;</span><span class="p">)</span>
<span class="k">JOIN</span><span class="w"> </span><span class="n">xref</span><span class="w"> </span><span class="n">x</span><span class="w"> </span><span class="k">ON</span><span class="w"> </span><span class="p">(</span><span class="n">ox</span><span class="p">.</span><span class="n">xref_id</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">x</span><span class="p">.</span><span class="n">xref_id</span><span class="p">)</span>
<span class="k">LEFT</span><span class="w"> </span><span class="k">JOIN</span><span class="w"> </span><span class="n">external_db</span><span class="w"> </span><span class="n">edb</span><span class="w"> </span><span class="k">ON</span><span class="w"> </span><span class="p">(</span><span class="n">x</span><span class="p">.</span><span class="n">external_db_id</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">edb</span><span class="p">.</span><span class="n">external_db_id</span><span class="p">)</span>
<span class="k">LEFT</span><span class="w"> </span><span class="k">JOIN</span><span class="w"> </span><span class="n">identity_xref</span><span class="w"> </span><span class="n">ix</span><span class="w"> </span><span class="k">ON</span><span class="w"> </span><span class="p">(</span><span class="n">ox</span><span class="p">.</span><span class="n">object_xref_id</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">ix</span><span class="p">.</span><span class="n">object_xref_id</span><span class="p">)</span>
<span class="k">LEFT</span><span class="w"> </span><span class="k">JOIN</span><span class="w"> </span><span class="n">external_synonym</span><span class="w"> </span><span class="n">es</span><span class="w"> </span><span class="k">ON</span><span class="w"> </span><span class="p">(</span><span class="n">x</span><span class="p">.</span><span class="n">xref_id</span><span class="w"> </span><span class="o">=</span><span class="w"> </span><span class="n">es</span><span class="p">.</span><span class="n">xref_id</span><span class="p">)</span>
<span class="k">LIMIT</span><span class="w"> </span><span class="mi">10</span><span class="p">;</span>
</pre></div>
</div>
<p>When tighter genomic scoping is required the <em>gene</em> table can be prefixed with <em>coord_system</em> and <em>seq_region</em>:</p>
<div class="highlight-sql notranslate"><div class="highlight"><pre><span></span><span class="k">FROM</span><span class="w"> </span><span class="n">coord_system</span><span class="w"> </span><span class="n">cs</span>
<span class="k">JOIN</span><span class="w"> </span><span class="n">seq_region</span><span class="w">  </span><span class="n">sr</span><span class="w"> </span><span class="k">USING</span><span class="w"> </span><span class="p">(</span><span class="n">coord_system_id</span><span class="p">)</span>
<span class="k">JOIN</span><span class="w"> </span><span class="n">gene</span><span class="w">        </span><span class="k">g</span><span class="w">  </span><span class="k">USING</span><span class="w"> </span><span class="p">(</span><span class="n">seq_region_id</span><span class="p">)</span>
</pre></div>
</div>
<p>You can experiment interactively against the public Ensembl MySQL mirror:</p>
<div class="highlight-bash notranslate"><div class="highlight"><pre><span></span>mysql<span class="w"> </span>--user<span class="o">=</span>anonymous<span class="w"> </span>--host<span class="o">=</span>ensembldb.ensembl.org<span class="w"> </span>-D<span class="w"> </span>homo_sapiens_core_105_38<span class="w"> </span>-A
<span class="c1"># Schema reference:</span>
<span class="c1"># https://m.ensembl.org/info/docs/api/core/core_schema.html</span>
</pre></div>
</div>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>filter_mode</strong> (<em>str</em>) – <p>Controls both the <strong>row subset</strong> and the <strong>output schema</strong>. Must be one of:</p>
<ul class="simple">
<li><p><code class="docutils literal notranslate"><span class="pre">&quot;all&quot;</span></code> - return <strong>every</strong> mapping found in MySQL, no post-filtering applied.</p></li>
<li><dl class="simple">
<dt><code class="docutils literal notranslate"><span class="pre">&quot;relevant&quot;</span></code> - return only mappings whose external database is marked <em>Include: true</em> in the</dt><dd><p><code class="xref py py-meth docutils literal notranslate"><span class="pre">ExternalDatabases.give_list_for_case()</span></code> YAML configuration.</p>
</dd>
</dl>
</li>
<li><p><code class="docutils literal notranslate"><span class="pre">&quot;database&quot;</span></code> - return a two-column summary (<code class="docutils literal notranslate"><span class="pre">name_db</span></code>, <code class="docutils literal notranslate"><span class="pre">count</span></code>) for <strong>all</strong> external databases.</p></li>
<li><dl class="simple">
<dt><code class="docutils literal notranslate"><span class="pre">&quot;relevant-database&quot;</span></code> - as above, but restricted to databases flagged <em>Include: true</em>.</dt><dd><p>The special values <code class="docutils literal notranslate"><span class="pre">&quot;relevant&quot;</span></code> and <code class="docutils literal notranslate"><span class="pre">&quot;relevant-database&quot;</span></code> implicitly consult the cached
<a class="reference internal" href="#idtrack._database_manager.DatabaseManager.external_inst" title="idtrack._database_manager.DatabaseManager.external_inst"><code class="xref py py-attr docutils literal notranslate"><span class="pre">external_inst</span></code></a> to honour the user’s curated allow-list.</p>
</dd>
</dl>
</li>
</ul>
</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><ul class="simple">
<li><dl class="simple">
<dt>For <code class="docutils literal notranslate"><span class="pre">&quot;all&quot;</span></code> / <code class="docutils literal notranslate"><span class="pre">&quot;relevant&quot;</span></code> - six-column frame</dt><dd><p><code class="docutils literal notranslate"><span class="pre">[&quot;release&quot;,</span> <span class="pre">&quot;graph_id&quot;,</span> <span class="pre">&quot;id_db&quot;,</span> <span class="pre">&quot;name_db&quot;,</span> <span class="pre">&quot;ensembl_identity&quot;,</span> <span class="pre">&quot;xref_identity&quot;]</span></code> holding one
row per Ensembl→external identifier edge. <code class="docutils literal notranslate"><span class="pre">graph_id</span></code> is the Ensembl stable ID (+version), while
the two <em>identity</em> columns store Smith-Waterman percent identities (<em>float16</em>) for QC.</p>
</dd>
</dl>
</li>
<li><dl class="simple">
<dt>For <code class="docutils literal notranslate"><span class="pre">&quot;database&quot;</span></code> / <code class="docutils literal notranslate"><span class="pre">&quot;relevant-database&quot;</span></code> - two-column frame</dt><dd><p><code class="docutils literal notranslate"><span class="pre">[&quot;name_db&quot;,</span> <span class="pre">&quot;count&quot;]</span></code> giving how many distinct <code class="docutils literal notranslate"><span class="pre">graph_id</span></code> values each external database
touches. <code class="docutils literal notranslate"><span class="pre">count</span></code> is an <code class="docutils literal notranslate"><span class="pre">int64</span></code>.</p>
</dd>
</dl>
</li>
</ul>
</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>pandas.DataFrame</p>
</dd>
<dt class="field-even">Raises<span class="colon">:</span></dt>
<dd class="field-even"><p><strong>ValueError</strong> – If <em>filter_mode</em> is not one of the accepted literals <strong>or</strong> if the YAML allow-list claims a
    database that is absent from the retrieved mappings—indicating the configuration and MySQL data are
    out of sync.</p>
</dd>
</dl>
<p class="rubric">Notes</p>
<p><em>Synonym handling</em> - any synonym brought in from <code class="docutils literal notranslate"><span class="pre">external_synonym</span></code> is prefixed with
<code class="xref py py-data docutils literal notranslate"><span class="pre">DB.synonym_id_nodes_prefix</span></code>, and its <code class="docutils literal notranslate"><span class="pre">name_db</span></code> is likewise prefixed so that synonym nodes remain
distinguishable during graph building.
<em>Caching</em> - the heavy MySQL queries are executed only if the processed frame is not already present in the
manager’s per-organism HDF5 cache; otherwise the cached frame is read from disk, ensuring repeat calls are
inexpensive.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="idtrack._database_manager.DatabaseManager.create_id_history">
<span class="sig-name descname"><span class="pre">create_id_history</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">narrow</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/idtrack/_database_manager.html#DatabaseManager.create_id_history"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#idtrack._database_manager.DatabaseManager.create_id_history" title="Link to this definition"></a></dt>
<dd><p>Retrieve historical relationships between successive Ensembl stable IDs.</p>
<p>Build a cross-release lineage table mapping every obsolete ID version to its immediate successor for the
configured <em>organism</em>, <em>form</em>, and release window.  The information is assembled from the Core tables
<code class="docutils literal notranslate"><span class="pre">stable_id_event</span></code> and <code class="docutils literal notranslate"><span class="pre">mapping_session</span></code> and then normalised so that all identifiers follow the canonical
<code class="docutils literal notranslate"><span class="pre">&lt;stable_id&gt;.&lt;version&gt;</span></code> convention.  Downstream graph-construction utilities depend on this table to
reconstruct how genes, transcripts, or translations evolve across Ensembl releases.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>narrow</strong> (<em>bool</em>) – If <code class="docutils literal notranslate"><span class="pre">True</span></code> drop auxiliary columns (mapping session metadata, assembly labels, creation
timestamps, etc.) to minimise on-disk footprint; otherwise return the full schema for exploratory
analyses.</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><p>Seven-column table with the following fields, ordered as listed—</p>
<blockquote>
<div><ul class="simple">
<li><p><code class="docutils literal notranslate"><span class="pre">old_stable_id</span></code> - obsolete identifier (empty string for “birth” events).</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">old_version</span></code>   - version number paired with <em>old_stable_id</em>.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">new_stable_id</span></code> - successor identifier (empty string for “retirement” events).</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">new_version</span></code>   - version paired with <em>new_stable_id</em>.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">score</span></code>         - homology score reported by Ensembl (<code class="docutils literal notranslate"><span class="pre">NaN</span></code> if unavailable).</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">old_release</span></code>   - Ensembl release where the <em>old</em> identifier last appeared.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">new_release</span></code>   - release where the <em>new</em> identifier first appeared.</p></li>
</ul>
</div></blockquote>
</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>pandas.DataFrame</p>
</dd>
<dt class="field-even">Raises<span class="colon">:</span></dt>
<dd class="field-even"><p><strong>ValueError</strong> – If the identifier delimiter
    <a class="reference internal" href="#idtrack._db.DB.id_ver_delimiter" title="idtrack._db.DB.id_ver_delimiter"><code class="xref py py-data docutils literal notranslate"><span class="pre">idtrack._db.DB.id_ver_delimiter</span></code></a> is found inside any <code class="docutils literal notranslate"><span class="pre">*_stable_id</span></code> field, indicating
    malformed input.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="idtrack._database_manager.DatabaseManager.create_id_history_fixed">
<span class="sig-name descname"><span class="pre">create_id_history_fixed</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">narrow</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">inspect</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/idtrack/_database_manager.html#DatabaseManager.create_id_history_fixed"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#idtrack._database_manager.DatabaseManager.create_id_history_fixed" title="Link to this definition"></a></dt>
<dd><p>Create a corrected ID-history table that repairs cyclic or duplicated version transitions (deprecated).</p>
<p>Certain edge cases in the raw <code class="docutils literal notranslate"><span class="pre">idhistory</span></code> extraction—e.g. <em>Homo sapiens</em> <code class="docutils literal notranslate"><span class="pre">ENSG00000232423</span></code> at release 105—
produce sequences like <code class="docutils literal notranslate"><span class="pre">1</span> <span class="pre">→</span> <span class="pre">2,</span> <span class="pre">2</span> <span class="pre">→</span> <span class="pre">3,</span> <span class="pre">1</span> <span class="pre">→</span> <span class="pre">2</span></code> where an already retired version resurfaces later on.
Such cycles violate the monotonic version semantics assumed by graph algorithms.  This helper rewrites the
offending rows so that once a version is superseded it never reappears, transforming the above sequence into
the logically consistent <code class="docutils literal notranslate"><span class="pre">3</span> <span class="pre">→</span> <span class="pre">2</span></code>.  The routine is retained for reproducibility but superseded by
<a class="reference internal" href="#idtrack._database_manager.DatabaseManager.create_id_history" title="idtrack._database_manager.DatabaseManager.create_id_history"><code class="xref py py-meth docutils literal notranslate"><span class="pre">DatabaseManager.create_id_history()</span></code></a>.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>narrow</strong> (<em>bool</em>) – Propagated to the underlying data fetch—when <code class="docutils literal notranslate"><span class="pre">True</span></code> start from the column-reduced
<code class="docutils literal notranslate"><span class="pre">idhistory_narrow</span></code> view instead of the full table.</p></li>
<li><p><strong>inspect</strong> (<em>bool</em>) – When <code class="docutils literal notranslate"><span class="pre">True</span></code> add diagnostic columns (e.g. <code class="docutils literal notranslate"><span class="pre">changed_old</span></code> and <code class="docutils literal notranslate"><span class="pre">changed_new</span></code>) to aid
manual auditing of the corrections; when <code class="docutils literal notranslate"><span class="pre">False</span></code> return only the cleaned canonical schema.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><dl class="simple">
<dt>Corrected seven-column table <code class="docutils literal notranslate"><span class="pre">old_stable_id</span></code>, <code class="docutils literal notranslate"><span class="pre">old_version</span></code>, <code class="docutils literal notranslate"><span class="pre">new_stable_id</span></code>,</dt><dd><p><code class="docutils literal notranslate"><span class="pre">new_version</span></code>, <code class="docutils literal notranslate"><span class="pre">score</span></code>, <code class="docutils literal notranslate"><span class="pre">old_release</span></code>, <code class="docutils literal notranslate"><span class="pre">new_release</span></code>—ready for serialization and downstream use.</p>
</dd>
</dl>
</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>pandas.DataFrame</p>
</dd>
</dl>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>This function is deprecated and will be removed in a future major release once the core extractor fully
addresses the ordering anomaly.</p>
</div>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="idtrack._database_manager.DatabaseManager.create_ids">
<span class="sig-name descname"><span class="pre">create_ids</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">form</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/idtrack/_database_manager.html#DatabaseManager.create_ids"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#idtrack._database_manager.DatabaseManager.create_ids" title="Link to this definition"></a></dt>
<dd><p>Retrieve and normalise raw Ensembl identifier records for the requested molecular form.</p>
<p>This method pulls the appropriate MySQL table(s) for <em>form</em>, copes with schema differences
across Ensembl releases (e.g. the historical <code class="docutils literal notranslate"><span class="pre">*_stable_id</span></code> split tables), coerces data
types, and standardises column names so that downstream graph-building steps all consume the
same shape.  It finishes by delegating to <a class="reference internal" href="#idtrack._database_manager.DatabaseManager.version_uniformize" title="idtrack._database_manager.DatabaseManager.version_uniformize"><code class="xref py py-meth docutils literal notranslate"><span class="pre">DatabaseManager.version_uniformize()</span></code></a> to
ensure the <em>Version</em> field is either a proper integer or <code class="docutils literal notranslate"><span class="pre">NaN</span></code> across the entire DataFrame.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>form</strong> (<em>str</em>) – Target molecular form - <code class="docutils literal notranslate"><span class="pre">&quot;gene&quot;</span></code>, <code class="docutils literal notranslate"><span class="pre">&quot;transcript&quot;</span></code>, or <code class="docutils literal notranslate"><span class="pre">&quot;translation&quot;</span></code>.
Anything else triggers a <code class="xref py py-class docutils literal notranslate"><span class="pre">ValueError</span></code>.</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><p>A de-duplicated, index-reset table whose columns depend on <em>form</em>:</p>
<blockquote>
<div><ul class="simple">
<li><p><strong>gene</strong> - <code class="docutils literal notranslate"><span class="pre">gene_id</span></code>, <code class="docutils literal notranslate"><span class="pre">gene_stable_id</span></code>, <code class="docutils literal notranslate"><span class="pre">gene_version</span></code></p></li>
<li><p><strong>transcript</strong> - <code class="docutils literal notranslate"><span class="pre">transcript_id</span></code>, <code class="docutils literal notranslate"><span class="pre">gene_id</span></code>, <code class="docutils literal notranslate"><span class="pre">transcript_stable_id</span></code>,
<code class="docutils literal notranslate"><span class="pre">transcript_version</span></code></p></li>
<li><p><strong>translation</strong> - <code class="docutils literal notranslate"><span class="pre">translation_id</span></code>, <code class="docutils literal notranslate"><span class="pre">transcript_id</span></code>, <code class="docutils literal notranslate"><span class="pre">translation_stable_id</span></code>,
<code class="docutils literal notranslate"><span class="pre">translation_version</span></code></p></li>
</ul>
<p>All ID columns are <code class="docutils literal notranslate"><span class="pre">int64</span></code> except the <code class="docutils literal notranslate"><span class="pre">*_stable_id</span></code> strings; version columns are
<code class="docutils literal notranslate"><span class="pre">int64</span></code> or <code class="docutils literal notranslate"><span class="pre">float64</span></code> (with <code class="docutils literal notranslate"><span class="pre">NaN</span></code> when absent).</p>
</div></blockquote>
</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>pandas.DataFrame</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="idtrack._database_manager.DatabaseManager.create_relation_archive">
<span class="sig-name descname"><span class="pre">create_relation_archive</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/idtrack/_database_manager.html#DatabaseManager.create_relation_archive"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#idtrack._database_manager.DatabaseManager.create_relation_archive" title="Link to this definition"></a></dt>
<dd><p>Retrieve a cross-release gene-transcript-translation mapping table.</p>
<p>This legacy helper pulls the Ensembl <code class="docutils literal notranslate"><span class="pre">gene_archive</span></code> table—spanning <em>all</em>
releases for the current organism—via
<a class="reference internal" href="#idtrack._database_manager.DatabaseManager.get_table" title="idtrack._database_manager.DatabaseManager.get_table"><code class="xref py py-meth docutils literal notranslate"><span class="pre">DatabaseManager.get_table()</span></code></a>, drops columns unrelated to identifier
mapping, and passes the result to
<a class="reference internal" href="#idtrack._database_manager.DatabaseManager._create_relation_helper" title="idtrack._database_manager.DatabaseManager._create_relation_helper"><code class="xref py py-meth docutils literal notranslate"><span class="pre">DatabaseManager._create_relation_helper()</span></code></a>.  <strong>Because the archive
contains known gaps, the preferred workflow is to call</strong>
<a class="reference internal" href="#idtrack._database_manager.DatabaseManager.create_relation_current" title="idtrack._database_manager.DatabaseManager.create_relation_current"><code class="xref py py-meth docutils literal notranslate"><span class="pre">DatabaseManager.create_relation_current()</span></code></a> <strong>once per release and
concatenate the outputs.</strong></p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p><dl class="simple">
<dt>Same schema as</dt><dd><p><a class="reference internal" href="#idtrack._database_manager.DatabaseManager.create_relation_current" title="idtrack._database_manager.DatabaseManager.create_relation_current"><code class="xref py py-meth docutils literal notranslate"><span class="pre">DatabaseManager.create_relation_current()</span></code></a>—<code class="docutils literal notranslate"><span class="pre">gene</span></code>,
<code class="docutils literal notranslate"><span class="pre">transcript</span></code>, <code class="docutils literal notranslate"><span class="pre">translation</span></code>—but potentially with missing rows
because Ensembl did not always back-populate older releases.</p>
</dd>
</dl>
</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p>pandas.DataFrame</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="idtrack._database_manager.DatabaseManager.create_relation_current">
<span class="sig-name descname"><span class="pre">create_relation_current</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/idtrack/_database_manager.html#DatabaseManager.create_relation_current"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#idtrack._database_manager.DatabaseManager.create_relation_current" title="Link to this definition"></a></dt>
<dd><p>Build a current-release gene-transcript-translation mapping table.</p>
<p>The routine fetches the <em>raw</em> stable-ID/​version tables for genes,
transcripts and translations via <a class="reference internal" href="#idtrack._database_manager.DatabaseManager.get_db" title="idtrack._database_manager.DatabaseManager.get_db"><code class="xref py py-meth docutils literal notranslate"><span class="pre">DatabaseManager.get_db()</span></code></a>, merges
them into a single wide frame, and then delegates to
<a class="reference internal" href="#idtrack._database_manager.DatabaseManager._create_relation_helper" title="idtrack._database_manager.DatabaseManager._create_relation_helper"><code class="xref py py-meth docutils literal notranslate"><span class="pre">DatabaseManager._create_relation_helper()</span></code></a> to harmonise version
columns and compress the information into three canonical node labels
(<code class="docutils literal notranslate"><span class="pre">&quot;&lt;stable_id&gt;.&lt;version&gt;&quot;</span></code>).  The resulting mapping is the authoritative
per-release link between molecular forms and is consumed by downstream
graph-building utilities such as <code class="xref py py-meth docutils literal notranslate"><span class="pre">DatabaseManager.create_graph()</span></code>.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>Three columns—<code class="docutils literal notranslate"><span class="pre">gene</span></code>, <code class="docutils literal notranslate"><span class="pre">transcript</span></code>, and
<code class="docutils literal notranslate"><span class="pre">translation</span></code>—with one row per transcript.  The <code class="docutils literal notranslate"><span class="pre">translation</span></code> column
may contain empty strings where non-coding transcripts have no
peptide.  All data are UTF-8 strings; duplicates are removed and the
index is reset.</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p>pandas.DataFrame</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="idtrack._database_manager.DatabaseManager.create_release_id">
<span class="sig-name descname"><span class="pre">create_release_id</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/idtrack/_database_manager.html#DatabaseManager.create_release_id"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#idtrack._database_manager.DatabaseManager.create_release_id" title="Link to this definition"></a></dt>
<dd><p>Return deduplicated stable-identifier/version pairs for the current form and release.</p>
<p>Raw identifiers are fetched via <a class="reference internal" href="#idtrack._database_manager.DatabaseManager.get_db" title="idtrack._database_manager.DatabaseManager.get_db"><code class="xref py py-meth docutils literal notranslate"><span class="pre">DatabaseManager.get_db()</span></code></a>, normalised with
<a class="reference internal" href="#idtrack._database_manager.DatabaseManager.version_fix" title="idtrack._database_manager.DatabaseManager.version_fix"><code class="xref py py-meth docutils literal notranslate"><span class="pre">DatabaseManager.version_fix()</span></code></a>, trimmed to the canonical columns, and sanity-checked.  Two
integrity rules are enforced: (1) the delimiter
<a class="reference internal" href="#idtrack._db.DB.id_ver_delimiter" title="idtrack._db.DB.id_ver_delimiter"><code class="xref py py-data docutils literal notranslate"><span class="pre">idtrack._db.DB.id_ver_delimiter</span></code></a> must <strong>not</strong> appear inside any stable identifier, and
(2) every stable identifier must be unique after deduplication.  Violations raise
<code class="xref py py-class docutils literal notranslate"><span class="pre">ValueError</span></code>.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>Two-column dataframe <code class="docutils literal notranslate"><span class="pre">[{form}_stable_id,</span> <span class="pre">{form}_version]</span></code> with duplicates removed.</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p>pandas.DataFrame</p>
</dd>
<dt class="field-odd">Raises<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>ValueError</strong> – If the delimiter is present inside any stable identifier or if identifiers are not
    unique after deduplication.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="idtrack._database_manager.DatabaseManager.create_version_info">
<span class="sig-name descname"><span class="pre">create_version_info</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/idtrack/_database_manager.html#DatabaseManager.create_version_info"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#idtrack._database_manager.DatabaseManager.create_version_info" title="Link to this definition"></a></dt>
<dd><p>Determine whether each Ensembl release stores identifiers with or without version suffixes.</p>
<p>Ensembl stable identifiers can appear either <em>with</em> a <code class="docutils literal notranslate"><span class="pre">.version</span></code> facet (e.g. <em>ENSG00000139618.17</em>)
or <em>without</em> it (e.g. <em>YAL001C</em> in <em>S. cerevisiae</em>).  For robust cross-release tracking the package
needs to know which convention applies to every release of the current organism.  The method loops
over <a class="reference internal" href="#idtrack._database_manager.DatabaseManager.available_releases" title="idtrack._database_manager.DatabaseManager.available_releases"><code class="xref py py-attr docutils literal notranslate"><span class="pre">available_releases</span></code></a>, downloads the raw identifier table for <em>self.form</em>, and inspects
the <code class="docutils literal notranslate"><span class="pre">&lt;form&gt;_version</span></code> column:</p>
<ul class="simple">
<li><p><strong>All values NaN</strong> → the release uses <em>unversioned</em> identifiers.</p></li>
<li><p><strong>No values NaN</strong>  → the release uses <em>versioned</em> identifiers.</p></li>
<li><p><strong>Mixed NaN / non-NaN</strong> → unsupported; raises <code class="xref py py-class docutils literal notranslate"><span class="pre">NotImplementedError</span></code>.</p></li>
</ul>
<p>The outcome is encoded as a Boolean flag per release and later consumed by
<a class="reference internal" href="#idtrack._database_manager.DatabaseManager.check_version_info" title="idtrack._database_manager.DatabaseManager.check_version_info"><code class="xref py py-meth docutils literal notranslate"><span class="pre">check_version_info()</span></code></a> to decide whether version strings should be kept,
stripped, or synthesised.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p><dl class="simple">
<dt>Two-column table with:</dt><dd><ul class="simple">
<li><p><code class="docutils literal notranslate"><span class="pre">ensembl_release</span></code> - integer release number.</p></li>
<li><dl class="simple">
<dt><code class="docutils literal notranslate"><span class="pre">version_info</span></code>   - <code class="docutils literal notranslate"><span class="pre">True</span></code> if <em>all</em> identifiers <strong>lack</strong> a version suffix, <code class="docutils literal notranslate"><span class="pre">False</span></code> if</dt><dd><p><em>all</em> identifiers <strong>include</strong> a version suffix.</p>
</dd>
</dl>
</li>
</ul>
</dd>
</dl>
</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p>pandas.DataFrame</p>
</dd>
<dt class="field-odd">Raises<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>NotImplementedError</strong> – If any individual release contains a mixture of versioned and unversioned
    identifiers, indicating an inconsistent upstream annotation.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="idtrack._database_manager.DatabaseManager.download_table">
<span class="sig-name descname"><span class="pre">download_table</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">table_key</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">usecols</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/idtrack/_database_manager.html#DatabaseManager.download_table"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#idtrack._database_manager.DatabaseManager.download_table" title="Link to this definition"></a></dt>
<dd><p>Download a raw Ensembl MySQL table and return it as a DataFrame.</p>
<p>The method forms the low-level backbone of all table acquisition in <em>IDTrackDocs</em>.  It opens a
direct connection to the Ensembl Core (or comparable) MySQL schema configured on the current
<a class="reference internal" href="#idtrack._database_manager.DatabaseManager" title="idtrack._database_manager.DatabaseManager"><code class="xref py py-class docutils literal notranslate"><span class="pre">DatabaseManager</span></code></a> instance, issues a <cite>SELECT</cite> statement against <em>table_key</em>, converts
the results into a <code class="xref py py-class docutils literal notranslate"><span class="pre">pandas.DataFrame</span></code>, and performs a minimal sanitisation pass
(bytes-to-string decoding, column subset validation, logging).  Public code is expected to call
<a class="reference internal" href="#idtrack._database_manager.DatabaseManager.get_table" title="idtrack._database_manager.DatabaseManager.get_table"><code class="xref py py-meth docutils literal notranslate"><span class="pre">DatabaseManager.get_table()</span></code></a>, which wraps this helper with caching and post-processing,
but keeping this routine separate allows fine-grained testing, mocking, and reuse in advanced
workflows.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>table_key</strong> (<em>str</em>) – Name of the raw table as it appears in the remote Ensembl database
(e.g. <code class="docutils literal notranslate"><span class="pre">'gene'</span></code>, <code class="docutils literal notranslate"><span class="pre">'mapping_session'</span></code>, <code class="docutils literal notranslate"><span class="pre">'xref'</span></code>).  Must exist in the schema
returned by <a class="reference internal" href="#idtrack._database_manager.DatabaseManager.mysql_database" title="idtrack._database_manager.DatabaseManager.mysql_database"><code class="xref py py-meth docutils literal notranslate"><span class="pre">DatabaseManager.mysql_database()</span></code></a>.</p></li>
<li><p><strong>usecols</strong> (<em>Optional</em><em>[</em><em>list</em><em>[</em><em>str</em><em>]</em><em>]</em>) – Sequence of column names to project; <em>None</em> retrieves the
entire table.  Column order is preserved.  An empty list is treated the same as <em>None</em>.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><dl class="simple">
<dt>A frame containing the requested columns in the exact order supplied via</dt><dd><p><em>usecols</em> (or all columns if <em>usecols</em> is <em>None</em>).  Index is monotonic and zero-based.</p>
</dd>
</dl>
</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>pandas.DataFrame</p>
</dd>
<dt class="field-even">Raises<span class="colon">:</span></dt>
<dd class="field-even"><p><strong>ValueError</strong> – If any element of <em>usecols</em> is missing from <em>table_key</em>, or if the query returns
    binary payloads that cannot be coerced into native Python types.</p>
</dd>
</dl>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="idtrack._database_manager.DatabaseManager.external_inst">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">external_inst</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><a class="reference internal" href="#idtrack._external_databases.ExternalDatabases" title="idtrack._external_databases.ExternalDatabases"><span class="pre">ExternalDatabases</span></a></em><a class="headerlink" href="#idtrack._database_manager.DatabaseManager.external_inst" title="Link to this definition"></a></dt>
<dd><p>Instantiate and cache an <code class="xref py py-class docutils literal notranslate"><span class="pre">ExternalDatabases</span></code> helper for this manager.</p>
<p>The instance mirrors the configuration of the surrounding <a class="reference internal" href="#idtrack._database_manager.DatabaseManager" title="idtrack._database_manager.DatabaseManager"><code class="xref py py-class docutils literal notranslate"><span class="pre">DatabaseManager</span></code></a>—organism,
Ensembl release, identifier form, local repository path, and genome assembly—so that all
interactions with external data sources remain consistent throughout the session.  Because the
property is backed by <code class="xref py py-data docutils literal notranslate"><span class="pre">functools.cached_property</span></code>, the helper is created exactly once
and reused on subsequent accesses, eliminating redundant network or file-system look-ups.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>A lazily created, configuration-matched helper object.</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><a class="reference internal" href="#idtrack._external_databases.ExternalDatabases" title="idtrack._external_databases.ExternalDatabases">ExternalDatabases</a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="idtrack._database_manager.DatabaseManager.file_name">
<span class="sig-name descname"><span class="pre">file_name</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">df_type</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">*</span></span><span class="n"><span class="pre">args</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ensembl_release</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/idtrack/_database_manager.html#DatabaseManager.file_name"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#idtrack._database_manager.DatabaseManager.file_name" title="Link to this definition"></a></dt>
<dd><p>Resolve HDF5 hierarchy key and absolute file path for a dataframe request.</p>
<p>This internal helper centralises every rule that <a class="reference internal" href="#idtrack._database_manager.DatabaseManager" title="idtrack._database_manager.DatabaseManager"><code class="xref py py-class docutils literal notranslate"><span class="pre">DatabaseManager</span></code></a> uses to build HDF5
<em>hierarchy</em> keys and their corresponding on-disk filenames, ensuring that any two call-sites
confronted with the same combination of organism, genome assembly, Ensembl release, dataframe
<em>kind</em>, and optional column subset produce <strong>identical</strong> results.  By funnelling every I/O
operation through this method the wider package avoids silent cache misses, duplicate downloads,
and hard-to-trace inconsistencies in downstream analytics.  Public code is expected to invoke
higher-level wrappers such as <a class="reference internal" href="#idtrack._database_manager.DatabaseManager.get_db" title="idtrack._database_manager.DatabaseManager.get_db"><code class="xref py py-meth docutils literal notranslate"><span class="pre">DatabaseManager.get_db()</span></code></a>; use this routine only when
implementing new caching utilities or in low-level tests.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>df_type</strong> (<em>str</em>) – Category of dataframe whose name is required. Accepted values are
<code class="docutils literal notranslate"><span class="pre">&quot;processed&quot;</span></code>, <code class="docutils literal notranslate"><span class="pre">&quot;mysql&quot;</span></code>, and <code class="docutils literal notranslate"><span class="pre">&quot;common&quot;</span></code>; any other string triggers <code class="xref py py-class docutils literal notranslate"><span class="pre">ValueError</span></code>.</p></li>
<li><p><strong>ensembl_release</strong> (<em>int</em><em>, </em><em>optional</em>) – Ensembl release to encode in the filename.  If <em>None</em>, the
current <code class="xref py py-attr docutils literal notranslate"><span class="pre">DatabaseManager.ensembl_release</span></code> is used instead.</p></li>
<li><p><strong>kwargs</strong> – Additional keyword arguments forwarded to the helper that handles the selected
<em>df_type</em> (currently only <code class="docutils literal notranslate"><span class="pre">usecols</span></code> for the <em>mysql</em> path).</p></li>
<li><p><strong>args</strong> – <p>Positional arguments interpreted according to <em>df_type</em>:</p>
<ul>
<li><p><strong>processed</strong> - <code class="docutils literal notranslate"><span class="pre">df_indicator</span></code> (str): symbolic label such as <code class="docutils literal notranslate"><span class="pre">&quot;idhistory&quot;</span></code> or
<code class="docutils literal notranslate"><span class="pre">&quot;idsraw_gene&quot;</span></code>.  The manager appends <code class="xref py py-attr docutils literal notranslate"><span class="pre">DatabaseManager.form</span></code> so that artefacts
for different biological forms do not collide.</p></li>
<li><p><strong>mysql</strong> - <code class="docutils literal notranslate"><span class="pre">table_key</span></code> (str): raw MySQL table name (e.g. <code class="docutils literal notranslate"><span class="pre">&quot;gene&quot;</span></code>, <code class="docutils literal notranslate"><span class="pre">&quot;exon&quot;</span></code>).
An optional <code class="docutils literal notranslate"><span class="pre">usecols</span></code> (list[str]) must then be supplied via <em>kwargs</em>; the column list
is embedded in the hierarchy using the delimiter held in
<code class="xref py py-attr docutils literal notranslate"><span class="pre">DatabaseManager._column_sep</span></code>.</p></li>
<li><p><strong>common</strong> - <code class="docutils literal notranslate"><span class="pre">df_indicator</span></code> (str): same as the processed case <strong>but without</strong> the form
suffix, allowing cross-form artefacts (e.g. <code class="docutils literal notranslate"><span class="pre">&quot;availabledatabases&quot;</span></code>) to share a single key.</p></li>
</ul>
</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><dl class="simple">
<dt>Two-element tuple <code class="docutils literal notranslate"><span class="pre">(hierarchy_key,</span> <span class="pre">file_path)</span></code> where <em>hierarchy_key</em> is</dt><dd><p>the internal node path (e.g. <code class="docutils literal notranslate"><span class="pre">&quot;ens111_mysql_gene_COL_gene_id&quot;</span></code>) and <em>file_path</em> is the
absolute path to <code class="docutils literal notranslate"><span class="pre">&lt;local_repository&gt;/&lt;organism&gt;_assembly-&lt;assembly&gt;.h5</span></code>.  The path is
<strong>not</strong> created on disk—callers remain responsible for reading or writing the HDF5 file.</p>
</dd>
</dl>
</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>tuple[str, str]</p>
</dd>
<dt class="field-even">Raises<span class="colon">:</span></dt>
<dd class="field-even"><p><strong>ValueError</strong> – If <em>df_type</em> is not one of the accepted categories or if the positional/keyword
    argument combination does not satisfy the expectations for that category (e.g. missing
    <code class="docutils literal notranslate"><span class="pre">table_key</span></code> when <em>df_type</em> is <code class="docutils literal notranslate"><span class="pre">&quot;mysql&quot;</span></code>).</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="idtrack._database_manager.DatabaseManager.get_db">
<span class="sig-name descname"><span class="pre">get_db</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">df_indicator</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">create_even_if_exist</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">save_after_calculation</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">overwrite_even_if_exist</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/idtrack/_database_manager.html#DatabaseManager.get_db"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#idtrack._database_manager.DatabaseManager.get_db" title="Link to this definition"></a></dt>
<dd><p>Retrieve or create a cached data table defined by an indicator string.</p>
<p>This method is the central gateway for <em>all</em> tabular resources managed by
<a class="reference internal" href="#idtrack._database_manager.DatabaseManager" title="idtrack._database_manager.DatabaseManager"><code class="xref py py-class docutils literal notranslate"><span class="pre">DatabaseManager</span></code></a>.  It interprets a compact <em>indicator</em> string,
decides whether the requested table already exists in the local HDF5
repository, and either loads the cached copy or triggers the appropriate
builder (<code class="docutils literal notranslate"><span class="pre">create_*</span></code> helper) to download/assemble it.  A consistent naming
convention is maintained so that subsequent calls with the same indicator
transparently reuse the on-disk cache, ensuring reproducible builds and
minimal network traffic.</p>
<p><strong>Supported base indicators</strong></p>
<ul class="simple">
<li><dl class="simple">
<dt><code class="docutils literal notranslate"><span class="pre">external</span></code> — cross-reference database registry; optional qualifier</dt><dd><p><code class="docutils literal notranslate"><span class="pre">relevant</span></code> | <code class="docutils literal notranslate"><span class="pre">database</span></code> | <code class="docutils literal notranslate"><span class="pre">relevant-database</span></code> narrows the view.</p>
</dd>
</dl>
</li>
<li><p><code class="docutils literal notranslate"><span class="pre">idsraw</span></code> — raw Ensembl identifiers for a given form <em>(``gene``,
``transcript``, ``translation``)</em>; requires the form as qualifier.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">ids</span></code> — release-specific identifier table (no qualifier).</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">externalcontent</span></code> — summary of per-database content.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">relationcurrent</span></code> — current gene/ID relationships.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">relationarchive</span></code> — historical gene/ID relationships across releases.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">idhistory</span></code> — full ID history; qualifier <code class="docutils literal notranslate"><span class="pre">narrow</span></code> restricts to current IDs.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">versioninfo</span></code> — version comparison across releases.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">availabledatabases</span></code> — list of locally cacheable resources.</p></li>
</ul>
<p>Additional indicators may be introduced by subclass extensions; consult the
module documentation for the authoritative list.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>df_indicator</strong> (<em>str</em>) – Compact descriptor of the table to retrieve.  Must
follow the <code class="docutils literal notranslate"><span class="pre">base[qualifier]</span></code> pattern described above.</p></li>
<li><p><strong>create_even_if_exist</strong> (<em>bool</em>) – Force a rebuild/download even if
a cached copy is present.  Defaults to <code class="docutils literal notranslate"><span class="pre">False</span></code>.</p></li>
<li><p><strong>save_after_calculation</strong> (<em>bool</em>) – Persist a newly created table
to the local HDF5 store.  Has no effect when the table is merely
loaded from disk.  Defaults to <code class="docutils literal notranslate"><span class="pre">True</span></code>.</p></li>
<li><p><strong>overwrite_even_if_exist</strong> (<em>bool</em>) – When saving, replace an
existing HDF5 key with the same hierarchy (file-internal path).
Defaults to <code class="docutils literal notranslate"><span class="pre">False</span></code>.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>The requested dataset.  The exact
shape, index, and column layout depend on <code class="docutils literal notranslate"><span class="pre">df_indicator</span></code>; see the
indicator list above for semantic details.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>pandas.DataFrame | pandas.Series</p>
</dd>
<dt class="field-even">Raises<span class="colon">:</span></dt>
<dd class="field-even"><p><strong>ValueError</strong> – If <em>df_indicator</em> is malformed, references an unsupported
    resource, or its qualifier violates the expected pattern (e.g.,
    missing form for <code class="docutils literal notranslate"><span class="pre">idsraw</span></code>).</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="idtrack._database_manager.DatabaseManager.get_release_date">
<span class="sig-name descname"><span class="pre">get_release_date</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/idtrack/_database_manager.html#DatabaseManager.get_release_date"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#idtrack._database_manager.DatabaseManager.get_release_date" title="Link to this definition"></a></dt>
<dd><p>Return a mapping of Ensembl release numbers to their publication dates.</p>
<p>The future implementation will query the <code class="docutils literal notranslate"><span class="pre">meta</span></code> table of each reachable
release—or fall back to the Ensembl REST API—to build a dictionary such as
<code class="docutils literal notranslate"><span class="pre">{105:</span> <span class="pre">date(2022,</span> <span class="pre">11,</span> <span class="pre">1),</span> <span class="pre">106:</span> <span class="pre">date(2023,</span> <span class="pre">2,</span> <span class="pre">7),</span> <span class="pre">…}</span></code>.  Down-stream
routines can then translate between absolute dates and release numbers,
enabling chronology-aware analyses and reporting.</p>
<dl class="field-list simple">
<dt class="field-odd">Raises<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>NotImplementedError</strong> – Always - date discovery is not yet implemented.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="idtrack._database_manager.DatabaseManager.get_table">
<span class="sig-name descname"><span class="pre">get_table</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">table_key</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">usecols</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">create_even_if_exist</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">save_after_calculation</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">overwrite_even_if_exist</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/idtrack/_database_manager.html#DatabaseManager.get_table"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#idtrack._database_manager.DatabaseManager.get_table" title="Link to this definition"></a></dt>
<dd><p>Download, cache, or read a raw MySQL table for the current release.</p>
<p>A high-level wrapper that coordinates three steps:</p>
<ol class="arabic simple">
<li><p><strong>Path resolution</strong> - determines the HDF5 file and internal key under the local repository that
belong to <em>table_key</em> (and <em>usecols</em>, if provided).</p></li>
<li><p><strong>Fetch or reuse</strong> - if the target key is absent, unreadable, or forcibly refreshed, delegates to
<a class="reference internal" href="#idtrack._database_manager.DatabaseManager.download_table" title="idtrack._database_manager.DatabaseManager.download_table"><code class="xref py py-meth docutils literal notranslate"><span class="pre">download_table()</span></code></a> to query the MySQL server; otherwise loads the dataframe from disk.</p></li>
<li><p><strong>Persistence</strong> - optionally stores the freshly downloaded dataframe back to disk, shrinking the
number of future network calls.</p></li>
</ol>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>table_key</strong> (<em>str</em>) – Name of the MySQL table (e.g. <code class="docutils literal notranslate"><span class="pre">&quot;gene&quot;</span></code>, <code class="docutils literal notranslate"><span class="pre">&quot;xref&quot;</span></code>, <code class="docutils literal notranslate"><span class="pre">&quot;mapping_session&quot;</span></code>).</p></li>
<li><p><strong>usecols</strong> (<em>list</em><em>[</em><em>str</em><em>] </em><em>| </em><em>None</em>) – Column subset to retrieve. <code class="docutils literal notranslate"><span class="pre">None</span></code> (default) selects <em>all</em> columns.</p></li>
<li><p><strong>create_even_if_exist</strong> (<em>bool</em>) – Ignore any on-disk cache and re-download the table unconditionally.</p></li>
<li><p><strong>save_after_calculation</strong> (<em>bool</em>) – Persist the dataframe to the computed HDF5 path when <code class="docutils literal notranslate"><span class="pre">True</span></code>.</p></li>
<li><p><strong>overwrite_even_if_exist</strong> (<em>bool</em>) – Replace an existing HDF5 key even when it is already present.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><dl class="simple">
<dt>The requested raw table with column order mirroring <em>usecols</em> when supplied,</dt><dd><p>otherwise the server’s natural order.</p>
</dd>
</dl>
</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>pandas.DataFrame</p>
</dd>
<dt class="field-even">Raises<span class="colon">:</span></dt>
<dd class="field-even"><p><strong>ValueError</strong> – If <em>usecols</em> is an empty list, not a list, or otherwise fails basic validation.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="idtrack._database_manager.DatabaseManager.id_ver_from_df">
<span class="sig-name descname"><span class="pre">id_ver_from_df</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">dbm_the_ids</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/idtrack/_database_manager.html#DatabaseManager.id_ver_from_df"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#idtrack._database_manager.DatabaseManager.id_ver_from_df" title="Link to this definition"></a></dt>
<dd><p>Assemble fully qualified node names from a <em>stable-ID / version</em> DataFrame.</p>
<p>This convenience routine converts a two-column frame—usually produced by
<a class="reference internal" href="#idtrack._database_manager.DatabaseManager.get_db" title="idtrack._database_manager.DatabaseManager.get_db"><code class="xref py py-meth docutils literal notranslate"><span class="pre">DatabaseManager.get_db()</span></code></a> with the <em>ids</em> form—into the canonical node
labels used throughout ID-track graphs (e.g. <code class="docutils literal notranslate"><span class="pre">ENSG00000000001.1</span></code>).  It first
validates that the input columns match <code class="xref py py-data docutils literal notranslate"><span class="pre">self._identifiers</span></code>
(typically <code class="docutils literal notranslate"><span class="pre">[&quot;gene_stable_id&quot;,</span> <span class="pre">&quot;gene_version&quot;]</span></code> or analogous for the
current <code class="docutils literal notranslate"><span class="pre">form</span></code>), then delegates per-row processing to
<a class="reference internal" href="#idtrack._database_manager.DatabaseManager.node_dict_maker" title="idtrack._database_manager.DatabaseManager.node_dict_maker"><code class="xref py py-meth docutils literal notranslate"><span class="pre">DatabaseManager.node_dict_maker()</span></code></a> and
<a class="reference internal" href="#idtrack._database_manager.DatabaseManager.node_name_maker" title="idtrack._database_manager.DatabaseManager.node_name_maker"><code class="xref py py-meth docutils literal notranslate"><span class="pre">DatabaseManager.node_name_maker()</span></code></a>.  The resulting list may be fed
directly into downstream graph builders or written to disk for later reuse.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>dbm_the_ids</strong> (<em>pandas.DataFrame</em>) – Two-column frame containing the stable
identifiers and their Ensembl version numbers.  The column order and
names <strong>must</strong> exactly match <code class="docutils literal notranslate"><span class="pre">self._identifiers</span></code>; otherwise an
exception is raised.</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><dl class="simple">
<dt>Ordered list where each element is either</dt><dd><p><code class="docutils literal notranslate"><span class="pre">&quot;&lt;ID&gt;.&lt;version&gt;&quot;</span></code> when a valid numeric version is present or simply
<code class="docutils literal notranslate"><span class="pre">&quot;&lt;ID&gt;&quot;</span></code> when the version is <em>None</em> / <em>NaN</em> / an alternative marker
(see <a class="reference internal" href="#idtrack._db.DB.alternative_versions" title="idtrack._db.DB.alternative_versions"><code class="xref py py-data docutils literal notranslate"><span class="pre">idtrack._db.DB.alternative_versions</span></code></a>).</p>
</dd>
</dl>
</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>list[str]</p>
</dd>
<dt class="field-even">Raises<span class="colon">:</span></dt>
<dd class="field-even"><p><strong>ValueError</strong> – If <code class="docutils literal notranslate"><span class="pre">dbm_the_ids</span></code> does not contain the expected column
    names stored in <code class="xref py py-data docutils literal notranslate"><span class="pre">self._identifiers</span></code>.</p>
</dd>
</dl>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="idtrack._database_manager.DatabaseManager.mysql_database">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">mysql_database</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="pre">str</span></em><a class="headerlink" href="#idtrack._database_manager.DatabaseManager.mysql_database" title="Link to this definition"></a></dt>
<dd><p>Return the canonical Ensembl Core schema name for the current organism, release, and assembly.</p>
<p>The manager must resolve exactly one MySQL schema on the remote Ensembl server whose name encodes the
<em>organism</em> (e.g. <code class="docutils literal notranslate"><span class="pre">homo_sapiens</span></code>), <em>release</em> (e.g. <code class="docutils literal notranslate"><span class="pre">111</span></code>), and <em>genome assembly</em> (e.g. <code class="docutils literal notranslate"><span class="pre">38</span></code>).  This
helper centralises that lookup, guaranteeing that downstream calls—such as <a class="reference internal" href="#idtrack._database_manager.DatabaseManager.download_table" title="idtrack._database_manager.DatabaseManager.download_table"><code class="xref py py-meth docutils literal notranslate"><span class="pre">download_table()</span></code></a>—always
talk to the correct database.  The search relies on <a class="reference internal" href="#idtrack._database_manager.DatabaseManager.get_db" title="idtrack._database_manager.DatabaseManager.get_db"><code class="xref py py-meth docutils literal notranslate"><span class="pre">get_db()</span></code></a> to fetch the server’s catalogue and
then filters it by a strict regular expression; any ambiguity is treated as fatal because it would break
reproducibility.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p><dl class="simple">
<dt>A single schema name like <code class="docutils literal notranslate"><span class="pre">&quot;homo_sapiens_core_111_38&quot;</span></code> that uniquely matches the manager’s</dt><dd><p>configuration.</p>
</dd>
</dl>
</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p>str</p>
</dd>
<dt class="field-odd">Raises<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>ValueError</strong> – If zero <strong>or</strong> more than one schema satisfies the search criteria, signalling server
    misconfiguration or an invalid combination of <em>organism</em> and <em>ensembl_release</em>.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="idtrack._database_manager.DatabaseManager.node_dict_maker">
<em class="property"><span class="pre">static</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">node_dict_maker</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">id_entry</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">version_entry</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/idtrack/_database_manager.html#DatabaseManager.node_dict_maker"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#idtrack._database_manager.DatabaseManager.node_dict_maker" title="Link to this definition"></a></dt>
<dd><p>Return a normalized <em>ID/Version</em> dictionary from raw column values.</p>
<p>This helper creates the canonical structure consumed by
<a class="reference internal" href="#idtrack._database_manager.DatabaseManager.node_name_maker" title="idtrack._database_manager.DatabaseManager.node_name_maker"><code class="xref py py-meth docutils literal notranslate"><span class="pre">DatabaseManager.node_name_maker()</span></code></a> and higher-level graph utilities,
ensuring that version numbers are strictly integers whenever possible.  It
also recognises special placeholders defined in
<a class="reference internal" href="#idtrack._db.DB.alternative_versions" title="idtrack._db.DB.alternative_versions"><code class="xref py py-data docutils literal notranslate"><span class="pre">idtrack._db.DB.alternative_versions</span></code></a> (e.g. <code class="docutils literal notranslate"><span class="pre">&quot;Retired&quot;</span></code> or
<code class="docutils literal notranslate"><span class="pre">&quot;Void&quot;</span></code>) and passes them through unchanged so that downstream code can
handle deprecated or missing entries appropriately.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>id_entry</strong> (<em>str</em>) – Stable identifier portion preceding the delimiter
(e.g. <code class="docutils literal notranslate"><span class="pre">&quot;ENSG00000000001&quot;</span></code>).</p></li>
<li><p><strong>version_entry</strong> (<em>Any</em>) – Raw version value following the delimiter
(e.g. <code class="docutils literal notranslate"><span class="pre">1</span></code> in <code class="docutils literal notranslate"><span class="pre">&quot;ENSG00000000001.1&quot;</span></code>).  May be <em>float</em>, <em>int</em>,
<em>str</em>, <em>None</em>, <em>NaN</em>, or an alternative placeholder such as
<code class="docutils literal notranslate"><span class="pre">&quot;Retired&quot;</span></code>.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><code class="docutils literal notranslate"><span class="pre">{&quot;ID&quot;:</span> <span class="pre">id_entry,</span> <span class="pre">&quot;Version&quot;:</span> <span class="pre">version_entry}</span></code> with
<em>Version</em> coerced to <em>int</em> when it represents a whole number.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>dict[str, Any]</p>
</dd>
<dt class="field-even">Raises<span class="colon">:</span></dt>
<dd class="field-even"><p><strong>ValueError</strong> – If <code class="docutils literal notranslate"><span class="pre">version_entry</span></code> is numeric but contains a fractional
    component (e.g. <code class="docutils literal notranslate"><span class="pre">1.2</span></code>), indicating a malformed identifier that cannot
    be represented as an integer version.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="idtrack._database_manager.DatabaseManager.node_name_maker">
<em class="property"><span class="pre">static</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">node_name_maker</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">node_dict</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/idtrack/_database_manager.html#DatabaseManager.node_name_maker"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#idtrack._database_manager.DatabaseManager.node_name_maker" title="Link to this definition"></a></dt>
<dd><p>Concatenate <em>ID</em> and <em>Version</em> into a single node label.</p>
<p>Given the miniature dictionary returned by
<a class="reference internal" href="#idtrack._database_manager.DatabaseManager.node_dict_maker" title="idtrack._database_manager.DatabaseManager.node_dict_maker"><code class="xref py py-meth docutils literal notranslate"><span class="pre">DatabaseManager.node_dict_maker()</span></code></a>, this helper builds the string
representation that uniquely identifies a biological entity within the graph
layer.  When a numeric version is available, it appends that value to the
stable ID using <a class="reference internal" href="#idtrack._db.DB.id_ver_delimiter" title="idtrack._db.DB.id_ver_delimiter"><code class="xref py py-data docutils literal notranslate"><span class="pre">idtrack._db.DB.id_ver_delimiter</span></code></a> (<code class="docutils literal notranslate"><span class="pre">&quot;.&quot;</span></code> by
default).  For organisms or datasets lacking versioned identifiers, it falls
back to the bare stable ID to preserve compatibility.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>node_dict</strong> (<em>dict</em><em>[</em><em>str</em><em>, </em><em>Any</em><em>]</em>) – Mapping with exactly two keys, <code class="docutils literal notranslate"><span class="pre">&quot;ID&quot;</span></code> and
<code class="docutils literal notranslate"><span class="pre">&quot;Version&quot;</span></code>, as produced by
<a class="reference internal" href="#idtrack._database_manager.DatabaseManager.node_dict_maker" title="idtrack._database_manager.DatabaseManager.node_dict_maker"><code class="xref py py-meth docutils literal notranslate"><span class="pre">DatabaseManager.node_dict_maker()</span></code></a>.</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><dl class="simple">
<dt>Either <code class="docutils literal notranslate"><span class="pre">&quot;&lt;ID&gt;.&lt;version&gt;&quot;</span></code> or <code class="docutils literal notranslate"><span class="pre">&quot;&lt;ID&gt;&quot;</span></code> depending on whether a</dt><dd><p>non-null, non-alternative version is present.</p>
</dd>
</dl>
</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>str</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="idtrack._database_manager.DatabaseManager.tables_in_disk">
<span class="sig-name descname"><span class="pre">tables_in_disk</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/idtrack/_database_manager.html#DatabaseManager.tables_in_disk"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#idtrack._database_manager.DatabaseManager.tables_in_disk" title="Link to this definition"></a></dt>
<dd><p>List all dataframes cached for this manager on local disk.</p>
<p>The helper inspects the HDF5 file located at the path generated by
<a class="reference internal" href="#idtrack._database_manager.DatabaseManager.file_name" title="idtrack._database_manager.DatabaseManager.file_name"><code class="xref py py-meth docutils literal notranslate"><span class="pre">file_name()</span></code></a> (<code class="docutils literal notranslate"><span class="pre">df_type=&quot;common&quot;</span></code>) and returns every key it contains.
When the file does not exist yet, an empty list is returned instead of raising.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>Sorted HDF5 keys corresponding to dataframes already materialised for this manager.</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p>list[str]</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="idtrack._database_manager.DatabaseManager.version_fix">
<span class="sig-name descname"><span class="pre">version_fix</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">df</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">version_str</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">version_info</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/idtrack/_database_manager.html#DatabaseManager.version_fix"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#idtrack._database_manager.DatabaseManager.version_fix" title="Link to this definition"></a></dt>
<dd><p>Apply a <em>global</em> ID-version policy to a DataFrame.</p>
<p>Depending on the organism and its historical annotation quirks, identifiers may (1) <strong>never</strong>
include a version, (2) <strong>always</strong> include a version, or (3) require a <em>synthetic</em> version when
mixing cross-release data.  The <em>version_info</em> flag encodes that policy:</p>
<ul class="simple">
<li><p><code class="docutils literal notranslate"><span class="pre">&quot;without_version&quot;</span></code> — strip all versions (set column to <code class="docutils literal notranslate"><span class="pre">NaN</span></code>).</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">&quot;with_version&quot;</span></code>    — cast column to <code class="docutils literal notranslate"><span class="pre">int64</span></code> (all values must exist).</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">&quot;add_version&quot;</span></code>     — fill missing entries with <code class="xref py py-data docutils literal notranslate"><span class="pre">DB.first_version</span></code>.</p></li>
</ul>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>df</strong> (<em>pandas.DataFrame</em>) – Frame whose <em>version_str</em> column needs harmonising.</p></li>
<li><p><strong>version_str</strong> (<em>str</em>) – Name of the column that stores version numbers.</p></li>
<li><p><strong>version_info</strong> (<em>Optional</em><em>[</em><em>str</em><em>]</em>) – One of <code class="docutils literal notranslate"><span class="pre">&quot;add_version&quot;</span></code>, <code class="docutils literal notranslate"><span class="pre">&quot;without_version&quot;</span></code>, or
<code class="docutils literal notranslate"><span class="pre">&quot;with_version&quot;</span></code>.  When <code class="docutils literal notranslate"><span class="pre">None</span></code> (default) the method calls
<a class="reference internal" href="#idtrack._database_manager.DatabaseManager.check_version_info" title="idtrack._database_manager.DatabaseManager.check_version_info"><code class="xref py py-meth docutils literal notranslate"><span class="pre">check_version_info()</span></code></a> to determine the correct policy automatically.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>Same object <em>df</em> with <em>version_str</em> updated in-place.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>pandas.DataFrame</p>
</dd>
<dt class="field-even">Raises<span class="colon">:</span></dt>
<dd class="field-even"><p><strong>ValueError</strong> – If <em>version_info</em> is not recognised.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="idtrack._database_manager.DatabaseManager.version_fix_incomplete">
<span class="sig-name descname"><span class="pre">version_fix_incomplete</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">df_fx</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">id_col_fx</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ver_col_fx</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/idtrack/_database_manager.html#DatabaseManager.version_fix_incomplete"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#idtrack._database_manager.DatabaseManager.version_fix_incomplete" title="Link to this definition"></a></dt>
<dd><p>Clean up version columns when <em>some</em> identifiers are entirely missing.</p>
<p>Ensembl <em>translation</em> tables occasionally encode parent IDs without a version while descendants
retain one, producing frames where <em>id_col_fx</em> is <code class="docutils literal notranslate"><span class="pre">NaN</span></code> but <em>ver_col_fx</em> contains a number.
This helper splits the frame, delegates to <a class="reference internal" href="#idtrack._database_manager.DatabaseManager.version_fix" title="idtrack._database_manager.DatabaseManager.version_fix"><code class="xref py py-meth docutils literal notranslate"><span class="pre">version_fix()</span></code></a> for each subset, then stitches
the pieces back together so that every row obeys a single “with/without/add version” policy.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>df_fx</strong> (<em>pandas.DataFrame</em>) – Data to harmonise.  The frame <strong>must</strong> include <em>id_col_fx</em> and <em>ver_col_fx</em>.</p></li>
<li><p><strong>id_col_fx</strong> (<em>str</em>) – Column holding the <em>stable</em> part of the identifier (e.g. <code class="docutils literal notranslate"><span class="pre">&quot;translation_id&quot;</span></code>).</p></li>
<li><p><strong>ver_col_fx</strong> (<em>str</em>) – Column holding the integer version suffix.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><dl class="simple">
<dt>Frame whose <em>ver_col_fx</em> is consistent with the organism-level policy</dt><dd><p>determined by <a class="reference internal" href="#idtrack._database_manager.DatabaseManager.check_version_info" title="idtrack._database_manager.DatabaseManager.check_version_info"><code class="xref py py-meth docutils literal notranslate"><span class="pre">check_version_info()</span></code></a>.</p>
</dd>
</dl>
</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>pandas.DataFrame</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="idtrack._database_manager.DatabaseManager.version_uniformize">
<span class="sig-name descname"><span class="pre">version_uniformize</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">df</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">version_str</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/idtrack/_database_manager.html#DatabaseManager.version_uniformize"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#idtrack._database_manager.DatabaseManager.version_uniformize" title="Link to this definition"></a></dt>
<dd><p>Normalise a <em>Version</em> column so every entry is either an <code class="docutils literal notranslate"><span class="pre">int</span></code> or <code class="docutils literal notranslate"><span class="pre">NaN</span></code>.</p>
<p>This post-processing helper finalises the output of <a class="reference internal" href="#idtrack._database_manager.DatabaseManager.create_ids" title="idtrack._database_manager.DatabaseManager.create_ids"><code class="xref py py-meth docutils literal notranslate"><span class="pre">DatabaseManager.create_ids()</span></code></a>.
Ensembl releases differ: some assign an explicit integer version to every stable identifier,
whereas others omit the suffix entirely.  Downstream code expects a <em>uniform</em> dtype, so this
routine coerces the designated column to a proper integer when <em>all</em> entries are present or
fills the entire column with <code class="docutils literal notranslate"><span class="pre">np.nan</span></code> when <em>none</em> are.  Mixed presence is forbidden because
it would break the ID-version pairing logic used by <a class="reference internal" href="#idtrack._database_manager.DatabaseManager.node_name_maker" title="idtrack._database_manager.DatabaseManager.node_name_maker"><code class="xref py py-meth docutils literal notranslate"><span class="pre">DatabaseManager.node_name_maker()</span></code></a>.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>df</strong> (<em>pandas.DataFrame</em>) – Frame returned by <a class="reference internal" href="#idtrack._database_manager.DatabaseManager.create_ids" title="idtrack._database_manager.DatabaseManager.create_ids"><code class="xref py py-meth docutils literal notranslate"><span class="pre">create_ids()</span></code></a>; must already contain a
column named <em>version_str</em>.</p></li>
<li><p><strong>version_str</strong> (<em>str</em>) – Name of the column that holds version information (e.g. <code class="docutils literal notranslate"><span class="pre">&quot;gene_version&quot;</span></code>).</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><dl class="simple">
<dt>Same object <em>df</em> with <em>version_str</em> either cast to <code class="docutils literal notranslate"><span class="pre">int64</span></code> or overwritten</dt><dd><p>with <code class="docutils literal notranslate"><span class="pre">np.nan</span></code> for every row.</p>
</dd>
</dl>
</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>pandas.DataFrame</p>
</dd>
<dt class="field-even">Raises<span class="colon">:</span></dt>
<dd class="field-even"><p><strong>NotImplementedError</strong> – If some rows have a version and others do not, indicating an Ensembl
    release with inconsistent schema.  Such a release is currently unsupported.</p>
</dd>
</dl>
</dd></dl>

</section>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>organism</strong> (<em>str</em>)</p></li>
<li><p><strong>form</strong> (<em>str</em>)</p></li>
<li><p><strong>local_repository</strong> (<em>str</em>)</p></li>
<li><p><strong>ensembl_release</strong> (<em>int</em><em> | </em><em>None</em>)</p></li>
<li><p><strong>ignore_before</strong> (<em>int</em><em> | </em><em>None</em>)</p></li>
<li><p><strong>ignore_after</strong> (<em>int</em><em> | </em><em>float</em><em> | </em><em>None</em>)</p></li>
<li><p><strong>store_raw_always</strong> (<em>bool</em>)</p></li>
<li><p><strong>genome_assembly</strong> (<em>int</em><em> | </em><em>None</em>)</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py class" id="module-idtrack._external_databases">
<dt class="sig sig-object py" id="idtrack._external_databases.ExternalDatabases">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">ExternalDatabases</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">organism</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ensembl_release</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">form</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">local_repository</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">genome_assembly</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/idtrack/_external_databases.html#ExternalDatabases"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#idtrack._external_databases.ExternalDatabases" title="Link to this definition"></a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></p>
<p>Manage third-party metadata for Ensembl entities through YAML side-car files.</p>
<p>This helper encapsulates everything related to the <em>external</em> (i.e. non-Ensembl) databases that
can be linked to a given organism, genome assembly, release, and biological form
(<em>gene</em> / <em>transcript</em> / <em>translation</em>).  Examples of such resources include ArrayExpress,
RefSeq, Uniprot, HGNC, and dozens of smaller annotation providers.  Rather than hard-coding
those relationships, the wider <em>ID-Track</em> toolkit stores them in a human-readable YAML file that
lives next to the local data cache managed by <code class="xref py py-class docutils literal notranslate"><span class="pre">_database_manager.DatabaseManager</span></code>.</p>
<p>The YAML workflow is:</p>
<ol class="arabic simple">
<li><p><a class="reference internal" href="#idtrack._external_databases.ExternalDatabases.create_template_yaml" title="idtrack._external_databases.ExternalDatabases.create_template_yaml"><code class="xref py py-meth docutils literal notranslate"><span class="pre">create_template_yaml()</span></code></a> enumerates every known combination and writes a <em>template</em>
where each entry is marked <code class="docutils literal notranslate"><span class="pre">Include:</span> <span class="pre">false</span></code>.</p></li>
<li><p>A user (or an automated post-processing step) reviews the template, toggling <code class="docutils literal notranslate"><span class="pre">Include</span></code> to
<code class="docutils literal notranslate"><span class="pre">true</span></code> for the resources they need.</p></li>
<li><p>The modified YAML is saved under <em>local_repository</em>; subsequent calls to
<a class="reference internal" href="#idtrack._external_databases.ExternalDatabases.load_modified_yaml" title="idtrack._external_databases.ExternalDatabases.load_modified_yaml"><code class="xref py py-meth docutils literal notranslate"><span class="pre">load_modified_yaml()</span></code></a> return it as a plain <code class="xref py py-class docutils literal notranslate"><span class="pre">dict</span></code> for downstream logic.</p></li>
<li><p><a class="reference internal" href="#idtrack._external_databases.ExternalDatabases.validate_yaml_file_up_to_date" title="idtrack._external_databases.ExternalDatabases.validate_yaml_file_up_to_date"><code class="xref py py-meth docutils literal notranslate"><span class="pre">validate_yaml_file_up_to_date()</span></code></a> warns if the user file lags behind a newer template
(e.g. because a later Ensembl release introduced extra tables).</p></li>
<li><p>Utility helpers such as <a class="reference internal" href="#idtrack._external_databases.ExternalDatabases.give_list_for_case" title="idtrack._external_databases.ExternalDatabases.give_list_for_case"><code class="xref py py-meth docutils literal notranslate"><span class="pre">give_list_for_case()</span></code></a> expose convenient filtered views—e.g.
<em>all databases that should be downloaded for the current form</em>, or <em>all releases supported by
assembly 38</em>.</p></li>
</ol>
<p>In short, <em>ExternalDatabases</em> provides a single, version-controlled “contract” describing which
third-party tables belong in an ID-track run, while granting users explicit opt-in control over
optional resources.</p>
<p>Instantiate a YAML controller tied to a specific organism, release, and assembly.</p>
<p>The constructor mirrors the core configuration of <code class="xref py py-class docutils literal notranslate"><span class="pre">_database_manager.DatabaseManager</span></code>
so that both objects operate on the exact same coordinate system.  No I/O is performed at
construction time; paths are merely recorded, and loggers are configured.  Heavy-weight
actions—such as scanning the cache for existing YAMLs or writing new ones—happen lazily when
the corresponding methods are called.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>organism</strong> (<em>str</em>) – Canonical Ensembl species identifier in <em>snake_case</em>
(e.g. <code class="docutils literal notranslate"><span class="pre">&quot;homo_sapiens&quot;</span></code>).  Case-insensitive but must match Ensembl conventions.</p></li>
<li><p><strong>ensembl_release</strong> (<em>int</em>) – Target Ensembl release number (e.g. <code class="docutils literal notranslate"><span class="pre">110</span></code>).  Must be ≥ 79 and
correspond to a release that actually exists for <em>organism</em> and <em>assembly</em>.</p></li>
<li><p><strong>form</strong> (<em>str</em>) – Entity level—<code class="docutils literal notranslate"><span class="pre">&quot;gene&quot;</span></code>, <code class="docutils literal notranslate"><span class="pre">&quot;transcript&quot;</span></code>, or <code class="docutils literal notranslate"><span class="pre">&quot;translation&quot;</span></code>.  Any other
value raises <code class="xref py py-class docutils literal notranslate"><span class="pre">ValueError</span></code> in higher-level validation.</p></li>
<li><p><strong>local_repository</strong> (<em>str</em>) – Writable directory where YAML files and downloads are cached.
The directory need not pre-exist; if missing, most public methods will attempt to create it.</p></li>
<li><p><strong>genome_assembly</strong> (<em>int</em>) – NCBI genome assembly version (e.g. <code class="docutils literal notranslate"><span class="pre">38</span></code> for GRCh38).  Used to
disambiguate multiple assemblies available for the same <em>organism</em>/<em>release</em> pair.</p></li>
</ul>
</dd>
</dl>
<dl class="py method">
<dt class="sig sig-object py" id="idtrack._external_databases.ExternalDatabases.create_template_yaml">
<span class="sig-name descname"><span class="pre">create_template_yaml</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">df</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/idtrack/_external_databases.html#ExternalDatabases.create_template_yaml"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#idtrack._external_databases.ExternalDatabases.create_template_yaml" title="Link to this definition"></a></dt>
<dd><p>Generate a template YAML enumerating external-database options.</p>
<p>This helper scans <em>df</em>—typically the dataframe returned by
<a class="reference internal" href="#idtrack._database_manager.DatabaseManager.create_database_content" title="idtrack._database_manager.DatabaseManager.create_database_content"><code class="xref py py-meth docutils literal notranslate"><span class="pre">idtrack._database_manager.DatabaseManager.create_database_content()</span></code></a>—and
writes a scaffold configuration file to
<a class="reference internal" href="#idtrack._external_databases.ExternalDatabases.file_name_template_yaml" title="idtrack._external_databases.ExternalDatabases.file_name_template_yaml"><code class="xref py py-meth docutils literal notranslate"><span class="pre">file_name_template_yaml()</span></code></a>.
The file lists every <em>organism</em> → <em>form</em> → database combination observed in
<em>df</em>, grouped by genome assembly and Ensembl release.  For each entry the
template records whether the database should be <em>included</em> when building an
ID-history graph, its integer <strong>Database Index</strong>, and an empty
<strong>Potential Synonymous</strong> placeholder that future versions may use to flag
overlapping resources.</p>
<p>Users are <strong>expected to edit</strong> the generated file—changing <code class="docutils literal notranslate"><span class="pre">Include</span></code> from
<code class="docutils literal notranslate"><span class="pre">false</span></code> to <code class="docutils literal notranslate"><span class="pre">true</span></code> where appropriate—and <strong>rename</strong> it by appending
<code class="docutils literal notranslate"><span class="pre">_modified</span></code> to the filename before the package will load it.  A warning to
that effect is emitted via <code class="xref py py-meth docutils literal notranslate"><span class="pre">logging.Logger.warning()</span></code>.</p>
<p>The resulting YAML resembles the structure below (truncated for brevity):</p>
<div class="highlight-yaml notranslate"><div class="highlight"><pre><span></span><span class="nt">homo_sapiens</span><span class="p">:</span>
<span class="w">    </span><span class="nt">gene</span><span class="p">:</span>
<span class="w">        </span><span class="nt">ArrayExpress</span><span class="p">:</span>
<span class="w">            </span><span class="nt">Assembly</span><span class="p">:</span>
<span class="w">                </span><span class="s">&quot;37&quot;</span><span class="p p-Indicator">:</span>
<span class="w">                    </span><span class="nt">Ensembl release</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">79,80,81,82,83,84,85,86,87,88,89</span>
<span class="w">                    </span><span class="nt">Include</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">false</span>
<span class="w">                </span><span class="s">&quot;38&quot;</span><span class="p p-Indicator">:</span>
<span class="w">                    </span><span class="nt">Ensembl release</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">79,80,81,82,83,84,85,86,87,88,89</span>
<span class="w">                    </span><span class="nt">Include</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">false</span>
<span class="w">            </span><span class="nt">Database Index</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">0</span>
<span class="w">            </span><span class="nt">Potential Synonymous</span><span class="p">:</span><span class="w"> </span><span class="s">&quot;&quot;</span>
<span class="w">        </span><span class="nt">Clone-based (Ensembl)</span><span class="p">:</span>
<span class="w">            </span><span class="nt">Assembly</span><span class="p">:</span>
<span class="w">                </span><span class="s">&quot;37&quot;</span><span class="p p-Indicator">:</span>
<span class="w">                    </span><span class="nt">Ensembl release</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">79,80,81,82,83,84,85</span>
<span class="w">                    </span><span class="nt">Include</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">false</span>
<span class="w">                </span><span class="s">&quot;38&quot;</span><span class="p p-Indicator">:</span>
<span class="w">                    </span><span class="nt">Ensembl release</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">79,80,81,82,83,84,85</span>
<span class="w">                    </span><span class="nt">Include</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">false</span>
<span class="w">            </span><span class="nt">Database Index</span><span class="p">:</span><span class="w"> </span><span class="l l-Scalar l-Scalar-Plain">5</span>
<span class="w">            </span><span class="nt">Potential Synonymous</span><span class="p">:</span><span class="w"> </span><span class="s">&quot;&quot;</span>
</pre></div>
</div>
<p><strong>Editing guidelines</strong></p>
<ul class="simple">
<li><p>Set <code class="docutils literal notranslate"><span class="pre">Include</span></code> to <em>true</em> for every assembly of the databases you need.</p></li>
<li><p>Save the edited file with <code class="docutils literal notranslate"><span class="pre">_modified</span></code> appended to the base name so that
downstream routines load the customised version.</p></li>
</ul>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>df</strong> (<em>pandas.DataFrame</em>) – Dataframe containing at least the columns
<code class="docutils literal notranslate"><span class="pre">[&quot;organism&quot;,</span> <span class="pre">&quot;form&quot;,</span> <span class="pre">&quot;name_db&quot;,</span> <span class="pre">&quot;assembly&quot;,</span> <span class="pre">&quot;release&quot;]</span></code>.  It
should be produced by
<a class="reference internal" href="#idtrack._database_manager.DatabaseManager.create_database_content" title="idtrack._database_manager.DatabaseManager.create_database_content"><code class="xref py py-meth docutils literal notranslate"><span class="pre">idtrack._database_manager.DatabaseManager.create_database_content()</span></code></a>
so that the expected schema is guaranteed.</p>
</dd>
<dt class="field-even">Raises<span class="colon">:</span></dt>
<dd class="field-even"><p><strong>ValueError</strong> – If <em>df</em> contains duplicate <em>assembly</em> entries for the same
    organism/form/database triple, causing an internal consistency check
    to fail.</p>
</dd>
</dl>
<p class="rubric">Notes</p>
<p>The <strong>Potential Synonymous</strong> is now all empty. In the following versions, it is aimed to integrate
a feature that prevent to heve synonymous databases in the list. Likewise, <cite>Database Index</cite> has now no
use case, in the program.
It is important to follow the final warning raised by the method. ‘’Please edit the file based on
requested external databases and add ‘_modified’ to the file name.’’. The editing should be done by
converting <cite>Include</cite> sections from <cite>false</cite> to <cite>true</cite>. It is recommended to make the change for each
assembly for a given database.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="idtrack._external_databases.ExternalDatabases.file_name_modified_yaml">
<span class="sig-name descname"><span class="pre">file_name_modified_yaml</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">mode</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/idtrack/_external_databases.html#ExternalDatabases.file_name_modified_yaml"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#idtrack._external_databases.ExternalDatabases.file_name_modified_yaml" title="Link to this definition"></a></dt>
<dd><p>Resolve the path to a <em>modified</em> YAML file customised by the user or shipped with the package.</p>
<p>The method supports two <em>modes</em> that map to different storage locations:</p>
<ul class="simple">
<li><p><code class="docutils literal notranslate"><span class="pre">&quot;configured&quot;</span></code> - the user-edited file living in <code class="xref py py-attr docutils literal notranslate"><span class="pre">ExternalDatabases.local_repository</span></code>.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">&quot;default&quot;</span></code> - the read-only fallback bundled under <code class="docutils literal notranslate"><span class="pre">&lt;package_root&gt;/default_config</span></code> for quick starts
and unit tests.</p></li>
</ul>
<p>By funnelling every lookup through this routine, higher-level helpers such as
<a class="reference internal" href="#idtrack._external_databases.ExternalDatabases.load_modified_yaml" title="idtrack._external_databases.ExternalDatabases.load_modified_yaml"><code class="xref py py-meth docutils literal notranslate"><span class="pre">ExternalDatabases.load_modified_yaml()</span></code></a> remain agnostic about the
underlying directory structure and can focus on validation and parsing
instead.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>mode</strong> (<em>str</em>) – Either <code class="docutils literal notranslate"><span class="pre">&quot;configured&quot;</span></code> or <code class="docutils literal notranslate"><span class="pre">&quot;default&quot;</span></code> selecting the
corresponding search location.</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>Absolute path of the requested YAML file.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>str</p>
</dd>
<dt class="field-even">Raises<span class="colon">:</span></dt>
<dd class="field-even"><p><strong>ValueError</strong> – If <em>mode</em> is not one of the recognised values.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="idtrack._external_databases.ExternalDatabases.file_name_template_yaml">
<span class="sig-name descname"><span class="pre">file_name_template_yaml</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/idtrack/_external_databases.html#ExternalDatabases.file_name_template_yaml"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#idtrack._external_databases.ExternalDatabases.file_name_template_yaml" title="Link to this definition"></a></dt>
<dd><p>Return absolute path to the <em>template</em> YAML configuration file.</p>
<p>A helper that deterministically builds the filename used by
<a class="reference internal" href="#idtrack._external_databases.ExternalDatabases.create_template_yaml" title="idtrack._external_databases.ExternalDatabases.create_template_yaml"><code class="xref py py-meth docutils literal notranslate"><span class="pre">ExternalDatabases.create_template_yaml()</span></code></a> when it first scaffolds the
external-database configuration for <em>organism</em>.  Centralising the logic here
keeps every component of <em>idtrack</em> that may need the path (tests, CLI tools,
future maintenance scripts) in perfect sync with a single implementation.  The
method performs <strong>no</strong> I/O; it merely concatenates
<code class="xref py py-attr docutils literal notranslate"><span class="pre">ExternalDatabases.local_repository</span></code> and the conventional filename
pattern <code class="docutils literal notranslate"><span class="pre">&quot;&lt;organism&gt;_externals_template.yml&quot;</span></code> so callers can decide whether
to create, read, or overwrite the file.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>Absolute path of <code class="docutils literal notranslate"><span class="pre">&lt;organism&gt;_externals_template.yml</span></code> located inside
<code class="xref py py-attr docutils literal notranslate"><span class="pre">ExternalDatabases.local_repository</span></code>.</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p>str</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="idtrack._external_databases.ExternalDatabases.give_list_for_case">
<span class="sig-name descname"><span class="pre">give_list_for_case</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">give_type</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/idtrack/_external_databases.html#ExternalDatabases.give_list_for_case"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#idtrack._external_databases.ExternalDatabases.give_list_for_case" title="Link to this definition"></a></dt>
<dd><p>Return database names or assembly codes extracted from the external-DB YAML file.</p>
<p>The helper provides a lightweight way for higher-level components (e.g.
<a class="reference internal" href="#idtrack._database_manager.DatabaseManager" title="idtrack._database_manager.DatabaseManager"><code class="xref py py-class docutils literal notranslate"><span class="pre">DatabaseManager</span></code></a>) to discover which
external resources—or which genome assemblies—are currently <em>eligible</em>
according to the user-editable YAML configuration created by
<a class="reference internal" href="#idtrack._external_databases.ExternalDatabases.create_template_yaml" title="idtrack._external_databases.ExternalDatabases.create_template_yaml"><code class="xref py py-meth docutils literal notranslate"><span class="pre">ExternalDatabases.create_template_yaml()</span></code></a>.  Instead of forcing the
caller to parse the YAML structure manually, the method filters the entries
for the manager’s <em>organism</em>, <em>form</em>, <em>Ensembl release</em> and
<em>genome assembly</em> and returns the requested slice.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>give_type</strong> (<em>str</em>) – <p>Kind of list to return.  Accepted values are</p>
<ul class="simple">
<li><p><code class="docutils literal notranslate"><span class="pre">&quot;db&quot;</span></code>
external-database <strong>names</strong> (<code class="docutils literal notranslate"><span class="pre">str</span></code>) whose
<code class="docutils literal notranslate"><span class="pre">Include</span></code> flag is <code class="docutils literal notranslate"><span class="pre">true</span></code> for the current organism, form,
assembly and Ensembl release.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">&quot;assembly&quot;</span></code>
genome-assembly <strong>codes</strong> (<code class="docutils literal notranslate"><span class="pre">int</span></code>) that
contain <em>any</em> external database entry for the current organism
and form.  Assemblies are returned even if their <code class="docutils literal notranslate"><span class="pre">Include</span></code>
flags are still <code class="docutils literal notranslate"><span class="pre">false</span></code>.</p></li>
</ul>
</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><ul class="simple">
<li><p>When <em>give_type</em> is <code class="docutils literal notranslate"><span class="pre">&quot;db&quot;</span></code>, a list of database names.</p></li>
<li><p>When <em>give_type</em> is <code class="docutils literal notranslate"><span class="pre">&quot;assembly&quot;</span></code>, a list of assembly codes.</p></li>
</ul>
</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>list[str] | list[int]</p>
</dd>
<dt class="field-even">Raises<span class="colon">:</span></dt>
<dd class="field-even"><p><strong>ValueError</strong> – If <em>give_type</em> is not <code class="docutils literal notranslate"><span class="pre">&quot;db&quot;</span></code> nor <code class="docutils literal notranslate"><span class="pre">&quot;assembly&quot;</span></code> or if an
    unexpected internal inconsistency is encountered while traversing
    the YAML structure.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="idtrack._external_databases.ExternalDatabases.load_modified_yaml">
<span class="sig-name descname"><span class="pre">load_modified_yaml</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/idtrack/_external_databases.html#ExternalDatabases.load_modified_yaml"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#idtrack._external_databases.ExternalDatabases.load_modified_yaml" title="Link to this definition"></a></dt>
<dd><p>Load the user-edited or default YAML configuration and verify release compatibility.</p>
<p>This convenience wrapper searches for the <em>configured</em> YAML file first; if it
does not exist or lacks read permissions a warning is logged and the
<em>default</em> YAML file shipped with the package is tried instead.  Failure to
locate <strong>either</strong> file aborts the process with <code class="xref py py-class docutils literal notranslate"><span class="pre">FileNotFoundError</span></code>.
After loading, the method delegates to
<a class="reference internal" href="#idtrack._external_databases.ExternalDatabases.validate_yaml_file_up_to_date" title="idtrack._external_databases.ExternalDatabases.validate_yaml_file_up_to_date"><code class="xref py py-meth docutils literal notranslate"><span class="pre">ExternalDatabases.validate_yaml_file_up_to_date()</span></code></a> to ensure that the
currently requested Ensembl release is represented in the configuration.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>Parsed YAML content keyed by <code class="docutils literal notranslate"><span class="pre">{organism</span> <span class="pre">→</span> <span class="pre">form</span> <span class="pre">→</span> <span class="pre">database</span> <span class="pre">→</span> <span class="pre">Assembly</span> <span class="pre">→</span> <span class="pre">{...}}</span></code>.</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p>dict</p>
</dd>
<dt class="field-odd">Raises<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>FileNotFoundError</strong> – If neither the <em>configured</em> nor the <em>default</em> YAML file can be accessed.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="idtrack._external_databases.ExternalDatabases.validate_yaml_file_up_to_date">
<span class="sig-name descname"><span class="pre">validate_yaml_file_up_to_date</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">read_yaml_file</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/idtrack/_external_databases.html#ExternalDatabases.validate_yaml_file_up_to_date"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#idtrack._external_databases.ExternalDatabases.validate_yaml_file_up_to_date" title="Link to this definition"></a></dt>
<dd><p>Assert that the YAML configuration lists the active Ensembl release.</p>
<p>The external-database mapping evolves with each Ensembl release.  This helper
extracts the set of releases encoded in <em>read_yaml_file</em>—no matter how deeply
nested—and verifies that <code class="xref py py-attr docutils literal notranslate"><span class="pre">ExternalDatabases.ensembl_release</span></code> is
present.  Triggering an exception here prevents downstream graph-construction
logic from silently operating on incomplete or outdated metadata, prompting
users to regenerate or update the YAML file before proceeding.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>read_yaml_file</strong> (<em>dict</em>) – Dictionary produced by
<a class="reference internal" href="#idtrack._external_databases.ExternalDatabases.load_modified_yaml" title="idtrack._external_databases.ExternalDatabases.load_modified_yaml"><code class="xref py py-meth docutils literal notranslate"><span class="pre">ExternalDatabases.load_modified_yaml()</span></code></a> containing the loaded YAML structure.</p>
</dd>
<dt class="field-even">Raises<span class="colon">:</span></dt>
<dd class="field-even"><p><strong>ValueError</strong> – If the current Ensembl release is absent from the YAML configuration.</p>
</dd>
</dl>
</dd></dl>

</dd></dl>

<dl class="py class" id="module-idtrack._the_graph">
<dt class="sig sig-object py" id="idtrack._the_graph.TheGraph">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">TheGraph</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="o"><span class="pre">*</span></span><span class="n"><span class="pre">args</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/idtrack/_the_graph.html#TheGraph"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#idtrack._the_graph.TheGraph" title="Link to this definition"></a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">MultiDiGraph</span></code></p>
<p>Represent a bio-identifier multigraph with IDTrack-specific helpers.</p>
<p>The class extends <code class="xref py py-class docutils literal notranslate"><span class="pre">networkx.MultiDiGraph</span></code> to model historical and
cross-reference relationships between Ensembl identifiers (genes,
transcripts, translations) and third-party database accessions
(UniProt, RefSeq, …).  It is built by
<a class="reference internal" href="#idtrack._graph_maker.GraphMaker" title="idtrack._graph_maker.GraphMaker"><code class="xref py py-class docutils literal notranslate"><span class="pre">idtrack._graph_maker.GraphMaker</span></code></a>, then queried by
<code class="xref py py-class docutils literal notranslate"><span class="pre">idtrack.Track</span></code> for high-performance path-finding across Ensembl
releases and external resources.</p>
<p>Additional cached properties (e.g. <a class="reference internal" href="#idtrack._the_graph.TheGraph.rev" title="idtrack._the_graph.TheGraph.rev"><code class="xref py py-attr docutils literal notranslate"><span class="pre">rev</span></code></a>,
<a class="reference internal" href="#idtrack._the_graph.TheGraph.combined_edges" title="idtrack._the_graph.TheGraph.combined_edges"><code class="xref py py-attr docutils literal notranslate"><span class="pre">combined_edges</span></code></a>, and <a class="reference internal" href="#idtrack._the_graph.TheGraph.hyperconnective_nodes" title="idtrack._the_graph.TheGraph.hyperconnective_nodes"><code class="xref py py-attr docutils literal notranslate"><span class="pre">hyperconnective_nodes</span></code></a>) collapse
expensive aggregate calculations into single attribute look-ups, while
helpers such as <code class="xref py py-meth docutils literal notranslate"><span class="pre">attach_included_forms()</span></code> record which biological
forms were merged into a particular instance.  Together these conveniences
allow downstream algorithms to traverse millions of edges without the
memory overhead of duplicating graphs or recomputing summaries.</p>
<p>Instantiate the multigraph and configure package logging.</p>
<p>All positional and keyword arguments are forwarded verbatim to
<code class="xref py py-class docutils literal notranslate"><span class="pre">networkx.MultiDiGraph</span></code>, allowing callers to pre-seed the graph
with nodes, edges, or name/metadata attributes exactly as they would with a
vanilla NetworkX constructor.  After delegating to <code class="docutils literal notranslate"><span class="pre">super().__init__</span></code>,
the method initialises two convenience attributes:</p>
<ul class="simple">
<li><dl class="simple">
<dt><code class="xref py py-data docutils literal notranslate"><span class="pre">log</span></code> — a dedicated <code class="docutils literal notranslate"><span class="pre">logging.Logger</span></code> named <code class="docutils literal notranslate"><span class="pre">&quot;the_graph&quot;</span></code> for</dt><dd><p>structured, per-instance diagnostics.</p>
</dd>
</dl>
</li>
<li><dl class="simple">
<dt><code class="xref py py-data docutils literal notranslate"><span class="pre">available_forms</span></code> — a placeholder set to <code class="docutils literal notranslate"><span class="pre">None</span></code> until</dt><dd><p><code class="xref py py-meth docutils literal notranslate"><span class="pre">attach_included_forms()</span></code> is called by <a class="reference internal" href="#idtrack._graph_maker.GraphMaker" title="idtrack._graph_maker.GraphMaker"><code class="xref py py-class docutils literal notranslate"><span class="pre">idtrack._graph_maker.GraphMaker</span></code></a>.</p>
</dd>
</dl>
</li>
</ul>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>args</strong> (<em>Any</em>) – Positional arguments accepted by <code class="xref py py-meth docutils literal notranslate"><span class="pre">networkx.MultiDiGraph.__init__()</span></code>.</p></li>
<li><p><strong>kwargs</strong> (<em>Any</em>) – Keyword arguments accepted by <code class="xref py py-meth docutils literal notranslate"><span class="pre">networkx.MultiDiGraph.__init__()</span></code>.</p></li>
</ul>
</dd>
</dl>
<dl class="py method">
<dt class="sig sig-object py" id="idtrack._the_graph.TheGraph._attach_included_forms">
<span class="sig-name descname"><span class="pre">_attach_included_forms</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">available_forms</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/idtrack/_the_graph.html#TheGraph._attach_included_forms"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#idtrack._the_graph.TheGraph._attach_included_forms" title="Link to this definition"></a></dt>
<dd><p>Record which Ensembl forms are present in the merged graph.</p>
<p>Graphs for <em>gene</em>, <em>transcript</em>, and <em>protein</em> are first built
independently by <a class="reference internal" href="#idtrack._graph_maker.GraphMaker" title="idtrack._graph_maker.GraphMaker"><code class="xref py py-class docutils literal notranslate"><span class="pre">GraphMaker</span></code></a> and then
merged into a single <a class="reference internal" href="#idtrack._the_graph.TheGraph" title="idtrack._the_graph.TheGraph"><code class="xref py py-class docutils literal notranslate"><span class="pre">TheGraph</span></code></a> instance.  This helper runs
<strong>after</strong> that merge to store the subset of forms that actually made it
into the final graph—information required by several cached properties
(e.g. <a class="reference internal" href="#idtrack._the_graph.TheGraph.available_external_databases" title="idtrack._the_graph.TheGraph.available_external_databases"><code class="xref py py-data docutils literal notranslate"><span class="pre">available_external_databases</span></code></a>) for consistency checks
and downstream analyses.  Calling the method <strong>before</strong> the merge would
mis-report available forms and corrupt those caches.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>available_forms</strong> (<em>list</em><em>[</em><em>str</em><em>]</em>) – Exact list of included forms
(typically <code class="docutils literal notranslate"><span class="pre">[&quot;gene&quot;,</span> <span class="pre">&quot;transcript&quot;,</span> <span class="pre">&quot;protein&quot;]</span></code>).  Order is
preserved so callers can rely on a deterministic iteration
sequence.</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p>None</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="idtrack._the_graph.TheGraph._combined_edges">
<em class="property"><span class="pre">static</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">_combined_edges</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">node_list</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">the_graph</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/idtrack/_the_graph.html#TheGraph._combined_edges"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#idtrack._the_graph.TheGraph._combined_edges" title="Link to this definition"></a></dt>
<dd><p>Aggregate database/assembly/release metadata for the edges of <em>node_list</em>.</p>
<p>The routine is the work-horse behind the
<a class="reference internal" href="#idtrack._the_graph.TheGraph.combined_edges" title="idtrack._the_graph.TheGraph.combined_edges"><code class="xref py py-attr docutils literal notranslate"><span class="pre">TheGraph.combined_edges</span></code></a> family of cached properties.
It iterates over every node in <em>node_list</em>, inspects each outgoing
(or, when <em>the_graph</em> is a reversed view, incoming) edge, and builds a
deterministic description of which external database, genome assembly,
and Ensembl release the connection originates from.</p>
<p>Edges that link two nodes of the <strong>same</strong> node-type are ignored so that
backbone history links (gene ↔ gene, transcript ↔ transcript, …) do not
pollute the output (as tested in
<code class="xref py py-meth docutils literal notranslate"><span class="pre">idtrack._track_tests.TrackTest.is_edge_with_same_nts_only_at_backbone_nodes()</span></code>).
For edges whose database key is one of the generic Ensembl forms
(<code class="docutils literal notranslate"><span class="pre">ensembl_gene</span></code>, <code class="docutils literal notranslate"><span class="pre">ensembl_transcript</span></code>, …) the key is rewritten to the
assembly-specific variant (e.g. <code class="docutils literal notranslate"><span class="pre">assembly_38_ensembl_gene</span></code>) to keep
assemblies logically separate in downstream analyses.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>node_list</strong> (<em>NodeView</em><em> | </em><em>list</em><em>[</em><em>str</em><em>]</em>) – Nodes whose edge metadata will be
consolidated.  Accepts either a plain list or the
<code class="xref py py-class docutils literal notranslate"><span class="pre">networkx</span></code> view returned by <code class="docutils literal notranslate"><span class="pre">graph.nodes</span></code>.</p></li>
<li><p><strong>the_graph</strong> (<em>nx.MultiDiGraph</em>) – Graph to inspect.  Pass <code class="docutils literal notranslate"><span class="pre">self</span></code> for the
native orientation or <code class="docutils literal notranslate"><span class="pre">self.rev</span></code> when a reverse walk is
required.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><dl class="simple">
<dt>Mapping  <code class="docutils literal notranslate"><span class="pre">{node:</span> <span class="pre">{database:</span> <span class="pre">{assembly:</span> <span class="pre">set[int]}}}</span></code> that summarises every</dt><dd><p>admissible edge attached to the requested nodes.</p>
</dd>
</dl>
</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>dict</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="idtrack._the_graph.TheGraph._combined_edges_genes_helper">
<em class="property"><span class="pre">static</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">_combined_edges_genes_helper</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">the_result</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/idtrack/_the_graph.html#TheGraph._combined_edges_genes_helper"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#idtrack._the_graph.TheGraph._combined_edges_genes_helper" title="Link to this definition"></a></dt>
<dd><p>Merge per-neighbour edge metadata for gene-centric queries.</p>
<p>This helper is used exclusively by <a class="reference internal" href="#idtrack._the_graph.TheGraph.combined_edges_genes" title="idtrack._the_graph.TheGraph.combined_edges_genes"><code class="xref py py-meth docutils literal notranslate"><span class="pre">TheGraph.combined_edges_genes()</span></code></a> and
<a class="reference internal" href="#idtrack._the_graph.TheGraph.combined_edges_assembly_specific_genes" title="idtrack._the_graph.TheGraph.combined_edges_assembly_specific_genes"><code class="xref py py-meth docutils literal notranslate"><span class="pre">TheGraph.combined_edges_assembly_specific_genes()</span></code></a> to post-process
the dictionaries returned by <a class="reference internal" href="#idtrack._the_graph.TheGraph._combined_edges" title="idtrack._the_graph.TheGraph._combined_edges"><code class="xref py py-meth docutils literal notranslate"><span class="pre">TheGraph._combined_edges()</span></code></a>.
Because backbone <em>gene</em> nodes have no outgoing edges except to other gene
nodes, the caller invokes <a class="reference internal" href="#idtrack._the_graph.TheGraph._combined_edges" title="idtrack._the_graph.TheGraph._combined_edges"><code class="xref py py-meth docutils literal notranslate"><span class="pre">TheGraph._combined_edges()</span></code></a> on a
<strong>reversed</strong> graph and receives one nested dictionary per neighbour.
The present routine</p>
<ol class="arabic simple">
<li><dl class="simple">
<dt>Flattens those per-neighbour sub-dicts so that information from</dt><dd><p>multiple neighbours of the same external database and assembly is unified.</p>
</dd>
</dl>
</li>
<li><dl class="simple">
<dt>Re-labels the generic <code class="docutils literal notranslate"><span class="pre">ensembl_gene</span></code> key to the assembly-qualified</dt><dd><p>form <code class="docutils literal notranslate"><span class="pre">assembly_&lt;N&gt;_ensembl_gene</span></code> so that the provenance of every
entry remains explicit and consistent with the rest of the code base.</p>
</dd>
</dl>
</li>
</ol>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>the_result</strong> (<em>dict</em>) – Nested mapping produced by
<a class="reference internal" href="#idtrack._the_graph.TheGraph._combined_edges" title="idtrack._the_graph.TheGraph._combined_edges"><code class="xref py py-meth docutils literal notranslate"><span class="pre">TheGraph._combined_edges()</span></code></a> for a <em>single</em> gene node.
The structure is  <code class="docutils literal notranslate"><span class="pre">{neighbour:</span> <span class="pre">{database:</span> <span class="pre">{assembly:</span> <span class="pre">set[int]}}}</span></code>.</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><dl class="simple">
<dt>Collapsed mapping  <code class="docutils literal notranslate"><span class="pre">{database:</span> <span class="pre">{assembly:</span> <span class="pre">set[int]}}</span></code> where all neighbour-level</dt><dd><p>dictionaries have been merged and database names have been renamed
to their assembly-specific counterparts when appropriate.</p>
</dd>
</dl>
</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>dict</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="idtrack._the_graph.TheGraph._get_active_ranges_of_id">
<span class="sig-name descname"><span class="pre">_get_active_ranges_of_id</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">input_id</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/idtrack/_the_graph.html#TheGraph._get_active_ranges_of_id"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#idtrack._the_graph.TheGraph._get_active_ranges_of_id" title="Link to this definition"></a></dt>
<dd><p>Calculating the range of active ensembl releases of nodes separately for backbone nodes and others.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>input_id</strong> (<em>str</em>) – Query ID.</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>Ranges of the ID as list of lists. Outputs should always be increasing, inclusive ranges.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>list[list]</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="idtrack._the_graph.TheGraph._node_trios">
<span class="sig-name descname"><span class="pre">_node_trios</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">the_id</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/idtrack/_the_graph.html#TheGraph._node_trios"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#idtrack._the_graph.TheGraph._node_trios" title="Link to this definition"></a></dt>
<dd><p>Calculates the unique tuple called <cite>trios</cite> (database, assembly, Ensembl release) for a given ID.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>the_id</strong> (<em>str</em>) – Query ID.</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>Set of trios.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>set[tuple]</p>
</dd>
</dl>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="idtrack._the_graph.TheGraph.available_external_databases">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">available_external_databases</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="pre">set</span></em><a class="headerlink" href="#idtrack._the_graph.TheGraph.available_external_databases" title="Link to this definition"></a></dt>
<dd><p>Find the available external databases found in the graph.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>Set of all external databases in the graph.</p>
</dd>
</dl>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="idtrack._the_graph.TheGraph.available_external_databases_assembly">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">available_external_databases_assembly</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="pre">dict</span><span class="p"><span class="pre">[</span></span><span class="pre">int</span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="pre">set</span><span class="p"><span class="pre">]</span></span></em><a class="headerlink" href="#idtrack._the_graph.TheGraph.available_external_databases_assembly" title="Link to this definition"></a></dt>
<dd><p>Find the available external databases found in the graph for each assembly separately.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>Dict of all external databases in the graph, assemblies as keys.</p>
</dd>
</dl>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="idtrack._the_graph.TheGraph.available_genome_assemblies">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">available_genome_assemblies</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="pre">set</span><span class="p"><span class="pre">[</span></span><span class="pre">int</span><span class="p"><span class="pre">]</span></span></em><a class="headerlink" href="#idtrack._the_graph.TheGraph.available_genome_assemblies" title="Link to this definition"></a></dt>
<dd><p>Find the genome assemblies found in the graph by iterating through all nodes.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>Genome assemblies, which is also found in <code class="xref py py-attr docutils literal notranslate"><span class="pre">DB.assembly_mysqlport_priority</span></code>.</p>
</dd>
</dl>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="idtrack._the_graph.TheGraph.available_releases_given_database_assembly">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">available_releases_given_database_assembly</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="pre">dict</span><span class="p"><span class="pre">[</span></span><span class="pre">tuple</span><span class="p"><span class="pre">[</span></span><span class="pre">str</span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="pre">int</span><span class="p"><span class="pre">]</span></span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="pre">set</span><span class="p"><span class="pre">]</span></span></em><a class="headerlink" href="#idtrack._the_graph.TheGraph.available_releases_given_database_assembly" title="Link to this definition"></a></dt>
<dd><p>Todo.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="idtrack._the_graph.TheGraph.calculate_caches">
<span class="sig-name descname"><span class="pre">calculate_caches</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">for_test</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/idtrack/_the_graph.html#TheGraph.calculate_caches"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#idtrack._the_graph.TheGraph.calculate_caches" title="Link to this definition"></a></dt>
<dd><p>Eagerly materialise every <code class="docutils literal notranslate"><span class="pre">&#64;cached_property</span></code> to prime the cache.</p>
<p>Accessing a cached property for the first time triggers an expensive
computation. Batch-loading all of them up-front improves latency for
subsequent graph queries and simplifies unit-test expectations because
no additional properties are computed lazily in the background.</p>
<p>The optional <em>for_test</em> flag activates a few heavyweight diagnostics
that are normally skipped in production but useful for test suites and
profiling.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>for_test</strong> (<em>bool</em>) – If <code class="docutils literal notranslate"><span class="pre">True</span></code> (default), also compute
caches that exist solely for testing or sanity-check purposes
(e.g. <a class="reference internal" href="#idtrack._the_graph.TheGraph.external_database_connection_form" title="idtrack._the_graph.TheGraph.external_database_connection_form"><code class="xref py py-data docutils literal notranslate"><span class="pre">external_database_connection_form</span></code></a>).  Set to
<code class="docutils literal notranslate"><span class="pre">False</span></code> to warm only the properties required at run-time.</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p>None</p>
</dd>
</dl>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="idtrack._the_graph.TheGraph.combined_edges">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">combined_edges</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="pre">dict</span></em><a class="headerlink" href="#idtrack._the_graph.TheGraph.combined_edges" title="Link to this definition"></a></dt>
<dd><p>Aggregate outgoing-edge metadata for every non-gene node in the graph.</p>
<p>This cached view pre-computes, for each backbone or external identifier, which external databases,
genome assemblies, and Ensembl releases are reachable through <em>outgoing</em> edges—while purposely
excluding Ensembl gene and assembly-specific gene nodes.  The summary accelerates synonym search
and other traversal routines in <code class="xref py py-meth docutils literal notranslate"><span class="pre">idtrack.track.Track.pathfinder()</span></code> because consumers can
consult a compact dictionary instead of repeatedly iterating raw NetworkX edges and attributes.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p><p>Nested mapping of the form <code class="docutils literal notranslate"><span class="pre">{node_name:</span> <span class="pre">{database_name:</span> <span class="pre">{assembly:</span> <span class="pre">set[int]}}}</span></code>, where</p>
<blockquote>
<div><ul class="simple">
<li><p><strong>node_name</strong> (<em>str</em>) - Identifier of the start node whose edges were inspected.</p></li>
<li><dl class="simple">
<dt><strong>database_name</strong> (<em>str</em>) - Canonical name of the external database or Ensembl sub-type</dt><dd><p>(e.g. <code class="docutils literal notranslate"><span class="pre">uniprot</span></code>, <code class="docutils literal notranslate"><span class="pre">refseq_rna</span></code>, <code class="docutils literal notranslate"><span class="pre">assembly_x_ensembl_gene</span></code>).</p>
</dd>
</dl>
</li>
<li><dl class="simple">
<dt><strong>assembly</strong> (<em>str</em>) - UCSC-style assembly label (e.g. <code class="docutils literal notranslate"><span class="pre">GRCh38</span></code>); <code class="docutils literal notranslate"><span class="pre">None</span></code> when the edge</dt><dd><p>is not assembly-scoped.</p>
</dd>
</dl>
</li>
<li><p><strong>set[int]</strong> - Collection of Ensembl release numbers in which the connection is valid.</p></li>
</ul>
</div></blockquote>
</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p>dict</p>
</dd>
</dl>
<p class="rubric">Notes</p>
<p><em>Edges that link two nodes of the **same*</em> node-type are ignored,* ensuring the dictionary
focuses on cross-type relationships that matter for ID translation.</p>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="idtrack._the_graph.TheGraph.combined_edges_assembly_specific_genes">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">combined_edges_assembly_specific_genes</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="pre">dict</span></em><a class="headerlink" href="#idtrack._the_graph.TheGraph.combined_edges_assembly_specific_genes" title="Link to this definition"></a></dt>
<dd><p>Aggregate incoming‐edge metadata for assembly-specific Ensembl gene nodes.</p>
<p>Assembly-specific gene identifiers (e.g. <code class="docutils literal notranslate"><span class="pre">GRCh37:ENSG00000123456</span></code>) represent loci that differ
between reference builds.  This property mirrors the logic of
<a class="reference internal" href="#idtrack._the_graph.TheGraph.combined_edges_genes" title="idtrack._the_graph.TheGraph.combined_edges_genes"><code class="xref py py-meth docutils literal notranslate"><span class="pre">combined_edges_genes()</span></code></a> but targets nodes <em>not</em> captured by that property,
ensuring the three cached dictionaries are mutually exclusive and collectively exhaustive.
Because each such gene belongs to exactly <strong>one</strong> assembly, the returned structure always contains
a single assembly key per outer node.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p><dl class="simple">
<dt>Mapping <code class="docutils literal notranslate"><span class="pre">{assembly_specific_gene_id:</span> <span class="pre">{database_name:</span> <span class="pre">{assembly:</span> <span class="pre">set[int]}}}</span></code> where the sole</dt><dd><p><em>assembly</em> key matches the assembly implied by the node’s own identifier.</p>
</dd>
</dl>
</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p>dict</p>
</dd>
</dl>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="idtrack._the_graph.TheGraph.combined_edges_genes">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">combined_edges_genes</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="pre">dict</span></em><a class="headerlink" href="#idtrack._the_graph.TheGraph.combined_edges_genes" title="Link to this definition"></a></dt>
<dd><p>Aggregate incoming-edge metadata for Ensembl gene nodes.</p>
<p>Gene nodes only possess <strong>incoming</strong> edges (toward the gene); therefore the calculation traverses
the graph in reverse (<code class="docutils literal notranslate"><span class="pre">self.rev</span></code>) to collect equivalent information to
<a class="reference internal" href="#idtrack._the_graph.TheGraph.combined_edges" title="idtrack._the_graph.TheGraph.combined_edges"><code class="xref py py-meth docutils literal notranslate"><span class="pre">combined_edges()</span></code></a>, but restricted solely to nodes whose
<a class="reference internal" href="#idtrack._db.DB.node_type_str" title="idtrack._db.DB.node_type_str"><code class="xref py py-data docutils literal notranslate"><span class="pre">idtrack._db.DB.node_type_str</span></code></a> is <code class="docutils literal notranslate"><span class="pre">DB.nts_ensembl[&quot;gene&quot;]</span></code>.  The result merges edge
data from <em>all</em> contributing external databases so that downstream callers receive one consolidated
view per gene.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p><dl class="simple">
<dt>Nested mapping <code class="docutils literal notranslate"><span class="pre">{gene_id:</span> <span class="pre">{database_name:</span> <span class="pre">{assembly:</span> <span class="pre">set[int]}}}</span></code>.</dt><dd><p>A single gene may appear under multiple assemblies when reference genomes share that transcript locus.</p>
</dd>
</dl>
</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p>dict</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="idtrack._the_graph.TheGraph.compact_ranges">
<em class="property"><span class="pre">static</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">compact_ranges</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">list_of_ranges</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/idtrack/_the_graph.html#TheGraph.compact_ranges"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#idtrack._the_graph.TheGraph.compact_ranges" title="Link to this definition"></a></dt>
<dd><p>Reduce the list of ranges into least possible number of ranges.</p>
<p>O(n) time and space complexity: a forward in place compaction and copying back the elements,
as then each inner step is O(1) (get/set instead of del).</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>list_of_ranges</strong> (<em>list</em><em>[</em><em>list</em><em>]</em>) – List of increasing, inclusive ranges. The elements are positive integers. Note that there
should be no overlapping elements.
The output of <a class="reference internal" href="#idtrack._the_graph.TheGraph.list_to_ranges" title="idtrack._the_graph.TheGraph.list_to_ranges"><code class="xref py py-meth docutils literal notranslate"><span class="pre">TheGraph.list_to_ranges()</span></code></a> or
<a class="reference internal" href="#idtrack._the_graph.TheGraph._get_active_ranges_of_id" title="idtrack._the_graph.TheGraph._get_active_ranges_of_id"><code class="xref py py-meth docutils literal notranslate"><span class="pre">TheGraph._get_active_ranges_of_id()</span></code></a> are a perfect input here.</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>List of ranges as defined in <code class="xref py py-attr docutils literal notranslate"><span class="pre">TheGraph.compact_ranges.list_of_ranges</span></code>.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>list[list]</p>
</dd>
</dl>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="idtrack._the_graph.TheGraph.external_database_connection_form">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">external_database_connection_form</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="pre">dict</span><span class="p"><span class="pre">[</span></span><span class="pre">str</span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="pre">str</span><span class="p"><span class="pre">]</span></span></em><a class="headerlink" href="#idtrack._the_graph.TheGraph.external_database_connection_form" title="Link to this definition"></a></dt>
<dd><p>Finds which form of Ensembl ID the external database identifiers are connected to.</p>
<p>Each external database connects to a specific form (gene, transcript, translation) Ensembl ID. The relevant
form is chosen by <code class="xref py py-class docutils literal notranslate"><span class="pre">ExternalDatabases</span></code> class.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>Dictionary mapping external database name into the associated form (gene, transcript, translation).</p>
</dd>
<dt class="field-even">Raises<span class="colon">:</span></dt>
<dd class="field-even"><p><strong>ValueError</strong> – If non-Ensembl node is connected.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="idtrack._the_graph.TheGraph.get_active_ranges_of_base_id_alternative">
<span class="sig-name descname"><span class="pre">get_active_ranges_of_base_id_alternative</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">base_id</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/idtrack/_the_graph.html#TheGraph.get_active_ranges_of_base_id_alternative"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#idtrack._the_graph.TheGraph.get_active_ranges_of_base_id_alternative" title="Link to this definition"></a></dt>
<dd><p>Get the range of an base ID based on the child IDs it is connected to.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>base_id</strong> (<em>str</em>) – Query ID. Should be with node type of Ensembl base ID.</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>Ranges of the ID as list of lists. Outputs should always be increasing, inclusive ranges.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>list[list]</p>
</dd>
</dl>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="idtrack._the_graph.TheGraph.get_active_ranges_of_id">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">get_active_ranges_of_id</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="pre">dict</span><span class="p"><span class="pre">[</span></span><span class="pre">str</span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="pre">list</span><span class="p"><span class="pre">[</span></span><span class="pre">list</span><span class="p"><span class="pre">]</span></span><span class="p"><span class="pre">]</span></span></em><a class="headerlink" href="#idtrack._the_graph.TheGraph.get_active_ranges_of_id" title="Link to this definition"></a></dt>
<dd><p>Returns the range of active ensembl releases of nodes, ignoring which assembly the release is coming from.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>A dictionary with following format
<code class="docutils literal notranslate"><span class="pre">{id:</span> <span class="pre">list_of_ranges</span></code>.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="idtrack._the_graph.TheGraph.get_active_ranges_of_id_ensembl_all_inclusive">
<span class="sig-name descname"><span class="pre">get_active_ranges_of_id_ensembl_all_inclusive</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">the_id</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/idtrack/_the_graph.html#TheGraph.get_active_ranges_of_id_ensembl_all_inclusive"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#idtrack._the_graph.TheGraph.get_active_ranges_of_id_ensembl_all_inclusive" title="Link to this definition"></a></dt>
<dd><p>Generate active ranges of Ensembl gene nodes with all assemblies.</p>
<p>Note that <a class="reference internal" href="#idtrack._the_graph.TheGraph.get_active_ranges_of_id" title="idtrack._the_graph.TheGraph.get_active_ranges_of_id"><code class="xref py py-meth docutils literal notranslate"><span class="pre">TheGraph.get_active_ranges_of_id()</span></code></a> method provided the range for main assembly
that the graph is build on. This method combines the other assemblies together. Also, it verifies whether
the <a class="reference internal" href="#idtrack._the_graph.TheGraph.combined_edges" title="idtrack._the_graph.TheGraph.combined_edges"><code class="xref py py-meth docutils literal notranslate"><span class="pre">TheGraph.combined_edges()</span></code></a> and <a class="reference internal" href="#idtrack._the_graph.TheGraph.get_active_ranges_of_id" title="idtrack._the_graph.TheGraph.get_active_ranges_of_id"><code class="xref py py-meth docutils literal notranslate"><span class="pre">TheGraph.get_active_ranges_of_id()</span></code></a>
methods provides consistent results.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>the_id</strong> (<em>str</em>) – Query ID. Should be either Ensembl gene or assembly specific Ensembl gene IDs.</p>
</dd>
<dt class="field-even">Raises<span class="colon">:</span></dt>
<dd class="field-even"><p><strong>ValueError</strong> – If there is inconsistency between the outputs of these two functions.
    If the query is not one of the specified node type.</p>
</dd>
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>Ranges of the ID as list of lists. Outputs should always be increasing, inclusive ranges.</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p>list[list]</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="idtrack._the_graph.TheGraph.get_external_database_nodes">
<span class="sig-name descname"><span class="pre">get_external_database_nodes</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">database_name</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/idtrack/_the_graph.html#TheGraph.get_external_database_nodes"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#idtrack._the_graph.TheGraph.get_external_database_nodes" title="Link to this definition"></a></dt>
<dd><p>For a given external database, returns set of all node names defined at least once in that database.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>database_name</strong> (<em>str</em>) – External database, it should be one of the item from
<a class="reference internal" href="#idtrack._the_graph.TheGraph.available_external_databases" title="idtrack._the_graph.TheGraph.available_external_databases"><code class="xref py py-meth docutils literal notranslate"><span class="pre">TheGraph.available_external_databases()</span></code></a>.</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>Node names (identifiers) set.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>set[str]</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="idtrack._the_graph.TheGraph.get_id_list">
<span class="sig-name descname"><span class="pre">get_id_list</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">database</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">assembly</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">release</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/idtrack/_the_graph.html#TheGraph.get_id_list"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#idtrack._the_graph.TheGraph.get_id_list" title="Link to this definition"></a></dt>
<dd><p>Given a trio (database, assembly, release), generates a list of node names (identifiers).</p>
<p>Similar to <a class="reference internal" href="#idtrack._the_graph.TheGraph.available_releases_given_database_assembly" title="idtrack._the_graph.TheGraph.available_releases_given_database_assembly"><code class="xref py py-meth docutils literal notranslate"><span class="pre">TheGraph.available_releases_given_database_assembly()</span></code></a>, the method uses
<cite>node_trios</cite> unnecessarily method, which consumes a lot of memory and hinders high
computational efficiency. However, this method is used only in testing purposes, when the speed and memory is
not of a concern.</p>
<p>It is imporant to note that the Ensembl IDs with versions <code class="xref py py-attr docutils literal notranslate"><span class="pre">DB.alternative_versions</span></code> will be also
returned if the database is the Ensembl gene with the main assembly, and the assembly is the main assembly.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>database</strong> (<em>str</em>) – External database name if the <cite>node type</cite> is <cite>external</cite>, else <cite>node type</cite>. The <cite>node type</cite>
for Ensembl should follow <code class="xref py py-attr docutils literal notranslate"><span class="pre">DB.nts_assembly</span></code> or <code class="xref py py-attr docutils literal notranslate"><span class="pre">DB.nts_base_ensembl</span></code>.</p></li>
<li><p><strong>assembly</strong> (<em>int</em>) – Assembly, it should be one of the item from
<a class="reference internal" href="#idtrack._the_graph.TheGraph.available_genome_assemblies" title="idtrack._the_graph.TheGraph.available_genome_assemblies"><code class="xref py py-meth docutils literal notranslate"><span class="pre">TheGraph.available_genome_assemblies()</span></code></a>.</p></li>
<li><p><strong>release</strong> (<em>int</em>) – Requested Ensembl releases as an integer.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>Node names (identifiers) list.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>list[str]</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="idtrack._the_graph.TheGraph.get_intersecting_ranges">
<em class="property"><span class="pre">static</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">get_intersecting_ranges</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">lor1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">lor2</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">compact</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/idtrack/_the_graph.html#TheGraph.get_intersecting_ranges"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#idtrack._the_graph.TheGraph.get_intersecting_ranges" title="Link to this definition"></a></dt>
<dd><p>As the name suggest, calculates the intersecting ranges of two list of ranges.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>lor1</strong> (<em>list</em><em>[</em><em>list</em><em>]</em>) – List of ranges as defined in <code class="xref py py-attr docutils literal notranslate"><span class="pre">TheGraph.compact_ranges.list_of_ranges</span></code>.</p></li>
<li><p><strong>lor2</strong> (<em>list</em><em>[</em><em>list</em><em>]</em>) – List of ranges as defined in <code class="xref py py-attr docutils literal notranslate"><span class="pre">TheGraph.compact_ranges.list_of_ranges</span></code>.</p></li>
<li><p><strong>compact</strong> (<em>bool</em>) – If set <cite>True</cite>, returns the reduced list of ranges via
<a class="reference internal" href="#idtrack._the_graph.TheGraph.compact_ranges" title="idtrack._the_graph.TheGraph.compact_ranges"><code class="xref py py-meth docutils literal notranslate"><span class="pre">TheGraph.compact_ranges()</span></code></a> method.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>List of ranges as defined in <code class="xref py py-attr docutils literal notranslate"><span class="pre">TheGraph.compact_ranges.list_of_ranges</span></code>.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>list[list]</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="idtrack._the_graph.TheGraph.get_next_edge_releases">
<span class="sig-name descname"><span class="pre">get_next_edge_releases</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">from_id</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">reverse</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/idtrack/_the_graph.html#TheGraph.get_next_edge_releases"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#idtrack._the_graph.TheGraph.get_next_edge_releases" title="Link to this definition"></a></dt>
<dd><p>Retrieves the next edge releases from a node, depending on the directionality of the graph.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>from_id</strong> (<em>str</em>) – Query ID. Should be with node type of Ensembl gene.</p></li>
<li><p><strong>reverse</strong> (<em>bool</em>) – The direction of desired next edges.
If <code class="docutils literal notranslate"><span class="pre">True</span></code>, previous edges are returned. If <code class="docutils literal notranslate"><span class="pre">False</span></code>, next edges are returned.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>The Ensembl releases of next (or previous if reverse is <code class="docutils literal notranslate"><span class="pre">True</span></code>) edges.</p>
</dd>
<dt class="field-odd">Raises<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>ValueError</strong> – If the query ID is not with the node type of graph backbone.</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p>list</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="idtrack._the_graph.TheGraph.get_two_nodes_coinciding_releases">
<span class="sig-name descname"><span class="pre">get_two_nodes_coinciding_releases</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">id1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">id2</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">compact</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/idtrack/_the_graph.html#TheGraph.get_two_nodes_coinciding_releases"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#idtrack._the_graph.TheGraph.get_two_nodes_coinciding_releases" title="Link to this definition"></a></dt>
<dd><p>Find the intersecting range of two nodes in the graph.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>id1</strong> (<em>str</em>) – First Query ID.</p></li>
<li><p><strong>id2</strong> (<em>str</em>) – Second Query ID.</p></li>
<li><p><strong>compact</strong> (<em>bool</em>) – Parameter to pass into <a class="reference internal" href="#idtrack._the_graph.TheGraph.get_intersecting_ranges" title="idtrack._the_graph.TheGraph.get_intersecting_ranges"><code class="xref py py-meth docutils literal notranslate"><span class="pre">TheGraph.get_intersecting_ranges()</span></code></a> method.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>List of ranges as defined in <code class="xref py py-attr docutils literal notranslate"><span class="pre">TheGraph.compact_ranges.list_of_ranges</span></code>.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>list[list]</p>
</dd>
</dl>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="idtrack._the_graph.TheGraph.hyperconnective_nodes">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">hyperconnective_nodes</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="pre">dict</span><span class="p"><span class="pre">[</span></span><span class="pre">str</span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="pre">int</span><span class="p"><span class="pre">]</span></span></em><a class="headerlink" href="#idtrack._the_graph.TheGraph.hyperconnective_nodes" title="Link to this definition"></a></dt>
<dd><p>Return hyper-connective external nodes and their out-degree counts.</p>
<p>Hyper-connective nodes are external identifiers whose out-degree (number of outgoing edges) exceeds
<a class="reference internal" href="#idtrack._db.DB.hyperconnecting_threshold" title="idtrack._db.DB.hyperconnecting_threshold"><code class="xref py py-attr docutils literal notranslate"><span class="pre">idtrack._db.DB.hyperconnecting_threshold</span></code></a>. Because such nodes may participate in tens of thousands
of mappings, they explode the breadth-first frontier of the synonym <em>pathfinder</em> algorithm and become a major
performance bottleneck.  The algorithm therefore <strong>ignores</strong> these nodes, sacrificing a small amount of
theoretical precision for a substantial speed-up.</p>
<p>In practice the precision penalty is negligible: hyper-connective nodes tend to be coarse-grained identifiers
that already suffer from low mapping specificity (for example, generic protein or transcript accessions
re-used across many unrelated biological entities).  Meaningful, one-to-one synonym relationships are almost
always reachable through alternative external identifiers.  Consequently, ignoring hyper-connective nodes both
accelerates the search and often improves the overall relevance of the results.</p>
<p>The value is computed lazily on first access and memoised via <code class="xref py py-meth docutils literal notranslate"><span class="pre">functools.cached_property()</span></code>, so the
underlying query runs at most once per <a class="reference internal" href="#idtrack._the_graph.TheGraph" title="idtrack._the_graph.TheGraph"><code class="xref py py-class docutils literal notranslate"><span class="pre">TheGraph</span></code></a> instance.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p><dl class="simple">
<dt>Mapping from external node identifier to its out-degree, limited to nodes whose</dt><dd><p>out-degree is greater than <a class="reference internal" href="#idtrack._db.DB.hyperconnecting_threshold" title="idtrack._db.DB.hyperconnecting_threshold"><code class="xref py py-attr docutils literal notranslate"><span class="pre">idtrack._db.DB.hyperconnecting_threshold</span></code></a> and whose
<a class="reference internal" href="#idtrack._db.DB.node_type_str" title="idtrack._db.DB.node_type_str"><code class="xref py py-data docutils literal notranslate"><span class="pre">idtrack._db.DB.node_type_str</span></code></a> equals <a class="reference internal" href="#idtrack._db.DB.nts_external" title="idtrack._db.DB.nts_external"><code class="xref py py-data docutils literal notranslate"><span class="pre">idtrack._db.DB.nts_external</span></code></a>.</p>
</dd>
</dl>
</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p>dict[str, int]</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="idtrack._the_graph.TheGraph.is_point_in_range">
<em class="property"><span class="pre">static</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">is_point_in_range</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">lor</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">p</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/idtrack/_the_graph.html#TheGraph.is_point_in_range"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#idtrack._the_graph.TheGraph.is_point_in_range" title="Link to this definition"></a></dt>
<dd><p>Simple method to determine whether a given integer is covered by list of ranges.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>lor</strong> (<em>list</em><em>[</em><em>list</em><em>]</em>) – List of ranges as defined in <code class="xref py py-attr docutils literal notranslate"><span class="pre">TheGraph.compact_ranges.list_of_ranges</span></code>.</p></li>
<li><p><strong>p</strong> (<em>int</em>) – A positive integer.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><code class="docutils literal notranslate"><span class="pre">True</span></code> if it is in the range, <code class="docutils literal notranslate"><span class="pre">False</span></code> otherwise.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>bool</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="idtrack._the_graph.TheGraph.list_to_ranges">
<em class="property"><span class="pre">static</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">list_to_ranges</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">lst</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/idtrack/_the_graph.html#TheGraph.list_to_ranges"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#idtrack._the_graph.TheGraph.list_to_ranges" title="Link to this definition"></a></dt>
<dd><p>Convert sorted non-repeating list of integers into list of inclusive non-overlapping ranges.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>lst</strong> (<em>list</em><em>[</em><em>int</em><em>]</em>) – List of integers. It should be sorted in increasing order. Repeating element is not allowed.
The output of <a class="reference internal" href="#idtrack._the_graph.TheGraph.ranges_to_list" title="idtrack._the_graph.TheGraph.ranges_to_list"><code class="xref py py-meth docutils literal notranslate"><span class="pre">TheGraph.ranges_to_list()</span></code></a> is a perfect input here.</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>Ranges as list of lists. Outputs should always be increasing, inclusive ranges. With positive integers.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>list[list]</p>
</dd>
</dl>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="idtrack._the_graph.TheGraph.lower_chars_graph">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">lower_chars_graph</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="pre">dict</span></em><a class="headerlink" href="#idtrack._the_graph.TheGraph.lower_chars_graph" title="Link to this definition"></a></dt>
<dd><p>A simple dictionary mapping the node name with its lower characters into the original node name.</p>
<dl class="field-list simple">
<dt class="field-odd">Raises<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>ValueError</strong> – If there are multiple nodes which becomes the same after lower character conversion.</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>A dictionary with following format
<code class="docutils literal notranslate"><span class="pre">{lower_char_id:</span> <span class="pre">id}</span></code></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="idtrack._the_graph.TheGraph.node_name_alternatives">
<span class="sig-name descname"><span class="pre">node_name_alternatives</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">identifier</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/idtrack/_the_graph.html#TheGraph.node_name_alternatives"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#idtrack._the_graph.TheGraph.node_name_alternatives" title="Link to this definition"></a></dt>
<dd><p>Matching a query ID into the ID found in the graph based on some criteria and priorities.</p>
<p>A query ID is sometimes not found exactly in the graph due to the format it has. However, very slight
modifications of the string of query ID could help the pathfinder locate the ID of interest. For example,
<cite>actb</cite> can be queried for the pathfinder, but nothing is found as upper character version of the query,
<cite>ACTB</cite>, is found in the graph instead.</p>
<p>Priority is as follows: (1) try to find directly in the graph (2) look for lower-char version (3) look for
initial substring before separators (4) check all possible variations of dash and underscore for each of which
follow the priority list above.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>identifier</strong> (<em>str</em>) – A bio-ID of interest. This could be an ID in the graph or some other ID to be matched
with the corresponding ID in the graph.</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><dl class="simple">
<dt>A tuple with the first element is the bio-ID in the graph if the query is found somehow, else <code class="docutils literal notranslate"><span class="pre">None</span></code>.</dt><dd><p>The second element is to show whether the query ID is found without any modifications or not.</p>
</dd>
</dl>
</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>tuple[str | None, bool]</p>
</dd>
</dl>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="idtrack._the_graph.TheGraph.node_trios">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">node_trios</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><span class="pre">dict</span><span class="p"><span class="pre">[</span></span><span class="pre">str</span><span class="p"><span class="pre">,</span></span><span class="w"> </span><span class="pre">set</span><span class="p"><span class="pre">[</span></span><span class="pre">tuple</span><span class="p"><span class="pre">]</span></span><span class="p"><span class="pre">]</span></span></em><a class="headerlink" href="#idtrack._the_graph.TheGraph.node_trios" title="Link to this definition"></a></dt>
<dd><p>Creates a dict for all nodes with <cite>node_trios</cite> calculated by <a class="reference internal" href="#idtrack._the_graph.TheGraph._node_trios" title="idtrack._the_graph.TheGraph._node_trios"><code class="xref py py-meth docutils literal notranslate"><span class="pre">TheGraph._node_trios()</span></code></a>.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>A memory intensive dictionary with node name as the key and calculated <cite>node_trios</cite> as the value.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="idtrack._the_graph.TheGraph.ranges_to_list">
<span class="sig-name descname"><span class="pre">ranges_to_list</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">lor</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/idtrack/_the_graph.html#TheGraph.ranges_to_list"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#idtrack._the_graph.TheGraph.ranges_to_list" title="Link to this definition"></a></dt>
<dd><p>Convert list of inclusive non-overlapping ranges into sorted sorted non-repeating list of integers.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>lor</strong> (<em>list</em><em>[</em><em>list</em><em>]</em>) – Ranges as list of lists. Should always be increasing, inclusive ranges. With positive integers.
The output of <a class="reference internal" href="#idtrack._the_graph.TheGraph.list_to_ranges" title="idtrack._the_graph.TheGraph.list_to_ranges"><code class="xref py py-meth docutils literal notranslate"><span class="pre">TheGraph.list_to_ranges()</span></code></a> is a perfect input here.</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>Sorted non-repeating list of integers.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>list[int]</p>
</dd>
</dl>
</dd></dl>

<dl class="py property">
<dt class="sig sig-object py" id="idtrack._the_graph.TheGraph.rev">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">rev</span></span><em class="property"><span class="p"><span class="pre">:</span></span><span class="w"> </span><a class="reference internal" href="#idtrack._the_graph.TheGraph" title="idtrack._the_graph.TheGraph"><span class="pre">TheGraph</span></a></em><a class="headerlink" href="#idtrack._the_graph.TheGraph.rev" title="Link to this definition"></a></dt>
<dd><p>Return a view of the same graph with all edge directions reversed.</p>
<p>The call delegates to <code class="xref py py-meth docutils literal notranslate"><span class="pre">networkx.MultiDiGraph.reverse()</span></code> with
<code class="docutils literal notranslate"><span class="pre">copy=False</span></code>, meaning the returned object re-uses the underlying data
structures and therefore consumes <strong>no additional memory</strong>.  Use this
property whenever a temporal walk must proceed <em>backwards</em> in history
(e.g. when resolving identifiers from a newer to an older Ensembl
release).</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p><dl class="simple">
<dt>A non-copying, lazily constructed reverse-orientation</dt><dd><p>view that honours every node and edge attribute of the original graph.</p>
</dd>
</dl>
</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><a class="reference internal" href="#idtrack._the_graph.TheGraph" title="idtrack._the_graph.TheGraph">TheGraph</a></p>
</dd>
</dl>
</dd></dl>

</dd></dl>

<dl class="py class" id="module-idtrack._graph_maker">
<dt class="sig sig-object py" id="idtrack._graph_maker.GraphMaker">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">GraphMaker</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">db_manager</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/idtrack/_graph_maker.html#GraphMaker"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#idtrack._graph_maker.GraphMaker" title="Link to this definition"></a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></p>
<p>Creates ID history graph.</p>
<p>It includes Ensembl gene ID history. Ensembl ID history is obtained from Ensembl
resources, which shows the connection between different Ensembl base IDs or different versions of the same Ensembl
base ID. Ensembl transcripts (with base IDs and versions) are connected to gene, and Ensembl proteins are
connected to transcripts. Additionally, a selected set of external databases are connected to the related Ensembl
IDs: for example UniProt IDs are associated with proteins, while RefSeq transcript IDs are associated with
transcripts. The <code class="docutils literal notranslate"><span class="pre">GraphMaker</span></code> class also saves the resulting graph into the defined temporary directory for later
calculations.</p>
<p>Class initialization.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>db_manager</strong> (<a class="reference internal" href="#idtrack._database_manager.DatabaseManager" title="idtrack._database_manager.DatabaseManager"><em>DatabaseManager</em></a>) – Needed to download all necessary tables and data frames.
It contains the temporary directory to save the resultant graph.</p>
</dd>
<dt class="field-even">Raises<span class="colon">:</span></dt>
<dd class="field-even"><p><strong>ValueError</strong> – <code class="docutils literal notranslate"><span class="pre">GraphMaker</span></code> has to be created with the latest release possible in <code class="docutils literal notranslate"><span class="pre">db_manager</span></code>. If not,
    the exception is raised.</p>
</dd>
</dl>
<dl class="py method">
<dt class="sig sig-object py" id="idtrack._graph_maker.GraphMaker.construct_graph">
<span class="sig-name descname"><span class="pre">construct_graph</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">narrow</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">form_list</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">narrow_external</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/idtrack/_graph_maker.html#GraphMaker.construct_graph"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#idtrack._graph_maker.GraphMaker.construct_graph" title="Link to this definition"></a></dt>
<dd><p>Main method to construct the graph.</p>
<p>It creates the graph with Ensembl gene, transcript and protein information. It also adds
<code class="docutils literal notranslate"><span class="pre">DB.nts_base_ensembl[f]</span></code> nodes into the graph, which has only base Ensembl gene ID (no version).
External database entries described in <code class="docutils literal notranslate"><span class="pre">ExternalDatabases</span></code> will be part of the graph. Normally, user
is not expected to use this method, as the method is utilized in <code class="docutils literal notranslate"><span class="pre">get_graph</span></code> method.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>narrow</strong> (<em>bool</em>) – Determine whether a some more information should be added between Ensembl gene IDs. For example,
which genome assembly is used, or when was the connection is established. For usual uses, no need to
set it <code class="docutils literal notranslate"><span class="pre">True</span></code>.</p></li>
<li><p><strong>form_list</strong> (<em>list</em><em> | </em><em>None</em>) – Determine which forms (transcript, translation, gene) should be included. If <code class="docutils literal notranslate"><span class="pre">None</span></code>, then
include all possible forms defined in <code class="docutils literal notranslate"><span class="pre">DatabaseManager</span></code>.
It has to be list composed of following strings: ‘gene’, ‘transcript’, ‘translation’.</p></li>
<li><p><strong>narrow_external</strong> (<em>bool</em>) – If set <code class="docutils literal notranslate"><span class="pre">False</span></code>, all possible external databases defined in Ensembl MySQL server will be
included into the graph. The graph will be immensely larger, and the ID history travel calculation will
be very slow. Additionally, the success of ID conversion under such a setting it has not been
tested yet.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>Resultant multiedge directed graph.</p>
</dd>
<dt class="field-odd">Raises<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>ValueError</strong> – Unexpected error.</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><a class="reference internal" href="#idtrack._the_graph.TheGraph" title="idtrack._the_graph.TheGraph"><em>TheGraph</em></a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="idtrack._graph_maker.GraphMaker.construct_graph_form">
<span class="sig-name descname"><span class="pre">construct_graph_form</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">narrow</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">db_manager</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/idtrack/_graph_maker.html#GraphMaker.construct_graph_form"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#idtrack._graph_maker.GraphMaker.construct_graph_form" title="Link to this definition"></a></dt>
<dd><p>Creates a graph with connected nodes based on historical relationships between each Ensembl IDs.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>narrow</strong> (<em>bool</em>) – See parameter in <code class="xref py py-attr docutils literal notranslate"><span class="pre">Graph.construct_graph.narrow</span></code></p></li>
<li><p><strong>db_manager</strong> (<a class="reference internal" href="#idtrack._database_manager.DatabaseManager" title="idtrack._database_manager.DatabaseManager"><em>DatabaseManager</em></a>) – The method reads ID history dataframe, and Ensembl IDs lists at each Ensembl release,
provided by <code class="docutils literal notranslate"><span class="pre">DatabaseManager</span></code>.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>Resultant multi edge directed graph.</p>
</dd>
<dt class="field-odd">Raises<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>ValueError</strong> – Unexpected error.</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><a class="reference internal" href="#idtrack._the_graph.TheGraph" title="idtrack._the_graph.TheGraph"><em>TheGraph</em></a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="idtrack._graph_maker.GraphMaker.create_file_name">
<span class="sig-name descname"><span class="pre">create_file_name</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">narrow</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/idtrack/_graph_maker.html#GraphMaker.create_file_name"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#idtrack._graph_maker.GraphMaker.create_file_name" title="Link to this definition"></a></dt>
<dd><p>File name creator which includes some information regarding the construction process.</p>
<p>Facilitates to recognize the graph based on file name.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>narrow</strong> (<em>bool</em>) – See parameter in <code class="xref py py-attr docutils literal notranslate"><span class="pre">Graph.construct_graph.narrow</span></code></p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>Absolute file path in the temporary directory provided by <code class="docutils literal notranslate"><span class="pre">DatabaseManager</span></code>.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>str</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="idtrack._graph_maker.GraphMaker.export_disk">
<span class="sig-name descname"><span class="pre">export_disk</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">g</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">file_path</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">overwrite</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/idtrack/_graph_maker.html#GraphMaker.export_disk"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#idtrack._graph_maker.GraphMaker.export_disk" title="Link to this definition"></a></dt>
<dd><p>Write the <cite>pickle</cite> file in the provided file path, which contains the graph.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>g</strong> (<a class="reference internal" href="#idtrack._the_graph.TheGraph" title="idtrack._the_graph.TheGraph"><em>TheGraph</em></a>) – Multi edge directed graph object to stor in the disk.</p></li>
<li><p><strong>file_path</strong> (<em>str</em>) – Absolute target path, provided by <code class="xref py py-meth docutils literal notranslate"><span class="pre">Graph.create_file_name()</span></code></p></li>
<li><p><strong>overwrite</strong> (<em>bool</em>) – See parameter in <code class="xref py py-attr docutils literal notranslate"><span class="pre">Graph.get_graph.overwrite_even_if_exist</span></code></p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="idtrack._graph_maker.GraphMaker.get_graph">
<span class="sig-name descname"><span class="pre">get_graph</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">narrow</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">create_even_if_exist</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">save_after_calculation</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">overwrite_even_if_exist</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/idtrack/_graph_maker.html#GraphMaker.get_graph"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#idtrack._graph_maker.GraphMaker.get_graph" title="Link to this definition"></a></dt>
<dd><p>Simplifies the graph construction process.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>narrow</strong> (<em>bool</em>) – See parameter in <code class="xref py py-attr docutils literal notranslate"><span class="pre">Graph.construct_graph.narrow</span></code></p></li>
<li><p><strong>create_even_if_exist</strong> (<em>bool</em>) – Determine whether create the graph even if it exists. If there is no graph in the
provided temporary directory, the graph will be created regardless.</p></li>
<li><p><strong>save_after_calculation</strong> (<em>bool</em>) – Determine whether resultant graph will be saved or not.</p></li>
<li><p><strong>overwrite_even_if_exist</strong> (<em>bool</em>) – If the graph will be saved, determine whether the program should overwrite.
If <code class="docutils literal notranslate"><span class="pre">False</span></code>, it does not re-saves the calculated (or loaded) graph.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>Resultant multi edge directed graph, which can be used in all future calculations.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference internal" href="#idtrack._the_graph.TheGraph" title="idtrack._the_graph.TheGraph"><em>TheGraph</em></a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="idtrack._graph_maker.GraphMaker.initialize_downloads">
<span class="sig-name descname"><span class="pre">initialize_downloads</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/idtrack/_graph_maker.html#GraphMaker.initialize_downloads"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#idtrack._graph_maker.GraphMaker.initialize_downloads" title="Link to this definition"></a></dt>
<dd><p>Initialize the external database downloads.</p>
<dl class="field-list simple">
<dt class="field-odd">Raises<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>NotImplementedError</strong> – Not implemented yet. Currently, the necessary data sources are downloaded when needed
    during the graph construction process.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="idtrack._graph_maker.GraphMaker.read_exported">
<em class="property"><span class="pre">static</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">read_exported</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">file_path</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/idtrack/_graph_maker.html#GraphMaker.read_exported"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#idtrack._graph_maker.GraphMaker.read_exported" title="Link to this definition"></a></dt>
<dd><p>Read the <cite>pickle</cite> file in the provided file path, which contains the graph.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>file_path</strong> (<em>str</em>) – Absolute path of the file of interest.</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>Resultant multi edge directed graph.</p>
</dd>
<dt class="field-odd">Raises<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>FileNotFoundError</strong> – When there is no file in the provided directory.</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p><a class="reference internal" href="#idtrack._the_graph.TheGraph" title="idtrack._the_graph.TheGraph"><em>TheGraph</em></a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="idtrack._graph_maker.GraphMaker.remove_non_gene_trees">
<em class="property"><span class="pre">static</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">remove_non_gene_trees</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">graph</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">forms_remove</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/idtrack/_graph_maker.html#GraphMaker.remove_non_gene_trees"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#idtrack._graph_maker.GraphMaker.remove_non_gene_trees" title="Link to this definition"></a></dt>
<dd><p>Removes the edges between the nodes with the same <cite>node type</cite> and removes abstract nodes (Void and Retired).</p>
<p>The nodes between two the same <code class="xref py py-attr docutils literal notranslate"><span class="pre">DB.node_type_str</span></code> will be removed. Also, the nodes with versions
<code class="xref py py-attr docutils literal notranslate"><span class="pre">DB.no_new_node_id</span></code> and <code class="xref py py-attr docutils literal notranslate"><span class="pre">DB.no_old_node_id</span></code> will be also removed.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>graph</strong> (<a class="reference internal" href="#idtrack._the_graph.TheGraph" title="idtrack._the_graph.TheGraph"><em>TheGraph</em></a>) – The output of <code class="xref py py-attr docutils literal notranslate"><span class="pre">Graph.construct_graph</span></code> or
<code class="xref py py-attr docutils literal notranslate"><span class="pre">Graph.construct_graph_form</span></code>.</p></li>
<li><p><strong>forms_remove</strong> (<em>list</em><em> | </em><em>None</em>) – Determine which <cite>node type</cite> are of interest.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>Resultant multi edge directed graph.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p><a class="reference internal" href="#idtrack._the_graph.TheGraph" title="idtrack._the_graph.TheGraph"><em>TheGraph</em></a></p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="idtrack._graph_maker.GraphMaker.split_id">
<em class="property"><span class="pre">static</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">split_id</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">id_to_split</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">which_part</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/idtrack/_graph_maker.html#GraphMaker.split_id"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#idtrack._graph_maker.GraphMaker.split_id" title="Link to this definition"></a></dt>
<dd><p>Simpler method to retrieve ID or Version part of a node name.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>id_to_split</strong> (<em>str</em>) – Query node name.</p></li>
<li><p><strong>which_part</strong> (<em>str</em>) – Either ‘Version’ or ‘ID’.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>The requested substring of the node name.</p>
</dd>
<dt class="field-odd">Raises<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>ValueError</strong> – If ‘which_part’ is assigned to some other value than ‘Version’, or ‘ID’.</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p>str | float</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="idtrack._graph_maker.GraphMaker.update_graph_with_the_new_release">
<span class="sig-name descname"><span class="pre">update_graph_with_the_new_release</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/idtrack/_graph_maker.html#GraphMaker.update_graph_with_the_new_release"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#idtrack._graph_maker.GraphMaker.update_graph_with_the_new_release" title="Link to this definition"></a></dt>
<dd><p>When new release arrive, just add new nodes.</p>
<dl class="field-list simple">
<dt class="field-odd">Raises<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>NotImplementedError</strong> – Not implemented yet. Currently, the user is expected to recreate whole graph using
    <code class="docutils literal notranslate"><span class="pre">get_graph</span></code> method. Note that not all databases need to be re-downloaded, the program will only
    download the new release, and re-construct the graph.</p>
</dd>
</dl>
</dd></dl>

</dd></dl>

<dl class="py class" id="module-idtrack._verify_organism">
<dt class="sig sig-object py" id="idtrack._verify_organism.VerifyOrganism">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">VerifyOrganism</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">organism_query</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/idtrack/_verify_organism.html#VerifyOrganism"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#idtrack._verify_organism.VerifyOrganism" title="Link to this definition"></a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></p>
<p>Finds the formal organism name and associated latest Ensembl release given a tentative organism name.</p>
<p>This class is designed to aid the user regarding the organism name accepted by the program and latest Ensembl
release possible without manually looking online resources. Every time of use, the class uses Ensembl REST API to
get the information regarding the organisms and associated Ensembl releases, so the class requires an active
Internet connection.</p>
<p>Class initialization.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>organism_query</strong> (<em>str</em>) – Organism of interest.
This does not have to be formal name. For example, ‘human’, ‘hsapiens’ as well as formal name
‘homo_sapiens’ is accepted by the program.</p>
</dd>
</dl>
<dl class="py method">
<dt class="sig sig-object py" id="idtrack._verify_organism.VerifyOrganism.fetch_organism_and_latest_release">
<span class="sig-name descname"><span class="pre">fetch_organism_and_latest_release</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">connect_timeout</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">read_timeout</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/idtrack/_verify_organism.html#VerifyOrganism.fetch_organism_and_latest_release"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#idtrack._verify_organism.VerifyOrganism.fetch_organism_and_latest_release" title="Link to this definition"></a></dt>
<dd><p>Creates a query and process the data to construct two dataframes for Ensembl release and organism names.</p>
<p>The query to get the name of the organisms defined in the database, but also to get the latest
ensembl release to go through. This method is actually main functionality of the class; in general uses, the
user is not expected to use this method directly.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>connect_timeout</strong> (<em>int</em>) – The number of seconds the requests will wait to establish a connection to a remote machine.</p></li>
<li><p><strong>read_timeout</strong> (<em>int</em>) – The number of seconds the client will wait for the server to send a response</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>Two dataframes to be used by the instance methods later on.</p>
</dd>
<dt class="field-odd">Raises<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>ValueError</strong> – Unexpected error.</p></li>
<li><p><strong>TimeoutError</strong> – If allowed time is passed.</p></li>
</ul>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p>tuple[<em>DataFrame</em>, <em>DataFrame</em>]</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="idtrack._verify_organism.VerifyOrganism.get_formal_name">
<span class="sig-name descname"><span class="pre">get_formal_name</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/idtrack/_verify_organism.html#VerifyOrganism.get_formal_name"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#idtrack._verify_organism.VerifyOrganism.get_formal_name" title="Link to this definition"></a></dt>
<dd><p>Using the API response and queried organism name, find the formal organism name to be used for the program.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>Formal organism name.</p>
</dd>
<dt class="field-even">Raises<span class="colon">:</span></dt>
<dd class="field-even"><ul class="simple">
<li><p><strong>KeyError</strong> – If the query is not defined in the pre-created dataframes.</p></li>
<li><p><strong>ValueError</strong> – If there are multiple entries for query and hence the query is ambiguous.”</p></li>
</ul>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>str</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="idtrack._verify_organism.VerifyOrganism.get_latest_release">
<span class="sig-name descname"><span class="pre">get_latest_release</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/idtrack/_verify_organism.html#VerifyOrganism.get_latest_release"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#idtrack._verify_organism.VerifyOrganism.get_latest_release" title="Link to this definition"></a></dt>
<dd><p>Using the API response and queried organism name, find the latest Ensembl release to be used for the program.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>Latest Ensembl release.</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p>int</p>
</dd>
</dl>
</dd></dl>

</dd></dl>

<dl class="py class" id="module-idtrack._track">
<dt class="sig sig-object py" id="idtrack._track.Track">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">Track</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">db_manager</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/idtrack/_track.html#Track"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#idtrack._track.Track" title="Link to this definition"></a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></p>
<p>Bidirectional path-finding resolver for biological identifiers.</p>
<p><cite>Track</cite> builds and queries a <em>bio-ID</em> multigraph that stitches together
Ensembl history edges (genes, transcripts, proteins) and cross-reference
edges to external databases (UniProt, RefSeq, …).  Given a <strong>source
identifier</strong>, a <strong>target Ensembl release</strong>, and/or a <strong>target database</strong>,
the class:</p>
<ol class="arabic simple">
<li><p><strong>Normalises</strong> the source to an Ensembl <em>gene</em> node when necessary.</p></li>
<li><p><strong>Time-travels</strong> through historical edges—forward or backward—until it
reaches the requested release, optionally “beaming-up” through external
IDs when the backbone is disconnected.</p></li>
<li><p><strong>Converts</strong> the resolved Ensembl gene into the requested external
database (or returns the gene itself) while annotating the result with
confidence scores and the full traversal path.</p></li>
</ol>
<p>Two mutually-recursive engines power the search:</p>
<ul class="simple">
<li><p><cite>_recursive_function</cite>   — depth-first search along temporal edges.</p></li>
<li><p><cite>_recursive_synonymous</cite> — search for synonymous nodes at a single
release to enable the external “beam-up”.</p></li>
</ul>
<dl class="py attribute">
<dt class="sig sig-object py" id="idtrack._track.Track.graph">
<span class="sig-name descname"><span class="pre">graph</span></span><a class="headerlink" href="#idtrack._track.Track.graph" title="Link to this definition"></a></dt>
<dd><p>The pre-computed bio-ID graph produced by
<code class="xref py py-class docutils literal notranslate"><span class="pre">_graph_maker.GraphMaker</span></code>.</p>
<dl class="field-list simple">
<dt class="field-odd">Type<span class="colon">:</span></dt>
<dd class="field-odd"><p>networkx.MultiDiGraph</p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="idtrack._track.Track.version_info">
<span class="sig-name descname"><span class="pre">version_info</span></span><a class="headerlink" href="#idtrack._track.Track.version_info" title="Link to this definition"></a></dt>
<dd><p>Metadata about the graph build (Ensembl releases included, build date,
Git commit, etc.).</p>
<dl class="field-list simple">
<dt class="field-odd">Type<span class="colon">:</span></dt>
<dd class="field-odd"><p>dict</p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="idtrack._track.Track._external_entrance_placeholder">
<span class="sig-name descname"><span class="pre">_external_entrance_placeholder</span></span><a class="headerlink" href="#idtrack._track.Track._external_entrance_placeholder" title="Link to this definition"></a></dt>
<dd><p>Sentinel node IDs that mark artificial edges used when an external ID
is pulled onto the Ensembl backbone (<cite>False → -1</cite>, <cite>True → 10001</cite>).</p>
<dl class="field-list simple">
<dt class="field-odd">Type<span class="colon">:</span></dt>
<dd class="field-odd"><p>dict[bool, int]</p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="idtrack._track.Track._external_entrance_placeholders">
<span class="sig-name descname"><span class="pre">_external_entrance_placeholders</span></span><a class="headerlink" href="#idtrack._track.Track._external_entrance_placeholders" title="Link to this definition"></a></dt>
<dd><p>Sorted list of the sentinel values above.</p>
<dl class="field-list simple">
<dt class="field-odd">Type<span class="colon">:</span></dt>
<dd class="field-odd"><p>list[int]</p>
</dd>
</dl>
</dd></dl>

<p>Create a <cite>Track</cite> resolver and load (or build) its graph.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>db_manager</strong> (<a class="reference internal" href="#idtrack._database_manager.DatabaseManager" title="idtrack._database_manager.DatabaseManager"><em>DatabaseManager</em></a>) – Connection manager that knows how to fetch Ensembl and
cross-reference tables from a local cache or a live MySQL mirror.
The same instance is forwarded to <code class="xref py py-class docutils literal notranslate"><span class="pre">_graph_maker.GraphMaker</span></code>.</p></li>
<li><p><strong>kwargs</strong> – Additional keyword arguments forwarded verbatim to
<code class="xref py py-meth docutils literal notranslate"><span class="pre">_graph_maker.GraphMaker.get_graph()</span></code>.  Common flags include
<cite>force_rebuild</cite> (recompute the graph from scratch), <cite>species</cite>
(restrict to one taxon), and <cite>cache_dir</cite> (override on-disk cache location).</p></li>
</ul>
</dd>
</dl>
<dl class="py property">
<dt class="sig sig-object py" id="idtrack._track.Track._calculate_node_scores_helper">
<em class="property"><span class="pre">property</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">_calculate_node_scores_helper</span></span><a class="headerlink" href="#idtrack._track.Track._calculate_node_scores_helper" title="Link to this definition"></a></dt>
<dd><p>Build and cache helper look-ups for node-scoring.</p>
<p>The property constructs two complementary data structures:</p>
<ul class="simple">
<li><dl class="simple">
<dt><strong>filter_set</strong> - the union of</dt><dd><p>(1) every external-database node-type present in the graph, and (2) every Ensembl-specific
node-type (gene, transcript, translation, …) across <em>all</em> assemblies.
This set can therefore be passed unmodified to
<a class="reference internal" href="#idtrack._track.Track.synonymous_nodes" title="idtrack._track.Track.synonymous_nodes"><code class="xref py py-meth docutils literal notranslate"><span class="pre">synonymous_nodes()</span></code></a> to ask for “anything that is not an assembly-less backbone gene”.</p>
</dd>
</dl>
</li>
<li><dl class="simple">
<dt><strong>ensembl_include</strong> - a mapping</dt><dd><p><cite>{form → set(node_type_str)}</cite> where each value lists the node-types
that should be considered equivalent to that <em>form</em> (e.g. <em>gene</em>, <em>transcript</em>, <em>translation</em>) when
computing richness metrics.</p>
</dd>
</dl>
</li>
</ul>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p><cite>(filter_set, ensembl_include)</cite> exactly as described above.</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p>tuple[set[str], dict[str, set[str]]]</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="idtrack._track.Track._choose_relevant_synonym_helper">
<span class="sig-name descname"><span class="pre">_choose_relevant_synonym_helper</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">from_id</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">synonym_ids</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">to_release</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">from_release</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">mode</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/idtrack/_track.html#Track._choose_relevant_synonym_helper"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#idtrack._track.Track._choose_relevant_synonym_helper" title="Link to this definition"></a></dt>
<dd><p>Select the <strong>most temporally relevant</strong> synonym(s) for an Ensembl gene-ID family.</p>
<p>The method evaluates each candidate in <cite>synonym_ids</cite> against the
<em>target</em> release <cite>to_release</cite> and, when applicable, the <em>source</em>
release <cite>from_release</cite>.  Its job is to decide <strong>where</strong> the path
should <em>enter</em> the Ensembl backbone and <strong>whether</strong> the remainder of
the traversal must run in <em>reverse</em> (new → old) order.</p>
<section id="selection-strategy">
<h2>Selection strategy<a class="headerlink" href="#selection-strategy" title="Link to this heading"></a></h2>
<ol class="arabic simple">
<li><p><strong>Fixed `from_release`</strong> - If the caller already knows the release
of the starting node, every candidate is paired with that same
release and the correct <em>reverse</em> flag is derived trivially.</p></li>
<li><p><strong>Non-backbone start</strong> - When the starting node is not an
Ensembl-gene backbone ID, the synonym whose <strong>active range edge</strong> is
<em>closest</em> (or <em>farthest</em>, per <cite>mode</cite>) to <cite>to_release</cite> is chosen.</p></li>
<li><p><strong>Backbone start</strong> - If the query is itself an Ensembl-gene, the
algorithm first looks for <em>overlapping</em> ranges between the query
and each synonym; if none overlap, it falls back to the distance
rule described in step 2.</p></li>
</ol>
<dl class="field-list simple">
<dt class="field-odd">param from_id<span class="colon">:</span></dt>
<dd class="field-odd"><p>Identifier from which the path search will start.</p>
</dd>
<dt class="field-even">type from_id<span class="colon">:</span></dt>
<dd class="field-even"><p>str</p>
</dd>
<dt class="field-odd">param synonym_ids<span class="colon">:</span></dt>
<dd class="field-odd"><p>Ensembl IDs considered synonyms of
<cite>from_id</cite> (typically the <em>same</em> gene with different version
numbers).</p>
</dd>
<dt class="field-even">type synonym_ids<span class="colon">:</span></dt>
<dd class="field-even"><p>Sequence[str]</p>
</dd>
<dt class="field-odd">param to_release<span class="colon">:</span></dt>
<dd class="field-odd"><p>Target Ensembl release that the overall conversion
aims for.</p>
</dd>
<dt class="field-even">type to_release<span class="colon">:</span></dt>
<dd class="field-even"><p>int</p>
</dd>
<dt class="field-odd">param from_release<span class="colon">:</span></dt>
<dd class="field-odd"><p>Release in which <cite>from_id</cite> is known to
be active.  If <em>None</em>, the method infers a suitable release for
each candidate.</p>
</dd>
<dt class="field-even">type from_release<span class="colon">:</span></dt>
<dd class="field-even"><p>int | None</p>
</dd>
<dt class="field-odd">param mode<span class="colon">:</span></dt>
<dd class="field-odd"><p>Either <cite>‘closest’</cite> or <cite>‘distant’</cite>—controls whether
the synonym chosen should minimise or maximise its distance to
<cite>to_release</cite>.</p>
</dd>
<dt class="field-even">type mode<span class="colon">:</span></dt>
<dd class="field-even"><p>str</p>
</dd>
<dt class="field-odd">returns<span class="colon">:</span></dt>
<dd class="field-odd"><p>One or more triplets of the form
<cite>[synonym_id, entry_release, reverse]</cite> where:</p>
<ul class="simple">
<li><p><cite>synonym_id</cite> - the chosen synonym,</p></li>
<li><p><cite>entry_release</cite> - release at which to join the backbone, and</p></li>
<li><p><cite>reverse</cite> - <em>True</em> if the subsequent history walk must run
backwards in time.</p></li>
</ul>
</dd>
<dt class="field-even">rtype<span class="colon">:</span></dt>
<dd class="field-even"><p>list[list[Union[str, int, bool]]]</p>
</dd>
<dt class="field-odd">raises ValueError<span class="colon">:</span></dt>
<dd class="field-odd"><p>If no synonym satisfies the distance/overlap criteria
    or if <cite>mode</cite> is invalid.</p>
</dd>
</dl>
</section>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>to_release</strong> (<em>int</em>)</p></li>
<li><p><strong>from_release</strong> (<em>int</em><em> | </em><em>None</em>)</p></li>
<li><p><strong>mode</strong> (<em>str</em>)</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="idtrack._track.Track._create_priority_list_ensembl">
<span class="sig-name descname"><span class="pre">_create_priority_list_ensembl</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">from_id</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">to_release</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/idtrack/_track.html#Track._create_priority_list_ensembl"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#idtrack._track.Track._create_priority_list_ensembl" title="Link to this definition"></a></dt>
<dd><p>Build a priority list of assemblies in which <cite>from_id</cite> is active.</p>
<p>The priorities are the numeric <strong>assembly rankings</strong> defined in
<code class="xref py py-attr docutils literal notranslate"><span class="pre">DB.assembly_mysqlport_priority</span></code> (smaller numbers mean
higher priority).</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>from_id</strong> (<em>str</em>) – Ensembl gene identifier.</p></li>
<li><p><strong>to_release</strong> (<em>int</em>) – Target Ensembl release; only assemblies that
contain this release are considered.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>Sorted list of priority values (ascending).</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>list[int]</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="idtrack._track.Track._final_conversion">
<span class="sig-name descname"><span class="pre">_final_conversion</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">converted</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">cnvt</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">final_database</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ens_release</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">return_path</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">return_ensembl_alternative</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">prevent_assembly_jumps</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">account_for_hyperconnected_nodes</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/idtrack/_track.html#Track._final_conversion"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#idtrack._track.Track._final_conversion" title="Link to this definition"></a></dt>
<dd><p>Convert an Ensembl gene node to the requested external database.</p>
<p>Convert an Ensembl gene node to the requested external database and
merge the result back into <cite>converted</cite>.</p>
<p>The routine:</p>
<ol class="arabic simple">
<li><p>Builds every legal synonym path from <cite>cnvt</cite> to
<cite>final_database</cite> that is active in <cite>ens_release</cite> (or in any
release as a fallback).</p></li>
<li><p>Computes assembly-jump penalties for each path.</p></li>
<li><p>Calls <a class="reference internal" href="#idtrack._track.Track._final_conversion_dict_prepare" title="idtrack._track.Track._final_conversion_dict_prepare"><code class="xref py py-meth docutils literal notranslate"><span class="pre">_final_conversion_dict_prepare()</span></code></a> to create the
conversion sub-dict.</p></li>
<li><p>Optionally falls back to returning the Ensembl gene itself when no
synonym exists and <cite>return_ensembl_alternative</cite> is True.</p></li>
</ol>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>converted</strong> (<em>dict</em>) – The current accumulator being built by
<a class="reference internal" href="#idtrack._track.Track.convert" title="idtrack._track.Track.convert"><code class="xref py py-meth docutils literal notranslate"><span class="pre">convert()</span></code></a>.</p></li>
<li><p><strong>cnvt</strong> (<em>str</em>) – Ensembl gene identifier that is undergoing final
conversion.</p></li>
<li><p><strong>final_database</strong> (<em>str</em>) – Target external database.</p></li>
<li><p><strong>ens_release</strong> (<em>int</em>) – Target Ensembl release.</p></li>
<li><p><strong>return_path</strong> (<em>bool</em>) – If True, embed the path(s) that lead to each
synonym.</p></li>
<li><p><strong>return_ensembl_alternative</strong> (<em>bool</em>) – When no synonym can be found,
add a fallback entry that keeps the Ensembl gene.</p></li>
<li><p><strong>prevent_assembly_jumps</strong> (<em>bool</em>) – Todo.</p></li>
<li><p><strong>account_for_hyperconnected_nodes</strong> (<em>bool</em>) – Todo.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>The same <cite>converted</cite> dict, updated in place (and also
returned for convenience).</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>dict</p>
</dd>
<dt class="field-even">Raises<span class="colon">:</span></dt>
<dd class="field-even"><p><a class="reference internal" href="#idtrack._db.EmptyConversionMetricsError" title="idtrack._db.EmptyConversionMetricsError"><strong>EmptyConversionMetricsError</strong></a> – Todo.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="idtrack._track.Track._final_conversion_dict_prepare">
<em class="property"><span class="pre">static</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">_final_conversion_dict_prepare</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">confidence</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">sysns</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">paths</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">min_priority_list</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">len_priority_list</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">add_ass_jump_list</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">final_database</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/idtrack/_track.html#Track._final_conversion_dict_prepare"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#idtrack._track.Track._final_conversion_dict_prepare" title="Link to this definition"></a></dt>
<dd><p>Assemble the final-conversion section that will be attached to a candidate path.</p>
<p>The section contains a global conversion-confidence flag plus one
entry per synonym that survived the path-finding stage.  When
<cite>paths</cite> is None the structure is identical but omits the
<cite>‘the_path’</cite> member to save memory.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>confidence</strong> (<em>int</em><em> | </em><em>float</em>) – Heuristic confidence for the <em>whole</em>
conversion step - <cite>0</cite> for “perfect”, larger values for
fallback scenarios, <cite>np.inf</cite> when no conversion was
possible.</p></li>
<li><p><strong>sysns</strong> (<em>list</em>) – List of synonym identifiers <em>in the same order</em>
as the metric lists below.</p></li>
<li><p><strong>paths</strong> (<em>list</em><em>[</em><em>list</em><em>] </em><em>| </em><em>None</em>) – One walk (edge list) per synonym, or
<em>None</em> if the caller does not want to expose paths.</p></li>
<li><p><strong>min_priority_list</strong> (<em>list</em>) – Minimum assembly priority reached
by each walk.</p></li>
<li><p><strong>len_priority_list</strong> (<em>list</em>) – Number of distinct assembly
priorities encountered by each walk.</p></li>
<li><p><strong>add_ass_jump_list</strong> (<em>list</em>) – Additional assembly-jump penalty
incurred during the synonym hop itself.</p></li>
<li><p><strong>final_database</strong> (<em>str</em>) – Name of the database these synonyms belong
to (e.g. <cite>‘uniprot’</cite> or <cite>DB.nts_ensembl[DB.backbone_form]</cite>).</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>Nested dictionary ready to be stored under the key
<cite>‘final_conversion’</cite>.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>dict</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="idtrack._track.Track._minimum_assembly_jumps_helper">
<em class="property"><span class="pre">static</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">_minimum_assembly_jumps_helper</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">step_pri</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">current_priority</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">priorities</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/idtrack/_track.html#Track._minimum_assembly_jumps_helper"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#idtrack._track.Track._minimum_assembly_jumps_helper" title="Link to this definition"></a></dt>
<dd><p>Internal worker for <a class="reference internal" href="#idtrack._track.Track.minimum_assembly_jumps" title="idtrack._track.Track.minimum_assembly_jumps"><code class="xref py py-meth docutils literal notranslate"><span class="pre">minimum_assembly_jumps()</span></code></a>.</p>
<p>Given the priority sets for the remaining edges, iterate until all have
been consumed while updating the <em>current</em> assembly priority and counting
how often it must drop.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>step_pri</strong> (<em>list</em><em>[</em><em>int</em><em>]</em>) – Priority values of the edge currently under
consideration.</p></li>
<li><p><strong>current_priority</strong> (<em>int</em>) – Priority value inherited from previous steps.</p></li>
<li><p><strong>priorities</strong> (<em>list</em><em>[</em><em>list</em><em>[</em><em>int</em><em>]</em><em>]</em>) – Priority lists for the <em>rest</em> of the path,
<strong>already sorted</strong> for correct bisecting.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>Same three-tuple as documented in
<a class="reference internal" href="#idtrack._track.Track.minimum_assembly_jumps" title="idtrack._track.Track.minimum_assembly_jumps"><code class="xref py py-meth docutils literal notranslate"><span class="pre">minimum_assembly_jumps()</span></code></a>.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>tuple[int, list[int], int]</p>
</dd>
<dt class="field-even">Raises<span class="colon">:</span></dt>
<dd class="field-even"><p><strong>ValueError</strong> – If the priority lattice enters an impossible state.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="idtrack._track.Track._path_score_sorter_all_targets">
<span class="sig-name descname"><span class="pre">_path_score_sorter_all_targets</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">dict_of_dict</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">from_id</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">to_release</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/idtrack/_track.html#Track._path_score_sorter_all_targets"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#idtrack._track.Track._path_score_sorter_all_targets" title="Link to this definition"></a></dt>
<dd><p>Select the overall best target(s).</p>
<p>Select the <strong>overall best target(s)</strong> once every candidate Ensembl node
has itself been reduced to its single best path.</p>
<p>The method linearises several per-path metrics into an <em>importance
order</em> (see the tuple at the top of the function), then:</p>
<ol class="arabic simple">
<li><p>Computes that ordered score for <strong>each</strong> pair <cite>(ensembl_gene, final_target)</cite>.</p></li>
<li><dl class="simple">
<dt>Finds the global minimum; if multiple pairs tie:</dt><dd><ul class="simple">
<li><p>Prefer the target whose identifier is identical to <cite>from_id</cite>.</p></li>
<li><dl class="simple">
<dt>If more than one Ensembl gene still tie, fall back on</dt><dd><p><a class="reference internal" href="#idtrack._track.Track.calculate_node_scores" title="idtrack._track.Track.calculate_node_scores"><code class="xref py py-meth docutils literal notranslate"><span class="pre">calculate_node_scores()</span></code></a> to favour the “richer” node.</p>
</dd>
</dl>
</li>
</ul>
</dd>
</dl>
</li>
<li><p>Returns a <em>pruned</em> copy of <cite>dict_of_dict</cite> that contains only the
surviving Ensembl genes, each with only the winning <cite>final_elements</cite> entry.
Additional provenance is written to <cite>filter_scores</cite>.</p></li>
</ol>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>dict_of_dict</strong> (<em>dict</em>) – Nested result of
<a class="reference internal" href="#idtrack._track.Track.calculate_score_and_select" title="idtrack._track.Track.calculate_score_and_select"><code class="xref py py-meth docutils literal notranslate"><span class="pre">calculate_score_and_select()</span></code></a>.  Keys are candidate
Ensembl genes; values are dictionaries that already contain
<em>one</em> best path per final target.</p></li>
<li><p><strong>from_id</strong> (<em>str</em>) – Original query identifier; used to break ties in
favour of “same as input”.</p></li>
<li><p><strong>to_release</strong> (<em>int</em>) – Target Ensembl release; forwarded to
<a class="reference internal" href="#idtrack._track.Track.calculate_node_scores" title="idtrack._track.Track.calculate_node_scores"><code class="xref py py-meth docutils literal notranslate"><span class="pre">calculate_node_scores()</span></code></a> during tie-breaking.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><dl class="simple">
<dt>A reduced version of <cite>dict_of_dict</cite> holding only the winner(s) and enriched with a</dt><dd><p><cite>final_elements[*][‘filter_scores’]</cite> sub-dict that records the filters applied.</p>
</dd>
</dl>
</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>dict</p>
</dd>
<dt class="field-even">Raises<span class="colon">:</span></dt>
<dd class="field-even"><p><strong>ValueError</strong> – If <cite>dict_of_dict</cite> is empty.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="idtrack._track.Track._path_score_sorter_single_target">
<em class="property"><span class="pre">static</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">_path_score_sorter_single_target</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">lst_of_dict</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/idtrack/_track.html#Track._path_score_sorter_single_target"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#idtrack._track.Track._path_score_sorter_single_target" title="Link to this definition"></a></dt>
<dd><p>Select the best score dictionary for <em>one</em> conversion target.</p>
<p>The input is a list of dictionaries produced by
<a class="reference internal" href="#idtrack._track.Track.calculate_score_and_select" title="idtrack._track.Track.calculate_score_and_select"><code class="xref py py-meth docutils literal notranslate"><span class="pre">calculate_score_and_select()</span></code></a>.  Each dictionary is converted into a
tuple according to the <em>lexicographic importance order</em></p>
<p><cite>(“assembly_jump”, “external_jump”, “external_step”,
“edge_scores_reduced”, “ensembl_step”)</cite></p>
<p>and the dictionary with the <strong>smallest</strong> tuple is returned.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>lst_of_dict</strong> (<em>list</em><em>[</em><em>dict</em><em>]</em>) – Candidate score dictionaries for this target.</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>The chosen “winner” score dictionary.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>dict</p>
</dd>
<dt class="field-even">Raises<span class="colon">:</span></dt>
<dd class="field-even"><p><strong>ValueError</strong> – If the input list is empty.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="idtrack._track.Track._recursive_synonymous">
<span class="sig-name descname"><span class="pre">_recursive_synonymous</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">_the_id</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">synonymous_ones</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">synonymous_ones_db</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">filter_node_type</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">the_path</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">the_path_db</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">depth_max</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">from_release</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ensembl_backbone_shallow_search</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">account_for_hyperconnected_nodes</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/idtrack/_track.html#Track._recursive_synonymous"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#idtrack._track.Track._recursive_synonymous" title="Link to this definition"></a></dt>
<dd><p>Helper method to be used in <code class="xref py py-meth docutils literal notranslate"><span class="pre">_graph.Track.synonymous_nodes()</span></code>.</p>
<p>Recursively explore the bio-ID graph to collect <strong>synonymous paths</strong>
starting at <cite>_the_id</cite> and ending on a node whose <em>type</em> is a member of <cite>filter_node_type</cite>.</p>
<p>A <em>path</em> is a list of node identifiers (<cite>_the_path</cite>) together with a
parallel list of their node-type strings (<cite>_the_path_db</cite>).
The search is <strong>breadth-limited</strong>: the <em>depth</em> of a path is defined as
the <strong>maximum count of any single node-type</strong> it contains
(e.g. a path with three <cite>external</cite> nodes has depth 3).  Recursion
stops when that depth would exceed <cite>depth_max</cite>.</p>
<p>Additional pruning rules:</p>
<ul class="simple">
<li><p>The walk never visits the same node twice (no cycles).</p></li>
<li><dl class="simple">
<dt>It never traverses two consecutive edges whose source and target</dt><dd><p>share the <strong>same node-type</strong>—this prevents “time-travel” within the
Ensembl history backbone.</p>
</dd>
</dl>
</li>
<li><dl class="simple">
<dt>When <cite>ensembl_backbone_shallow_search</cite> is <em>True</em>, the search is</dt><dd><p>restricted to the <strong>reverse</strong> direction except for node-types listed
in <code class="xref py py-attr docutils literal notranslate"><span class="pre">DB.nts_bidirectional_synonymous_search</span></code>.</p>
</dd>
</dl>
</li>
</ul>
<p>On reaching a terminating node the method <em>appends</em> the discovered
paths to <cite>synonymous_ones</cite> and <cite>synonymous_ones_db</cite>. It does not
return anything. Results are accumulated in <cite>synonymous_ones</cite> and <cite>synonymous_ones_db</cite>.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>_the_id</strong> (<em>str</em>) – Identifier of the starting node (Ensembl or
external).</p></li>
<li><p><strong>synonymous_ones</strong> (<em>list</em>) – Mutable list that will receive
each successful identifier path.</p></li>
<li><p><strong>synonymous_ones_db</strong> (<em>list</em>) – Mutable list that will receive
the corresponding node-type paths.</p></li>
<li><p><strong>filter_node_type</strong> (<em>set</em><em>[</em><em>str</em><em>]</em>) – Allowed node-types for the <strong>final</strong>
node of a path (e.g. <cite>{‘ensembl_gene’}</cite>).</p></li>
<li><p><strong>the_path</strong> (<em>list</em><em> | </em><em>None</em>) – Current path leading to
<cite>_the_id</cite>; <em>None</em> for the root invocation.</p></li>
<li><p><strong>the_path_db</strong> (<em>list</em><em> | </em><em>None</em>) – Node-type counterpart of
<cite>the_path</cite>; <em>None</em> for the root invocation.</p></li>
<li><p><strong>depth_max</strong> (<em>int</em>) – Maximum allowed depth as defined above.</p></li>
<li><p><strong>from_release</strong> (<em>int</em><em> | </em><em>None</em>) – If given, only keep terminal
nodes that are <em>active</em> in this Ensembl release.</p></li>
<li><p><strong>ensembl_backbone_shallow_search</strong> (<em>bool</em>) – Activate the
shallow, mostly-reverse search mode described above.</p></li>
<li><p><strong>account_for_hyperconnected_nodes</strong> (<em>bool</em>) – Todo.</p></li>
</ul>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="idtrack._track.Track.calculate_node_scores">
<span class="sig-name descname"><span class="pre">calculate_node_scores</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">the_id</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ens_release</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/idtrack/_track.html#Track.calculate_node_scores"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#idtrack._track.Track.calculate_node_scores" title="Link to this definition"></a></dt>
<dd><p>Rank competing Ensembl targets by the “richness” of their synonyms.</p>
<p>The method counts, within a radius of <em>two</em> synonym hops, how many unique
identifiers of various categories point to each candidate and returns the
counts as <strong>negative integers</strong> so that <em>smaller</em> is <em>better</em> for the
up-stream sorter.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>the_id</strong> (<em>str</em>) – Identifier that is being converted.</p></li>
<li><p><strong>ens_release</strong> (<em>int</em>) – Target Ensembl release; only synonyms active in this
release are considered.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><dl class="simple">
<dt><cite>[-ext, -form₁, -form₂]</cite> where</dt><dd><ul class="simple">
<li><p><cite>ext</cite>  - number of distinct <strong>external-database</strong> synonyms.</p></li>
<li><p><cite>form₁</cite> - number of distinct synonyms of the most important Ensembl form (typically <em>gene</em>).</p></li>
<li><p><cite>form₂</cite> - number of distinct synonyms of the second form (typically <em>transcript</em> or <em>translation</em>).</p></li>
</ul>
</dd>
</dl>
</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>list</p>
</dd>
<dt class="field-even">Raises<span class="colon">:</span></dt>
<dd class="field-even"><p><strong>ValueError</strong> – If the graph does not expose exactly the two expected
    non-backbone forms, or if a synonym node’s type cannot be mapped to
    <em>external</em>, <em>form₁</em>, or <em>form₂</em>.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="idtrack._track.Track.calculate_score_and_select">
<span class="sig-name descname"><span class="pre">calculate_score_and_select</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">all_possible_paths</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">reduction</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">remove_na</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">from_releases</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">to_release</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">score_of_the_queried_item</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">return_path</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">from_id</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/idtrack/_track.html#Track.calculate_score_and_select"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#idtrack._track.Track.calculate_score_and_select" title="Link to this definition"></a></dt>
<dd><p>Collapse a set of candidate paths into the single best path per target.</p>
<p>For each path produced by the search engine the function:</p>
<p>1. Computes an <strong>edge-score aggregate</strong> using <cite>reduction</cite> while handling
missing values as directed by <cite>remove_na</cite>.
2. Tallies <em>external</em> statistics (steps, jumps, initial conversion
confidence) and <em>assembly</em> statistics (number of priority drops, final
priority).
3. Packs all metrics into a dictionary and stores it under the key of the
path’s final destination node.
4. Keeps only the lexicographically “smallest” dictionary per destination
via <a class="reference internal" href="#idtrack._track.Track._path_score_sorter_single_target" title="idtrack._track.Track._path_score_sorter_single_target"><code class="xref py py-meth docutils literal notranslate"><span class="pre">_path_score_sorter_single_target()</span></code></a>.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>all_possible_paths</strong> (<em>tuple</em>) – Sequence of edge-lists representing
every admissible walk returned by the path-finder.</p></li>
<li><p><strong>reduction</strong> (<em>Callable</em>) – Function such as
<cite>np.mean</cite> or <cite>sum</cite> used to collapse edge weights into one number.</p></li>
<li><p><strong>remove_na</strong> (<em>str</em>) – How to treat <em>NaN</em> edge weights - one of
<cite>‘omit’</cite>, <cite>‘to_1’</cite>, <cite>‘to_0’</cite>.</p></li>
<li><p><strong>from_releases</strong> (<em>Iterable</em><em>[</em><em>int</em><em>]</em>) – Release that each path starts from;
must align with <cite>all_possible_paths</cite>.</p></li>
<li><p><strong>to_release</strong> (<em>int</em>) – Target release - needed to know whether an edge is
traversed forward or reverse.</p></li>
<li><p><strong>score_of_the_queried_item</strong> (<em>float</em>) – Fallback weight for the implicit edge
that represents the <em>query ID</em> itself.</p></li>
<li><p><strong>return_path</strong> (<em>bool</em>) – If <em>True</em>, embed the full edge-list inside each
score dict under the key <cite>‘the_path’</cite>.</p></li>
<li><p><strong>from_id</strong> (<em>str</em>) – Original identifier being converted - echoed back in the
score dict for traceability.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>Mapping
<cite>{destination_id → best_score_dict}</cite>.
Each <em>score dict</em> contains (inter alia)
<cite>assembly_jump</cite>, <cite>external_jump</cite>, <cite>external_step</cite>,
<cite>edge_scores_reduced</cite>, and <cite>ensembl_step</cite>.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>dict</p>
</dd>
<dt class="field-even">Raises<span class="colon">:</span></dt>
<dd class="field-even"><p><strong>ValueError</strong> – If an unexpected edge encoding is encountered, if an edge
    score is invalid/∞, or if <cite>remove_na</cite> is set to an unknown mode.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="idtrack._track.Track.choose_relevant_synonym">
<span class="sig-name descname"><span class="pre">choose_relevant_synonym</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">the_id</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">depth_max</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">to_release</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">filter_node_type</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">from_release</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/idtrack/_track.html#Track.choose_relevant_synonym"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#idtrack._track.Track.choose_relevant_synonym" title="Link to this definition"></a></dt>
<dd><p>Wrapper that <strong>discovers, clusters, and ranks</strong> synonymous Ensembl candidates for a given identifier.</p>
<p>The function performs three steps:</p>
<ol class="arabic simple">
<li><p><strong>Discover</strong> paths to <em>all</em> Ensembl-gene nodes that share the same
biological identity (<cite>synonymous_nodes</cite>).</p></li>
<li><p><strong>Cluster</strong> those paths by <strong>gene ID</strong> (ignoring version).</p></li>
<li><p><strong>Rank</strong> each cluster with
<a class="reference internal" href="#idtrack._track.Track._choose_relevant_synonym_helper" title="idtrack._track.Track._choose_relevant_synonym_helper"><code class="xref py py-meth docutils literal notranslate"><span class="pre">_choose_relevant_synonym_helper()</span></code></a>, selecting the entry release
(and direction) that best suits <cite>to_release</cite>.</p></li>
</ol>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>the_id</strong> (<em>str</em>) – Source identifier (Ensembl or external).</p></li>
<li><p><strong>depth_max</strong> (<em>int</em>) – Maximum depth passed to
<a class="reference internal" href="#idtrack._track.Track.synonymous_nodes" title="idtrack._track.Track.synonymous_nodes"><code class="xref py py-meth docutils literal notranslate"><span class="pre">synonymous_nodes()</span></code></a>; governs how far the synonym search is
allowed to roam through external nodes.</p></li>
<li><p><strong>to_release</strong> (<em>int</em>) – Target Ensembl release required by the overall
conversion.</p></li>
<li><p><strong>filter_node_type</strong> (<em>set</em><em>[</em><em>str</em><em>]</em>) – Node-types that the synonym search
must terminate on (usually <cite>{‘ensembl_gene’}</cite>).</p></li>
<li><p><strong>from_release</strong> (<em>int</em><em> | </em><em>None</em>) – Known active release of <cite>the_id</cite>.  If
<em>None</em>, the helper will infer one.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><p>A list whose elements are</p>
<p><cite>[synonym_id, entry_release, reverse, identifier_path, node_type_path]</cite></p>
<p>where the last two items reproduce the path returned by
<a class="reference internal" href="#idtrack._track.Track.synonymous_nodes" title="idtrack._track.Track.synonymous_nodes"><code class="xref py py-meth docutils literal notranslate"><span class="pre">synonymous_nodes()</span></code></a>.</p>
</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>list[list[Any]]</p>
</dd>
</dl>
<p class="rubric">Notes</p>
<p><em>The method purposefully keeps **all*</em> equally-ranked candidates;
further tie-breaking is deferred to the main path-scoring routine.*</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="idtrack._track.Track.convert">
<span class="sig-name descname"><span class="pre">convert</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">from_id</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">from_release=None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">to_release=None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">final_database=None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">reduction=&lt;function</span> <span class="pre">mean&gt;</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">remove_na='omit'</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">score_of_the_queried_item=nan</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">go_external=True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">prioritize_to_one_filter=False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">return_path=False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">deprioritize_lrg_genes=True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">return_ensembl_alternative=True</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/idtrack/_track.html#Track.convert"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#idtrack._track.Track.convert" title="Link to this definition"></a></dt>
<dd><p>End-to-end ID conversion workflow.</p>
<p>Starting from <cite>from_id</cite> the routine</p>
<ol class="arabic simple">
<li><p>Determines the correct <strong>time-travel direction</strong> if
<cite>from_release</cite> is unspecified.</p></li>
<li><p>Enumerates <em>all</em> admissible paths with
<a class="reference internal" href="#idtrack._track.Track.get_possible_paths" title="idtrack._track.Track.get_possible_paths"><code class="xref py py-meth docutils literal notranslate"><span class="pre">get_possible_paths()</span></code></a> (forward and/or reverse).</p></li>
<li><p>Collapses those paths with <a class="reference internal" href="#idtrack._track.Track.calculate_score_and_select" title="idtrack._track.Track.calculate_score_and_select"><code class="xref py py-meth docutils literal notranslate"><span class="pre">calculate_score_and_select()</span></code></a>.</p></li>
<li><p>Optionally converts the surviving Ensembl gene(s) into
<cite>final_database</cite> via <a class="reference internal" href="#idtrack._track.Track._final_conversion" title="idtrack._track.Track._final_conversion"><code class="xref py py-meth docutils literal notranslate"><span class="pre">_final_conversion()</span></code></a>.</p></li>
<li><p>Optionally applies a final global selection with
<a class="reference internal" href="#idtrack._track.Track._path_score_sorter_all_targets" title="idtrack._track.Track._path_score_sorter_all_targets"><code class="xref py py-meth docutils literal notranslate"><span class="pre">_path_score_sorter_all_targets()</span></code></a>.</p></li>
</ol>
<p>The output structure mirrors this decision tree and, when
<cite>return_path</cite> is <em>True</em>, embeds the full edge list so that callers
can audit every hop.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>from_id</strong> (<em>str</em>) – Source identifier (Ensembl, UniProt, RefSeq, …).</p></li>
<li><p><strong>from_release</strong> (<em>int</em><em> | </em><em>None</em>) – Starting Ensembl release.
<em>None</em> → infer from the graph.</p></li>
<li><p><strong>to_release</strong> (<em>int</em><em> | </em><em>None</em>) – Target Ensembl release.
Defaults to the newest release contained in the graph.</p></li>
<li><p><strong>final_database</strong> (<em>str</em><em> | </em><em>None</em>) – External database to
convert into.  <em>None</em> → stay on the Ensembl gene.</p></li>
<li><p><strong>reduction</strong> (<em>Callable</em>) – Function (e.g. <cite>numpy.mean</cite>)
used to collapse per-edge weights.  Must accept an iterable of
floats and return a float.</p></li>
<li><p><strong>remove_na</strong> (<em>str</em>) – Strategy for <em>NaN</em> edge weights -
<cite>‘omit’</cite>, <cite>‘to_1’</cite>, or <cite>‘to_0’</cite>.</p></li>
<li><p><strong>score_of_the_queried_item</strong> (<em>float</em>) – Weight assigned to
the implicit edge that represents <cite>from_id</cite> itself.</p></li>
<li><p><strong>go_external</strong> (<em>bool</em>) – Allow jumps through external
databases when the backbone is disconnected.</p></li>
<li><p><strong>prioritize_to_one_filter</strong> (<em>bool</em>) – After all scoring,
keep only the single globally best target.</p></li>
<li><p><strong>return_path</strong> (<em>bool</em>) – Embed the full edge list(s) in the
returned dictionary.</p></li>
<li><p><strong>deprioritize_lrg_genes</strong> (<em>bool</em>) – If <em>True</em> and other
results exist, drop LRG_* genomic regions from the final set.</p></li>
<li><p><strong>return_ensembl_alternative</strong> (<em>bool</em>) – When converting to an
external database, also return the Ensembl gene as a fallback.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><ul class="simple">
<li><p><strong>dict</strong> - Structured result as described above.</p></li>
<li><p><strong>None</strong> - No admissible path was found.</p></li>
</ul>
</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>dict | None</p>
</dd>
<dt class="field-even">Raises<span class="colon">:</span></dt>
<dd class="field-even"><p><strong>ValueError</strong> – For non-callable <cite>reduction</cite>, unsupported
    <cite>remove_na</cite> modes, unknown <cite>final_database</cite> values, or
    logical inconsistencies detected during processing.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="idtrack._track.Track.convert_optimized_multiple">
<span class="sig-name descname"><span class="pre">convert_optimized_multiple</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/idtrack/_track.html#Track.convert_optimized_multiple"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#idtrack._track.Track.convert_optimized_multiple" title="Link to this definition"></a></dt>
<dd><p>Placeholder for a <strong>batch-optimised</strong> converter.</p>
<p>The intended behaviour is to accept <em>multiple</em> query IDs and choose a
conversion target for each such that cross-sample clashes (e.g.
duplicate loci) are minimised.</p>
<dl class="field-list simple">
<dt class="field-odd">Raises<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>NotImplementedError</strong> – Always - the optimisation strategy is not yet implemented.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="idtrack._track.Track.edge_key_orientor">
<span class="sig-name descname"><span class="pre">edge_key_orientor</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">n1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">n2</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">n3</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/idtrack/_track.html#Track.edge_key_orientor"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#idtrack._track.Track.edge_key_orientor" title="Link to this definition"></a></dt>
<dd><p>Return the stored orientation of a multigraph edge.</p>
<p>For multigraphs every logical edge is stored <em>once</em>, but the caller may
hold <cite>(u, v, k)</cite> or <cite>(v, u, k)</cite>.
This helper resolves the ambiguity so that subsequent attribute look-ups
succeed.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>n1</strong> (<em>str</em>) – One endpoint of the edge.</p></li>
<li><p><strong>n2</strong> (<em>str</em>) – The other endpoint.</p></li>
<li><p><strong>n3</strong> (<em>int</em>) – Edge key (index) within the <cite>networkx</cite> multi-edge.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>A triple that is guaranteed to exist as written
in <cite>self.graph</cite>.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>tuple[str, str, int]</p>
</dd>
<dt class="field-even">Raises<span class="colon">:</span></dt>
<dd class="field-even"><p><strong>AssertionError</strong> – If neither orientation is present in the graph.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="idtrack._track.Track.get_from_release_and_reverse_vars">
<em class="property"><span class="pre">static</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">get_from_release_and_reverse_vars</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">lor</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">p</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">mode</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/idtrack/_track.html#Track.get_from_release_and_reverse_vars"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#idtrack._track.Track.get_from_release_and_reverse_vars" title="Link to this definition"></a></dt>
<dd><p>Derive a list of <cite>(release, reverse)</cite> tuples.</p>
<p>Derive a list of <cite>(release, reverse)</cite> tuples that indicate <em>which</em>
Ensembl release to start the graph walk from and <em>whether</em> that walk
should move <strong>backwards in time</strong>.</p>
<p>Given a collection of active-range intervals <cite>lor</cite> and a pivot
release <cite>p</cite>, the algorithm selects one or two release points per
interval depending on <cite>mode</cite>:</p>
<ul class="simple">
<li><dl class="simple">
<dt>‘closest’ - choose the release <strong>nearest</strong> to <cite>p</cite> within or at</dt><dd><p>the ends of the interval.</p>
</dd>
</dl>
</li>
<li><dl class="simple">
<dt>‘distant’ - choose the release <strong>farthest</strong> from <cite>p</cite> within the</dt><dd><p>interval.</p>
</dd>
</dl>
</li>
</ul>
<p>The boolean in each tuple is <em>True</em> when the walk should start <strong>after</strong>
the selected release and move backwards (i.e. “reverse mode”), and
<em>False</em> when it should move forwards.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>lor</strong> (<em>list</em>) – List of inclusive
<cite>(first_release, last_release)</cite> intervals in ascending order.</p></li>
<li><p><strong>p</strong> (<em>int</em>) – Pivot release around which “closest” or “distant” is
evaluated.</p></li>
<li><p><strong>mode</strong> (<em>str</em>) – Either <cite>‘closest’</cite> or <cite>‘distant’</cite>.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>Release / reverse-flag pairs, ordered in
the sequence they should be tried by the path-finder.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>list[tuple[int, bool]]</p>
</dd>
<dt class="field-even">Raises<span class="colon">:</span></dt>
<dd class="field-even"><p><strong>ValueError</strong> – If an interval in <cite>lor</cite> is malformed, <cite>mode</cite> is
    not recognised, or internal consistency checks fail.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="idtrack._track.Track.get_next_edges">
<span class="sig-name descname"><span class="pre">get_next_edges</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">from_id</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">from_release</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">reverse</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">debugging</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/idtrack/_track.html#Track.get_next_edges"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#idtrack._track.Track.get_next_edges" title="Link to this definition"></a></dt>
<dd><p>Enumerate <strong>chronologically admissible</strong> history edges from a node.</p>
<p>Starting at <cite>from_id</cite> and release <cite>from_release</cite>, the method scans
outgoing (or incoming, when <cite>reverse</cite> is <em>True</em>) edges whose
timestamps allow the path to <strong>advance</strong> in the desired temporal
direction.  It collapses duplicate “same-ID” transitions and flags
self-loops so that later heuristics can treat branch points and tips
differently.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>from_id</strong> (<em>str</em>) – Current node from which the search will step.</p></li>
<li><p><strong>from_release</strong> (<em>int</em>) – Release at which the current node is known to
exist.</p></li>
<li><p><strong>reverse</strong> (<em>bool</em>) – <em>False</em> to walk <strong>forward</strong> in history
(old → new), <em>True</em> to walk <strong>backward</strong> (new → old).</p></li>
<li><p><strong>debugging</strong> (<em>bool</em>) – If set, disables the duplicate-edge
collapse so that unit tests can inspect the raw edge set.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><p>Sorted list of edge
descriptors, each of which is</p>
<p><cite>[edge_release, is_self_loop, src_node, dst_node, multiedge_key]</cite>.</p>
</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>list[list[Union[int, bool, str, int]]]</p>
</dd>
<dt class="field-even">Raises<span class="colon">:</span></dt>
<dd class="field-even"><p><strong>ValueError</strong> – If inconsistent multi-edges (same nodes, same release)
    are detected—this signals a corrupted graph build.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="idtrack._track.Track.get_possible_paths">
<span class="sig-name descname"><span class="pre">get_possible_paths</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">from_id</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">from_release</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">to_release</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">reverse</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">go_external</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">increase_depth_until</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">2</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">increase_jump_until</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">from_release_inferred</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/idtrack/_track.html#Track.get_possible_paths"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#idtrack._track.Track.get_possible_paths" title="Link to this definition"></a></dt>
<dd><p>Run <a class="reference internal" href="#idtrack._track.Track.path_search" title="idtrack._track.Track.path_search"><code class="xref py py-meth docutils literal notranslate"><span class="pre">path_search()</span></code></a> under <strong>progressively relaxed settings</strong>.</p>
<p>Run <a class="reference internal" href="#idtrack._track.Track.path_search" title="idtrack._track.Track.path_search"><code class="xref py py-meth docutils literal notranslate"><span class="pre">path_search()</span></code></a> under <strong>progressively relaxed settings</strong> until at
least one viable path is found—or every relaxation level is exhausted.</p>
<p>Four search stages are attempted in order:</p>
<ol class="arabic simple">
<li><p><strong>Backbone-only</strong> - external jumps disabled.</p></li>
<li><dl class="simple">
<dt><strong>External enabled</strong> - allow external jumps; increment synonym depth</dt><dd><p>and jump limit after each failure up to
<cite>increase_depth_until</cite>/<cite>increase_jump_until</cite>.</p>
</dd>
</dl>
</li>
<li><dl class="simple">
<dt><strong>Backbone with multiple-Ensembl transition</strong> - external disabled but</dt><dd><p>permit starting release to shift on external nodes.</p>
</dd>
</dl>
</li>
<li><dl class="simple">
<dt><strong>External + multiple-transition</strong> - most permissive search, with</dt><dd><p>iterative depth/jump relaxation as in stage 2.</p>
</dd>
</dl>
</li>
</ol>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>from_id</strong> (<em>str</em>) – Identifier to convert.</p></li>
<li><p><strong>from_release</strong> (<em>int</em>) – Release at which the search begins.</p></li>
<li><p><strong>to_release</strong> (<em>int</em>) – Desired target release.</p></li>
<li><p><strong>reverse</strong> (<em>bool</em>) – Traverse the Ensembl history backwards if <em>True</em>,
forwards otherwise.</p></li>
<li><p><strong>go_external</strong> (<em>bool</em>) – If <em>False</em>, skip any stage that
requires external jumps.</p></li>
<li><p><strong>increase_depth_until</strong> (<em>int</em>) – Additional synonym-search
depth to allow beyond the default.</p></li>
<li><p><strong>increase_jump_until</strong> (<em>int</em>) – Additional external-jump count
to allow beyond the default.</p></li>
<li><p><strong>from_release_inferred</strong> (<em>bool</em>) – <em>Reserved for future use.</em>
Indicates that <em>from_release</em> was chosen automatically rather
than provided by the user.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>All paths discovered by the most restrictive stage that yielded
<strong>at least one</strong> result, returned as an immutable tuple.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>tuple[tuple[tuple[str, str, int]]]</p>
</dd>
</dl>
<p class="rubric">Notes</p>
<p>The function copies and mutates
<code class="xref py py-attr docutils literal notranslate"><span class="pre">DB.external_search_settings</span></code> internally; the caller’s
copy is not modified.</p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="idtrack._track.Track.identify_source">
<span class="sig-name descname"><span class="pre">identify_source</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">dataset_ids</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">mode</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/idtrack/_track.html#Track.identify_source"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#idtrack._track.Track.identify_source" title="Link to this definition"></a></dt>
<dd><p>Infer the most likely origin (assembly and/or Ensembl release) of a heterogeneous identifier list.</p>
<p>The function tallies how often each origin triple appears among
<cite>dataset_ids</cite> and returns the counts sorted in descending order.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>dataset_ids</strong> (<em>list</em><em>[</em><em>str</em><em>]</em>) – Collection of identifiers to analyse.</p></li>
<li><p><strong>mode</strong> (<em>str</em>) – Granularity of the origin to extract - one of
- ‘complete’                  → (assembly, db, release)
- ‘ensembl_release’           → release only
- ‘assembly’                  → assembly only
- ‘assembly_ensembl_release’  → (assembly, release)</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>Pairs <cite>(origin, count)</cite> sorted by
frequency.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>list[tuple[Any, int]]</p>
</dd>
<dt class="field-even">Raises<span class="colon">:</span></dt>
<dd class="field-even"><p><strong>ValueError</strong> – If <cite>mode</cite> is not one of the recognised values.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="idtrack._track.Track.minimum_assembly_jumps">
<span class="sig-name descname"><span class="pre">minimum_assembly_jumps</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">the_path</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">step_pri</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">current_priority</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/idtrack/_track.html#Track.minimum_assembly_jumps"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#idtrack._track.Track.minimum_assembly_jumps" title="Link to this definition"></a></dt>
<dd><p>Compute the penalty incurred by assembly downgrades along a path.</p>
<p>Each path step may be annotated with one or more candidate assemblies.
These are translated into priority values via
<cite>DB.assembly_mysqlport_priority</cite>.  The algorithm walks the path,
tracking the current priority and counting how many times it must drop
to a lower priority value—each drop constitutes an “assembly jump”
penalty.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>the_path</strong> (<em>Iterable</em><em>[</em><em>tuple</em><em>]</em>) – Sequence of edge descriptors; each element is
either <cite>(n1, n2, k)</cite> or <cite>(n1, n2, k, release)</cite>.</p></li>
<li><p><strong>step_pri</strong> (<em>list</em><em>[</em><em>int</em><em>] </em><em>| </em><em>None</em><em>, </em><em>optional</em>) – Priority list for the first edge.
If None, it is derived from <cite>the_path</cite>.</p></li>
<li><p><strong>current_priority</strong> (<em>int</em><em> | </em><em>None</em><em>, </em><em>optional</em>) – Starting priority.  If None,
initialised to <cite>max(step_pri)</cite>.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><ul class="simple">
<li><p><cite>assembly_jump</cite> - total number of priority drops.</p></li>
<li><p><cite>step_pri</cite> - priority list of the last processed edge.</p></li>
<li><p><cite>current_priority</cite> - priority value after the final edge.</p></li>
</ul>
</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>tuple[int, list[int], int]</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="idtrack._track.Track.path_search">
<span class="sig-name descname"><span class="pre">path_search</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">from_id</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">from_release</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">to_release</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">reverse</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">external_settings</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">external_jump</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">multiple_ensembl_transition</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/idtrack/_track.html#Track.path_search"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#idtrack._track.Track.path_search" title="Link to this definition"></a></dt>
<dd><p>Enumerate every admissible <strong>history path</strong> from <em>from_id</em> at <em>from_release</em> to <em>to_release</em>.</p>
<p>The algorithm performs a depth-first traversal of the Ensembl history
edges.  Whenever it becomes stranded on a non-backbone node it may
“beam-up” via a <em>synonym path</em> through an external database, subject to
the constraints in <cite>external_settings</cite>:</p>
<ul class="simple">
<li><p><cite>synonymous_max_depth</cite> - maximum depth of a synonym search.</p></li>
<li><p><cite>jump_limit</cite> - maximum number of external “beam-up” jumps allowed.</p></li>
<li><p><cite>nts_backbone</cite> - canonical node-type of the Ensembl backbone.</p></li>
</ul>
<p>Additional flags control the initial conditions:</p>
<ul class="simple">
<li><dl class="simple">
<dt>Setting <cite>external_jump</cite> to <em>np.inf</em> <strong>disables</strong> external jumps.</dt><dd><p>Setting it to <em>None</em> <strong>enables</strong> them with the counter reset to <cite>0</cite>.</p>
</dd>
</dl>
</li>
<li><dl class="simple">
<dt><cite>multiple_ensembl_transition</cite> allows the algorithm to time-travel to</dt><dd><p>a <em>different</em> release while still on an external node; this is useful
when <em>from_release</em> was inferred and might not actually connect.</p>
</dd>
</dl>
</li>
</ul>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>from_id</strong> (<em>str</em>) – Identifier to start the search from.</p></li>
<li><p><strong>from_release</strong> (<em>int</em>) – Release number where <em>from_id</em> is considered
active.</p></li>
<li><p><strong>to_release</strong> (<em>int</em>) – Target Ensembl release.</p></li>
<li><p><strong>reverse</strong> (<em>bool</em>) – If <em>True</em>, traverse the graph <strong>backwards</strong> in time;
otherwise forwards.</p></li>
<li><p><strong>external_settings</strong> (<em>dict</em>) – Copy of
<code class="xref py py-attr docutils literal notranslate"><span class="pre">DB.external_search_settings</span></code> that governs depth,
jump limits, and backbone node-type.</p></li>
<li><p><strong>external_jump</strong> (<em>float</em><em> | </em><em>None</em>) – Current external-jump count
(<em>None</em> starts from zero, <em>np.inf</em> forbids any jump).</p></li>
<li><p><strong>multiple_ensembl_transition</strong> (<em>bool</em>) – Permit the synonym
engine to select a <em>different</em> release for an external node when
no path exists at <em>from_release</em>.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>A set of <strong>edge-lists</strong>.  Each edge is stored as
<cite>(src, dst, key)</cite>; an empty walk that terminates immediately
is represented by <cite>((None, from_id, None),)</cite>.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>set[tuple[tuple[str, str, int]]]</p>
</dd>
</dl>
<p class="rubric">Notes</p>
<p><em>The method is intentionally side-effect free; it constructs all
intermediate data on the stack and returns a fresh set.</em></p>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="idtrack._track.Track.path_step_possible_assembly_jumps">
<span class="sig-name descname"><span class="pre">path_step_possible_assembly_jumps</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">n1</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">n2</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">n3</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">n4</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/idtrack/_track.html#Track.path_step_possible_assembly_jumps"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#idtrack._track.Track.path_step_possible_assembly_jumps" title="Link to this definition"></a></dt>
<dd><p>Return the genome assemblies that can legally be used for a single edge.</p>
<p>The helper inspects the edge that connects <cite>n1</cite> → <cite>n2</cite> and filters the
assemblies recorded on that edge against the <em>release</em> constraint <cite>n4</cite>:</p>
<ul class="simple">
<li><dl class="simple">
<dt><strong>None</strong> - the edge is treated as backbone history; the result is the</dt><dd><p>graph-wide default assembly (usually the build on which the backbone was
constructed).</p>
</dd>
</dl>
</li>
<li><p><strong>int</strong> - keep only assemblies whose <em>release set</em> contains that single release.</p></li>
<li><p><strong>set[int]</strong> - keep assemblies whose release set intersects the provided set.</p></li>
</ul>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>n1</strong> (<em>str</em>) – Source node identifier.</p></li>
<li><p><strong>n2</strong> (<em>str</em>) – Destination node identifier.</p></li>
<li><p><strong>n3</strong> (<em>int</em>) – Edge key within the NetworkX multigraph.</p></li>
<li><p><strong>n4</strong> (<em>int</em><em> | </em><em>set</em><em>[</em><em>int</em><em>] </em><em>| </em><em>None</em><em>, </em><em>optional</em>) – Release filter as described above.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>Sorted list of assembly names (e.g. <cite>[“GRCh37”, “GRCh38”]</cite>).</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>list[str]</p>
</dd>
<dt class="field-even">Raises<span class="colon">:</span></dt>
<dd class="field-even"><p><strong>ValueError</strong> – If <cite>n4</cite> is of an unsupported type.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="idtrack._track.Track.should_graph_reversed">
<span class="sig-name descname"><span class="pre">should_graph_reversed</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">from_id</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">to_release</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/idtrack/_track.html#Track.should_graph_reversed"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#idtrack._track.Track.should_graph_reversed" title="Link to this definition"></a></dt>
<dd><p>Determine the <strong>temporal orientation</strong> of the graph walk.</p>
<p>Given an identifier that is active in one or more release intervals,
the routine decides whether the subsequent path-finder must move
<strong>forward in time</strong>, <strong>backward in time</strong>, or explore <em>both</em> directions
in parallel in order to reach the target release.</p>
<p>The decision is based on the <em>closest</em> boundary of every active
interval returned by
<a class="reference internal" href="#idtrack._track.Track.get_from_release_and_reverse_vars" title="idtrack._track.Track.get_from_release_and_reverse_vars"><code class="xref py py-meth docutils literal notranslate"><span class="pre">Track.get_from_release_and_reverse_vars()</span></code></a> (<cite>mode=’closest’</cite>).</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>from_id</strong> (<em>str</em>) – The starting identifier (Ensembl gene, transcript,
protein, or external ID).</p></li>
<li><p><strong>to_release</strong> (<em>int</em>) – The Ensembl release the user wishes to convert
<em>to</em>.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><ul class="simple">
<li><dl class="simple">
<dt><cite>‘forward’</cite> - walk old → new, starting at the <strong>earliest</strong></dt><dd><p>release in which <em>from_id</em> is active
&amp;nbsp;&amp;nbsp;&amp;nbsp;→ return <cite>(‘forward’, start_release)</cite></p>
</dd>
</dl>
</li>
<li><dl class="simple">
<dt><cite>’reverse’</cite> - walk new → old, starting at the <strong>latest</strong></dt><dd><p>active release
&amp;nbsp;&amp;nbsp;&amp;nbsp;→ return <cite>(‘reverse’, start_release)</cite></p>
</dd>
</dl>
</li>
<li><dl class="simple">
<dt><cite>’both’</cite> - split search: one forward walk and one reverse</dt><dd><p>walk
&amp;nbsp;&amp;nbsp;&amp;nbsp;→ return <cite>(‘both’, (forward_start,
reverse_start))</cite></p>
</dd>
</dl>
</li>
</ul>
</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>tuple[str, Union[int, tuple[int, int]]]</p>
</dd>
<dt class="field-even">Raises<span class="colon">:</span></dt>
<dd class="field-even"><p><strong>ValueError</strong> – If <em>from_id</em> is never active in or around
    <cite>to_release</cite> (i.e. no viable starting release can be found).</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="idtrack._track.Track.synonymous_nodes">
<span class="sig-name descname"><span class="pre">synonymous_nodes</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">the_id</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">depth_max</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">filter_node_type</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">from_release</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ensembl_backbone_shallow_search</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">account_for_hyperconnected_nodes</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/idtrack/_track.html#Track.synonymous_nodes"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#idtrack._track.Track.synonymous_nodes" title="Link to this definition"></a></dt>
<dd><p>Public wrapper around <a class="reference internal" href="#idtrack._track.Track._recursive_synonymous" title="idtrack._track.Track._recursive_synonymous"><code class="xref py py-meth docutils literal notranslate"><span class="pre">_recursive_synonymous()</span></code></a>.</p>
<p>The method returns <strong>all minimal-length synonym paths</strong> emanating from <cite>the_id</cite>.</p>
<p>The function first runs a <em>default</em> depth search determined by
<cite>DB.external_search_settings[‘synonymous_max_depth’]</cite>.
If no synonym is found and <cite>depth_max</cite> is greater than that default,
a second, deeper search is attempted.</p>
<p>For every distinct <em>target</em> node the shortest path is kept; longer
paths to the same target are discarded.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>the_id</strong> (<em>str</em>) – Source identifier.</p></li>
<li><p><strong>depth_max</strong> (<em>int</em>) – Maximum search depth to try if the default search fails.</p></li>
<li><p><strong>filter_node_type</strong> (<em>set</em><em>[</em><em>str</em><em>]</em>) – Node-types that are acceptable for the
<em>target</em> node(s).  Must <em>not</em> include the generic
<cite>‘external’</cite> type—specify the concrete external DB instead.</p></li>
<li><p><strong>from_release</strong> (<em>int</em><em> | </em><em>None</em>) – Constrain targets to those active in this Ensembl release.</p></li>
<li><p><strong>ensembl_backbone_shallow_search</strong> (<em>bool</em>) – If <em>True</em>,
restricts the graph traversal as explained in <a class="reference internal" href="#idtrack._track.Track._recursive_synonymous" title="idtrack._track.Track._recursive_synonymous"><code class="xref py py-meth docutils literal notranslate"><span class="pre">_recursive_synonymous()</span></code></a>.</p></li>
<li><p><strong>account_for_hyperconnected_nodes</strong> (<em>bool</em>) – Todo.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><dl class="simple">
<dt>A list whose elements are <cite>[identifier_path, node_type_path]</cite> pairs,</dt><dd><p>each representing the minimal route to one synonymous node.</p>
</dd>
</dl>
</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>list[list[list[str]]]</p>
</dd>
<dt class="field-even">Raises<span class="colon">:</span></dt>
<dd class="field-even"><p><strong>ValueError</strong> – If <cite>filter_node_type</cite> improperly contains the generic
    external type, or if <cite>depth_max</cite> is incompatible with <cite>ensembl_backbone_shallow_search</cite>.</p>
</dd>
</dl>
</dd></dl>

</dd></dl>

<dl class="py class" id="module-idtrack._track_tests">
<dt class="sig sig-object py" id="idtrack._track_tests.TrackTests">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">TrackTests</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="o"><span class="pre">*</span></span><span class="n"><span class="pre">args</span></span></em>, <em class="sig-param"><span class="o"><span class="pre">**</span></span><span class="n"><span class="pre">kwargs</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/idtrack/_track_tests.html#TrackTests"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#idtrack._track_tests.TrackTests" title="Link to this definition"></a></dt>
<dd><p>Bases: <a class="reference internal" href="#idtrack._track.Track" title="idtrack._track.Track"><code class="xref py py-class docutils literal notranslate"><span class="pre">Track</span></code></a>, <code class="xref py py-class docutils literal notranslate"><span class="pre">ABC</span></code></p>
<p>Developer-facing integrity-test harness for <code class="xref py py-class docutils literal notranslate"><span class="pre">Track</span></code>.</p>
<p>This module defines <a class="reference internal" href="#idtrack._track_tests.TrackTests" title="idtrack._track_tests.TrackTests"><code class="xref py py-class docutils literal notranslate"><span class="pre">TrackTests</span></code></a>, a mix-in that adds an extensive
white-box test suite to a populated <code class="xref py py-class docutils literal notranslate"><span class="pre">idtrack.Track</span></code> instance.  The
class is <strong>for developers only</strong>; it should never be used in production
pipelines.  Every public method beginning with <code class="docutils literal notranslate"><span class="pre">is_</span></code> returns a boolean
that tells whether a specific invariant holds.  Methods beginning with
<code class="docutils literal notranslate"><span class="pre">history_</span></code> execute heavier, end-to-end conversions and collect rich
statistics. The class is intended to be <em>mixin-ed</em> into a concrete Track
subclass—or instantiated standalone—<strong>after</strong> the underlying graph and
lookup tables have been fully built.  It performs purely read-only
operations and therefore imposes no risk of mutating state.</p>
<p>All test methods share the following contract:</p>
<ul class="simple">
<li><p>They never raise on failure—<em>return-value only</em>—so they can be run in
bulk without interrupting your session.</p></li>
<li><p>A return value of <code class="docutils literal notranslate"><span class="pre">True</span></code> means the invariant holds; <code class="docutils literal notranslate"><span class="pre">False</span></code> means
a violation was detected.</p></li>
<li><p>Where useful, a <em>verbose</em> flag gives a tqdm progress bar so long-
running checks remain user-friendly.</p></li>
</ul>
<p>Typical use:</p>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span><span class="n">tests</span> <span class="o">=</span> <span class="n">TrackTests</span><span class="p">(</span><span class="o">...</span><span class="p">)</span>
<span class="n">tests</span><span class="o">.</span><span class="n">is_id_functions_consistent_ensembl</span><span class="p">()</span>  <span class="c1"># Raises if inconsistent.</span>
</pre></div>
</div>
<div class="admonition note">
<p class="admonition-title">Note</p>
<p>The class is <em>not</em> designed for production; instantiate it only in test
suites or interactive debugging sessions.</p>
</div>
<p>Initialize the test harness.</p>
<p>All positional and keyword arguments are forwarded verbatim to
<code class="xref py py-class docutils literal notranslate"><span class="pre">__init__</span></code>.  Besides constructing the underlying
graph, the initializer sets up a dedicated <code class="xref py py-data docutils literal notranslate"><span class="pre">logging.Logger</span></code>
named <code class="docutils literal notranslate"><span class="pre">&quot;track_tests&quot;</span></code> so individual test routines can emit structured
diagnostics without polluting the main application log.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>args</strong> – Positional arguments accepted by <code class="xref py py-class docutils literal notranslate"><span class="pre">__init__</span></code>.</p></li>
<li><p><strong>kwargs</strong> – Keyword arguments accepted by <code class="xref py py-class docutils literal notranslate"><span class="pre">__init__</span></code>.</p></li>
</ul>
</dd>
</dl>
<dl class="py method">
<dt class="sig sig-object py" id="idtrack._track_tests.TrackTests.format_history_travel_testing_report">
<span class="sig-name descname"><span class="pre">format_history_travel_testing_report</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">res</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">include_header</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">line_separation_at_end</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/idtrack/_track_tests.html#TrackTests.format_history_travel_testing_report"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#idtrack._track_tests.TrackTests.format_history_travel_testing_report" title="Link to this definition"></a></dt>
<dd><p>Todo.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>res</strong> (<em>dict</em><em>[</em><em>str</em><em>, </em><em>Any</em><em>]</em>) – Todo.</p></li>
<li><p><strong>include_header</strong> (<em>bool</em>) – Todo. Defaults to False.</p></li>
<li><p><strong>line_separation_at_end</strong> (<em>bool</em>) – Todo.. Defaults to True.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>Todo.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>str</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="idtrack._track_tests.TrackTests.format_history_travel_testing_report_header">
<span class="sig-name descname"><span class="pre">format_history_travel_testing_report_header</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">p</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/idtrack/_track_tests.html#TrackTests.format_history_travel_testing_report_header"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#idtrack._track_tests.TrackTests.format_history_travel_testing_report_header" title="Link to this definition"></a></dt>
<dd><p>Todo.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>p</strong> (<em>dict</em><em>[</em><em>str</em><em>, </em><em>Any</em><em>]</em>) – Todo.</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>Todo.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>str</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="idtrack._track_tests.TrackTests.history_travel_testing">
<span class="sig-name descname"><span class="pre">history_travel_testing</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">from_release</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">from_assembly</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">from_database</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">to_release</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">to_database</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">go_external</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">prioritize_to_one_filter</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">convert_using_release</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">from_fraction</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1.0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">verbose</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">verbose_detailed</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">return_ensembl_alternative</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/idtrack/_track_tests.html#TrackTests.history_travel_testing"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#idtrack._track_tests.TrackTests.history_travel_testing" title="Link to this definition"></a></dt>
<dd><p>Run an end-to-end Ensembl-history conversion and collect granular QA metrics.</p>
<p>The routine samples identifiers from <em>from_database</em>/<em>from_release</em>
(optionally down-sampling via <em>from_fraction</em>) and converts each one to
<em>to_database</em>/<em>to_release</em> using <code class="xref py py-meth docutils literal notranslate"><span class="pre">idtrack.Track.convert()</span></code>.  It
is intentionally <strong>non-fatal</strong>: every failure mode is caught, logged
and tallied so that large regression suites can run unattended.  All
results are returned in a single nested <em>metrics</em> dictionary whose
structure mirrors the printable report produced by
<a class="reference internal" href="#idtrack._track_tests.TrackTests.format_history_travel_testing_report" title="idtrack._track_tests.TrackTests.format_history_travel_testing_report"><code class="xref py py-meth docutils literal notranslate"><span class="pre">format_history_travel_testing_report()</span></code></a>.</p>
<p>The statistics fall into <strong>four conceptual groups</strong>; each counter not only records an absolute event
count but also serves as a red-flag indicator for specific classes of mapping pathology.  Use the
guidelines below to interpret the numbers and decide whether a run is <em>healthy</em>, <em>questionable</em>, or
<em>action-required</em>.</p>
<ul>
<li><p><strong>Failure / anomaly counters</strong></p>
<ul class="simple">
<li><p><code class="docutils literal notranslate"><span class="pre">history_voyage_failed_gracefully</span></code> - the converter raised <code class="xref py py-data docutils literal notranslate"><span class="pre">EmptyConversionMetricsError</span></code>.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">history_voyage_failed_unknown</span></code> - any <em>other</em> unexpected exception.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">query_not_in_the_graph</span></code> - source ID absent from the graph.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">lost_item</span></code> - traversal finished but produced <strong>no</strong> final IDs.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">lost_item_but_the_same_id_exists</span></code> - special case of <em>lost_item</em>
when the target and source DB are both Ensembl-gene and the target
ID still exists in the graph.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">found_ids_not_accurate</span></code> - at least one returned target ID is not
part of the authoritative <em>ids_to</em> reference set.</p></li>
</ul>
</li>
<li><p><strong>Mapping quality</strong></p>
<ul class="simple">
<li><p><code class="docutils literal notranslate"><span class="pre">one_to_one_ids</span></code> - queries that resolved to <em>exactly</em> one target ID.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">one_to_multiple_ids</span></code> - queries with &gt; 1 admissible targets.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">one_to_multiple_final_conversion</span></code> - subset of the above where
<em>exactly one</em> traversal path was found (heuristics eliminated alternatives).</p></li>
</ul>
</li>
<li><p><strong>Collision analysis</strong></p>
<p>The list <code class="docutils literal notranslate"><span class="pre">clashing_id_type</span> <span class="pre">==</span> <span class="pre">[clash_one_one,</span> <span class="pre">clash_multi_multi,</span>
<span class="pre">clash_multi_one]</span></code> classifies <em>target IDs</em> that were reached by
more than one query:</p>
<ul class="simple">
<li><p><code class="docutils literal notranslate"><span class="pre">clash_one_one</span></code> - every colliding query was 1→1.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">clash_multi_multi</span></code> - every colliding query was 1→many.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">clash_multi_one</span></code> - mixture of 1→1 and 1→many queries (most
alarming category).</p></li>
</ul>
</li>
<li><p><strong>Timings &amp; book-keeping</strong></p>
<ul class="simple">
<li><p><code class="docutils literal notranslate"><span class="pre">time</span></code> - wall-clock runtime in seconds.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">conversion</span></code> - per-query mapping result for <em>all</em> successful traversals.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">converted_item_dict</span></code> / <code class="docutils literal notranslate"><span class="pre">converted_item_dict_reversed</span></code> - raw
per-ID caches used to derive the higher-level counters above.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">parameters</span></code> - echo of the function arguments.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">ids</span></code> - the sampled <em>from</em> and reference <em>to</em> ID sets.</p></li>
</ul>
</li>
</ul>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>from_release</strong> (<em>int</em>) – Ensembl release number of the <em>source</em> IDs.</p></li>
<li><p><strong>from_assembly</strong> (<em>int</em>) – Genome assembly code of the source IDs.</p></li>
<li><p><strong>from_database</strong> (<em>str</em>) – Node-type / database of the source IDs.</p></li>
<li><p><strong>to_release</strong> (<em>int</em>) – Ensembl release number of the <em>target</em> IDs.</p></li>
<li><p><strong>to_database</strong> (<em>str</em>) – Node-type / database to convert <strong>into</strong>.</p></li>
<li><p><strong>go_external</strong> (<em>bool</em>) – Permit temporary detours through external IDs when native Ensembl history edges break.</p></li>
<li><p><strong>prioritize_to_one_filter</strong> (<em>bool</em>) – Prefer 1→1 mappings over 1→many when multiple paths exist.</p></li>
<li><p><strong>convert_using_release</strong> (<em>bool</em>) – Pass <em>from_release</em> straight into
<code class="xref py py-meth docutils literal notranslate"><span class="pre">idtrack.Track.convert()</span></code> instead of letting it infer the starting point.</p></li>
<li><p><strong>from_fraction</strong> (<em>float</em>) – Fraction (0 &lt; x ≤ 1) of the
<em>ids_from</em> population to sample; speeds up smoke tests.</p></li>
<li><p><strong>verbose</strong> (<em>bool</em>) – Show tqdm progress bar (coarse).</p></li>
<li><p><strong>verbose_detailed</strong> (<em>bool</em>) – Embed live metric counters in the tqdm postfix.</p></li>
<li><p><strong>return_ensembl_alternative</strong> (<em>bool</em>) – Forwarded to <code class="xref py py-meth docutils literal notranslate"><span class="pre">idtrack.Track.convert()</span></code>.</p></li>
</ul>
</dd>
<dt class="field-even">Raises<span class="colon">:</span></dt>
<dd class="field-even"><p><strong>ValueError</strong> – If either database argument refers to an Ensembl
    node-type (must use backbone helpers instead) or if
    <em>from_fraction</em> is outside the open interval (0, 1].</p>
</dd>
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p><dl class="simple">
<dt>Nested <em>metrics</em> dictionary with the layout described above.</dt><dd><p>Use <a class="reference internal" href="#idtrack._track_tests.TrackTests.format_history_travel_testing_report" title="idtrack._track_tests.TrackTests.format_history_travel_testing_report"><code class="xref py py-meth docutils literal notranslate"><span class="pre">format_history_travel_testing_report()</span></code></a> for a
human-readable summary.</p>
</dd>
</dl>
</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p>dict</p>
</dd>
</dl>
<p class="rubric">Notes</p>
<ul class="simple">
<li><p>All counters are <strong>absolute counts</strong> - divide by
<code class="docutils literal notranslate"><span class="pre">len(metrics['ids']['from'])</span></code> to obtain rates.</p></li>
<li><p>The collision analysis is inspired by the <em>clash statistics</em>
logic implemented at the end of the function and helps spot
discrepant “unique” IDs that suddenly become ambiguous. Keeping
all three clash counters at zero is the gold standard for a
healthy build.</p></li>
</ul>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="idtrack._track_tests.TrackTests.history_travel_testing_random">
<span class="sig-name descname"><span class="pre">history_travel_testing_random</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">from_fraction</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">include_ensembl_source</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">include_external_source</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">include_ensembl_destination</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">include_external_destination</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">verbose</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">verbose_detailed</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">strict_forward</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">convert_using_release</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">prioritize_to_one_filter</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">return_result</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/idtrack/_track_tests.html#TrackTests.history_travel_testing_random"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#idtrack._track_tests.TrackTests.history_travel_testing_random" title="Link to this definition"></a></dt>
<dd><p>Convenience wrapper around <a class="reference internal" href="#idtrack._track_tests.TrackTests.history_travel_testing" title="idtrack._track_tests.TrackTests.history_travel_testing"><code class="xref py py-meth docutils literal notranslate"><span class="pre">history_travel_testing()</span></code></a>.</p>
<p>The routine generates a <em>random</em> but internally consistent test case
via <a class="reference internal" href="#idtrack._track_tests.TrackTests.history_travel_testing_random_arguments_generator" title="idtrack._track_tests.TrackTests.history_travel_testing_random_arguments_generator"><code class="xref py py-meth docutils literal notranslate"><span class="pre">history_travel_testing_random_arguments_generator()</span></code></a>, logs
the chosen parameters (unless <em>verbose</em> is <em>False</em>) and delegates the
heavy lifting to <a class="reference internal" href="#idtrack._track_tests.TrackTests.history_travel_testing" title="idtrack._track_tests.TrackTests.history_travel_testing"><code class="xref py py-meth docutils literal notranslate"><span class="pre">history_travel_testing()</span></code></a>.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>from_fraction</strong> (<em>float</em>) – Fraction of IDs to sample from the source set.</p></li>
<li><p><strong>strict_forward</strong> (<em>bool</em>) – Forwarded to the argument generator.</p></li>
<li><p><strong>convert_using_release</strong> (<em>bool</em>) – Forwarded to
<a class="reference internal" href="#idtrack._track_tests.TrackTests.history_travel_testing" title="idtrack._track_tests.TrackTests.history_travel_testing"><code class="xref py py-meth docutils literal notranslate"><span class="pre">history_travel_testing()</span></code></a>.</p></li>
<li><p><strong>prioritize_to_one_filter</strong> (<em>bool</em>) – Forwarded to
<a class="reference internal" href="#idtrack._track_tests.TrackTests.history_travel_testing" title="idtrack._track_tests.TrackTests.history_travel_testing"><code class="xref py py-meth docutils literal notranslate"><span class="pre">history_travel_testing()</span></code></a>.</p></li>
<li><p><strong>verbose</strong> (<em>bool</em>) – Show coarse progress information.</p></li>
<li><p><strong>verbose_detailed</strong> (<em>bool</em>) – Include extended per-ID counters in the progress bar.</p></li>
<li><p><strong>return_result</strong> (<em>bool</em>) – Todo.</p></li>
<li><p><strong>include_ensembl_source</strong> – Todo.</p></li>
<li><p><strong>include_external_source</strong> – Todo.</p></li>
<li><p><strong>include_ensembl_destination</strong> – Todo.</p></li>
<li><p><strong>include_external_destination</strong> – Todo.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>The <em>metrics</em> dictionary returned by
<a class="reference internal" href="#idtrack._track_tests.TrackTests.history_travel_testing" title="idtrack._track_tests.TrackTests.history_travel_testing"><code class="xref py py-meth docutils literal notranslate"><span class="pre">history_travel_testing()</span></code></a>.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>dict</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="idtrack._track_tests.TrackTests.history_travel_testing_random_arguments_generator">
<span class="sig-name descname"><span class="pre">history_travel_testing_random_arguments_generator</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">strict_forward</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">include_exclude_list</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/idtrack/_track_tests.html#TrackTests.history_travel_testing_random_arguments_generator"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#idtrack._track_tests.TrackTests.history_travel_testing_random_arguments_generator" title="Link to this definition"></a></dt>
<dd><p>Generate a plausible random parameter set for <a class="reference internal" href="#idtrack._track_tests.TrackTests.history_travel_testing" title="idtrack._track_tests.TrackTests.history_travel_testing"><code class="xref py py-meth docutils literal notranslate"><span class="pre">history_travel_testing()</span></code></a>.</p>
<p>The helper picks <strong>compatible</strong> source/target assemblies, releases and
databases so the subsequent conversion test has a realistic chance to
succeed.  When <em>strict_forward</em> is <em>True</em> the target release is
guaranteed to be <strong>≥</strong> the source release (no time-travel back).</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>strict_forward</strong> (<em>bool</em>) – Enforce a non-decreasing release direction.</p></li>
<li><p><strong>include_exclude_list</strong> (<em>list</em>) – Todo.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p>Keys <code class="docutils literal notranslate"><span class="pre">from_assembly</span></code>, <code class="docutils literal notranslate"><span class="pre">from_release</span></code>, <code class="docutils literal notranslate"><span class="pre">to_release</span></code>,
<code class="docutils literal notranslate"><span class="pre">from_database</span></code>, <code class="docutils literal notranslate"><span class="pre">to_database</span></code> ready to be splatted into
<a class="reference internal" href="#idtrack._track_tests.TrackTests.history_travel_testing" title="idtrack._track_tests.TrackTests.history_travel_testing"><code class="xref py py-meth docutils literal notranslate"><span class="pre">history_travel_testing()</span></code></a>.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>dict</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="idtrack._track_tests.TrackTests.how_many_corresponding_path_ensembl">
<span class="sig-name descname"><span class="pre">how_many_corresponding_path_ensembl</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">from_release</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">from_assembly</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">to_release</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">go_external</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">verbose</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/idtrack/_track_tests.html#TrackTests.how_many_corresponding_path_ensembl"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#idtrack._track_tests.TrackTests.how_many_corresponding_path_ensembl" title="Link to this definition"></a></dt>
<dd><p>Count history paths between two Ensembl releases.</p>
<p>The method iterates over <strong>all</strong> Ensembl-gene stable IDs that exist in
<em>from_release</em>/<em>from_assembly</em>.  For every ID that is present in the
graph it calls <code class="xref py py-meth docutils literal notranslate"><span class="pre">idtrack.Track.get_possible_paths()</span></code> and records
how many distinct paths the searcher finds to <em>to_release</em>.</p>
<p>The routine is <strong>non-destructive</strong>; it merely provides a quick way to
gauge the <cite>density</cite> of the history graph or to spot releases where
path-finding was unexpectedly difficult.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>from_release</strong> (<em>int</em>) – Source Ensembl release number.</p></li>
<li><p><strong>from_assembly</strong> (<em>int</em>) – Source genome assembly code.</p></li>
<li><p><strong>to_release</strong> (<em>int</em>) – Target Ensembl release number.</p></li>
<li><p><strong>go_external</strong> (<em>bool</em>) – If <em>True</em> history paths are allowed to temporarily
leave the Ensembl lineage via external databases.</p></li>
<li><p><strong>verbose</strong> (<em>bool</em>) – Show a tqdm progress bar (default <em>True</em>).</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><p>A list of two-element sub-lists
<code class="docutils literal notranslate"><span class="pre">[stable_id,</span> <span class="pre">n_paths]</span></code> where <code class="docutils literal notranslate"><span class="pre">n_paths</span></code> is</p>
<blockquote>
<div><ul class="simple">
<li><p>an <em>int</em> ≥ 0 when the ID was in the graph, or</p></li>
<li><p><em>None</em> when the source ID was absent.</p></li>
</ul>
</div></blockquote>
</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>list[list[Union[str, int, None]]]</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="idtrack._track_tests.TrackTests.is_base_is_range_correct">
<span class="sig-name descname"><span class="pre">is_base_is_range_correct</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">verbose</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/idtrack/_track_tests.html#TrackTests.is_base_is_range_correct"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#idtrack._track_tests.TrackTests.is_base_is_range_correct" title="Link to this definition"></a></dt>
<dd><p>Verify consistency of <em>base-gene</em> active-range calculations.</p>
<p>Each “base Ensembl gene” node (<code class="docutils literal notranslate"><span class="pre">node_type</span> <span class="pre">==</span> <span class="pre">'base_ensembl_gene'</span></code>)
has an <em>active release range</em>—the list of Ensembl releases during
which descendants of the gene were present.  There are two
independent ways to obtain this information:</p>
<ol class="arabic simple">
<li><p><strong>High-level helper</strong>
<code class="docutils literal notranslate"><span class="pre">graph.get_active_ranges_of_base_id_alternative</span></code> - a cached
convenience wrapper.</p></li>
<li><p><strong>Low-level reconstruction</strong> by aggregating the <em>combined_edges</em>
table and converting the set of releases into compact
<code class="docutils literal notranslate"><span class="pre">[start,</span> <span class="pre">end]</span></code> slices via <code class="docutils literal notranslate"><span class="pre">graph.list_to_ranges</span></code>.</p></li>
</ol>
<p>This test iterates through <strong>all</strong> base-gene nodes and asserts that
the two methods deliver byte-identical results.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>verbose</strong> (<em>bool</em>) – If <em>True</em> (default) show a tqdm progress
bar that updates with the current node under inspection.</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><code class="docutils literal notranslate"><span class="pre">True</span></code> if <em>every</em> base-gene has matching ranges; <code class="docutils literal notranslate"><span class="pre">False</span></code>
as soon as a single mismatch is encountered.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>bool</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="idtrack._track_tests.TrackTests.is_combined_edges_dicts_overlapping_and_complete">
<span class="sig-name descname"><span class="pre">is_combined_edges_dicts_overlapping_and_complete</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/idtrack/_track_tests.html#TrackTests.is_combined_edges_dicts_overlapping_and_complete"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#idtrack._track_tests.TrackTests.is_combined_edges_dicts_overlapping_and_complete" title="Link to this definition"></a></dt>
<dd><p>Check edge-dictionary partitioning invariants.</p>
<p>The Track graph materialises three <em>edge caches</em>—<code class="docutils literal notranslate"><span class="pre">combined_edges</span></code>
and its two specialised siblings—each storing adjacency and release
metadata for a different subset of nodes:</p>
<ul class="simple">
<li><p><code class="docutils literal notranslate"><span class="pre">combined_edges</span></code> - all nodes, including backbone genes.</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">combined_edges_genes</span></code> - stable Ensembl genes (non-assembly-specific).</p></li>
<li><p><code class="docutils literal notranslate"><span class="pre">combined_edges_assembly_specific_genes</span></code> - genes that exist only
on a single assembly.</p></li>
</ul>
<p>The design contract says:</p>
<ol class="arabic simple">
<li><p><strong>Disjointness</strong> - No node key may appear in more than one
dictionary.</p></li>
<li><p><strong>Completeness</strong> - The <em>union</em> of the dictionaries must cover all
graph nodes <strong>except</strong> those that represent <em>alternative database
versions</em> (e.g. “EnsemblMetazoa”) which are intentionally kept
separate.</p></li>
</ol>
<p>This routine enforces both rules.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p><code class="docutils literal notranslate"><span class="pre">True</span></code> if the dictionaries are pair-wise disjoint <strong>and</strong>
collectively cover every eligible node; <code class="docutils literal notranslate"><span class="pre">False</span></code> otherwise.</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p>bool</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="idtrack._track_tests.TrackTests.is_edge_with_same_nts_only_at_backbone_nodes">
<span class="sig-name descname"><span class="pre">is_edge_with_same_nts_only_at_backbone_nodes</span></span><span class="sig-paren">(</span><span class="sig-paren">)</span><a class="reference internal" href="_modules/idtrack/_track_tests.html#TrackTests.is_edge_with_same_nts_only_at_backbone_nodes"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#idtrack._track_tests.TrackTests.is_edge_with_same_nts_only_at_backbone_nodes" title="Link to this definition"></a></dt>
<dd><p>Assert <em>same-node-type</em> edges exist <strong>only</strong> between backbone genes.</p>
<p>The graph is a <em>multilayer</em> network where nodes of different</p>
<p>This method traverses every base-gene node and checks that condition.</p>
<dl class="field-list simple">
<dt class="field-odd">Returns<span class="colon">:</span></dt>
<dd class="field-odd"><p><code class="docutils literal notranslate"><span class="pre">True</span></code> when no overlaps are found; <code class="docutils literal notranslate"><span class="pre">False</span></code> otherwise.
Each offending base ID triggers a warning with the conflicting
ranges.</p>
</dd>
<dt class="field-even">Return type<span class="colon">:</span></dt>
<dd class="field-even"><p>bool</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="idtrack._track_tests.TrackTests.is_final_external_conversion_robust">
<span class="sig-name descname"><span class="pre">is_final_external_conversion_robust</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">convert_using_release</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">database</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">ens_rel</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">verbose</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">from_fraction</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">1.0</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">prioritize_to_one_filter</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">False</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/idtrack/_track_tests.html#TrackTests.is_final_external_conversion_robust"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#idtrack._track_tests.TrackTests.is_final_external_conversion_robust" title="Link to this definition"></a></dt>
<dd><p>Validate Ensembl→external conversion against MySQL ground truth.</p>
<p>A random external database is chosen for <strong>every</strong> genome assembly. For
the selected combination the method grabs the authoritative mapping
table (graph-ID → external ID set) from MySQL and converts the same
graph-IDs with <code class="xref py py-meth docutils literal notranslate"><span class="pre">idtrack.Track.convert()</span></code>.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>convert_using_release</strong> (<em>bool</em>) – Whether to pin the <em>from_release</em> when
calling the converter.  Keeping this <em>True</em> usually speeds up
the search and mimics user-facing behaviour.</p></li>
<li><p><strong>verbose</strong> (<em>bool</em>) – Print the current assembly/database/release being tested.</p></li>
<li><p><strong>prioritize_to_one_filter</strong> (<em>bool</em>) – Todo.</p></li>
<li><p><strong>ens_rel</strong> (<em>int</em><em> | </em><em>None</em>) – Todo.</p></li>
<li><p><strong>from_fraction</strong> (<em>float</em>) – Todo.</p></li>
<li><p><strong>database</strong> (<em>str</em><em> | </em><em>None</em>) – Todo.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><dl class="simple">
<dt><em>True</em> if every converted set equals the MySQL reference,</dt><dd><p><em>False</em> upon the first deviation.</p>
</dd>
</dl>
</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>bool</p>
</dd>
<dt class="field-even">Raises<span class="colon">:</span></dt>
<dd class="field-even"><p><strong>ValueError</strong> – Todo.</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="idtrack._track_tests.TrackTests.is_id_functions_consistent_ensembl">
<span class="sig-name descname"><span class="pre">is_id_functions_consistent_ensembl</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">verbose</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/idtrack/_track_tests.html#TrackTests.is_id_functions_consistent_ensembl"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#idtrack._track_tests.TrackTests.is_id_functions_consistent_ensembl" title="Link to this definition"></a></dt>
<dd><p>Ensure Ensembl ID-list helpers agree with the SQL back-end.</p>
<p>For every release listed in <code class="docutils literal notranslate"><span class="pre">graph.graph[&quot;confident_for_release&quot;]</span></code>
the test compares two independent sources of Ensembl-gene IDs for the
<em>current genome assembly</em>:</p>
<ol class="arabic simple">
<li><p>IDs retrieved directly from MySQL via
<code class="xref py py-class docutils literal notranslate"><span class="pre">DatabaseManager</span></code>.</p></li>
<li><p>IDs returned by <code class="xref py py-meth docutils literal notranslate"><span class="pre">idtrack.Track.get_id_list()</span></code> from the graph.</p></li>
</ol>
<p>A mismatch means either the graph was built incompletely or the
helper functions drift out of sync with the database schema.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>verbose</strong> (<em>bool</em>) – If <code class="docutils literal notranslate"><span class="pre">True</span></code> (default) show a tqdm progress-bar while
iterating through the releases.</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><code class="docutils literal notranslate"><span class="pre">True</span></code> when <strong>all</strong> releases produce identical sets; else
<code class="docutils literal notranslate"><span class="pre">False</span></code> (a descriptive warning is logged).</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>bool</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="idtrack._track_tests.TrackTests.is_id_functions_consistent_ensembl_2">
<span class="sig-name descname"><span class="pre">is_id_functions_consistent_ensembl_2</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">verbose</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/idtrack/_track_tests.html#TrackTests.is_id_functions_consistent_ensembl_2"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#idtrack._track_tests.TrackTests.is_id_functions_consistent_ensembl_2" title="Link to this definition"></a></dt>
<dd><p>Cross-check Ensembl ID <em>range</em> helpers against raw edge data.</p>
<p>For every <strong>backbone</strong> Ensembl-gene node the routine computes the list
of active release ranges in two distinct ways:</p>
<ol class="arabic simple">
<li><p><strong>Raw computation</strong> - by flattening
<code class="docutils literal notranslate"><span class="pre">graph.combined_edges_genes</span></code> and compacting the releases with
<code class="xref py py-meth docutils literal notranslate"><span class="pre">idtrack.Track.list_to_ranges()</span></code>.</p></li>
<li><p><strong>Cached lookup</strong> - via the lazily built dictionary
<code class="docutils literal notranslate"><span class="pre">graph.get_active_ranges_of_id</span></code>.</p></li>
</ol>
<p>The two lists must match exactly.  A divergence would indicate that
the cached helper is out of sync with the authoritative edge
structure.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>verbose</strong> (<em>bool</em>) – If <code class="docutils literal notranslate"><span class="pre">True</span></code> (default) wrap the iteration in a tqdm bar.</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><code class="docutils literal notranslate"><span class="pre">True</span></code> if <em>all</em> gene nodes pass; <code class="docutils literal notranslate"><span class="pre">False</span></code> after the first
failure (a warning is emitted).</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>bool</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="idtrack._track_tests.TrackTests.is_id_functions_consistent_external">
<span class="sig-name descname"><span class="pre">is_id_functions_consistent_external</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">verbose</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/idtrack/_track_tests.html#TrackTests.is_id_functions_consistent_external"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#idtrack._track_tests.TrackTests.is_id_functions_consistent_external" title="Link to this definition"></a></dt>
<dd><p>Check external-ID list helpers against the raw MySQL tables.</p>
<p>For <strong>every</strong> combination of <em>assembly</em>, <em>Ensembl release</em> (limited to
<code class="xref py py-attr docutils literal notranslate"><span class="pre">graph.graph[&quot;confident_for_release&quot;]</span></code>) and <em>external
database</em> this test performs the following steps:</p>
<ol class="arabic simple">
<li><p>Query the authoritative list of external IDs directly from the
MySQL snapshot via <code class="xref py py-class docutils literal notranslate"><span class="pre">DatabaseManager</span></code>.</p></li>
<li><p>Ask the in-memory graph for the same list via
<code class="xref py py-meth docutils literal notranslate"><span class="pre">idtrack.Track.get_id_list()</span></code>.</p></li>
<li><p>Normalise node names through
<code class="xref py py-meth docutils literal notranslate"><span class="pre">idtrack.Track.node_name_alternatives()</span></code> to cope with the
occasional “_1” suffix.</p></li>
<li><p>Compare the two sets.  A mismatch is logged and the method returns
<code class="docutils literal notranslate"><span class="pre">False</span></code> immediately.</p></li>
</ol>
<p>The exhaustive traversal is expensive (minutes for large genomes) but
ensures the graph`s indexing helpers never drift from the actual
database content.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>verbose</strong> (<em>bool</em>) – If <em>True</em> (default) display a tqdm progress bar and emit
log messages at INFO level.  When <em>False</em> the method runs
silently.</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><em>True</em> when every single comparison matched, <em>False</em> as soon
as an inconsistency is encountered.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>bool</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="idtrack._track_tests.TrackTests.is_node_consistency_robust">
<span class="sig-name descname"><span class="pre">is_node_consistency_robust</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">verbose</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/idtrack/_track_tests.html#TrackTests.is_node_consistency_robust"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#idtrack._track_tests.TrackTests.is_node_consistency_robust" title="Link to this definition"></a></dt>
<dd><p>Check for illegal neighbour relationships and multi-edges.</p>
<p>The graph may contain <strong>exactly one</strong> edge between nodes of <em>different</em>
node-types.  Nodes of the <em>same</em> node-type are only allowed when that
type is the Ensembl backbone (<code class="docutils literal notranslate"><span class="pre">ensembl_gene</span></code>).  Any deviation - a
lateral same-type connection or &gt;1 multi-edge - is logged and aborts
the test.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>verbose</strong> (<em>bool</em>) – Print offending nodes when a violation is detected.</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><em>True</em> when the graph satisfies the topology rules, <em>False</em>
otherwise.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>bool</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="idtrack._track_tests.TrackTests.is_range_functions_robust">
<span class="sig-name descname"><span class="pre">is_range_functions_robust</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">verbose</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">True</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/idtrack/_track_tests.html#TrackTests.is_range_functions_robust"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#idtrack._track_tests.TrackTests.is_range_functions_robust" title="Link to this definition"></a></dt>
<dd><p>Detect overlapping release ranges among sibling Ensembl IDs.</p>
<p>A <em>base</em> Ensembl-gene ID is the stable identifier that groups multiple
versioned Ensembl-gene records (siblings).  The gene-history model
requires that the release ranges of sibling IDs <strong>never overlap</strong> -
each release must be covered by <em>exactly one</em> child stable ID.</p>
<p>This method traverses every base-gene node and checks that condition.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><p><strong>verbose</strong> (<em>bool</em>) – If <code class="docutils literal notranslate"><span class="pre">True</span></code> (default) display a tqdm progress-bar.</p>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><code class="docutils literal notranslate"><span class="pre">True</span></code> when no overlaps are found; <code class="docutils literal notranslate"><span class="pre">False</span></code> otherwise.
Each offending base ID triggers a warning with the conflicting
ranges.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>bool</p>
</dd>
</dl>
</dd></dl>

<dl class="py method">
<dt class="sig sig-object py" id="idtrack._track_tests.TrackTests.random_dataset_source_generator">
<span class="sig-name descname"><span class="pre">random_dataset_source_generator</span></span><span class="sig-paren">(</span><em class="sig-param"><span class="n"><span class="pre">assembly</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">include_external</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">include_ensembl</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">for_final_database</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">only_backbone_tests</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">release_lower_limit</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em>, <em class="sig-param"><span class="n"><span class="pre">form</span></span><span class="o"><span class="pre">=</span></span><span class="default_value"><span class="pre">None</span></span></em><span class="sig-paren">)</span><a class="reference internal" href="_modules/idtrack/_track_tests.html#TrackTests.random_dataset_source_generator"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#idtrack._track_tests.TrackTests.random_dataset_source_generator" title="Link to this definition"></a></dt>
<dd><p>Pick a random (&lt;database&gt;, &lt;assembly&gt;, &lt;release&gt;) tuple.</p>
<p>The function guarantees that the triple actually exists in the graph
and - if <em>release_lower_limit</em> is provided - honours the minimum
release constraint.</p>
<dl class="field-list simple">
<dt class="field-odd">Parameters<span class="colon">:</span></dt>
<dd class="field-odd"><ul class="simple">
<li><p><strong>assembly</strong> (<em>int</em>) – NCBI assembly code (integer).</p></li>
<li><p><strong>include_ensembl</strong> (<em>bool</em>) – Whether Ensembl backbone databases may be returned
as <em>database</em>.</p></li>
<li><p><strong>release_lower_limit</strong> (<em>int</em><em> | </em><em>None</em>) – Smallest permissible Ensembl release number
for the returned triple.  <em>None</em> disables the filter.</p></li>
<li><p><strong>form</strong> (<em>str</em><em> | </em><em>None</em>) – Restrict the draw to a particular connection <em>form</em>
(protein/coding/gene).  <em>None</em> means no restriction.</p></li>
<li><p><strong>only_backbone_tests</strong> (<em>bool</em>) – Todo.</p></li>
<li><p><strong>include_external</strong> (<em>bool</em>) – Todo.</p></li>
<li><p><strong>for_final_database</strong> (<em>bool</em>) – Todo.</p></li>
</ul>
</dd>
<dt class="field-even">Returns<span class="colon">:</span></dt>
<dd class="field-even"><p><code class="docutils literal notranslate"><span class="pre">(&lt;database&gt;,</span> <span class="pre">&lt;assembly&gt;,</span> <span class="pre">&lt;release&gt;)</span></code> or <em>None</em> when
no matching release exists.</p>
</dd>
<dt class="field-odd">Return type<span class="colon">:</span></dt>
<dd class="field-odd"><p>tuple | None</p>
</dd>
<dt class="field-even">Raises<span class="colon">:</span></dt>
<dd class="field-even"><p><strong>ValueError</strong> – Todo.</p>
</dd>
</dl>
</dd></dl>

</dd></dl>

<dl class="py class" id="module-idtrack._db">
<dt class="sig sig-object py" id="idtrack._db.DB">
<em class="property"><span class="pre">class</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">DB</span></span><a class="reference internal" href="_modules/idtrack/_db.html#DB"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#idtrack._db.DB" title="Link to this definition"></a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">object</span></code></p>
<p>Store constants shared across IDTrack modules for Ensembl data access and graph construction.</p>
<p>This class centralizes every constant that multiple components (e.g. <code class="xref py py-class docutils literal notranslate"><span class="pre">idtrack.graph.GraphMaker</span></code>,
<code class="xref py py-class docutils literal notranslate"><span class="pre">idtrack.pathfinder.PathFinder</span></code>) rely on when talking to the Ensembl FTP mirror, REST API,
and public MySQL instances. Housing the values in one immutable namespace prevents circular imports,
ensures a single source of truth, and simplifies testing. <cite>DB</cite> is never instantiated; import the class
and reference its attributes directly.</p>
<dl class="py attribute">
<dt class="sig sig-object py" id="idtrack._db.DB.id_ver_delimiter">
<span class="sig-name descname"><span class="pre">id_ver_delimiter</span></span><a class="headerlink" href="#idtrack._db.DB.id_ver_delimiter" title="Link to this definition"></a></dt>
<dd><p>Character separating a stable identifier from its version suffix.</p>
<dl class="field-list simple">
<dt class="field-odd">Type<span class="colon">:</span></dt>
<dd class="field-odd"><p>str</p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="idtrack._db.DB.first_version">
<span class="sig-name descname"><span class="pre">first_version</span></span><a class="headerlink" href="#idtrack._db.DB.first_version" title="Link to this definition"></a></dt>
<dd><p>Default version assumed when an ID lacks an explicit version component.</p>
<dl class="field-list simple">
<dt class="field-odd">Type<span class="colon">:</span></dt>
<dd class="field-odd"><p>int</p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="idtrack._db.DB.connection_timeout">
<span class="sig-name descname"><span class="pre">connection_timeout</span></span><a class="headerlink" href="#idtrack._db.DB.connection_timeout" title="Link to this definition"></a></dt>
<dd><p>TCP connect timeout in seconds used by both FTP and REST clients.</p>
<dl class="field-list simple">
<dt class="field-odd">Type<span class="colon">:</span></dt>
<dd class="field-odd"><p>int</p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="idtrack._db.DB.reading_timeout">
<span class="sig-name descname"><span class="pre">reading_timeout</span></span><a class="headerlink" href="#idtrack._db.DB.reading_timeout" title="Link to this definition"></a></dt>
<dd><p>Socket read timeout in seconds applied to FTP and REST operations.</p>
<dl class="field-list simple">
<dt class="field-odd">Type<span class="colon">:</span></dt>
<dd class="field-odd"><p>int</p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="idtrack._db.DB.ensembl_ftp_base">
<span class="sig-name descname"><span class="pre">ensembl_ftp_base</span></span><a class="headerlink" href="#idtrack._db.DB.ensembl_ftp_base" title="Link to this definition"></a></dt>
<dd><p>Hostname of the Ensembl public FTP mirror.</p>
<dl class="field-list simple">
<dt class="field-odd">Type<span class="colon">:</span></dt>
<dd class="field-odd"><p>str</p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="idtrack._db.DB.rest_server_api">
<span class="sig-name descname"><span class="pre">rest_server_api</span></span><a class="headerlink" href="#idtrack._db.DB.rest_server_api" title="Link to this definition"></a></dt>
<dd><p>Root URL of the Ensembl REST API.</p>
<dl class="field-list simple">
<dt class="field-odd">Type<span class="colon">:</span></dt>
<dd class="field-odd"><p>str</p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="idtrack._db.DB.rest_server_ext">
<span class="sig-name descname"><span class="pre">rest_server_ext</span></span><a class="headerlink" href="#idtrack._db.DB.rest_server_ext" title="Link to this definition"></a></dt>
<dd><p>Resource path appended to <a class="reference internal" href="#idtrack._db.DB.rest_server_api" title="idtrack._db.DB.rest_server_api"><code class="xref py py-data docutils literal notranslate"><span class="pre">rest_server_api</span></code></a> to query species metadata.</p>
<dl class="field-list simple">
<dt class="field-odd">Type<span class="colon">:</span></dt>
<dd class="field-odd"><p>str</p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="idtrack._db.DB.mysql_host">
<span class="sig-name descname"><span class="pre">mysql_host</span></span><a class="headerlink" href="#idtrack._db.DB.mysql_host" title="Link to this definition"></a></dt>
<dd><p>Hostname of the Ensembl public MySQL server.</p>
<dl class="field-list simple">
<dt class="field-odd">Type<span class="colon">:</span></dt>
<dd class="field-odd"><p>str</p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="idtrack._db.DB.myqsl_user">
<span class="sig-name descname"><span class="pre">myqsl_user</span></span><a class="headerlink" href="#idtrack._db.DB.myqsl_user" title="Link to this definition"></a></dt>
<dd><p>Username for anonymous MySQL access.</p>
<dl class="field-list simple">
<dt class="field-odd">Type<span class="colon">:</span></dt>
<dd class="field-odd"><p>str</p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="idtrack._db.DB.mysql_togo">
<span class="sig-name descname"><span class="pre">mysql_togo</span></span><a class="headerlink" href="#idtrack._db.DB.mysql_togo" title="Link to this definition"></a></dt>
<dd><p>Placeholder string kept for backward compatibility when assembling connection URLs.</p>
<dl class="field-list simple">
<dt class="field-odd">Type<span class="colon">:</span></dt>
<dd class="field-odd"><p>str</p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="idtrack._db.DB.assembly_mysqlport_priority">
<span class="sig-name descname"><span class="pre">assembly_mysqlport_priority</span></span><a class="headerlink" href="#idtrack._db.DB.assembly_mysqlport_priority" title="Link to this definition"></a></dt>
<dd><p>Mapping of genome assembly identifiers
(e.g. <code class="docutils literal notranslate"><span class="pre">38</span></code> or <code class="docutils literal notranslate"><span class="pre">37</span></code>) to the port, priority rank (1 = highest), and minimum release available
on that port.</p>
<dl class="field-list simple">
<dt class="field-odd">Type<span class="colon">:</span></dt>
<dd class="field-odd"><p>dict[int, dict[str, int]]</p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="idtrack._db.DB.assembly_priority">
<span class="sig-name descname"><span class="pre">assembly_priority</span></span><a class="headerlink" href="#idtrack._db.DB.assembly_priority" title="Link to this definition"></a></dt>
<dd><p>Sorted list of priority integers derived from
<a class="reference internal" href="#idtrack._db.DB.assembly_mysqlport_priority" title="idtrack._db.DB.assembly_mysqlport_priority"><code class="xref py py-data docutils literal notranslate"><span class="pre">assembly_mysqlport_priority</span></code></a>, highest priority first.</p>
<dl class="field-list simple">
<dt class="field-odd">Type<span class="colon">:</span></dt>
<dd class="field-odd"><p>list[int]</p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="idtrack._db.DB.sorted_assemblies">
<span class="sig-name descname"><span class="pre">sorted_assemblies</span></span><a class="headerlink" href="#idtrack._db.DB.sorted_assemblies" title="Link to this definition"></a></dt>
<dd><p>Equivalent information sorted by ascending
priority, making <code class="docutils literal notranslate"><span class="pre">sorted_assemblies[0]</span></code> the assembly with rank 1.</p>
<dl class="field-list simple">
<dt class="field-odd">Type<span class="colon">:</span></dt>
<dd class="field-odd"><p>list[tuple[int, dict[str, int]]]</p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="idtrack._db.DB.main_assembly">
<span class="sig-name descname"><span class="pre">main_assembly</span></span><a class="headerlink" href="#idtrack._db.DB.main_assembly" title="Link to this definition"></a></dt>
<dd><p>Assembly identifier chosen as the preferred default for queries.</p>
<dl class="field-list simple">
<dt class="field-odd">Type<span class="colon">:</span></dt>
<dd class="field-odd"><p>int</p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="idtrack._db.DB.synonym_id_nodes_prefix">
<span class="sig-name descname"><span class="pre">synonym_id_nodes_prefix</span></span><a class="headerlink" href="#idtrack._db.DB.synonym_id_nodes_prefix" title="Link to this definition"></a></dt>
<dd><p>Prefix inserted before node identifiers that represent synonym edges.</p>
<dl class="field-list simple">
<dt class="field-odd">Type<span class="colon">:</span></dt>
<dd class="field-odd"><p>str</p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="idtrack._db.DB.no_old_node_id">
<span class="sig-name descname"><span class="pre">no_old_node_id</span></span><a class="headerlink" href="#idtrack._db.DB.no_old_node_id" title="Link to this definition"></a></dt>
<dd><p>Sentinel used when a historical ID is retired.</p>
<dl class="field-list simple">
<dt class="field-odd">Type<span class="colon">:</span></dt>
<dd class="field-odd"><p>str</p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="idtrack._db.DB.no_new_node_id">
<span class="sig-name descname"><span class="pre">no_new_node_id</span></span><a class="headerlink" href="#idtrack._db.DB.no_new_node_id" title="Link to this definition"></a></dt>
<dd><p>Sentinel used when no future successor exists.</p>
<dl class="field-list simple">
<dt class="field-odd">Type<span class="colon">:</span></dt>
<dd class="field-odd"><p>str</p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="idtrack._db.DB.alternative_versions">
<span class="sig-name descname"><span class="pre">alternative_versions</span></span><a class="headerlink" href="#idtrack._db.DB.alternative_versions" title="Link to this definition"></a></dt>
<dd><p>Two sentinels—<a class="reference internal" href="#idtrack._db.DB.no_new_node_id" title="idtrack._db.DB.no_new_node_id"><code class="xref py py-data docutils literal notranslate"><span class="pre">no_new_node_id</span></code></a> and <a class="reference internal" href="#idtrack._db.DB.no_old_node_id" title="idtrack._db.DB.no_old_node_id"><code class="xref py py-data docutils literal notranslate"><span class="pre">no_old_node_id</span></code></a>.</p>
<dl class="field-list simple">
<dt class="field-odd">Type<span class="colon">:</span></dt>
<dd class="field-odd"><p>set[str]</p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="idtrack._db.DB.hyperconnecting_threshold">
<span class="sig-name descname"><span class="pre">hyperconnecting_threshold</span></span><a class="headerlink" href="#idtrack._db.DB.hyperconnecting_threshold" title="Link to this definition"></a></dt>
<dd><p>Maximum allowable out-degree before a node is considered
hyper-connected and ignored by breadth-first expansions.</p>
<dl class="field-list simple">
<dt class="field-odd">Type<span class="colon">:</span></dt>
<dd class="field-odd"><p>int</p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="idtrack._db.DB.node_type_str">
<span class="sig-name descname"><span class="pre">node_type_str</span></span><a class="headerlink" href="#idtrack._db.DB.node_type_str" title="Link to this definition"></a></dt>
<dd><p>Edge/Node attribute key holding the node type value.</p>
<dl class="field-list simple">
<dt class="field-odd">Type<span class="colon">:</span></dt>
<dd class="field-odd"><p>str</p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="idtrack._db.DB.nts_external">
<span class="sig-name descname"><span class="pre">nts_external</span></span><a class="headerlink" href="#idtrack._db.DB.nts_external" title="Link to this definition"></a></dt>
<dd><p>Canonical node type assigned to entities originating outside Ensembl.</p>
<dl class="field-list simple">
<dt class="field-odd">Type<span class="colon">:</span></dt>
<dd class="field-odd"><p>str</p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="idtrack._db.DB.forms_in_order">
<span class="sig-name descname"><span class="pre">forms_in_order</span></span><a class="headerlink" href="#idtrack._db.DB.forms_in_order" title="Link to this definition"></a></dt>
<dd><p>Stable ordering of Ensembl entity forms (<code class="docutils literal notranslate"><span class="pre">gene</span></code>, <code class="docutils literal notranslate"><span class="pre">transcript</span></code>,
<code class="docutils literal notranslate"><span class="pre">translation</span></code>). Order matters when inferring parent/child relationships.</p>
<dl class="field-list simple">
<dt class="field-odd">Type<span class="colon">:</span></dt>
<dd class="field-odd"><p>list[str]</p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="idtrack._db.DB.backbone_form">
<span class="sig-name descname"><span class="pre">backbone_form</span></span><a class="headerlink" href="#idtrack._db.DB.backbone_form" title="Link to this definition"></a></dt>
<dd><p>Form selected as backbone for graph traversals (always <code class="docutils literal notranslate"><span class="pre">gene</span></code>).</p>
<dl class="field-list simple">
<dt class="field-odd">Type<span class="colon">:</span></dt>
<dd class="field-odd"><p>str</p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="idtrack._db.DB.nts_ensembl">
<span class="sig-name descname"><span class="pre">nts_ensembl</span></span><a class="headerlink" href="#idtrack._db.DB.nts_ensembl" title="Link to this definition"></a></dt>
<dd><p>Map each canonical form to its namespaced node type
(<code class="docutils literal notranslate"><span class="pre">gene</span></code> → <code class="docutils literal notranslate"><span class="pre">ensembl_gene</span></code>, etc.).</p>
<dl class="field-list simple">
<dt class="field-odd">Type<span class="colon">:</span></dt>
<dd class="field-odd"><p>dict[str, str]</p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="idtrack._db.DB.nts_ensembl_reverse">
<span class="sig-name descname"><span class="pre">nts_ensembl_reverse</span></span><a class="headerlink" href="#idtrack._db.DB.nts_ensembl_reverse" title="Link to this definition"></a></dt>
<dd><p>Reverse mapping of <a class="reference internal" href="#idtrack._db.DB.nts_ensembl" title="idtrack._db.DB.nts_ensembl"><code class="xref py py-data docutils literal notranslate"><span class="pre">nts_ensembl</span></code></a>.</p>
<dl class="field-list simple">
<dt class="field-odd">Type<span class="colon">:</span></dt>
<dd class="field-odd"><p>dict[str, str]</p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="idtrack._db.DB.nts_assembly">
<span class="sig-name descname"><span class="pre">nts_assembly</span></span><a class="headerlink" href="#idtrack._db.DB.nts_assembly" title="Link to this definition"></a></dt>
<dd><p>Form-to-assembly-specific node type map.</p>
<dl class="field-list simple">
<dt class="field-odd">Type<span class="colon">:</span></dt>
<dd class="field-odd"><p>dict[str, dict[str, str]]</p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="idtrack._db.DB.nts_assembly_reverse">
<span class="sig-name descname"><span class="pre">nts_assembly_reverse</span></span><a class="headerlink" href="#idtrack._db.DB.nts_assembly_reverse" title="Link to this definition"></a></dt>
<dd><p>Reverse mapping of <a class="reference internal" href="#idtrack._db.DB.nts_assembly" title="idtrack._db.DB.nts_assembly"><code class="xref py py-data docutils literal notranslate"><span class="pre">nts_assembly</span></code></a>.</p>
<dl class="field-list simple">
<dt class="field-odd">Type<span class="colon">:</span></dt>
<dd class="field-odd"><p>dict[str, dict[str, str]]</p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="idtrack._db.DB.nts_base_ensembl">
<span class="sig-name descname"><span class="pre">nts_base_ensembl</span></span><a class="headerlink" href="#idtrack._db.DB.nts_base_ensembl" title="Link to this definition"></a></dt>
<dd><p>Reduced node type names stripped of assembly suffixes.</p>
<dl class="field-list simple">
<dt class="field-odd">Type<span class="colon">:</span></dt>
<dd class="field-odd"><p>dict[str, str]</p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="idtrack._db.DB.nts_base_ensembl_reverse">
<span class="sig-name descname"><span class="pre">nts_base_ensembl_reverse</span></span><a class="headerlink" href="#idtrack._db.DB.nts_base_ensembl_reverse" title="Link to this definition"></a></dt>
<dd><p>Reverse mapping of <a class="reference internal" href="#idtrack._db.DB.nts_base_ensembl" title="idtrack._db.DB.nts_base_ensembl"><code class="xref py py-data docutils literal notranslate"><span class="pre">nts_base_ensembl</span></code></a>.</p>
<dl class="field-list simple">
<dt class="field-odd">Type<span class="colon">:</span></dt>
<dd class="field-odd"><p>dict[str, str]</p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="idtrack._db.DB.nts_bidirectional_synonymous_search">
<span class="sig-name descname"><span class="pre">nts_bidirectional_synonymous_search</span></span><a class="headerlink" href="#idtrack._db.DB.nts_bidirectional_synonymous_search" title="Link to this definition"></a></dt>
<dd><p>Node types for which synonym searches are
performed bidirectionally.</p>
<dl class="field-list simple">
<dt class="field-odd">Type<span class="colon">:</span></dt>
<dd class="field-odd"><p>set[str]</p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="idtrack._db.DB.nts_assembly_gene">
<span class="sig-name descname"><span class="pre">nts_assembly_gene</span></span><a class="headerlink" href="#idtrack._db.DB.nts_assembly_gene" title="Link to this definition"></a></dt>
<dd><p>Every node type that represents a gene, regardless of assembly.</p>
<dl class="field-list simple">
<dt class="field-odd">Type<span class="colon">:</span></dt>
<dd class="field-odd"><p>set[str]</p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="idtrack._db.DB.connection_dict">
<span class="sig-name descname"><span class="pre">connection_dict</span></span><a class="headerlink" href="#idtrack._db.DB.connection_dict" title="Link to this definition"></a></dt>
<dd><p>Edge attribute key whose value stores connection metadata dictionaries.</p>
<dl class="field-list simple">
<dt class="field-odd">Type<span class="colon">:</span></dt>
<dd class="field-odd"><p>str</p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="idtrack._db.DB.conn_dict_str_ensembl_base">
<span class="sig-name descname"><span class="pre">conn_dict_str_ensembl_base</span></span><a class="headerlink" href="#idtrack._db.DB.conn_dict_str_ensembl_base" title="Link to this definition"></a></dt>
<dd><p>Constant placed under <a class="reference internal" href="#idtrack._db.DB.connection_dict" title="idtrack._db.DB.connection_dict"><code class="xref py py-data docutils literal notranslate"><span class="pre">connection_dict</span></code></a> when the edge
points to an Ensembl data source.</p>
<dl class="field-list simple">
<dt class="field-odd">Type<span class="colon">:</span></dt>
<dd class="field-odd"><p>str</p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="idtrack._db.DB.external_search_settings">
<span class="sig-name descname"><span class="pre">external_search_settings</span></span><a class="headerlink" href="#idtrack._db.DB.external_search_settings" title="Link to this definition"></a></dt>
<dd><p>Default limits for outward traversal into external
databases. Keys are <code class="docutils literal notranslate"><span class="pre">jump_limit</span></code>, <code class="docutils literal notranslate"><span class="pre">synonymous_max_depth</span></code>, and <code class="docutils literal notranslate"><span class="pre">nts_backbone</span></code>.</p>
<dl class="field-list simple">
<dt class="field-odd">Type<span class="colon">:</span></dt>
<dd class="field-odd"><p>dict[str, Any]</p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="idtrack._db.DB.placeholder_na">
<span class="sig-name descname"><span class="pre">placeholder_na</span></span><a class="headerlink" href="#idtrack._db.DB.placeholder_na" title="Link to this definition"></a></dt>
<dd><p>Sentinel stored in HDF5 datasets where a true NA/None value is not
permitted or would break downstream type expectations.</p>
<dl class="field-list simple">
<dt class="field-odd">Type<span class="colon">:</span></dt>
<dd class="field-odd"><p>str</p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="idtrack._db.DB.UTF8">
<span class="sig-name descname"><span class="pre">UTF8</span></span><a class="headerlink" href="#idtrack._db.DB.UTF8" title="Link to this definition"></a></dt>
<dd><p>The literal string <code class="docutils literal notranslate"><span class="pre">&quot;utf-8&quot;</span></code>—a canonical spelling of the UTF-8 encoding name
used when writing variable-length strings to HDF5 files.</p>
<dl class="field-list simple">
<dt class="field-odd">Type<span class="colon">:</span></dt>
<dd class="field-odd"><p>str</p>
</dd>
</dl>
</dd></dl>

<dl class="py attribute">
<dt class="sig sig-object py" id="idtrack._db.DB.UTF8_STR">
<span class="sig-name descname"><span class="pre">UTF8_STR</span></span><a class="headerlink" href="#idtrack._db.DB.UTF8_STR" title="Link to this definition"></a></dt>
<dd><p>Pre-configured variable-length UTF-8 string dtype created via
<code class="xref py py-func docutils literal notranslate"><span class="pre">h5py.string_dtype()</span></code>.  Pass this value when creating HDF5 datasets that should
hold arbitrary Unicode text to avoid hard-coding datatypes throughout the codebase.</p>
<dl class="field-list simple">
<dt class="field-odd">Type<span class="colon">:</span></dt>
<dd class="field-odd"><p>h5py.Datatype</p>
</dd>
</dl>
</dd></dl>

</dd></dl>

<dl class="py exception">
<dt class="sig sig-object py" id="idtrack._db.EmptyConversionMetricsError">
<em class="property"><span class="pre">exception</span><span class="w"> </span></em><span class="sig-name descname"><span class="pre">EmptyConversionMetricsError</span></span><a class="reference internal" href="_modules/idtrack/_db.html#EmptyConversionMetricsError"><span class="viewcode-link"><span class="pre">[source]</span></span></a><a class="headerlink" href="#idtrack._db.EmptyConversionMetricsError" title="Link to this definition"></a></dt>
<dd><p>Bases: <code class="xref py py-class docutils literal notranslate"><span class="pre">ValueError</span></code></p>
<p>Raised when conversion_metrics is empty and no alternative conversion is possible.</p>
</dd></dl>

</section>


           </div>
          </div>
          <footer><div class="rst-footer-buttons" role="navigation" aria-label="Footer">
        <a href="_notebooks/comparison.html" class="btn btn-neutral float-left" title="Comparison with Available Tools" accesskey="p" rel="prev"><span class="fa fa-arrow-circle-left" aria-hidden="true"></span> Previous</a>
        <a href="contributing.html" class="btn btn-neutral float-right" title="Contributor Guide" accesskey="n" rel="next">Next <span class="fa fa-arrow-circle-right" aria-hidden="true"></span></a>
    </div>

  <hr/>

  <div role="contentinfo">
    <p>&#169; Copyright 2025, Kemal Inecik.</p>
  </div>

  Built with <a href="https://www.sphinx-doc.org/">Sphinx</a> using a
    <a href="https://github.com/readthedocs/sphinx_rtd_theme">theme</a>
    provided by <a href="https://readthedocs.org">Read the Docs</a>.
   

</footer>
        </div>
      </div>
    </section>
  </div>
  <script>
      jQuery(function () {
          SphinxRtdTheme.Navigation.enable(true);
      });
  </script> 

</body>
</html>